
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Ensembl Genes is a collection of Python tools to work with Ensembl gene annotation.">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genes/ensembl/genes/transcriptomic_data/select_transcriptomic_data/">
      
      
        <link rel="prev" href="../get_transcriptomic_data/">
      
      
      
        
      
      
      <link rel="icon" href="../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>select_transcriptomic_data - Ensembl Genes</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:300,300i,400,400i,700,700i%7CIBM+Plex+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Lato";--md-code-font:"IBM Plex Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="ensembl" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ensemblgenestranscriptomic_dataselect_transcriptomic_data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="Ensembl Genes" class="md-header__button md-logo" aria-label="Ensembl Genes" data-md-component="logo">
      
  <img src="../../../../img/ebang.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl Genes
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              select_transcriptomic_data
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genes" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub repository
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../.." class="md-tabs__link">
          
  
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../../code_of_conduct/" class="md-tabs__link">
          
  
  
  Development

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../automation/pre_release_ftp/" class="md-tabs__link">
          
  
  
  Code reference

        </a>
      </li>
    
  

    
  

    
  

    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="Ensembl Genes" class="md-nav__button md-logo" aria-label="Ensembl Genes" data-md-component="logo">
      
  <img src="../../../../img/ebang.png" alt="logo">

    </a>
    Ensembl Genes
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genes" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub repository
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Home
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../install/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Install
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Usage
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Development
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Development
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../code_of_conduct/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code of Conduct
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../coverage.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Coverage report
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    Code reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Code reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" checked>
        
          
          <label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    ensembl
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    ensembl
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1_1" checked>
        
          
          <label class="md-nav__link" for="__nav_3_1_1" id="__nav_3_1_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    genes
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_1_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_1_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    genes
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_1" id="__nav_3_1_1_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    automation
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    automation
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../automation/pre_release_ftp/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    pre_release_ftp
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_2" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_2" id="__nav_3_1_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    content
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    content
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../content/main_static_content/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    main_static_content
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../content/main_static_murine_content/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    main_static_murine_content
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_3" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_3" id="__nav_3_1_1_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    info_from_registry
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    info_from_registry
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/assign_species_prefix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    assign_species_prefix
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/assign_stable_space/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    assign_stable_space
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/build_anno_commands/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    build_anno_commands
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/check_if_annotated/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    check_if_annotated
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/check_stable_space_old_registry/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    check_stable_space_old_registry
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/create_config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    create_config
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/create_pipe_reg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    create_pipe_reg
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/genebuild_start_pipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    genebuild_start_pipeline
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/mysql_helper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    mysql_helper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/registry_helper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    registry_helper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/seed_nonvert/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    seed_nonvert
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/start_pipeline_from_registry/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    start_pipeline_from_registry
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/taxonomy_helper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    taxonomy_helper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/update_assembly_registry/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    update_assembly_registry
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/write_metrics_to_registry/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    write_metrics_to_registry
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_4" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_4" id="__nav_3_1_1_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    metadata
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    metadata
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metadata/beta_patcher/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    beta_patcher
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metadata/core_meta_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    core_meta_data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_5" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_5" id="__nav_3_1_1_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    metrics
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    metrics
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metrics/busco_lineage_selector/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    busco_lineage_selector
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metrics/busco_metakeys_patch/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    busco_metakeys_patch
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metrics/check_busco_score/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    check_busco_score
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_6" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_6" id="__nav_3_1_1_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    projects
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    projects
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../projects/write_yaml/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    write_yaml
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_7" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_7" id="__nav_3_1_1_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    stable_id
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_7">
            <span class="md-nav__icon md-icon"></span>
            
  
    stable_id
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stable_id/pid_overlap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    pid_overlap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_8" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_8" id="__nav_3_1_1_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    tracking
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_8">
            <span class="md-nav__icon md-icon"></span>
            
  
    tracking
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tracking/bioproject_tracking/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    bioproject_tracking
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tracking/live_tracking/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    live_tracking
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1_1_9" checked>
        
          
          <label class="md-nav__link" for="__nav_3_1_1_9" id="__nav_3_1_1_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    transcriptomic_data
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_9_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_1_1_9">
            <span class="md-nav__icon md-icon"></span>
            
  
    transcriptomic_data
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../check_for_transcriptomic/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    check_for_transcriptomic
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../get_transcriptomic_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    get_transcriptomic_data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    select_transcriptomic_data
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    select_transcriptomic_data
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;select_transcriptomic_data
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" select_transcriptomic_data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.check_fastqc_star_quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_fastqc_star_quality
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.clean_repeated_words" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;clean_repeated_words
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.connect_to_db" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;connect_to_db
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;create_report
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.fastqc_quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;fastqc_quality
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.filter_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;filter_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.prioritise_tissues" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;prioritise_tissues
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.star_quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;star_quality
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;select_transcriptomic_data
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" select_transcriptomic_data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.check_fastqc_star_quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_fastqc_star_quality
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.clean_repeated_words" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;clean_repeated_words
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.connect_to_db" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;connect_to_db
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;create_report
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.fastqc_quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;fastqc_quality
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.filter_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;filter_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.prioritise_tissues" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;prioritise_tissues
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.star_quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;star_quality
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="ensemblgenestranscriptomic_dataselect_transcriptomic_data"><code>ensembl.genes.transcriptomic_data.select_transcriptomic_data</code><a class="headerlink" href="#ensemblgenestranscriptomic_dataselect_transcriptomic_data" title="Permanent link">&para;</a></h1>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.transcriptomic_data.select_transcriptomic_data</code>


<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Select assessed data from the database and create a report.
This script connects to a MySQL database, retrieves data related to sequencing runs,
applies quality checks using FastQC and STAR, and generates a report summarizing the results.</p>










<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.check_fastqc_star_quality" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">check_fastqc_star_quality</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.check_fastqc_star_quality" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Apply FastQC and STAR quality to each row
Args:
    df (pd.DataFrame): DataFrame containing the data to be processed.
Returns:
    pd.DataFrame: DataFrame with additional columns for FastQC and STAR quality.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">check_fastqc_star_quality</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Apply FastQC and STAR quality to each row</span>
<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): DataFrame containing the data to be processed.</span>
<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: DataFrame with additional columns for FastQC and STAR quality.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Apply FastQC and STAR quality to each row</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;fastqc_pass&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">fastqc_quality</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;star_quality&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">star_quality</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Assign final quality labels</span>

    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">)[</span><span class="s2">&quot;fastqc_pass&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;Failed&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;Passed&quot;</span>
    <span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">)[</span><span class="s2">&quot;star_quality&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;Failed&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;Passed&quot;</span>
    <span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span>
    <span class="p">)</span>
    <span class="c1"># print(df.head(10))</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.clean_repeated_words" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">clean_repeated_words</span><span class="p">(</span><span class="n">text</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.clean_repeated_words" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Remove consecutive duplicate words from a string."</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>text</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The input string from which to remove</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A string with consecutive duplicate words removed,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>preserving the first occurrence of each word.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">clean_repeated_words</span><span class="p">(</span><span class="n">text</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove consecutive duplicate words from a string.&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        text (str): The input string from which to remove</span>
<span class="sd">        consecutive duplicate words.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: A string with consecutive duplicate words removed,</span>
<span class="sd">        preserving the first occurrence of each word.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">words</span> <span class="o">=</span> <span class="n">text</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">words</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>
    <span class="c1"># Keep only the first occurrence, skip consecutive duplicates</span>
    <span class="n">cleaned_words</span> <span class="o">=</span> <span class="p">[</span><span class="n">words</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="k">for</span> <span class="n">w</span> <span class="ow">in</span> <span class="n">words</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="k">if</span> <span class="n">w</span> <span class="o">!=</span> <span class="n">cleaned_words</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">cleaned_words</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">w</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cleaned_words</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.connect_to_db" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">connect_to_db</span><span class="p">(</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="mi">3306</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.connect_to_db" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Establishes a connection to the MySQL database.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>host</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The hostname or IP address of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>user</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The username to connect to the database.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>password</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The password for the user.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>db</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the database to connect to.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>port</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The port of the MySQL server (default is 3306).</p>
              </div>
            </td>
            <td>
                  <code>3306</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    A pymysql connection object.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">connect_to_db</span><span class="p">(</span>
    <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">password</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="mi">3306</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Establishes a connection to the MySQL database.</span>

<span class="sd">    Args:</span>
<span class="sd">        host: The hostname or IP address of the MySQL server.</span>
<span class="sd">        user: The username to connect to the database.</span>
<span class="sd">        password: The password for the user.</span>
<span class="sd">        db: The name of the database to connect to.</span>
<span class="sd">        port: The port of the MySQL server (default is 3306).</span>
<span class="sd">    Returns:</span>
<span class="sd">        A pymysql connection object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">connection</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="o">=</span><span class="n">password</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="n">port</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">connection</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">MySQLError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error connecting to MySQL: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.create_report" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">tissue_report_file</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.create_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create a report based on the DataFrame.
This function summarizes the FastQC and STAR quality results for each taxon_id and run_accession.
It calculates the number of runs that passed both quality checks and the percentage of runs that passed.
The report is printed to the console.
1. Group the DataFrame by taxon_id and run_accession.
2. Aggregate the final_fastqc_status and final_star_status columns.
3. Create a new column passed_both to indicate if both quality checks passed.
4. Group the DataFrame by taxon_id and aggregate the results.
5. Calculate the percentage of runs that passed both quality checks.
6. Print the report to the console.
7. Return the summarized DataFrame.
8. The report includes the following columns:
    - taxon_id: The taxon ID.
    - passed_fastqc: The number of runs that passed FastQC.
    - passed_star: The number of runs that passed STAR.
    - passed_both: The number of runs that passed both quality checks.
    - total_runs: The total number of runs for the taxon_id.
    - percentage_passed_both: The percentage of runs that passed both quality checks.
Args:
    df (pd.DataFrame): DataFrame containing the data to be processed.
Returns:
    pd.DataFrame: DataFrame with the summarized report.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_report</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">tissue_report_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a report based on the DataFrame.</span>
<span class="sd">    This function summarizes the FastQC and STAR quality results for each taxon_id and run_accession.</span>
<span class="sd">    It calculates the number of runs that passed both quality checks and the percentage of runs that passed.</span>
<span class="sd">    The report is printed to the console.</span>
<span class="sd">    1. Group the DataFrame by taxon_id and run_accession.</span>
<span class="sd">    2. Aggregate the final_fastqc_status and final_star_status columns.</span>
<span class="sd">    3. Create a new column passed_both to indicate if both quality checks passed.</span>
<span class="sd">    4. Group the DataFrame by taxon_id and aggregate the results.</span>
<span class="sd">    5. Calculate the percentage of runs that passed both quality checks.</span>
<span class="sd">    6. Print the report to the console.</span>
<span class="sd">    7. Return the summarized DataFrame.</span>
<span class="sd">    8. The report includes the following columns:</span>
<span class="sd">        - taxon_id: The taxon ID.</span>
<span class="sd">        - passed_fastqc: The number of runs that passed FastQC.</span>
<span class="sd">        - passed_star: The number of runs that passed STAR.</span>
<span class="sd">        - passed_both: The number of runs that passed both quality checks.</span>
<span class="sd">        - total_runs: The total number of runs for the taxon_id.</span>
<span class="sd">        - percentage_passed_both: The percentage of runs that passed both quality checks.</span>
<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): DataFrame containing the data to be processed.</span>
<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: DataFrame with the summarized report.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Collapse to one row per run_accession</span>
    <span class="n">run_summary</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span> <span class="s2">&quot;run_accession&quot;</span><span class="p">])</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
            <span class="n">final_fastqc_status</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="s2">&quot;first&quot;</span><span class="p">),</span>
            <span class="n">final_star_status</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">,</span> <span class="s2">&quot;first&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">run_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">run_summary</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span>
        <span class="n">run_summary</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span>
    <span class="p">)</span>
    <span class="c1"># REPORT</span>
    <span class="n">taxon_pass_summary</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">run_summary</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
            <span class="n">passed_fastqc</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
            <span class="n">passed_star</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
            <span class="n">passed_both</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;passed_both&quot;</span><span class="p">,</span> <span class="s2">&quot;sum&quot;</span><span class="p">),</span>
            <span class="n">total_runs</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">,</span> <span class="s2">&quot;nunique&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="c1"># Compute percentage outside of .agg()</span>
    <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;percentage_passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;total_runs&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_fastqc_only&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_fastqc&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_star_only&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_star&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="c1"># print(taxon_pass_summary.head())</span>
    <span class="c1"># print(&quot;\n=== Run Quality Summary by Taxon ===&quot;)</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">taxon_pass_summary</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Taxon ID: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">])</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;Total Runs: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;total_runs&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;Passed BOTH: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_both&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="c1"># f&quot;FastQC Passed: {row[&#39;passed_fastqc&#39;]} | &quot;</span>
            <span class="c1"># f&quot;STAR Passed: {row[&#39;passed_star&#39;]} | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;🧪 FastQC Only: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_fastqc_only&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;🚀 STAR Only: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_star_only&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;✔️ Both %: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;percentage_passed_both&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">%&quot;</span>
        <span class="p">)</span>
        <span class="c1"># Prepare tissue-level summary</span>
    <span class="k">if</span> <span class="s2">&quot;tissue_prediction&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">run_tissue_summary</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span> <span class="s2">&quot;tissue_prediction&quot;</span><span class="p">,</span> <span class="s2">&quot;run_accession&quot;</span><span class="p">])</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
                <span class="n">final_fastqc_status</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="s2">&quot;first&quot;</span><span class="p">),</span>
                <span class="n">final_star_status</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">,</span> <span class="s2">&quot;first&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="n">run_tissue_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">run_tissue_summary</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span>
        <span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">run_tissue_summary</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span>

        <span class="n">tissue_pass_summary</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">run_tissue_summary</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span> <span class="s2">&quot;tissue_prediction&quot;</span><span class="p">])</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
                <span class="n">passed_fastqc</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
                <span class="n">passed_star</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
                <span class="n">passed_both</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;passed_both&quot;</span><span class="p">,</span> <span class="s2">&quot;sum&quot;</span><span class="p">),</span>
                <span class="n">total_runs</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">,</span> <span class="s2">&quot;nunique&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;percentage_passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;total_runs&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>

        <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_fastqc_only&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_fastqc&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_star_only&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_star&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="c1"># Save to text file</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">tissue_report_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">tissue_pass_summary</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Taxon ID: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">])</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;Tissue: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;tissue_prediction&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;Total Runs: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;total_runs&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;Passed BOTH: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_both&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="c1"># f&quot;FastQC Passed: {row[&#39;passed_fastqc&#39;]} | &quot;</span>
                    <span class="c1"># f&quot;STAR Passed: {row[&#39;passed_star&#39;]} | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;FastQC Only: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_fastqc_only&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;STAR Only: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_star_only&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;BOTH %: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;percentage_passed_both&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">%</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.fastqc_quality" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">fastqc_quality</span><span class="p">(</span><span class="n">row</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.fastqc_quality" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate FastQC quality based on the criteria.
The function checks the FastQC quality criteria for each row and returns True if the criteria are met.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>row</code>
            </td>
            <td>
                  <code><span title="_type_">_type_</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe row</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the criteria are met, False otherwise.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">fastqc_quality</span><span class="p">(</span><span class="n">row</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate FastQC quality based on the criteria.</span>
<span class="sd">    The function checks the FastQC quality criteria for each row and returns True if the criteria are met.</span>


<span class="sd">    Args:</span>
<span class="sd">        row (_type_): dataframe row</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the criteria are met, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fastqc_criteria</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;per_base_sequence_quality&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">,</span> <span class="s2">&quot;WARN&quot;</span><span class="p">],</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;overrepresented_sequences&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">,</span> <span class="s2">&quot;WARN&quot;</span><span class="p">],</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;per_base_n_content&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">,</span> <span class="s2">&quot;WARN&quot;</span><span class="p">],</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;per_sequence_quality_scores&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">,</span> <span class="s2">&quot;WARN&quot;</span><span class="p">],</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">fastqc_criteria</span><span class="p">)</span> <span class="o">==</span> <span class="mi">4</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.filter_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">filter_data</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.filter_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Filter the DataFrame to select tissue-annotated samples based on quality criteria.</p>
<p>Steps:
1. Remove duplicate run_accessions.
2. Filter out rows with null tissue predictions.
3. Exclude rows with tumor/cancer-related tissue labels.
4. Prioritize specific tissue types (e.g., brain, heart, lung).
5. Group by tissue and select samples up to 250M total reads.
6. If fewer than 500 samples selected, add more high-quality unannotated samples.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input DataFrame with tissue predictions and QC info.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>debug</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, write intermediate CSVs for debugging.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[str]: List of selected run_accessions.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">filter_data</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter the DataFrame to select tissue-annotated samples based on quality criteria.</span>

<span class="sd">    Steps:</span>
<span class="sd">    1. Remove duplicate run_accessions.</span>
<span class="sd">    2. Filter out rows with null tissue predictions.</span>
<span class="sd">    3. Exclude rows with tumor/cancer-related tissue labels.</span>
<span class="sd">    4. Prioritize specific tissue types (e.g., brain, heart, lung).</span>
<span class="sd">    5. Group by tissue and select samples up to 250M total reads.</span>
<span class="sd">    6. If fewer than 500 samples selected, add more high-quality unannotated samples.</span>

<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): Input DataFrame with tissue predictions and QC info.</span>
<span class="sd">        debug (bool): If True, write intermediate CSVs for debugging.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[str]: List of selected run_accessions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;run_accession&quot;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>
    <span class="n">df_tissue</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">notnull</span><span class="p">()]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df_tissue</span> <span class="o">=</span> <span class="n">df_tissue</span><span class="p">[</span>
        <span class="o">~</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span>
            <span class="sa">r</span><span class="s2">&quot;tumor|cancer|melanoma|cancer/disease&quot;</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
    <span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Prioritise tissue types</span>
    <span class="n">priority_tissues</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;heart&quot;</span><span class="p">,</span>
        <span class="s2">&quot;lung&quot;</span><span class="p">,</span>
        <span class="s2">&quot;brain&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ovary&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ovaries&quot;</span><span class="p">,</span>
        <span class="s2">&quot;testes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;testis&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gonad&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gonads&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">df_tissue</span> <span class="o">=</span> <span class="n">prioritise_tissues</span><span class="p">(</span><span class="n">df_tissue</span><span class="p">,</span> <span class="n">priority_tissues</span><span class="p">)</span>
    <span class="n">df_tissue</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;tissueall.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">df_tissue</span> <span class="o">=</span> <span class="n">df_tissue</span><span class="p">[</span>
        <span class="o">~</span><span class="p">(</span>
            <span class="p">(</span><span class="o">~</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Failed&quot;</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Failed&quot;</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">]</span>
    <span class="n">df_tissue</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;tissueless.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">df_no_tissue</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">run_accessions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Filter tissue-annotated samples</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">df_tissue</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">):</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">group</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
            <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;priority&quot;</span><span class="p">,</span> <span class="s2">&quot;passed_both&quot;</span><span class="p">,</span> <span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="s2">&quot;final_star_status&quot;</span><span class="p">],</span>
            <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;group.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">total_length</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">group</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">total_length</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;total_sequences&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">250</span> <span class="o">*</span> <span class="mi">10</span><span class="o">**</span><span class="mi">6</span><span class="p">:</span>
                <span class="n">run_accessions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">])</span>
                <span class="n">total_length</span> <span class="o">+=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;total_sequences&quot;</span><span class="p">]</span>

    <span class="n">df_tissue</span><span class="p">[</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">run_accessions</span><span class="p">)]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="s2">&quot;tissuedanio.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">run_accessions</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">500</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">df_no_tissue</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">):</span>
            <span class="n">total_length</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">group</span><span class="p">[</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                    <span class="n">run_accessions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">run_accessions</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Module's entry-point.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Module&#39;s entry-point.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">prog</span><span class="o">=</span><span class="s2">&quot;llm_prediction.py&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Predict tissue using LLMs&quot;</span>
    <span class="p">)</span>

    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--taxon_id&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;10116&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxonomy ID&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--host&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;mysql-ens-genebuild-prod-1&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Host&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--user&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;ensadmin&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;User&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--password&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;ensembl&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Password&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--database&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;gb_transcriptomic_registry&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Database&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--port&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">4527</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Port&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--file_name&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output file name&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--read_type&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;short&quot;</span><span class="p">,</span> <span class="s2">&quot;long&quot;</span><span class="p">],</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;short&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Read type short/long&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--csv_for_main&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if true will produce the csv we need for main pipeline otherwise </span><span class="se">\</span>
<span class="s2">            the one for the full alignment pipeline&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">engine</span> <span class="o">=</span> <span class="n">create_engine</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;mysql+pymysql://</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">user</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">password</span><span class="si">}</span><span class="s2">@</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">host</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">port</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">database</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Load the rows to fix (e.g., 150k rows)</span>
    <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">read_type</span> <span class="o">==</span> <span class="s2">&quot;short&quot;</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;SELECT r.taxon_id, r.run_accession, r.sample_accession, r.platform, r.paired, </span><span class="se">\</span>
<span class="s2">            r.tissue_prediction, d.file_name, d.file_url,d.md5,d.per_base_sequence_quality, </span><span class="se">\</span>
<span class="s2">            d.per_sequence_quality_scores, d.per_base_n_content, d.overrepresented_sequences,</span><span class="se">\</span>
<span class="s2">            d.total_sequences, a.uniquely_mapped_reads_percentage, </span><span class="se">\</span>
<span class="s2">            a.percentage_reads_mapped_to_multiple_loci,a.percentage_reads_unmapped_too_short, </span><span class="se">\</span>
<span class="s2">            a.assembly_accession </span><span class="se">\</span>
<span class="s2">            FROM run r INNER JOIN data_files d ON r.run_id=d.run_id INNER JOIN align a ON r.run_id=a.run_id </span><span class="se">\</span>
<span class="s2">            WHERE  r.qc_status=&#39;ALIGNED&#39; AND r.paired=1 AND r.platform=&#39;ILLUMINA&#39; AND r.taxon_id = &quot;</span>
            <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">read_type</span> <span class="o">==</span> <span class="s2">&quot;long&quot;</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;SELECT r.taxon_id, r.run_accession, r.sample_accession, r.platform, r.paired, </span><span class="se">\</span>
<span class="s2">            r.tissue_prediction, d.file_name, d.file_url,d.md5,d.per_base_sequence_quality, </span><span class="se">\</span>
<span class="s2">            d.per_sequence_quality_scores, d.per_base_n_content, d.overrepresented_sequences, </span><span class="se">\</span>
<span class="s2">            d.total_sequences, a.uniquely_mapped_reads_percentage, </span><span class="se">\</span>
<span class="s2">            a.percentage_reads_mapped_to_multiple_loci,a.percentage_reads_unmapped_too_short </span><span class="se">\</span>
<span class="s2">            FROM run r INNER JOIN data_files d ON r.run_id=d.run_id INNER JOIN align a ON r.run_id=a.run_id </span><span class="se">\</span>
<span class="s2">            WHERE  r.qc_status=&#39;ALIGNED&#39; AND r.paired=0 AND r.platform=&#39;PACBIO_SMRT&#39; AND r.taxon_id = &quot;</span>
            <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span>
        <span class="p">)</span>
    <span class="c1"># Connect to the database</span>
    <span class="c1"># db_connection = connect_to_db(**db_config)</span>
    <span class="c1"># Clean the input text</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># if db_connection:</span>
        <span class="c1"># df = pd.read_sql(query, db_connection)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_sql</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">engine</span><span class="p">)</span>
        <span class="c1"># print(f&quot;Loaded {len(df)} rows.&quot;)</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">MySQLError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error connecting to MySQL: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">check_fastqc_star_quality</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="c1"># df[&quot;tissue_prediction&quot;] = df[&quot;tissue_prediction&quot;].apply(</span>
        <span class="c1">#    lambda x: re.search(r&quot;^(.*)&quot;, str(x)).group(1) if re.search(r&quot;^(.*)&quot;, str(x)) else x</span>
        <span class="c1"># )</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span>
        <span class="p">)</span>
        <span class="c1"># df[&quot;tissue_prediction&quot;] = df[&quot;tissue_prediction&quot;].apply(</span>
        <span class="c1">#    lambda x: (m.group(1) if (m := re.search(r&quot;^(.*)&quot;, str(x))) else x)</span>
        <span class="c1"># )</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span>
            <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Answer: &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Output: &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
                <span class="sa">r</span><span class="s2">&quot;[, ]+&quot;</span><span class="p">,</span>
                <span class="s2">&quot;_&quot;</span><span class="p">,</span>  <span class="c1"># replace space/comma with underscore</span>
                <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
                    <span class="sa">r</span><span class="s2">&quot;\b\d+\b|\bday\b|\b[a-zA-Z]\b&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;&quot;</span><span class="p">,</span>  <span class="c1"># remove numbers, &quot;day&quot;, and single letters</span>
                    <span class="n">clean_repeated_words</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)),</span>
                <span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span>
                <span class="s2">&quot;_&quot;</span>
            <span class="p">)</span>  <span class="c1"># remove leading/trailing underscores</span>
        <span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>

        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span>
            <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;NONE&quot;</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
            <span class="s2">&quot;tissue_prediction&quot;</span><span class="p">,</span>
        <span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span>
            <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
            <span class="s2">&quot;tissue_prediction&quot;</span><span class="p">,</span>
        <span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">create_report</span><span class="p">(</span>
            <span class="n">df</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">file_name</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">_tissue_summary.txt&quot;</span>
        <span class="p">)</span>
        <span class="c1"># print(df.head())</span>
        <span class="c1"># Remove duplicates based on &#39;run_accession&#39; while keeping the first row for each</span>
        <span class="n">df_original</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">run_accessions</span> <span class="o">=</span> <span class="n">filter_data</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="c1"># Build a regex pattern from run_accession list</span>
        <span class="n">selected_accessions</span> <span class="o">=</span> <span class="n">run_accessions</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">25000</span><span class="p">]</span>
        <span class="c1"># df_final = df_original[df_original[&quot;run_accession&quot;].isin(run_accessions[0:250])].copy()</span>
        <span class="n">df_final</span> <span class="o">=</span> <span class="n">df_original</span><span class="p">[</span>
            <span class="n">df_original</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">selected_accessions</span><span class="p">)</span>
        <span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">df_final</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
            <span class="n">df_final</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">],</span> <span class="n">categories</span><span class="o">=</span><span class="n">selected_accessions</span><span class="p">,</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="n">df_final</span> <span class="o">=</span> <span class="n">df_final</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
            <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">,</span> <span class="s2">&quot;file_name&quot;</span><span class="p">]</span>
        <span class="p">)</span>  <span class="c1"># &quot;run_accession&quot;)</span>

        <span class="n">df_final</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df_final</span><span class="p">[</span><span class="s2">&quot;predicted_tissue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_final</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="p">(</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;sample_accession&quot;</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">])</span>
                <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># print(df_final.head())</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">df_final</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">csv_for_main</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">df_final</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;file_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">df_final</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.fastq.gz&quot;</span>
                <span class="p">)</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;col1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;col_1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;col0&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;ENA&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;ENA&quot;</span>

                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">file_name</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">df_final</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                        <span class="n">f</span><span class="p">,</span>
                        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                        <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                        <span class="n">columns</span><span class="o">=</span><span class="p">[</span>
                            <span class="s2">&quot;predicted_tissue&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;run_accession&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;col1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;file_name&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;col_1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;col1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;col0&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;ENA&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;platform&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;sample_accession&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;file_url&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;md5&quot;</span><span class="p">,</span>
                        <span class="p">],</span>
                        <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                    <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># taxon_id,gca,platform,paired,tissue,run_accession,pair1,md5_1,pair2,md5_2</span>
                <span class="n">output_df</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">df_final</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                        <span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
                            <span class="p">{</span>
                                <span class="s2">&quot;taxon_id&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;assembly_accession&quot;</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                                    <span class="s2">&quot;assembly_accession&quot;</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="s2">&quot;NA&quot;</span><span class="p">])</span>
                                <span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span>
                                    <span class="mi">0</span>
                                <span class="p">],</span>  <span class="c1"># optional if gca exists</span>
                                <span class="s2">&quot;platform&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;platform&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;paired&quot;</span><span class="p">:</span> <span class="nb">bool</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;paired&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">])),</span>
                                <span class="c1"># &quot;paired&quot;: g[&quot;paired&quot;].iloc[0],</span>
                                <span class="s2">&quot;predicted_tissue&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;predicted_tissue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;pair1&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;_1&quot;</span><span class="p">)][</span>
                                    <span class="s2">&quot;file_url&quot;</span>
                                <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;md5_1&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;_1&quot;</span><span class="p">)][</span>
                                    <span class="s2">&quot;md5&quot;</span>
                                <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;pair2&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;_2&quot;</span><span class="p">)][</span>
                                    <span class="s2">&quot;file_url&quot;</span>
                                <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;md5_2&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;_2&quot;</span><span class="p">)][</span>
                                    <span class="s2">&quot;md5&quot;</span>
                                <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="p">}</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                    <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
                <span class="p">)</span>
                <span class="n">output_df</span> <span class="o">=</span> <span class="n">output_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tissue&quot;</span><span class="p">],</span> <span class="n">errors</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                    <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;predicted_tissue&quot;</span><span class="p">:</span> <span class="s2">&quot;tissue&quot;</span><span class="p">}</span>
                <span class="p">)</span>

                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">file_name</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">output_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                        <span class="n">f</span><span class="p">,</span>
                        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
                        <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                        <span class="n">columns</span><span class="o">=</span><span class="p">[</span>
                            <span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;assembly_accession&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;platform&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;paired&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;tissue&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;run_accession&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;pair1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;md5_1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;pair2&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;md5_2&quot;</span><span class="p">,</span>
                        <span class="p">],</span>
                        <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.prioritise_tissues" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">prioritise_tissues</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">priority_tissues</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.prioritise_tissues" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Assign a numeric priority to tissue predictions based on a predefined list.</p>
<p>This function adds a 'priority' column to the input DataFrame:
- Rows where the 'tissue_prediction' matches any of the <code>priority_tissues</code>
(case-insensitive, partial match) are assigned a priority of 1.
- All other rows receive a priority of 2.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing a 'tissue_prediction' column.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>priority_tissues</code>
            </td>
            <td>
                  <code><span title="List">List</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of tissue names to prioritize (e.g., ['brain', 'liver']).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: Modified DataFrame with an added 'priority' column.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">prioritise_tissues</span><span class="p">(</span>
    <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">priority_tissues</span><span class="p">:</span> <span class="n">typing</span><span class="o">.</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Assign a numeric priority to tissue predictions based on a predefined list.</span>

<span class="sd">    This function adds a &#39;priority&#39; column to the input DataFrame:</span>
<span class="sd">    - Rows where the &#39;tissue_prediction&#39; matches any of the `priority_tissues`</span>
<span class="sd">    (case-insensitive, partial match) are assigned a priority of 1.</span>
<span class="sd">    - All other rows receive a priority of 2.</span>

<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): DataFrame containing a &#39;tissue_prediction&#39; column.</span>
<span class="sd">        priority_tissues (List[str]): List of tissue names to prioritize (e.g., [&#39;brain&#39;, &#39;liver&#39;]).</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: Modified DataFrame with an added &#39;priority&#39; column.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">priority_regex</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">escape</span><span class="p">,</span> <span class="n">priority_tissues</span><span class="p">)),</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span>
    <span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;priority&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">1</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="n">priority_regex</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="mi">2</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.star_quality" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">star_quality</span><span class="p">(</span><span class="n">row</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.star_quality" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate STAR quality based on the criteria.
The function checks the STAR quality criteria for each row and returns True if the criteria are met.
Args:
    row (<em>type</em>): dataframe row
Returns:
    bool: True if the criteria are met, False otherwise.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">star_quality</span><span class="p">(</span><span class="n">row</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate STAR quality based on the criteria.</span>
<span class="sd">    The function checks the STAR quality criteria for each row and returns True if the criteria are met.</span>
<span class="sd">    Args:</span>
<span class="sd">        row (_type_): dataframe row</span>
<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the criteria are met, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;uniquely_mapped_reads_percentage&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">50</span>
        <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;percentage_reads_mapped_to_multiple_loci&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">30</span>
        <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;percentage_reads_unmapped_too_short&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">20</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-<script>document.write(new Date().getFullYear())</script>] <a href="https://www.ebi.ac.uk">EMBL-European Bioinformatics Institute</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../../..", "features": ["content.tooltips", "navigation.expand", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "search.highlight", "search.suggest", "toc.follow"], "search": "../../../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
    
  </body>
</html>