
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Ensembl Genes is a collection of Python tools to work with Ensembl gene annotation.">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genes/ensembl/genes/tracking/bioproject_tracking/">
      
      
        <link rel="prev" href="../../stable_id/pid_overlap/">
      
      
        <link rel="next" href="../live_tracking/">
      
      
        
      
      
      <link rel="icon" href="../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>bioproject_tracking - Ensembl Genes</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:300,300i,400,400i,700,700i%7CIBM+Plex+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Lato";--md-code-font:"IBM Plex Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="ensembl" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ensemblgenestrackingbioproject_tracking" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="Ensembl Genes" class="md-header__button md-logo" aria-label="Ensembl Genes" data-md-component="logo">
      
  <img src="../../../../img/ebang.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl Genes
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              bioproject_tracking
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genes" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub repository
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../.." class="md-tabs__link">
          
  
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../../code_of_conduct/" class="md-tabs__link">
          
  
  
  Development

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../automation/pre_release_ftp/" class="md-tabs__link">
          
  
  
  Code reference

        </a>
      </li>
    
  

    
  

    
  

    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="Ensembl Genes" class="md-nav__button md-logo" aria-label="Ensembl Genes" data-md-component="logo">
      
  <img src="../../../../img/ebang.png" alt="logo">

    </a>
    Ensembl Genes
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genes" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub repository
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Home
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../install/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Install
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Usage
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Development
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Development
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../code_of_conduct/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code of Conduct
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../coverage.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Coverage report
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    Code reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Code reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" checked>
        
          
          <label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    ensembl
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    ensembl
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1_1" checked>
        
          
          <label class="md-nav__link" for="__nav_3_1_1" id="__nav_3_1_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    genes
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_1_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_1_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    genes
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_1" id="__nav_3_1_1_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    automation
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    automation
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../automation/pre_release_ftp/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    pre_release_ftp
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../automation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    README
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_2" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_2" id="__nav_3_1_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    content
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    content
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../content/main_static_content/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    main_static_content
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../content/main_static_murine_content/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    main_static_murine_content
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_3" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_3" id="__nav_3_1_1_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    info_from_registry
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    info_from_registry
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/assign_species_prefix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    assign_species_prefix
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/assign_stable_space/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    assign_stable_space
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/build_anno_commands/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    build_anno_commands
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/check_if_annotated/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    check_if_annotated
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/check_stable_space_old_registry/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    check_stable_space_old_registry
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/create_config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    create_config
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/create_pipe_reg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    create_pipe_reg
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/genebuild_start_pipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    genebuild_start_pipeline
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/mysql_helper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    mysql_helper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/registry_helper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    registry_helper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/seed_nonvert/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    seed_nonvert
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/start_pipeline_from_registry/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    start_pipeline_from_registry
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/taxonomy_helper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    taxonomy_helper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/update_assembly_registry/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    update_assembly_registry
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/write_metrics_to_registry/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    write_metrics_to_registry
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../info_from_registry/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    README
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_4" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_4" id="__nav_3_1_1_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    metadata
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    metadata
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metadata/beta_patcher/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    beta_patcher
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metadata/core_meta_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    core_meta_data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metadata/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    README
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_5" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_5" id="__nav_3_1_1_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    metrics
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    metrics
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metrics/busco_lineage_selector/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    busco_lineage_selector
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metrics/busco_metakeys_patch/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    busco_metakeys_patch
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../metrics/check_busco_score/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    check_busco_score
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_6" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_6" id="__nav_3_1_1_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    projects
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    projects
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../projects/write_yaml/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    write_yaml
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../projects/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    README
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_7" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_7" id="__nav_3_1_1_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    stable_id
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_7">
            <span class="md-nav__icon md-icon"></span>
            
  
    stable_id
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stable_id/pid_overlap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    pid_overlap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1_1_8" checked>
        
          
          <label class="md-nav__link" for="__nav_3_1_1_8" id="__nav_3_1_1_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    tracking
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_8_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_1_1_8">
            <span class="md-nav__icon md-icon"></span>
            
  
    tracking
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    bioproject_tracking
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    bioproject_tracking
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;bioproject_tracking
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" bioproject_tracking">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.config" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;config
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.add_ftp" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_ftp
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_assembly_accessions" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_assembly_accessions
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_ensembl_live" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_ensembl_live
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_pre_release" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_pre_release
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_taxonomy_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_taxonomy_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.mysql_fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_fetch_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.write_report" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_report
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../live_tracking/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    live_tracking
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    README
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1_1_9" >
        
          
          <label class="md-nav__link" for="__nav_3_1_1_9" id="__nav_3_1_1_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    transcriptomic_data
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_1_1_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1_1_9">
            <span class="md-nav__icon md-icon"></span>
            
  
    transcriptomic_data
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../transcriptomic_data/check_for_transcriptomic/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    check_for_transcriptomic
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../transcriptomic_data/get_transcriptomic_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    get_transcriptomic_data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../transcriptomic_data/select_transcriptomic_data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    select_transcriptomic_data
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;bioproject_tracking
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" bioproject_tracking">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.config" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;config
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.add_ftp" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_ftp
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_assembly_accessions" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_assembly_accessions
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_ensembl_live" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_ensembl_live
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_pre_release" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_pre_release
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_taxonomy_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_taxonomy_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.mysql_fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_fetch_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.write_report" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_report
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="ensemblgenestrackingbioproject_tracking"><code>ensembl.genes.tracking.bioproject_tracking</code><a class="headerlink" href="#ensemblgenestrackingbioproject_tracking" title="Permanent link">&para;</a></h1>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.tracking.bioproject_tracking" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.tracking.bioproject_tracking</code>


<a href="#ensembl.genes.tracking.bioproject_tracking" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Script to track assemblies from NCBI BioProject/Taxon IDs and their presence in Ensembl.
Generates a report of assembly accessions, Ensembl database info, taxonomy classification,
and FTP links if available.</p>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.tracking.bioproject_tracking.config" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">config</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.tracking.bioproject_tracking.config" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.add_ftp" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">add_ftp</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">release_type</span><span class="o">=</span><span class="s1">&#39;live&#39;</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.add_ftp" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Adds FTP links to the annotations dictionary based on the release type.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>annotations</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary where each key is             an assembly accession and its value contains annotation details,             including matches with database names or genome UUIDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The type of release to determine FTP link construction.            Can be either "live" or "pre".</p>
              </div>
            </td>
            <td>
                  <code>&#39;live&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict[str, Dict[str, Any]]: The updated annotations dictionary with FTP links         added to each match and the top-level annotation where applicable.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">add_ftp</span><span class="p">(</span>  <span class="c1"># pylint:disable=too-many-locals, too-many-branches</span>
    <span class="n">annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span> <span class="n">release_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;live&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adds FTP links to the annotations dictionary based on the release type.</span>

<span class="sd">    Args:</span>
<span class="sd">        annotations (Dict[str, Dict[str, Any]]): A dictionary where each key is \</span>
<span class="sd">            an assembly accession and its value contains annotation details, \</span>
<span class="sd">            including matches with database names or genome UUIDs.</span>
<span class="sd">        release_type (str): The type of release to determine FTP link construction.\</span>
<span class="sd">            Can be either &quot;live&quot; or &quot;pre&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Dict[str, Any]]: The updated annotations dictionary with FTP links \</span>
<span class="sd">        added to each match and the top-level annotation where applicable.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;matches&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">release_type</span> <span class="o">==</span> <span class="s2">&quot;pre&quot;</span><span class="p">:</span>
            <span class="n">dbnames</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">):</span>
                <span class="n">dbnames</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">])</span>

            <span class="k">for</span> <span class="n">dbname</span> <span class="ow">in</span> <span class="n">dbnames</span><span class="p">:</span>
                <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                SELECT meta_value</span>
<span class="s2">                FROM meta</span>
<span class="s2">                WHERE meta_key = &#39;organism.scientific_name&#39;</span>
<span class="s2">                &quot;&quot;&quot;</span>
                <span class="n">result</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
                    <span class="n">query</span><span class="p">,</span>
                    <span class="p">(),</span>
                    <span class="n">server_group</span><span class="o">=</span><span class="s2">&quot;pre-release&quot;</span><span class="p">,</span>
                    <span class="n">server_name</span><span class="o">=</span><span class="s2">&quot;gb1&quot;</span><span class="p">,</span>
                    <span class="n">db_name</span><span class="o">=</span><span class="n">dbname</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span><span class="p">:</span>
                    <span class="k">continue</span>  <span class="c1"># skip if no scientific name found</span>

                <span class="n">scientific_name</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ftp_link</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://ftp.ebi.ac.uk/pub/databases/ensembl/pre-release/</span><span class="si">{</span><span class="n">scientific_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">&quot;</span>

                <span class="c1"># attach to matching entries</span>
                <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">matches</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">dbname</span><span class="p">:</span>
                        <span class="n">m</span><span class="p">[</span><span class="s2">&quot;ftp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ftp_link</span>
                <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">dbname</span><span class="p">:</span>
                    <span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;ftp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ftp_link</span>  <span class="c1"># top-level</span>

        <span class="k">elif</span> <span class="n">release_type</span> <span class="o">==</span> <span class="s2">&quot;live&quot;</span><span class="p">:</span>
            <span class="c1"># Build map guuid-&gt;match for easy updates</span>
            <span class="n">by_guuid</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">):</span> <span class="n">m</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">)}</span>
            <span class="n">guuids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">by_guuid</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
            <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">):</span>
                <span class="n">guuids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;guuid&quot;</span><span class="p">])</span>

            <span class="k">for</span> <span class="n">guuid</span> <span class="ow">in</span> <span class="n">guuids</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">guuid</span> <span class="ow">or</span> <span class="n">guuid</span> <span class="o">==</span> <span class="s2">&quot;unknown&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                SELECT genome.genebuild_date, organism.scientific_name, dataset_attribute.value</span>
<span class="s2">                FROM genome</span>
<span class="s2">                JOIN organism USING(organism_id)</span>
<span class="s2">                JOIN genome_dataset USING(genome_id) </span>
<span class="s2">                JOIN dataset_attribute USING(dataset_id)</span>
<span class="s2">                WHERE genome.genome_uuid = </span><span class="si">%s</span>
<span class="s2">                AND attribute_id=169</span>
<span class="s2">                &quot;&quot;&quot;</span>
                <span class="n">data_fetch</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="p">(</span><span class="n">guuid</span><span class="p">,))</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">data_fetch</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="n">date</span><span class="p">,</span> <span class="n">scientific_name</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="n">data_fetch</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">date</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">date</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
                <span class="n">scientific_name</span> <span class="o">=</span> <span class="n">scientific_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
                <span class="n">ftp_link</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://ftp.ebi.ac.uk/pub/ensemblorganisms/</span><span class="si">{</span><span class="n">scientific_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s2">/geneset/</span><span class="si">{</span><span class="n">date</span><span class="si">}</span><span class="s2">/&quot;</span>

                <span class="k">if</span> <span class="n">guuid</span> <span class="ow">in</span> <span class="n">by_guuid</span><span class="p">:</span>
                    <span class="n">by_guuid</span><span class="p">[</span><span class="n">guuid</span><span class="p">][</span><span class="s2">&quot;ftp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ftp_link</span>
                    <span class="n">by_guuid</span><span class="p">[</span><span class="n">guuid</span><span class="p">][</span><span class="s2">&quot;date&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">date</span>
                <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">guuid</span><span class="p">:</span>
                    <span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;ftp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ftp_link</span>
                    <span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;date&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">date</span>

    <span class="k">return</span> <span class="n">annotations</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.get_assembly_accessions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_assembly_accessions</span><span class="p">(</span><span class="n">query_id</span><span class="p">,</span> <span class="n">query_type</span><span class="p">,</span> <span class="n">only_haploid</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.get_assembly_accessions" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetches assembly accessions from NCBI Datasets API based on BioProject or Taxon ID.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>query_id</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The BioProject or Taxon ID to query.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>query_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The type of query, either "bioproject" or "taxon".</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>only_haploid</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, only include haploid assemblies. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="int">int</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict[str, Dict[str, int]]: A dictionary mapping assembly accessions to their taxon IDs.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_assembly_accessions</span><span class="p">(</span>  <span class="c1"># pylint:disable=too-many-branches, too-many-statements, too-many-locals</span>
    <span class="n">query_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">query_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">only_haploid</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches assembly accessions from NCBI Datasets API based on BioProject or Taxon ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        query_id (str): The BioProject or Taxon ID to query.</span>
<span class="sd">        query_type (str): The type of query, either &quot;bioproject&quot; or &quot;taxon&quot;.</span>
<span class="sd">        only_haploid (bool): If True, only include haploid assemblies. Defaults to False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Dict[str, int]]: A dictionary mapping assembly accessions to their taxon IDs.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_parse_reports</span><span class="p">(</span><span class="n">lines</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
        <span class="n">accs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">obj</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
            <span class="k">except</span> <span class="n">json</span><span class="o">.</span><span class="n">JSONDecodeError</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">report</span> <span class="o">=</span> <span class="n">obj</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;report&quot;</span><span class="p">,</span> <span class="n">obj</span><span class="p">)</span>
            <span class="n">assembly_info</span> <span class="o">=</span> <span class="n">report</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_info&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">{}</span>
            <span class="k">if</span> <span class="n">only_haploid</span> <span class="ow">and</span> <span class="n">assembly_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_type&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;haploid&quot;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">accession</span> <span class="o">=</span> <span class="n">report</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;accession&quot;</span><span class="p">)</span>
            <span class="n">taxon_id</span> <span class="o">=</span> <span class="p">(</span><span class="n">report</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;organism&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tax_id&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">accession</span><span class="p">:</span>
                <span class="n">accs</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">:</span> <span class="n">taxon_id</span><span class="p">}</span>
        <span class="k">return</span> <span class="n">accs</span>

    <span class="c1"># --- 1) Try CLI (gets ALL versions) ---</span>
    <span class="n">datasets_cli</span> <span class="o">=</span> <span class="n">shutil</span><span class="o">.</span><span class="n">which</span><span class="p">(</span><span class="s2">&quot;datasets&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">datasets_cli</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">query_type</span> <span class="o">==</span> <span class="s2">&quot;bioproject&quot;</span><span class="p">:</span>
                <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">datasets_cli</span><span class="p">,</span>
                    <span class="s2">&quot;summary&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;genome&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;accession&quot;</span><span class="p">,</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">query_id</span><span class="p">),</span>
                    <span class="s2">&quot;--assembly-version&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;all&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;--as-json-lines&quot;</span><span class="p">,</span>
                <span class="p">]</span>
            <span class="k">elif</span> <span class="n">query_type</span> <span class="o">==</span> <span class="s2">&quot;taxon&quot;</span><span class="p">:</span>
                <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">datasets_cli</span><span class="p">,</span>
                    <span class="s2">&quot;summary&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;genome&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;taxon&quot;</span><span class="p">,</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">query_id</span><span class="p">),</span>
                    <span class="s2">&quot;--assembly-version&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;all&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;--as-json-lines&quot;</span><span class="p">,</span>
                <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid query_type &#39;</span><span class="si">{</span><span class="n">query_type</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">{}</span>

            <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">parsed</span> <span class="o">=</span> <span class="n">_parse_reports</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">splitlines</span><span class="p">())</span>
            <span class="k">if</span> <span class="n">parsed</span><span class="p">:</span>  <span class="c1"># pylint:disable=no-else-return</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;[Datasets CLI] Retrieved </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">parsed</span><span class="p">)</span><span class="si">}</span><span class="s2"> assemblies (all versions).&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="n">parsed</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;[Datasets CLI] Parsed zero rows from stdout; falling back to API.&quot;</span>
                <span class="p">)</span>
        <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;[Datasets CLI] Exit </span><span class="si">{</span><span class="n">e</span><span class="o">.</span><span class="n">returncode</span><span class="si">}</span><span class="s2">. STDERR:</span><span class="se">\n</span><span class="si">{</span><span class="n">e</span><span class="o">.</span><span class="n">stderr</span><span class="si">}</span><span class="se">\n</span><span class="s2">Falling back to API.&quot;</span>
            <span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># pylint:disable=broad-exception-caught</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;[Datasets CLI] Unexpected error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">. Falling back to API.&quot;</span>
            <span class="p">)</span>

    <span class="c1"># --- 2) Fallback: API (latest only) ---</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">base_url</span> <span class="o">=</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;urls&quot;</span><span class="p">][</span><span class="s2">&quot;datasets&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">query_type</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>  <span class="c1"># pylint:disable=broad-exception-caught</span>
        <span class="n">base_url</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">base_url</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="s2">&quot;[Datasets API] Base URL not found in config; returning empty set.&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="p">{}</span>

    <span class="n">assembly_accessions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">page_size</span> <span class="o">=</span> <span class="mi">5000</span>
    <span class="n">next_page_token</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_url</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">query_id</span><span class="si">}</span><span class="s2">/dataset_report?page_size=</span><span class="si">{</span><span class="n">page_size</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="n">next_page_token</span><span class="p">:</span>
            <span class="n">url</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;&amp;page_token=</span><span class="si">{</span><span class="n">next_page_token</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>  <span class="c1"># pylint:disable=missing-timeout</span>
            <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">assembly</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;reports&quot;</span><span class="p">,</span> <span class="p">[]):</span>
                <span class="n">assembly_info</span> <span class="o">=</span> <span class="n">assembly</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_info&quot;</span><span class="p">,</span> <span class="p">{})</span> <span class="ow">or</span> <span class="p">{}</span>
                <span class="k">if</span> <span class="n">only_haploid</span> <span class="ow">and</span> <span class="n">assembly_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_type&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;haploid&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">accession</span> <span class="o">=</span> <span class="n">assembly</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;accession&quot;</span><span class="p">)</span>
                <span class="n">taxon_id</span> <span class="o">=</span> <span class="p">(</span><span class="n">assembly</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;organism&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tax_id&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">accession</span><span class="p">:</span>
                    <span class="n">assembly_accessions</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">:</span> <span class="n">taxon_id</span><span class="p">}</span>
            <span class="n">next_page_token</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;next_page_token&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">next_page_token</span><span class="p">:</span>
                <span class="k">break</span>
        <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Datasets API] Error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">break</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;[Datasets API] Retrieved </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">assembly_accessions</span><span class="p">)</span><span class="si">}</span><span class="s2"> assemblies (latest only).&quot;</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">assembly_accessions</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.get_ensembl_live" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_ensembl_live</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.get_ensembl_live" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetches Ensembl live database annotations for a list of assembly accessions.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>accessions_taxon</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="int">int</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary mapping assembly accessions
to basic taxon information, such as {"GCA_000001405.39": {"taxon_id": 9606}}.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Tuple[Dict[str, Dict[str, Any]], List[str]]: A tuple containing:</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>1) A dictionary where each key is an assembly accession and its value is another dictionary with:
- taxon_id (from input)
- matches: List[{"guuid","dbname"}]
- guuid, dbname: best match (latest core; prefer non-cm)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Tuple">Tuple</span>[<span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]], <span title="typing.List">List</span>[<span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>2) A list of accessions that were not found in the Ensembl live database.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_ensembl_live</span><span class="p">(</span>
    <span class="n">accessions_taxon</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches Ensembl live database annotations for a list of assembly accessions.</span>

<span class="sd">    Args:</span>
<span class="sd">        accessions_taxon (Dict[str, Dict[str, int]]): A dictionary mapping assembly accessions</span>
<span class="sd">            to basic taxon information, such as {&quot;GCA_000001405.39&quot;: {&quot;taxon_id&quot;: 9606}}.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[Dict[str, Dict[str, Any]], List[str]]: A tuple containing:</span>
<span class="sd">        1) A dictionary where each key is an assembly accession and its value is another dictionary with:</span>
<span class="sd">           - taxon_id (from input)</span>
<span class="sd">           - matches: List[{&quot;guuid&quot;,&quot;dbname&quot;}]</span>
<span class="sd">           - guuid, dbname: best match (latest core; prefer non-cm)</span>
<span class="sd">        2) A list of accessions that were not found in the Ensembl live database.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">accessions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">accessions</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">{},</span> <span class="p">[]</span>
    <span class="n">placeholders</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">accessions</span><span class="p">))</span>
    <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    SELECT DISTINCT</span>
<span class="s2">        assembly.accession,</span>
<span class="s2">        genome.genome_uuid,</span>
<span class="s2">        dataset_source.name AS database_name</span>
<span class="s2">    FROM genome</span>
<span class="s2">    JOIN assembly ON genome.assembly_id = assembly.assembly_id</span>
<span class="s2">    JOIN genome_dataset ON genome.genome_id = genome_dataset.genome_id</span>
<span class="s2">    JOIN dataset ON genome_dataset.dataset_id = dataset.dataset_id</span>
<span class="s2">    JOIN dataset_source ON dataset.dataset_source_id = dataset_source.dataset_source_id</span>
<span class="s2">    JOIN genome_release ON genome.genome_id = genome_release.genome_id</span>
<span class="s2">    JOIN ensembl_release ON genome_release.release_id = ensembl_release.release_id</span>
<span class="s2">    WHERE dataset.name = &#39;genebuild&#39;</span>
<span class="s2">      AND assembly.accession IN (</span><span class="si">{</span><span class="n">placeholders</span><span class="si">}</span><span class="s2">)</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="n">rows</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">accessions</span><span class="p">))</span>

    <span class="c1"># Seed with taxon info so missing ones can still be tracked</span>
    <span class="n">live_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">guuid</span><span class="p">,</span> <span class="n">dbname</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
        <span class="n">entry</span> <span class="o">=</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span>
            <span class="n">accession</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">accession</span><span class="p">,</span> <span class="p">{}))</span>
        <span class="p">)</span>
        <span class="n">entry</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;matches&quot;</span><span class="p">,</span> <span class="p">[])</span>
        <span class="c1"># Avoid duplicates across joins if any</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">any</span><span class="p">(</span>
            <span class="n">m</span><span class="p">[</span><span class="s2">&quot;guuid&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">guuid</span> <span class="ow">and</span> <span class="n">m</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">dbname</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">entry</span><span class="p">[</span><span class="s2">&quot;matches&quot;</span><span class="p">]</span>
        <span class="p">):</span>
            <span class="n">entry</span><span class="p">[</span><span class="s2">&quot;matches&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;guuid&quot;</span><span class="p">:</span> <span class="n">guuid</span><span class="p">,</span> <span class="s2">&quot;dbname&quot;</span><span class="p">:</span> <span class="n">dbname</span><span class="p">})</span>

    <span class="c1"># Choose best match for backward compatibility fields</span>
    <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="n">entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;matches&quot;</span><span class="p">,</span> <span class="p">[])</span>
        <span class="k">if</span> <span class="n">matches</span><span class="p">:</span>
            <span class="n">best</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">matches</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">m</span><span class="p">:</span> <span class="n">_score_dbname</span><span class="p">(</span><span class="n">m</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">]))</span>
            <span class="n">entry</span><span class="p">[</span><span class="s2">&quot;guuid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">best</span><span class="p">[</span><span class="s2">&quot;guuid&quot;</span><span class="p">]</span>
            <span class="n">entry</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">best</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">]</span>

    <span class="c1"># Compute missing (accessions with no rows at all)</span>
    <span class="n">matched_accessions</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">live_annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="n">missing_annotations</span> <span class="o">=</span> <span class="p">[</span><span class="n">acc</span> <span class="k">for</span> <span class="n">acc</span> <span class="ow">in</span> <span class="n">accessions</span> <span class="k">if</span> <span class="n">acc</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">matched_accessions</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">live_annotations</span><span class="p">,</span> <span class="n">missing_annotations</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.get_pre_release" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_pre_release</span><span class="p">(</span><span class="n">missing_annotations</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.get_pre_release" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Checks for the existence of pre-release Ensembl database schemas for a         list of missing accessions.
This function takes a list of GenBank/RefSeq assembly accessions that were not found in the     Ensembl live database lookup (e.g., from <code>get_ensembl_live</code>). It reformats each accession     to match the expected pre-release Ensembl database schema naming convention (e.g.,    'GCA_000001405.39' → 'gca000001405v39') and queries the MySQL server's     <code>information_schema.schemata</code> table to check if a database with that name exists.
If such a schema is found, the function records the corresponding accession and     database name in the returned dictionary.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>missing_annotations</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of assembly accessions (e.g., 'GCA_000001405.39')             for which no live annotation database was found.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict[str, Dict[str, str]]: A dictionary mapping each accession with a found pre-release </p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>database to a nested dictionary containing:
- "dbname": The name of the pre-release schema (e.g., 'gca000001405v39').</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>Example</code></td>            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>{
    "GCA_000001405.39": {
        "dbname": "gca000001405v39"
    },
    ...
}</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Notes:
    - Only accessions matching the pattern 'GCA/GCF_XXXXXXXXX.XX' are processed.
    - If no matching schema is found for an accession, it is omitted from the result.
    - The function uses <code>mysql_fetch_data</code> to query the MySQL server's schema metadata.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_pre_release</span><span class="p">(</span><span class="n">missing_annotations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Checks for the existence of pre-release Ensembl database schemas for a \</span>
<span class="sd">        list of missing accessions.</span>
<span class="sd">    This function takes a list of GenBank/RefSeq assembly accessions that were not found in the \</span>
<span class="sd">    Ensembl live database lookup (e.g., from `get_ensembl_live`). It reformats each accession \</span>
<span class="sd">    to match the expected pre-release Ensembl database schema naming convention (e.g.,\</span>
<span class="sd">    &#39;GCA_000001405.39&#39; → &#39;gca000001405v39&#39;) and queries the MySQL server&#39;s \</span>
<span class="sd">    `information_schema.schemata` table to check if a database with that name exists.</span>
<span class="sd">    If such a schema is found, the function records the corresponding accession and \</span>
<span class="sd">    database name in the returned dictionary.</span>

<span class="sd">    Args:</span>
<span class="sd">        missing_annotations (List[str]): A list of assembly accessions (e.g., &#39;GCA_000001405.39&#39;) \</span>
<span class="sd">            for which no live annotation database was found.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Dict[str, str]]: A dictionary mapping each accession with a found pre-release </span>
<span class="sd">        database to a nested dictionary containing:</span>
<span class="sd">            - &quot;dbname&quot;: The name of the pre-release schema (e.g., &#39;gca000001405v39&#39;).</span>
<span class="sd">        Example:</span>
<span class="sd">            {</span>
<span class="sd">                &quot;GCA_000001405.39&quot;: {</span>
<span class="sd">                    &quot;dbname&quot;: &quot;gca000001405v39&quot;</span>
<span class="sd">                },</span>
<span class="sd">                ...</span>
<span class="sd">            }</span>
<span class="sd">    Notes:</span>
<span class="sd">        - Only accessions matching the pattern &#39;GCA/GCF_XXXXXXXXX.XX&#39; are processed.</span>
<span class="sd">        - If no matching schema is found for an accession, it is omitted from the result.</span>
<span class="sd">        - The function uses `mysql_fetch_data` to query the MySQL server&#39;s schema metadata.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pre_release_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">missing_annotations</span><span class="p">:</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(GCA|GCF)_(\d+)\.(\d+)&quot;</span><span class="p">,</span> <span class="n">accession</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">match</span><span class="p">:</span>
            <span class="k">continue</span>  <span class="c1"># skip malformed accession</span>

        <span class="n">prefix</span><span class="p">,</span> <span class="n">number</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
        <span class="n">lookup_string</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;%</span><span class="si">{</span><span class="n">prefix</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="si">}{</span><span class="n">number</span><span class="si">}</span><span class="s2">v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">%&quot;</span>

        <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT SCHEMA_NAME</span>
<span class="s2">        FROM information_schema.schemata</span>
<span class="s2">        WHERE SCHEMA_NAME like </span><span class="si">%s</span>
<span class="s2">        &quot;&quot;&quot;</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">query</span><span class="p">,</span> <span class="p">(</span><span class="n">lookup_string</span><span class="p">,),</span> <span class="n">server_group</span><span class="o">=</span><span class="s2">&quot;pre-release&quot;</span><span class="p">,</span> <span class="n">server_name</span><span class="o">=</span><span class="s2">&quot;gb1&quot;</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">result</span><span class="p">:</span>
            <span class="n">pre_release_annotations</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;guuid&quot;</span><span class="p">:</span> <span class="s2">&quot;unknown&quot;</span><span class="p">,</span>  <span class="c1"># typo fixed from guiid -&gt; guuid</span>
                <span class="s2">&quot;dbname&quot;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
            <span class="p">}</span>
    <span class="k">return</span> <span class="n">pre_release_annotations</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.get_taxonomy_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_taxonomy_info</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">,</span> <span class="n">accessions_taxon</span><span class="p">,</span> <span class="n">rank</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.get_taxonomy_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetches taxonomy information for a given rank from the NCBI Datasets API.
For each accession in <code>live_annotations</code>, this function retrieves its <code>taxon_id</code>    and queries the NCBI taxonomy endpoint to get the classification at the specified rank.
The retrieved rank name is then added to the corresponding annotation data.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>live_annotations</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary where each key is             an assembly accession and its value contains annotation details (including             at least "taxon_id").</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>accessions_taxon</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="int">int</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary mapping accessions             to basic taxon information, such as {"GCA_000001405.39": {"taxon_id": 9606}}.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rank</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The taxonomic rank to retrieve (e.g., "order", "class", "phylum").</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict[str, Dict[str, str]]: The updated <code>live_annotations</code> dictionary with an</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>additional key for the requested rank. For example:
{
    "GCA_000001405.39": {
        "taxon_id": 9606,
        "guuid": "some-uuid",
        "dbname": "ensembl_core",
        "order": "Primates"
    },
    ...
}</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the API query fails or the rank is not found, the value for that rank is not added</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(or set to None/unknown).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_taxonomy_info</span><span class="p">(</span>
    <span class="n">live_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span>
    <span class="n">accessions_taxon</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
    <span class="n">rank</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches taxonomy information for a given rank from the NCBI Datasets API.</span>
<span class="sd">    For each accession in `live_annotations`, this function retrieves its `taxon_id`\</span>
<span class="sd">    and queries the NCBI taxonomy endpoint to get the classification at the specified rank.</span>
<span class="sd">    The retrieved rank name is then added to the corresponding annotation data.</span>

<span class="sd">    Args:</span>
<span class="sd">        live_annotations (Dict[str, Dict[str, str]]): A dictionary where each key is \</span>
<span class="sd">            an assembly accession and its value contains annotation details (including \</span>
<span class="sd">            at least &quot;taxon_id&quot;).</span>
<span class="sd">        accessions_taxon (Dict[str, Dict[str, int]]): A dictionary mapping accessions \</span>
<span class="sd">            to basic taxon information, such as {&quot;GCA_000001405.39&quot;: {&quot;taxon_id&quot;: 9606}}.</span>
<span class="sd">        rank (str): The taxonomic rank to retrieve (e.g., &quot;order&quot;, &quot;class&quot;, &quot;phylum&quot;).</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Dict[str, str]]: The updated `live_annotations` dictionary with an</span>
<span class="sd">        additional key for the requested rank. For example:</span>
<span class="sd">            {</span>
<span class="sd">                &quot;GCA_000001405.39&quot;: {</span>
<span class="sd">                    &quot;taxon_id&quot;: 9606,</span>
<span class="sd">                    &quot;guuid&quot;: &quot;some-uuid&quot;,</span>
<span class="sd">                    &quot;dbname&quot;: &quot;ensembl_core&quot;,</span>
<span class="sd">                    &quot;order&quot;: &quot;Primates&quot;</span>
<span class="sd">                },</span>
<span class="sd">                ...</span>
<span class="sd">            }</span>
<span class="sd">        If the API query fails or the rank is not found, the value for that rank is not added</span>
<span class="sd">        (or set to None/unknown).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="p">:</span>
        <span class="n">taxon_id</span> <span class="o">=</span> <span class="n">accessions_taxon</span><span class="p">[</span><span class="n">accession</span><span class="p">][</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span>
        <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://api.ncbi.nlm.nih.gov/datasets/v2alpha/taxonomy/taxon/</span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">/dataset_report&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>  <span class="c1"># pylint:disable=missing-timeout</span>
            <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>

            <span class="n">taxonomies</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;reports&quot;</span><span class="p">,</span> <span class="p">[])</span>
            <span class="k">for</span> <span class="n">taxonomy</span> <span class="ow">in</span> <span class="n">taxonomies</span><span class="p">:</span>
                <span class="n">rank_name</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">taxonomy</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;taxonomy&quot;</span><span class="p">,</span> <span class="p">{})</span>
                    <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;classification&quot;</span><span class="p">,</span> <span class="p">{})</span>
                    <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rank</span><span class="p">,</span> <span class="p">{})</span>
                    <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="n">taxonomy_info</span> <span class="o">=</span> <span class="p">{</span><span class="n">rank</span><span class="p">:</span> <span class="n">rank_name</span><span class="p">}</span>
                <span class="n">live_annotations</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">taxonomy_info</span><span class="p">)</span>

        <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">HTTPError</span> <span class="k">as</span> <span class="n">http_err</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HTTP error occurred: </span><span class="si">{</span><span class="n">http_err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>  <span class="c1"># pylint:disable=broad-exception-caught</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;An error occurred: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">live_annotations</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main entry point of the script.
Parses command-line arguments to determine if a BioProject ID or Taxon ID is provided,
whether to filter only haploid assemblies, the desired output file path, the taxonomic
rank to retrieve, and whether to include FTP links. It then:
  1. Fetches assembly accessions from the NCBI API based on the provided ID.
  2. Fetches corresponding Ensembl live database information from the local MySQL database.
  3. Retrieves taxonomy information for the specified rank from the NCBI API.
  4. Optionally adds FTP links if requested.
  5. Writes all collected data into a tab-separated report file.
Usage Example:
    python script_name.py --bioproject_id PRJNA12345 --haploid --rank order --ftp
Returns:
    None</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main entry point of the script.</span>
<span class="sd">    Parses command-line arguments to determine if a BioProject ID or Taxon ID is provided,</span>
<span class="sd">    whether to filter only haploid assemblies, the desired output file path, the taxonomic</span>
<span class="sd">    rank to retrieve, and whether to include FTP links. It then:</span>
<span class="sd">      1. Fetches assembly accessions from the NCBI API based on the provided ID.</span>
<span class="sd">      2. Fetches corresponding Ensembl live database information from the local MySQL database.</span>
<span class="sd">      3. Retrieves taxonomy information for the specified rank from the NCBI API.</span>
<span class="sd">      4. Optionally adds FTP links if requested.</span>
<span class="sd">      5. Writes all collected data into a tab-separated report file.</span>
<span class="sd">    Usage Example:</span>
<span class="sd">        python script_name.py --bioproject_id PRJNA12345 --haploid --rank order --ftp</span>
<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Fetch assembly data and report annotations.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--bioproject_id&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;NCBI BioProject ID&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--taxon_id&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxonomy ID&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--haploid&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Fetch only haploid assemblies&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--report_file&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;./report_file.tsv&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--classification&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Provide breakdown of taxonomic classification&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--rank&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;order&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--ftp&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Include FTP links in the report&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--pre_release&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Include list of pre-release databases in the report&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="c1"># Determine which ID is provided and which query type is relevant</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="p">:</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;You must provide either --bioproject_id or --taxon_id&quot;</span><span class="p">)</span>

    <span class="n">accessions_taxon</span> <span class="o">=</span> <span class="n">get_assembly_accessions</span><span class="p">(</span>
        <span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="p">,</span>
        <span class="s2">&quot;bioproject&quot;</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span> <span class="k">else</span> <span class="s2">&quot;taxon&quot;</span><span class="p">,</span>
        <span class="n">args</span><span class="o">.</span><span class="n">haploid</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Fetch annotations (live)</span>
    <span class="n">live_annotations</span><span class="p">,</span> <span class="n">missing_annotations</span> <span class="o">=</span> <span class="n">get_ensembl_live</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="p">)</span>

    <span class="c1"># Taxonomy classification (top-level only; rows inherit from details)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">classification</span><span class="p">:</span>
        <span class="n">get_taxonomy_info</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">,</span> <span class="n">accessions_taxon</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">rank</span><span class="p">)</span>

    <span class="c1"># Optionally add FTP links</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">ftp</span><span class="p">:</span>
        <span class="n">live_annotations</span> <span class="o">=</span> <span class="n">add_ftp</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">,</span> <span class="s2">&quot;live&quot;</span><span class="p">)</span>

    <span class="c1"># Unique species count from live annotations</span>
    <span class="n">unique_taxon_ids</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">details</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">details</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="k">if</span> <span class="s2">&quot;taxon_id&quot;</span> <span class="ow">in</span> <span class="n">details</span>
    <span class="p">}</span>

    <span class="c1"># Optionally add Pre-release data</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">pre_release</span> <span class="ow">and</span> <span class="n">missing_annotations</span><span class="p">:</span>
        <span class="n">pre_release_annotations</span> <span class="o">=</span> <span class="n">get_pre_release</span><span class="p">(</span><span class="n">missing_annotations</span><span class="p">)</span>

        <span class="k">for</span> <span class="p">(</span>
            <span class="n">accession</span>
        <span class="p">)</span> <span class="ow">in</span> <span class="n">pre_release_annotations</span><span class="p">:</span>  <span class="c1"># pylint:disable=consider-using-dict-items</span>
            <span class="k">if</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">accessions_taxon</span><span class="p">:</span>
                <span class="n">pre_release_annotations</span><span class="p">[</span><span class="n">accession</span><span class="p">][</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">accessions_taxon</span><span class="p">[</span>
                    <span class="n">accession</span>
                <span class="p">][</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">classification</span> <span class="ow">and</span> <span class="n">pre_release_annotations</span><span class="p">:</span>
            <span class="n">get_taxonomy_info</span><span class="p">(</span><span class="n">pre_release_annotations</span><span class="p">,</span> <span class="n">accessions_taxon</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">rank</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">ftp</span> <span class="ow">and</span> <span class="n">pre_release_annotations</span><span class="p">:</span>
            <span class="n">pre_release_annotations</span> <span class="o">=</span> <span class="n">add_ftp</span><span class="p">(</span><span class="n">pre_release_annotations</span><span class="p">,</span> <span class="s2">&quot;pre&quot;</span><span class="p">)</span>

        <span class="n">all_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="o">**</span><span class="n">live_annotations</span><span class="p">,</span>
            <span class="o">**</span><span class="n">pre_release_annotations</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">all_annotations</span> <span class="o">=</span> <span class="n">live_annotations</span>

    <span class="c1"># Console summary</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="p">)</span><span class="si">}</span><span class="s2"> assemblies under BioProject ID </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="p">)</span><span class="si">}</span><span class="s2"> assemblies for taxon ID </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">)</span><span class="si">}</span><span class="s2"> annotations in beta.ensembl.org </span><span class="se">\</span>
<span class="s2">            for </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_taxon_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> unique species&quot;</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">classification</span><span class="p">:</span>
        <span class="n">rank_values</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">rank</span><span class="p">,</span> <span class="s2">&quot;unknown&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">details</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="p">]</span>
        <span class="n">rank_counts</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">rank_values</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Breakdown:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">rank_counts</span><span class="p">)</span>

    <span class="c1"># Write final report</span>
    <span class="n">write_report</span><span class="p">(</span>
        <span class="n">all_annotations</span><span class="p">,</span>
        <span class="n">args</span><span class="o">.</span><span class="n">report_file</span><span class="p">,</span>
        <span class="n">args</span><span class="o">.</span><span class="n">rank</span><span class="p">,</span>
        <span class="n">include_class</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">classification</span><span class="p">,</span>
        <span class="n">include_ftp</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">ftp</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.mysql_fetch_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="p">(),</span> <span class="n">server_group</span><span class="o">=</span><span class="s1">&#39;meta&#39;</span><span class="p">,</span> <span class="n">server_name</span><span class="o">=</span><span class="s1">&#39;beta&#39;</span><span class="p">,</span> <span class="n">db_name</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.mysql_fetch_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Executes a SQL query with optional parameters to fetch results from a specified MySQL server.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>query</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The SQL query to execute.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>params</code>
            </td>
            <td>
                  <code><span title="typing.Tuple">Tuple</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Parameters to pass to the SQL query. Defaults to ().</p>
              </div>
            </td>
            <td>
                  <code>()</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_group</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The server group as defined in the config. Defaults to "meta".</p>
              </div>
            </td>
            <td>
                  <code>&#39;meta&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The server name within the group as defined in the config. Defaults to "beta".</p>
              </div>
            </td>
            <td>
                  <code>&#39;beta&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>db_name</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The database name to connect to. If None, uses the default from config. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Sequence">Sequence</span>[<span title="typing.Tuple">Tuple</span>[<span title="typing.Any">Any</span>, ...]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Sequence[Tuple[Any, ...]]: The fetched rows from the query result.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="KeyError">KeyError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the server group or name is not found in the config.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pymysql.Error">Error</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If there is an error connecting to the database or executing the query.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">mysql_fetch_data</span><span class="p">(</span>
    <span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">params</span><span class="p">:</span> <span class="n">Tuple</span> <span class="o">=</span> <span class="p">(),</span>
    <span class="n">server_group</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;meta&quot;</span><span class="p">,</span>
    <span class="n">server_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;beta&quot;</span><span class="p">,</span>
    <span class="n">db_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Sequence</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="o">...</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Executes a SQL query with optional parameters to fetch results from a specified MySQL server.</span>

<span class="sd">    Args:</span>
<span class="sd">        query (str): The SQL query to execute.</span>
<span class="sd">        params (Tuple, optional): Parameters to pass to the SQL query. Defaults to ().</span>
<span class="sd">        server_group (str, optional): The server group as defined in the config. Defaults to &quot;meta&quot;.</span>
<span class="sd">        server_name (str, optional): The server name within the group as defined in the config. Defaults to &quot;beta&quot;.</span>
<span class="sd">        db_name (Optional[str], optional): The database name to connect to. If None, uses the default from config. Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Sequence[Tuple[Any, ...]]: The fetched rows from the query result.</span>

<span class="sd">    Raises:</span>
<span class="sd">        KeyError: If the server group or name is not found in the config.</span>
<span class="sd">        pymysql.Error: If there is an error connecting to the database or executing the query.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">server_config</span> <span class="o">=</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="n">server_group</span><span class="p">][</span><span class="n">server_name</span><span class="p">]</span>
        <span class="n">connection</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_config</span><span class="p">[</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_config</span><span class="p">[</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_config</span><span class="p">[</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">database</span><span class="o">=</span><span class="n">db_name</span> <span class="ow">or</span> <span class="n">server_config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;db_name&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span><span class="p">)</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
        <span class="n">connection</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">result</span>

    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">key_err</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid server group or name in config: </span><span class="si">{</span><span class="n">key_err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">sql_err</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MySQL Error: </span><span class="si">{</span><span class="n">sql_err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">[]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.write_report" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">write_report</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">,</span> <span class="n">report_file</span><span class="p">,</span> <span class="n">rank</span><span class="p">,</span> <span class="n">include_class</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">include_ftp</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.write_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Writes a tab-separated report. If an accession has multiple matches,
produces one line per match; otherwise one line.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>live_annotations</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary where each key is             an assembly accession and its value contains annotation details,             including matches with database names or genome UUIDs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>report_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the output report file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rank</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The taxonomic rank to include if <code>include_class</code> is True.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>include_class</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, includes the taxonomic classification at the specified rank.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>include_ftp</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, includes FTP links in the report.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">write_report</span><span class="p">(</span>
    <span class="n">live_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span>
    <span class="n">report_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">rank</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">include_class</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">include_ftp</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Writes a tab-separated report. If an accession has multiple matches,</span>
<span class="sd">    produces one line per match; otherwise one line.</span>

<span class="sd">    Args:</span>
<span class="sd">        live_annotations (Dict[str, Dict[str, Any]]): A dictionary where each key is \</span>
<span class="sd">            an assembly accession and its value contains annotation details, \</span>
<span class="sd">            including matches with database names or genome UUIDs.</span>
<span class="sd">        report_file (str): The path to the output report file.</span>
<span class="sd">        rank (str): The taxonomic rank to include if `include_class` is True.</span>
<span class="sd">        include_class (bool): If True, includes the taxonomic classification at the specified rank.</span>
<span class="sd">        include_ftp (bool): If True, includes FTP links in the report.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">report_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">report_file</span><span class="p">)</span>
    <span class="n">lines_written</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">report_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">details</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">matches</span> <span class="o">=</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;matches&quot;</span><span class="p">)</span>

            <span class="c1"># If we have multiple, write one row per match</span>
            <span class="n">rows</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">matches</span>
                <span class="k">if</span> <span class="n">matches</span>
                <span class="k">else</span> <span class="p">[</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;guuid&quot;</span><span class="p">:</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">),</span>
                        <span class="s2">&quot;dbname&quot;</span><span class="p">:</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">),</span>
                        <span class="s2">&quot;ftp&quot;</span><span class="p">:</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ftp&quot;</span><span class="p">),</span>
                        <span class="s2">&quot;date&quot;</span><span class="p">:</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;date&quot;</span><span class="p">),</span>
                    <span class="p">}</span>
                <span class="p">]</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">accession</span><span class="p">),</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;unknown&quot;</span><span class="p">),</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;unknown&quot;</span><span class="p">),</span>
                <span class="p">]</span>
                <span class="k">if</span> <span class="n">include_class</span><span class="p">:</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rank</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;unknown&quot;</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">include_ftp</span><span class="p">:</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;date&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;date&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ftp&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ftp&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;N/A&quot;</span><span class="p">))</span>

                <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">lines_written</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Report written to </span><span class="si">{</span><span class="n">report_path</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span><span class="si">}</span><span class="s2"> with </span><span class="si">{</span><span class="n">lines_written</span><span class="si">}</span><span class="s2"> lines.&quot;</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-<script>document.write(new Date().getFullYear())</script>] <a href="https://www.ebi.ac.uk">EMBL-European Bioinformatics Institute</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../../..", "features": ["content.tooltips", "navigation.expand", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "search.highlight", "search.suggest", "toc.follow"], "search": "../../../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>