
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://ensembl.github.io/ensembl-genes/api/">
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>API Reference - Ensembl Genes</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:300,300i,400,400i,700,700i%7CIBM+Plex+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Lato";--md-code-font:"IBM Plex Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="ensembl" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#api-reference" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Ensembl Genes" class="md-header__button md-logo" aria-label="Ensembl Genes" data-md-component="logo">
      
  <img src="../img/ebang.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl Genes
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              API Reference
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Ensembl/ensembl-genes" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub repository
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../index.md" class="md-tabs__link">
          
  
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../code_of_conduct.md" class="md-tabs__link">
          
  
  
  Development

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../reference" class="md-tabs__link">
        
  
  
    
  
  Code reference

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Ensembl Genes" class="md-nav__button md-logo" aria-label="Ensembl Genes" data-md-component="logo">
      
  <img src="../img/ebang.png" alt="logo">

    </a>
    Ensembl Genes
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Ensembl/ensembl-genes" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub repository
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Home
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../index.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../install.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Install
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Usage
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Development
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Development
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../code_of_conduct.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code of Conduct
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../coverage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Coverage report
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code reference
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#automation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Automation
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;pre_release_ftp
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" pre_release_ftp">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.calculate_md5" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;calculate_md5
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.create_ftp_directory_structure" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;create_ftp_directory_structure
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.extract_species_from_filename" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;extract_species_from_filename
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.find_2bit" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;find_2bit
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.find_annotation_files" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;find_annotation_files
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.find_reheadered_fasta" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;find_reheadered_fasta
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.format_species_name" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;format_species_name
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.generate_md5_checksums" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;generate_md5_checksums
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.generate_new_filename" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;generate_new_filename
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.is_compressed_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;is_compressed_file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.parse_gca_id" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;parse_gca_id
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.automation.pre_release_ftp.setup_logging" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;setup_logging
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#content" class="md-nav__link">
    <span class="md-ellipsis">
      
        Content
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;main_static_content
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" main_static_content">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.CORE_QUERY" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;CORE_QUERY
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.args" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;args
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.assembly_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;assembly_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.core_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;core_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.core_meta" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;core_meta
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.db" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;db
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.output_dir" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;output_dir
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.parser" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parser
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.server_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;server_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.url_name" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;url_name
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.get_assembly_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_assembly_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.mysql_fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_fetch_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_content.write_content" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_content
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;main_static_murine_content
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" main_static_murine_content">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.CORE_QUERY" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;CORE_QUERY
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.args" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;args
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.assembly_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;assembly_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.core_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;core_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.core_meta" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;core_meta
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.db" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;db
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.output_dir" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;output_dir
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.parser" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parser
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.server_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;server_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.strain" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;strain
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.url_name" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;url_name
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.get_assembly_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_assembly_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.mysql_fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_fetch_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.content.main_static_murine_content.write_content" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_content
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#info-from-registry" class="md-nav__link">
    <span class="md-ellipsis">
      
        Info From Registry
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_species_prefix" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;assign_species_prefix
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" assign_species_prefix">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_species_prefix.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_species_prefix.create_prefix" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;create_prefix
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_species_prefix.exiting_prefix" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;exiting_prefix
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_species_prefix.generate_random_prefix" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;generate_random_prefix
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_species_prefix.get_special_cases" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_special_cases
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_species_prefix.get_species_prefix" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_species_prefix
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_species_prefix.insert_prefix_into_db" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;insert_prefix_into_db
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_stable_space" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;assign_stable_space
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" assign_stable_space">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_stable_space.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_stable_space.assign_stable_id" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;assign_stable_id
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_stable_space.get_stable_space" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_stable_space
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_stable_space.insert_to_db" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;insert_to_db
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_stable_space.stable_space_per_taxon" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;stable_space_per_taxon
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.assign_stable_space.stable_space_range" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;stable_space_range
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.build_anno_commands" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;build_anno_commands
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" build_anno_commands">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.build_anno_commands.build_annotation_commands" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;build_annotation_commands
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.check_if_annotated" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;check_if_annotated
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" check_if_annotated">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.check_if_annotated.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_if_annotated
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" check_if_annotated">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated--raises" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated--returns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.check_stable_space_old_registry" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;check_stable_space_old_registry
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" check_stable_space_old_registry">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.check_stable_space_old_registry.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.check_stable_space_old_registry.get_old_stable_space_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_old_stable_space_info
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;create_config
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" create_config">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.copy_config" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;copy_config
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" copy_config">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.copy_config--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.copy_config--returns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.copy_config--raises" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.edit_config_anno" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;edit_config_anno
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" edit_config_anno">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.edit_config_anno--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.edit_config_anno--raises" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.edit_config_main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;edit_config_main
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" edit_config_main">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.edit_config_main--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_config.edit_config_main--raises" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_pipe_reg" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;create_pipe_reg
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" create_pipe_reg">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_pipe_reg--functions" class="md-nav__link">
    <span class="md-ellipsis">
      
        Functions:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_pipe_reg.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;create_registry_entry
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" create_registry_entry">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry--returns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry--raises" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_pipe_reg.update_registry_path_in_pipedb" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;update_registry_path_in_pipedb
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" update_registry_path_in_pipedb">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.create_pipe_reg.update_registry_path_in_pipedb--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;genebuild_start_pipeline
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" genebuild_start_pipeline">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.args" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;args
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.ehive_urls" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;ehive_urls
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.parser" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parser
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.init_pipeline_anno" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;init_pipeline_anno
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.init_pipeline_main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;init_pipeline_main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.mysql_helper" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;mysql_helper
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" mysql_helper">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.mysql_helper.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.mysql_helper.mysql_fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_fetch_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.mysql_helper.mysql_get_connection" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_get_connection
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.mysql_helper.mysql_update" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_update
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.registry_helper" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;registry_helper
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" registry_helper">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.registry_helper.fetch_assembly_id" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;fetch_assembly_id
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.registry_helper.fetch_current_genebuild_record" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;fetch_current_genebuild_record
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.registry_helper.fetch_genebuild_status_id" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;fetch_genebuild_status_id
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.registry_helper.fetch_registry_ids" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;fetch_registry_ids
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.seed_nonvert" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;seed_nonvert
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" seed_nonvert">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.seed_nonvert.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.seed_nonvert.dict_to_perl_hash" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;dict_to_perl_hash
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.seed_nonvert.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.seed_nonvert.seed_jobs_from_json" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;seed_jobs_from_json
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;start_pipeline_from_registry
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" start_pipeline_from_registry">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.args" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;args
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.parser" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parser
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.add_generated_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_generated_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.copy_general_module" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;copy_general_module
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.create_dir" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;create_dir
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.current_projection_source_db" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;current_projection_source_db
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.custom_loading" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;custom_loading
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_info_for_pipeline_anno" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_info_for_pipeline_anno
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_info_for_pipeline_main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_info_for_pipeline_main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_metadata_from_registry" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_metadata_from_registry
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_rna_and_busco_check_threshold" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_rna_and_busco_check_threshold
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_server_settings_anno" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_server_settings_anno
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_server_settings_main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_server_settings_main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.load_anno_settings" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;load_anno_settings
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.load_main_settings" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;load_main_settings
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.load_settings" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;load_settings
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.taxonomy_helper" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;taxonomy_helper
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" taxonomy_helper">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.taxonomy_helper.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.taxonomy_helper.assign_clade" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;assign_clade
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.taxonomy_helper.assign_clade_info_custom_loading" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;assign_clade_info_custom_loading
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.taxonomy_helper.create_tax_dictionary_from_registry" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;create_tax_dictionary_from_registry
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.taxonomy_helper.get_parent_taxon" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_parent_taxon
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.taxonomy_helper.load_clade_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;load_clade_data
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.update_assembly_registry" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;update_assembly_registry
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" update_assembly_registry">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.update_assembly_registry.args" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;args
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.update_assembly_registry.parser" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parser
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.update_assembly_registry.ensure_genebuilder_exists" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;ensure_genebuilder_exists
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.update_assembly_registry.insert_new_record" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;insert_new_record
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.update_assembly_registry.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.update_assembly_registry.set_old_record_historical" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;set_old_record_historical
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.update_assembly_registry.update_existing_record" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;update_existing_record
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.write_metrics_to_registry" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;write_metrics_to_registry
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" write_metrics_to_registry">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.args" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;args
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.parser" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parser
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.fetch_core_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;fetch_core_metrics
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.partition_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;partition_metrics
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.write_assembly_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_assembly_metrics
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.write_genebuild_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_genebuild_metrics
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metadata" class="md-nav__link">
    <span class="md-ellipsis">
      
        Metadata
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;beta_patcher
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" beta_patcher">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.METADATA_API_AVAILABLE" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;METADATA_API_AVAILABLE
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.generate_all_patches" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;generate_all_patches
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.get_genome_adaptor" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_genome_adaptor
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.get_genome_by_production_name" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_genome_by_production_name
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.get_genome_by_uuid" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_genome_by_uuid
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.group_patches_by_genome" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;group_patches_by_genome
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.read_csv_patches" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;read_csv_patches
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.resolve_genome_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;resolve_genome_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.setup_logging" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;setup_logging
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.suggest_genomes_by_production_name" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;suggest_genomes_by_production_name
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.write_core_patch_for_genome" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_core_patch_for_genome
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.beta_patcher.write_metadata_patch_for_genome" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_metadata_patch_for_genome
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;core_meta_data
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" core_meta_data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.args" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;args
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.core_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;core_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.core_meta" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;core_meta
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.core_query" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;core_query
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.current_id" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;current_id
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.db" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;db
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.gca_accession" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;gca_accession
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.log_file_path" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;log_file_path
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.log_ini_path" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;log_ini_path
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.meta_keys_to_remove" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;meta_keys_to_remove
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.meta_value" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;meta_value
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.metadata_dir" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;metadata_dir
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.name_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;name_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.name_query" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;name_query
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.output_dir" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;output_dir
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.parent_rank_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parent_rank_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.parent_taxonomy_query" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parent_taxonomy_query
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.parser" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parser
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.parts" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;parts
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.provider_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;provider_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.provider_static_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;provider_static_file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.rank_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;rank_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.rank_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;rank_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.ref_accession" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;ref_accession
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.ref_list" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;ref_list
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.ref_static_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;ref_static_file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.required_meta_keys" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;required_meta_keys
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.s_name_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;s_name_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.s_name_query" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;s_name_query
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.server_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;server_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.snp" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;snp
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.snp_list" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;snp_list
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.snp_static_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;snp_static_file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.species_id" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;species_id
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.species_meta" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;species_meta
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.species_query" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;species_query
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.sql_out" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;sql_out
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.taxonomy_query" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;taxonomy_query
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.truth_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;truth_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.url" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;url
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.url_list" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;url_list
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.url_static_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;url_static_file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.get_biosample_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_biosample_metadata
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.get_ena_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_ena_metadata
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.get_ncbi_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_ncbi_metadata
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metadata.core_meta_data.mysql_fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_fetch_data
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metrics" class="md-nav__link">
    <span class="md-ellipsis">
      
        Metrics
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_lineage_selector" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;busco_lineage_selector
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" busco_lineage_selector">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_lineage_selector.get_dataset_match" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_dataset_match
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_lineage_selector.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_lineage_selector.parse_args" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;parse_args
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_metakeys_patch" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;busco_metakeys_patch
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" busco_metakeys_patch">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_metakeys_patch.execute_sql_patches" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;execute_sql_patches
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_metakeys_patch.generate_sql_patches" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;generate_sql_patches
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_metakeys_patch.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_metakeys_patch.parse_busco_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;parse_busco_file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.busco_metakeys_patch.process_busco_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;process_busco_file
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.check_busco_score" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;check_busco_score
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" check_busco_score">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.check_busco_score.evaluate_busco" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;evaluate_busco
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.check_busco_score.extract_completeness" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;extract_completeness
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.metrics.check_busco_score.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#projects" class="md-nav__link">
    <span class="md-ellipsis">
      
        Projects
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;write_yaml
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" write_yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;EnsemblFTP
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" EnsemblFTP">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.ebi_ftp" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;ebi_ftp
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.ebi_ftp_path" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;ebi_ftp_path
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.ensembl_ftp" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;ensembl_ftp
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.ensembl_ftp_path" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;ensembl_ftp_path
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.max_retries" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;max_retries
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.retry_sleep" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;retry_sleep
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.timeout" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;timeout
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.check_for_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_for_file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.check_pre_release_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;check_pre_release_file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.close_connections" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;close_connections
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.return_to_root" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-method"></code>&nbsp;return_to_root
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.check_database_on_server" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_database_on_server
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.check_url_status" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_url_status
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.find_database_server" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;find_database_server
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.mysql_fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_fetch_data
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" mysql_fetch_data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.mysql_fetch_data--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.mysql_fetch_data--returns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.write_yaml" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_yaml
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" write_yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.projects.write_yaml.write_yaml--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#stable-id" class="md-nav__link">
    <span class="md-ellipsis">
      
        Stable ID
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.stable_id.pid_overlap" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;pid_overlap
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" pid_overlap">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.stable_id.pid_overlap.gff3_file_new" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;gff3_file_new
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.stable_id.pid_overlap.gff3_file_old" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;gff3_file_old
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.stable_id.pid_overlap.output_report" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;output_report
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.stable_id.pid_overlap.compare_protein_sets" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;compare_protein_sets
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.stable_id.pid_overlap.extract_proteins_from_gff3" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;extract_proteins_from_gff3
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tracking" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tracking
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;bioproject_tracking
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" bioproject_tracking">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.config" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;config
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.add_ftp" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;add_ftp
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_assembly_accessions" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_assembly_accessions
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_ensembl_live" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_ensembl_live
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_pre_release" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_pre_release
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.get_taxonomy_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_taxonomy_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.mysql_fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_fetch_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.bioproject_tracking.write_report" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;write_report
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.live_tracking" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;live_tracking
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" live_tracking">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.live_tracking.config" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-attribute"></code>&nbsp;config
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.live_tracking.check_database_on_server" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_database_on_server
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.live_tracking.clean_server" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;clean_server
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.live_tracking.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.tracking.live_tracking.mysql_fetch_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;mysql_fetch_data
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#transcriptomic-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Transcriptomic Data
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;get_transcriptomic_data
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" get_transcriptomic_data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data.InputSchema" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;InputSchema
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data.get_data_from_ena" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_data_from_ena
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data.get_sample_info" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;get_sample_info
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;check_for_transcriptomic
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" check_for_transcriptomic">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic.InputSchema" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-class"></code>&nbsp;InputSchema
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic.check_data_from_ena" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_data_from_ena
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic.ena_rest_api" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;ena_rest_api
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-module"></code>&nbsp;select_transcriptomic_data
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label=" select_transcriptomic_data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.check_fastqc_star_quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;check_fastqc_star_quality
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.clean_repeated_words" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;clean_repeated_words
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.connect_to_db" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;connect_to_db
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.create_report" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;create_report
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.fastqc_quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;fastqc_quality
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.filter_data" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;filter_data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.main" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;main
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.prioritise_tissues" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;prioritise_tissues
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.star_quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        <code class="doc-symbol doc-symbol-toc doc-symbol-function"></code>&nbsp;star_quality
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="api-reference">API Reference<a class="headerlink" href="#api-reference" title="Permanent link">&para;</a></h1>
<p>This page exposes API reference pages generated from the <code>ensembl</code> Python package
under <code>src/python/ensembl</code> using <code>mkdocstrings</code>.</p>
<!-- Auto-generated: modules grouped by subpackage -->

<h2 id="automation">Automation<a class="headerlink" href="#automation" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.automation.pre_release_ftp" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.automation.pre_release_ftp</code>


<a href="#ensembl.genes.automation.pre_release_ftp" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Pre-release FTP processing script for Ensembl genebuild pipeline.</p>
<p>This script gathers GTF, GFF3, and reheadered toplevel FASTA files from a 
pre-release directory structure and organizes them into an FTP-ready directory
structure with standardized naming and MD5 checksums.</p>










<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.calculate_md5" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">calculate_md5</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.calculate_md5" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate MD5 checksum of a file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filepath</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the file to checksum</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Hexadecimal MD5 hash</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="IOError">IOError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If file cannot be read</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">calculate_md5</span><span class="p">(</span><span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate MD5 checksum of a file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filepath: Path to the file to checksum</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Hexadecimal MD5 hash</span>

<span class="sd">    Raises:</span>
<span class="sd">        IOError: If file cannot be read</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hash_md5</span> <span class="o">=</span> <span class="n">hashlib</span><span class="o">.</span><span class="n">md5</span><span class="p">()</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="nb">iter</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="mi">4096</span><span class="p">),</span> <span class="sa">b</span><span class="s2">&quot;&quot;</span><span class="p">):</span>
            <span class="n">hash_md5</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">hash_md5</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.create_ftp_directory_structure" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">create_ftp_directory_structure</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">species_name</span><span class="p">,</span> <span class="n">gca_string</span><span class="p">,</span> <span class="n">annotation_files</span><span class="p">,</span> <span class="n">fasta_file</span><span class="p">,</span> <span class="n">two_bit_file</span><span class="p">,</span> <span class="n">logger</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.create_ftp_directory_structure" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create FTP directory structure and copy files with standardized naming.</p>
<p>Creates directory structure: {output_path}/ftp_release/{Species_Name}/{GCA_ID}/
and copies files with standardized names.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>output_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Base output directory</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>species_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species name for directory creation</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gca_string</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Full GCA identifier (e.g., 'GCA_030222105.1')</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_files</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Optional">Optional</span>[<span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of annotation file paths</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>fasta_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to reheadered FASTA file</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>two_bit_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to 2bit file</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>logger</code>
            </td>
            <td>
                  <code><span title="logging.Logger">Logger</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Logger instance for status messages</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Tuple">Tuple</span>[<span title="str">str</span>, <span title="typing.List">List</span>[<span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Tuple[str, List[str]]: FTP directory path and list of copied file paths</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="OSError">OSError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If directory creation or file copying fails</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_ftp_directory_structure</span><span class="p">(</span>
    <span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">species_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gca_string</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">annotation_files</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
    <span class="n">fasta_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">two_bit_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">logger</span><span class="p">:</span> <span class="n">logging</span><span class="o">.</span><span class="n">Logger</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create FTP directory structure and copy files with standardized naming.</span>

<span class="sd">    Creates directory structure: {output_path}/ftp_release/{Species_Name}/{GCA_ID}/</span>
<span class="sd">    and copies files with standardized names.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_path: Base output directory</span>
<span class="sd">        species_name: Species name for directory creation</span>
<span class="sd">        gca_string: Full GCA identifier (e.g., &#39;GCA_030222105.1&#39;)</span>
<span class="sd">        annotation_files: Dictionary of annotation file paths</span>
<span class="sd">        fasta_file: Path to reheadered FASTA file</span>
<span class="sd">        two_bit_file: Path to 2bit file</span>
<span class="sd">        logger: Logger instance for status messages</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[str, List[str]]: FTP directory path and list of copied file paths</span>

<span class="sd">    Raises:</span>
<span class="sd">        OSError: If directory creation or file copying fails</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Parse GCA information</span>
    <span class="n">gca_number</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="n">parse_gca_id</span><span class="p">(</span><span class="n">gca_string</span><span class="p">)</span>

    <span class="c1"># Create FTP directory structure: species_name/GCA_ID/</span>
    <span class="n">species_title</span> <span class="o">=</span> <span class="n">format_species_name</span><span class="p">(</span><span class="n">species_name</span><span class="p">)</span>

    <span class="n">ftp_base</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;ftp_release&quot;</span><span class="p">)</span>
    <span class="n">ftp_species_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ftp_base</span><span class="p">,</span> <span class="n">species_title</span><span class="p">)</span>
    <span class="n">ftp_final_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ftp_species_dir</span><span class="p">,</span> <span class="n">gca_string</span><span class="p">)</span>

    <span class="c1"># Create directories</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">ftp_final_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Created FTP directory: </span><span class="si">{</span><span class="n">ftp_final_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">copied_files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Copy and rename annotation files</span>
    <span class="k">for</span> <span class="n">file_type</span><span class="p">,</span> <span class="n">source_file</span> <span class="ow">in</span> <span class="n">annotation_files</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">source_file</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">source_file</span><span class="p">):</span>
            <span class="n">new_filename</span> <span class="o">=</span> <span class="n">generate_new_filename</span><span class="p">(</span>
                <span class="n">source_file</span><span class="p">,</span> <span class="n">species_name</span><span class="p">,</span> <span class="n">gca_number</span><span class="p">,</span> <span class="n">version</span><span class="p">,</span> <span class="n">file_type</span>
            <span class="p">)</span>
            <span class="n">dest_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ftp_final_dir</span><span class="p">,</span> <span class="n">new_filename</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Copying </span><span class="si">{</span><span class="n">file_type</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">source_file</span><span class="p">)</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">new_filename</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">source_file</span><span class="p">,</span> <span class="n">dest_file</span><span class="p">)</span>
            <span class="n">copied_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dest_file</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No </span><span class="si">{</span><span class="n">file_type</span><span class="si">}</span><span class="s2"> file found to copy&quot;</span><span class="p">)</span>

    <span class="c1"># Copy and rename FASTA file</span>
    <span class="k">if</span> <span class="n">fasta_file</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">):</span>
        <span class="n">new_fasta_name</span> <span class="o">=</span> <span class="n">generate_new_filename</span><span class="p">(</span>
            <span class="n">fasta_file</span><span class="p">,</span> <span class="n">species_name</span><span class="p">,</span> <span class="n">gca_number</span><span class="p">,</span> <span class="n">version</span><span class="p">,</span> <span class="s2">&quot;dna.softmasked.fa&quot;</span>
        <span class="p">)</span>
        <span class="n">dest_fasta</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ftp_final_dir</span><span class="p">,</span> <span class="n">new_fasta_name</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Copying FASTA: </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">)</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">new_fasta_name</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="n">dest_fasta</span><span class="p">)</span>
        <span class="n">copied_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dest_fasta</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No reheadered FASTA file found to copy&quot;</span><span class="p">)</span>

    <span class="c1"># Copy 2bit file if it exists</span>
    <span class="k">if</span> <span class="n">two_bit_file</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">two_bit_file</span><span class="p">):</span>
        <span class="n">new_2bit_name</span> <span class="o">=</span> <span class="n">generate_new_filename</span><span class="p">(</span>
            <span class="n">two_bit_file</span><span class="p">,</span> <span class="n">species_name</span><span class="p">,</span> <span class="n">gca_number</span><span class="p">,</span> <span class="n">version</span><span class="p">,</span> <span class="s2">&quot;dna.toplevel.2bit&quot;</span>
        <span class="p">)</span>
        <span class="n">dest_2bit</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ftp_final_dir</span><span class="p">,</span> <span class="n">new_2bit_name</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Copying 2bit file: </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">two_bit_file</span><span class="p">)</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">new_2bit_name</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">two_bit_file</span><span class="p">,</span> <span class="n">dest_2bit</span><span class="p">)</span>
        <span class="n">copied_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dest_2bit</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No 2bit file found to copy&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ftp_final_dir</span><span class="p">,</span> <span class="n">copied_files</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.extract_species_from_filename" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">extract_species_from_filename</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.extract_species_from_filename" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extract species name from Ensembl filename format.</p>
<p>Parses filenames like:
- Bos_taurus_gca002263795v4.ARS-UCD2.0.114.primary_assembly.1.gff3.gz
- macaca_mulatta_gca030222105v1.gtf.gz</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filename</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Full filename or path to parse</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species name (e.g., 'Bos_taurus', 'macaca_mulatta')</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">extract_species_from_filename</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract species name from Ensembl filename format.</span>

<span class="sd">    Parses filenames like:</span>
<span class="sd">    - Bos_taurus_gca002263795v4.ARS-UCD2.0.114.primary_assembly.1.gff3.gz</span>
<span class="sd">    - macaca_mulatta_gca030222105v1.gtf.gz</span>

<span class="sd">    Args:</span>
<span class="sd">        filename: Full filename or path to parse</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Species name (e.g., &#39;Bos_taurus&#39;, &#39;macaca_mulatta&#39;)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">basename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="c1"># Split on first occurrence of version pattern (gca followed by numbers)</span>
    <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;([a-zA-Z_]+)_gca\d+&quot;</span><span class="p">,</span> <span class="n">basename</span><span class="p">,</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Fallback: take everything before the first dot</span>
    <span class="k">return</span> <span class="n">basename</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.find_2bit" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">find_2bit</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.find_2bit" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Find the 2bit file in the output directory.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>output_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the base output directory</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    Optional[str]: Path to the 2bit file or None if not found</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_2bit</span><span class="p">(</span><span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find the 2bit file in the output directory.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_path: Path to the base output directory</span>
<span class="sd">    Returns:</span>
<span class="sd">        Optional[str]: Path to the 2bit file or None if not found</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">two_bit_pattern</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;*_softmasked_toplevel.2bit&quot;</span><span class="p">)</span>
    <span class="n">two_bit_matches</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">two_bit_pattern</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">two_bit_matches</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">two_bit_matches</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Return first match</span>
    <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.find_annotation_files" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">find_annotation_files</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.find_annotation_files" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Find GFF3 and GTF annotation files in the clade (vertabrates/ etc) directory structure.
Prioritizes main annotation files over specialized variants like abinitio.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>output_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the base output directory</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Optional">Optional</span>[<span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict[str, Optional[str]]: Dictionary with 'gff3' and 'gtf' keys mapping
                     to file paths or None if not found</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_annotation_files</span><span class="p">(</span><span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find GFF3 and GTF annotation files in the clade (vertabrates/ etc) directory structure.</span>
<span class="sd">    Prioritizes main annotation files over specialized variants like abinitio.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_path: Path to the base output directory</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Optional[str]]: Dictionary with &#39;gff3&#39; and &#39;gtf&#39; keys mapping</span>
<span class="sd">                                 to file paths or None if not found</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">annotation_files</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;gff3&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;gtf&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>

    <span class="c1"># Look for GFF3 files</span>
    <span class="n">gff3_pattern</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="s2">&quot;gff3&quot;</span><span class="p">,</span> <span class="s2">&quot;**&quot;</span><span class="p">,</span> <span class="s2">&quot;*.gff3.gz&quot;</span><span class="p">)</span>
    <span class="n">gff3_matches</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">gff3_pattern</span><span class="p">,</span> <span class="n">recursive</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">gff3_matches</span><span class="p">:</span>
        <span class="c1"># Prioritize main GFF3 files over specialized variants</span>
        <span class="n">main_gff3</span> <span class="o">=</span> <span class="n">_select_main_annotation_file</span><span class="p">(</span><span class="n">gff3_matches</span><span class="p">,</span> <span class="s2">&quot;gff3&quot;</span><span class="p">)</span>
        <span class="n">annotation_files</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">main_gff3</span>

    <span class="c1"># Look for GTF files</span>
    <span class="n">gtf_pattern</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="s2">&quot;gtf&quot;</span><span class="p">,</span> <span class="s2">&quot;**&quot;</span><span class="p">,</span> <span class="s2">&quot;*.gtf.gz&quot;</span><span class="p">)</span>
    <span class="n">gtf_matches</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">gtf_pattern</span><span class="p">,</span> <span class="n">recursive</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">gtf_matches</span><span class="p">:</span>
        <span class="c1"># Prioritize main GTF files over specialized variants</span>
        <span class="n">main_gtf</span> <span class="o">=</span> <span class="n">_select_main_annotation_file</span><span class="p">(</span><span class="n">gtf_matches</span><span class="p">,</span> <span class="s2">&quot;gtf&quot;</span><span class="p">)</span>
        <span class="n">annotation_files</span><span class="p">[</span><span class="s2">&quot;gtf&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">main_gtf</span>

    <span class="k">return</span> <span class="n">annotation_files</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.find_reheadered_fasta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">find_reheadered_fasta</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.find_reheadered_fasta" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Find the reheadered toplevel FASTA file and its corresponding FAI index file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>output_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory path to search for FASTA files</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="tuple">tuple</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>tuple[str, str]: Full paths to (FASTA file, FAI file)</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If reheadered FASTA file or its FAI index is not found</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_reheadered_fasta</span><span class="p">(</span><span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find the reheadered toplevel FASTA file and its corresponding FAI index file.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_path: Directory path to search for FASTA files</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple[str, str]: Full paths to (FASTA file, FAI file)</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If reheadered FASTA file or its FAI index is not found</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Find FASTA file</span>
    <span class="n">fasta_pattern</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;*_softmasked_toplevel.fa.gz&quot;</span><span class="p">)</span>
    <span class="n">fasta_matches</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">fasta_pattern</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">fasta_matches</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;No reheadered toplevel FASTA file found in </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">fasta_matches</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.format_species_name" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">format_species_name</span><span class="p">(</span><span class="n">species_name</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.format_species_name" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Reformat species name into typical binomial nomenclature format.</p>
<p>Converts species names to the standard scientific naming convention where
the genus is capitalized, subsequent parts are lowercase, and all parts
are separated by underscores. Replaces spaces, hyphens, and periods with
underscores.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>species_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species name to format (e.g., 'homo sapiens', 'canis_lupus')</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Formatted species name (e.g., 'Homo_sapiens', 'Canis_lupus')</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">format_species_name</span><span class="p">(</span><span class="n">species_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reformat species name into typical binomial nomenclature format.</span>

<span class="sd">    Converts species names to the standard scientific naming convention where</span>
<span class="sd">    the genus is capitalized, subsequent parts are lowercase, and all parts</span>
<span class="sd">    are separated by underscores. Replaces spaces, hyphens, and periods with</span>
<span class="sd">    underscores.</span>

<span class="sd">    Args:</span>
<span class="sd">        species_name: Species name to format (e.g., &#39;homo sapiens&#39;, &#39;canis_lupus&#39;)</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Formatted species name (e.g., &#39;Homo_sapiens&#39;, &#39;Canis_lupus&#39;)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parts</span> <span class="o">=</span> <span class="n">species_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">species_title</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="c1"># Add any additional parts if they exist</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">species_title</span> <span class="o">+=</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">part</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="k">for</span> <span class="n">part</span> <span class="ow">in</span> <span class="n">parts</span><span class="p">[</span><span class="mi">2</span><span class="p">:])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">species_title</span> <span class="o">=</span> <span class="n">species_name</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">species_title</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.generate_md5_checksums" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">generate_md5_checksums</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="n">files</span><span class="p">,</span> <span class="n">logger</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.generate_md5_checksums" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate MD5 checksums for files and write to CHECKSUMS file.</p>
<p>Creates a CHECKSUMS file in the specified directory with MD5 hashes
for all provided files in the format: {hash}  {filename}</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>directory</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory where CHECKSUMS file will be created</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>files</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of file paths to checksum</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>logger</code>
            </td>
            <td>
                  <code><span title="logging.Logger">Logger</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Logger instance for status messages</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the created CHECKSUMS file</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="IOError">IOError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If CHECKSUMS file cannot be written</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_md5_checksums</span><span class="p">(</span>
    <span class="n">directory</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">logger</span><span class="p">:</span> <span class="n">logging</span><span class="o">.</span><span class="n">Logger</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate MD5 checksums for files and write to CHECKSUMS file.</span>

<span class="sd">    Creates a CHECKSUMS file in the specified directory with MD5 hashes</span>
<span class="sd">    for all provided files in the format: {hash}  {filename}</span>

<span class="sd">    Args:</span>
<span class="sd">        directory: Directory where CHECKSUMS file will be created</span>
<span class="sd">        files: List of file paths to checksum</span>
<span class="sd">        logger: Logger instance for status messages</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Path to the created CHECKSUMS file</span>

<span class="sd">    Raises:</span>
<span class="sd">        IOError: If CHECKSUMS file cannot be written</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">md5_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">directory</span><span class="p">,</span> <span class="s2">&quot;CHECKSUMS&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Generating MD5 checksums in: </span><span class="si">{</span><span class="n">md5_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">md5_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">filepath</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">files</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
                <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
                <span class="n">md5_hash</span> <span class="o">=</span> <span class="n">calculate_md5</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">md5_hash</span><span class="si">}</span><span class="s2">  </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">md5_hash</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">md5_file</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.generate_new_filename" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">generate_new_filename</span><span class="p">(</span><span class="n">original_file</span><span class="p">,</span> <span class="n">species_name</span><span class="p">,</span> <span class="n">gca_number</span><span class="p">,</span> <span class="n">version</span><span class="p">,</span> <span class="n">file_type</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.generate_new_filename" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate standardized filename for FTP release.</p>
<p>Creates filenames in the format: species_gca{number}v{version}.{type}.gz</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>original_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Original file path (used for reference)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>species_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species name (e.g., 'Macaca_mulatta')</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gca_number</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>GCA accession number (e.g., '030222105')</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>version</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Version number (e.g., '1')</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>file_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>File type (e.g., 'gff3', 'gtf', 'dna.toplevel.fa')</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Standardized filename</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_new_filename</span><span class="p">(</span>
    <span class="n">original_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">species_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">gca_number</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">version</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">file_type</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate standardized filename for FTP release.</span>

<span class="sd">    Creates filenames in the format: species_gca{number}v{version}.{type}.gz</span>

<span class="sd">    Args:</span>
<span class="sd">        original_file: Original file path (used for reference)</span>
<span class="sd">        species_name: Species name (e.g., &#39;Macaca_mulatta&#39;)</span>
<span class="sd">        gca_number: GCA accession number (e.g., &#39;030222105&#39;)</span>
<span class="sd">        version: Version number (e.g., &#39;1&#39;)</span>
<span class="sd">        file_type: File type (e.g., &#39;gff3&#39;, &#39;gtf&#39;, &#39;dna.toplevel.fa&#39;)</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Standardized filename</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">species_lower</span> <span class="o">=</span> <span class="n">species_name</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
    <span class="n">new_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">species_lower</span><span class="si">}</span><span class="s2">_gca</span><span class="si">{</span><span class="n">gca_number</span><span class="si">}</span><span class="s2">v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">file_type</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="k">if</span> <span class="n">is_compressed_file</span><span class="p">(</span><span class="n">original_file</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">new_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">):</span>
        <span class="n">new_name</span> <span class="o">+=</span> <span class="s2">&quot;.gz&quot;</span>
    <span class="k">return</span> <span class="n">new_name</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.is_compressed_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">is_compressed_file</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.is_compressed_file" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Check if a file is actually compressed by examining its magic bytes.</p>
<p>Checks for gzip magic bytes (0x1f, 0x8b) at the beginning of the file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filepath</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the file to check</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if file is actually gzip compressed</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">is_compressed_file</span><span class="p">(</span><span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if a file is actually compressed by examining its magic bytes.</span>

<span class="sd">    Checks for gzip magic bytes (0x1f, 0x8b) at the beginning of the file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filepath: Path to the file to check</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if file is actually gzip compressed</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">magic_bytes</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
            <span class="c1"># Check for gzip magic bytes (0x1f, 0x8b)</span>
            <span class="k">return</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">magic_bytes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span>
                <span class="ow">and</span> <span class="n">magic_bytes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mh">0x1F</span>
                <span class="ow">and</span> <span class="n">magic_bytes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mh">0x8B</span>
            <span class="p">)</span>
    <span class="k">except</span> <span class="p">(</span><span class="ne">IOError</span><span class="p">,</span> <span class="ne">OSError</span><span class="p">):</span>
        <span class="c1"># If we can&#39;t read the file, assume it&#39;s not compressed</span>
        <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main function to process pre-release FTP data.</p>
<p>Parses command line arguments, validates inputs, and orchestrates the
file gathering and FTP structure creation process.</p>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="SystemExit">SystemExit</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>On validation errors or processing failures</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main function to process pre-release FTP data.</span>

<span class="sd">    Parses command line arguments, validates inputs, and orchestrates the</span>
<span class="sd">    file gathering and FTP structure creation process.</span>

<span class="sd">    Raises:</span>
<span class="sd">        SystemExit: On validation errors or processing failures</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Gather GTF, GFF3, and FASTA files into FTP directory structure&quot;</span><span class="p">,</span>
        <span class="n">formatter_class</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">RawDescriptionHelpFormatter</span><span class="p">,</span>
        <span class="n">epilog</span><span class="o">=</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">Examples:</span>
<span class="s2">  </span><span class="si">%(prog)s</span><span class="s2"> -p /path/to/GCA_030222105.1/ -g GCA_030222105.1 -s Macaca_mulatta</span>
<span class="s2">  </span><span class="si">%(prog)s</span><span class="s2"> -p /path/to/data/ -g GCA_002263795.4</span>
<span class="s2">        &quot;&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-p&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--path&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output path containing the genome data directory structure&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-g&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--gca&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;GCA identifier in format GCA_NNNNNNNN.N (e.g., GCA_030222105.1)&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-s&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--species&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Species name (e.g., Macaca_mulatta). If not provided, will extract from filenames&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">logger</span> <span class="o">=</span> <span class="n">setup_logging</span><span class="p">()</span>

    <span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">path</span>
    <span class="n">gca_string</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">gca</span>
    <span class="n">species_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">species</span>

    <span class="c1"># Validate input path</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output_path</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Output path does not exist: </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">output_path</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Output path is not a directory: </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing pre-release FTP data in: </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GCA identifier: </span><span class="si">{</span><span class="n">gca_string</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Find reheadered FASTA file</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">fasta_file</span> <span class="o">=</span> <span class="n">find_reheadered_fasta</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found reheadered FASTA file: </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">FileNotFoundError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

        <span class="n">annotation_files</span> <span class="o">=</span> <span class="n">find_annotation_files</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found annotation files: </span><span class="si">{</span><span class="n">annotation_files</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Find 2bit file if it exists</span>
        <span class="n">two_bit_file</span> <span class="o">=</span> <span class="n">find_2bit</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">two_bit_file</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found 2bit file: </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">two_bit_file</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No 2bit file found in the output directory&quot;</span><span class="p">)</span>

        <span class="c1"># Extract species name if not provided</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">species_name</span><span class="p">:</span>
            <span class="c1"># Try to extract from GFF3 file first, then GTF</span>
            <span class="k">for</span> <span class="n">file_type</span><span class="p">,</span> <span class="n">filepath</span> <span class="ow">in</span> <span class="n">annotation_files</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">filepath</span><span class="p">:</span>
                    <span class="n">species_name</span> <span class="o">=</span> <span class="n">extract_species_from_filename</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Extracted species name from </span><span class="si">{</span><span class="n">file_type</span><span class="si">}</span><span class="s2"> file: </span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                    <span class="k">break</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">species_name</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                    <span class="s2">&quot;Could not determine species name. Please provide with -s option&quot;</span>
                <span class="p">)</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

        <span class="c1"># Create FTP directory structure and copy files</span>
        <span class="n">ftp_dir</span><span class="p">,</span> <span class="n">copied_files</span> <span class="o">=</span> <span class="n">create_ftp_directory_structure</span><span class="p">(</span>
            <span class="n">output_path</span><span class="p">,</span>
            <span class="n">species_name</span><span class="p">,</span>
            <span class="n">gca_string</span><span class="p">,</span>
            <span class="n">annotation_files</span><span class="p">,</span>
            <span class="n">fasta_file</span><span class="p">,</span>
            <span class="n">two_bit_file</span><span class="p">,</span>
            <span class="n">logger</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Generate MD5 checksums</span>
        <span class="k">if</span> <span class="n">copied_files</span><span class="p">:</span>
            <span class="n">checksum_file</span> <span class="o">=</span> <span class="n">generate_md5_checksums</span><span class="p">(</span><span class="n">ftp_dir</span><span class="p">,</span> <span class="n">copied_files</span><span class="p">,</span> <span class="n">logger</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MD5 checksums written to: </span><span class="si">{</span><span class="n">checksum_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;FTP structure created successfully in: </span><span class="si">{</span><span class="n">ftp_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Pre-release FTP processing completed successfully&quot;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error during processing: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.parse_gca_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">parse_gca_id</span><span class="p">(</span><span class="n">gca_string</span><span class="p">)</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.parse_gca_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parse GCA ID to extract accession number and version information.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gca_string</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>GCA identifier in format GCA_030222105.1</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Tuple">Tuple</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Tuple[str, str]: GCA number and version (e.g., ('030222105', '1'))</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If GCA format is invalid</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">parse_gca_id</span><span class="p">(</span><span class="n">gca_string</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse GCA ID to extract accession number and version information.</span>

<span class="sd">    Args:</span>
<span class="sd">        gca_string: GCA identifier in format GCA_030222105.1</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[str, str]: GCA number and version (e.g., (&#39;030222105&#39;, &#39;1&#39;))</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If GCA format is invalid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract GCA number and version from format like GCA_030222105.1</span>
    <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;GCA_(\d+)\.(\d+)&quot;</span><span class="p">,</span> <span class="n">gca_string</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">match</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Invalid GCA format: </span><span class="si">{</span><span class="n">gca_string</span><span class="si">}</span><span class="s2">. Expected format: GCA_NNNNNNNN.N&quot;</span>
        <span class="p">)</span>

    <span class="n">gca_number</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">version</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gca_number</span><span class="p">,</span> <span class="n">version</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.automation.pre_release_ftp.setup_logging" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">setup_logging</span><span class="p">()</span></code>

<a href="#ensembl.genes.automation.pre_release_ftp.setup_logging" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Set up logging configuration with timestamps and appropriate formatting.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="logging.Logger">Logger</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>logging.Logger: Configured logger instance</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/automation/pre_release_ftp.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">setup_logging</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">logging</span><span class="o">.</span><span class="n">Logger</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Set up logging configuration with timestamps and appropriate formatting.</span>

<span class="sd">    Returns:</span>
<span class="sd">        logging.Logger: Configured logger instance</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
        <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span>
        <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> - </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">datefmt</span><span class="o">=</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="content">Content<a class="headerlink" href="#content" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.content.main_static_content" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.content.main_static_content</code>


<a href="#ensembl.genes.content.main_static_content" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Module to create static content files for main release.
This module fetches assembly information from a MySQL database and the ENA API,
and generates HTML files containing assembly and annotation content.</p>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.CORE_QUERY" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">CORE_QUERY</span> <span class="o">=</span> <span class="s2">&quot;SELECT meta_key, meta_value FROM meta WHERE meta_key IN (&#39;assembly.accession&#39;, &#39;species.url&#39;);&quot;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.CORE_QUERY" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.assembly_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">assembly_info</span> <span class="o">=</span> <span class="n">get_assembly_info</span><span class="p">(</span><span class="n">core_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">])</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.assembly_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.core_dict" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">core_dict</span> <span class="o">=</span> <span class="p">{(</span><span class="n">meta_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span> <span class="p">(</span><span class="n">meta_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">meta_pair</span> <span class="ow">in</span> <span class="n">core_meta</span><span class="p">}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.core_dict" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.core_meta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">core_meta</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">CORE_QUERY</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;server&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]),</span> <span class="n">user</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;server&#39;</span><span class="p">][</span><span class="s1">&#39;db_user&#39;</span><span class="p">]),</span> <span class="n">port</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;server&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]),</span> <span class="n">database</span><span class="o">=</span><span class="n">db</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.core_meta" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.db" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">db</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">db_name</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.db" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.output_dir" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">output_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.output_dir" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.parser" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Prepare static content files for main release&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.parser" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.server_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">server_info</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;server&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;db_host&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">host</span><span class="p">,</span> <span class="s1">&#39;db_port&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">port</span><span class="p">,</span> <span class="s1">&#39;db_user&#39;</span><span class="p">:</span> <span class="s1">&#39;ensro&#39;</span><span class="p">}}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.server_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_content.url_name" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">url_name</span> <span class="o">=</span> <span class="n">core_dict</span><span class="p">[</span><span class="s1">&#39;species.url&#39;</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_content.url_name" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.content.main_static_content.get_assembly_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_assembly_info</span><span class="p">(</span><span class="n">accession</span><span class="p">)</span></code>

<a href="#ensembl.genes.content.main_static_content.get_assembly_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Retrieve assembly information from the ENA API for a given accession.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>accession</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The assembly accession number to fetch information for.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary containing assembly attributes such as name, level,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>submitter, and counts.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/content/main_static_content.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_assembly_info</span><span class="p">(</span><span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve assembly information from the ENA API for a given accession.</span>

<span class="sd">    Args:</span>
<span class="sd">        accession (str): The assembly accession number to fetch information for.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: A dictionary containing assembly attributes such as name, level,</span>
<span class="sd">        submitter, and counts.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">assembly_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/browser/api/xml/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">assembly_xml</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">assembly_url</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">assembly_dict</span> <span class="o">=</span> <span class="n">xmltodict</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">assembly_xml</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>
    <span class="n">assembly_attribs</span> <span class="o">=</span> <span class="n">assembly_dict</span><span class="p">[</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY_ATTRIBUTES&quot;</span><span class="p">][</span>
        <span class="s2">&quot;ASSEMBLY_ATTRIBUTE&quot;</span>
    <span class="p">]</span>

    <span class="n">return_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Retrieve specific attributes with error handling</span>
    <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assembly_dict</span><span class="p">[</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="s2">&quot;NAME&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
    <span class="p">)</span>
    <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assembly_dict</span><span class="p">[</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="s2">&quot;ASSEMBLY_LEVEL&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
    <span class="p">)</span>
    <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.submitter&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assembly_dict</span><span class="p">[</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">][</span>
        <span class="s2">&quot;IDENTIFIERS&quot;</span>
    <span class="p">][</span><span class="s2">&quot;SUBMITTER_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;@namespace&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="c1"># Fetch specific attributes based on tags</span>
    <span class="k">for</span> <span class="n">attrib_set</span> <span class="ow">in</span> <span class="n">assembly_attribs</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;count-contig&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;contig.count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;scaffold-count&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;scaffold.count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ENA-LAST-UPDATED&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.date&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;n50&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;scaffold.n50&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;contig-n50&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;contig.n50&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">return_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.content.main_static_content.mysql_fetch_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">)</span></code>

<a href="#ensembl.genes.content.main_static_content.mysql_fetch_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch data from a MySQL database based on a provided query.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>query</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>SQL query to execute.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>database</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Database name to connect to.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>host</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Host address of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>port</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Port number of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>user</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Username for the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>tuple</code></td>            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.Tuple">Tuple</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A tuple containing the fetched data.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/content/main_static_content.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mysql_fetch_data</span><span class="p">(</span>
    <span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">user</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Tuple</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch data from a MySQL database based on a provided query.</span>

<span class="sd">    Args:</span>
<span class="sd">        query (str): SQL query to execute.</span>
<span class="sd">        database (str): Database name to connect to.</span>
<span class="sd">        host (str): Host address of the MySQL server.</span>
<span class="sd">        port (int): Port number of the MySQL server.</span>
<span class="sd">        user (str): Username for the MySQL server.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing the fetched data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="n">port</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">cursor</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
        <span class="n">info</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Database error: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">info</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">finally</span><span class="p">:</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">info</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.content.main_static_content.write_content" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">write_content</span><span class="p">(</span><span class="n">info</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">,</span> <span class="n">url_path</span><span class="p">)</span></code>

<a href="#ensembl.genes.content.main_static_content.write_content" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Write assembly and annotation content to HTML files.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly information to be written.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_dir</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory path for output files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>url_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name used to title the output files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/content/main_static_content.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_content</span><span class="p">(</span><span class="n">info</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">url_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write assembly and annotation content to HTML files.</span>

<span class="sd">    Args:</span>
<span class="sd">        info (dict): Assembly information to be written.</span>
<span class="sd">        output_dir (Path): Directory path for output files.</span>
<span class="sd">        url_path (str): Name used to title the output files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span>  <span class="c1"># pylint: disable=unspecified-encoding</span>
        <span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">url_path</span><span class="si">}</span><span class="s2">_assembly.html&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">assembly_out</span><span class="p">:</span>  <span class="c1"># pylint: disable=unspecified-encoding</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;&lt;p&gt;The </span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;assembly.name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> assembly was submitted by &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;assembly.submitter&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> and last updated on &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;assembly.date&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">. The assembly is on the &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;assembly.level&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;level, consisting of </span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;contig.count&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> contigs assembled into &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;scaffold.count&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> scaffolds. &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;The N50 size is the length such that 50% of the assembled genome lies in &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;blocks of the N50 size or longer. &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;The N50 length for the contigs is </span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;contig.n50&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> while the &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;scaffold N50 is </span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;scaffold.n50&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">.&lt;/p&gt;&quot;</span><span class="p">,</span>
            <span class="n">file</span><span class="o">=</span><span class="n">assembly_out</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span>  <span class="c1"># pylint: disable=unspecified-encoding</span>
        <span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">url_path</span><span class="si">}</span><span class="s2">_annotation.html&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">annotation_out</span><span class="p">:</span>  <span class="c1"># pylint: disable=unspecified-encoding</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="s2">&quot;&lt;p&gt;Genome annotation was generated using the &quot;</span>
            <span class="s1">&#39;&lt;a href=&quot;https://beta.ensembl.org/help/articles/vertebrate-genome-annotation&quot;&gt;Ensembl vertebrate annotation pipeline&lt;/a&gt;. &#39;</span>  <span class="c1"># pylint: disable=line-too-long</span>
            <span class="s2">&quot;&lt;/p&gt;&lt;p&gt;In accordance with the &quot;</span>
            <span class="s1">&#39;&lt;a href=&quot;https://en.wikipedia.org/wiki/Fort_Lauderdale_Agreement&quot;&gt;Fort Lauderdale Agreement&lt;/a&gt;, please check the publication &#39;</span>  <span class="c1"># pylint: disable=line-too-long</span>
            <span class="s2">&quot;status of the genome/assembly before publishing any genome-wide analyses using these data.&lt;/p&gt;&quot;</span><span class="p">,</span>  <span class="c1"># pylint: disable=line-too-long</span>
            <span class="n">file</span><span class="o">=</span><span class="n">annotation_out</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.content.main_static_murine_content" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.content.main_static_murine_content</code>


<a href="#ensembl.genes.content.main_static_murine_content" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Prepare static content files for main release.
This script fetches assembly information from a MySQL database and the ENA API,
and generates HTML files containing assembly and annotation content.</p>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.CORE_QUERY" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">CORE_QUERY</span> <span class="o">=</span> <span class="s2">&quot;SELECT meta_key, meta_value FROM meta WHERE meta_key IN (&#39;assembly.accession&#39;, &#39;species.url&#39;, &#39;species.strain&#39;);&quot;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.CORE_QUERY" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.assembly_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">assembly_info</span> <span class="o">=</span> <span class="n">get_assembly_info</span><span class="p">(</span><span class="n">core_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">])</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.assembly_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.core_dict" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">core_dict</span> <span class="o">=</span> <span class="p">{(</span><span class="n">meta_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span> <span class="p">(</span><span class="n">meta_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">meta_pair</span> <span class="ow">in</span> <span class="n">core_meta</span><span class="p">}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.core_dict" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.core_meta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">core_meta</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">CORE_QUERY</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;server&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]),</span> <span class="n">user</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;server&#39;</span><span class="p">][</span><span class="s1">&#39;db_user&#39;</span><span class="p">]),</span> <span class="n">port</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;server&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]),</span> <span class="n">database</span><span class="o">=</span><span class="n">db</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.core_meta" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.db" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">db</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">db_name</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.db" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.output_dir" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">output_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.output_dir" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.parser" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Prepare static content files for main release&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.parser" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.server_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">server_info</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;server&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;db_host&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">host</span><span class="p">,</span> <span class="s1">&#39;db_port&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">port</span><span class="p">,</span> <span class="s1">&#39;db_user&#39;</span><span class="p">:</span> <span class="s1">&#39;ensro&#39;</span><span class="p">}}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.server_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.strain" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">strain</span> <span class="o">=</span> <span class="n">core_dict</span><span class="p">[</span><span class="s1">&#39;species.strain&#39;</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.strain" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.content.main_static_murine_content.url_name" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">url_name</span> <span class="o">=</span> <span class="n">core_dict</span><span class="p">[</span><span class="s1">&#39;species.url&#39;</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.content.main_static_murine_content.url_name" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.content.main_static_murine_content.get_assembly_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_assembly_info</span><span class="p">(</span><span class="n">accession</span><span class="p">)</span></code>

<a href="#ensembl.genes.content.main_static_murine_content.get_assembly_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Retrieve assembly information from the ENA API for a given accession.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>accession</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The assembly accession number to fetch information for.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary containing assembly attributes such as name, level,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>submitter, and counts.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/content/main_static_murine_content.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_assembly_info</span><span class="p">(</span><span class="n">accession</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve assembly information from the ENA API for a given accession.</span>

<span class="sd">    Args:</span>
<span class="sd">        accession (str): The assembly accession number to fetch information for.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: A dictionary containing assembly attributes such as name, level,</span>
<span class="sd">        submitter, and counts.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">assembly_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/browser/api/xml/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">assembly_xml</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">assembly_url</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">assembly_dict</span> <span class="o">=</span> <span class="n">xmltodict</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">assembly_xml</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>
    <span class="n">assembly_attribs</span> <span class="o">=</span> <span class="n">assembly_dict</span><span class="p">[</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY_ATTRIBUTES&quot;</span><span class="p">][</span>
        <span class="s2">&quot;ASSEMBLY_ATTRIBUTE&quot;</span>
    <span class="p">]</span>

    <span class="n">return_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Retrieve specific attributes with error handling</span>
    <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assembly_dict</span><span class="p">[</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="s2">&quot;NAME&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
    <span class="p">)</span>
    <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assembly_dict</span><span class="p">[</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="s2">&quot;ASSEMBLY_LEVEL&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
    <span class="p">)</span>
    <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.submitter&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assembly_dict</span><span class="p">[</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">][</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">][</span>
        <span class="s2">&quot;IDENTIFIERS&quot;</span>
    <span class="p">][</span><span class="s2">&quot;SUBMITTER_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;@namespace&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="c1"># Fetch specific attributes based on tags</span>
    <span class="k">for</span> <span class="n">attrib_set</span> <span class="ow">in</span> <span class="n">assembly_attribs</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;count-contig&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;contig.count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;scaffold-count&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;scaffold.count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ENA-LAST-UPDATED&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.date&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;n50&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;scaffold.n50&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;TAG&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;contig-n50&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;contig.n50&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="p">[</span><span class="s2">&quot;VALUE&quot;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">return_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.content.main_static_murine_content.mysql_fetch_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">)</span></code>

<a href="#ensembl.genes.content.main_static_murine_content.mysql_fetch_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch data from a MySQL database based on a provided query.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>query</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>SQL query to execute.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>database</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Database name to connect to.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>host</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Host address of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>port</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Port number of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>user</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Username for the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>tuple</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A tuple containing the fetched data.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/content/main_static_murine_content.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch data from a MySQL database based on a provided query.</span>

<span class="sd">    Args:</span>
<span class="sd">        query (str): SQL query to execute.</span>
<span class="sd">        database (str): Database name to connect to.</span>
<span class="sd">        host (str): Host address of the MySQL server.</span>
<span class="sd">        port (int): Port number of the MySQL server.</span>
<span class="sd">        user (str): Username for the MySQL server.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing the fetched data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="n">port</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">cursor</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
        <span class="n">info</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Database error: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">info</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">finally</span><span class="p">:</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">info</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.content.main_static_murine_content.write_content" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">write_content</span><span class="p">(</span><span class="n">info</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">,</span> <span class="n">url_path</span><span class="p">,</span> <span class="n">species_strain</span><span class="p">)</span></code>

<a href="#ensembl.genes.content.main_static_murine_content.write_content" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Write assembly and annotation content to HTML files.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly information to be written.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory path for output files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>url_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name used to title the output files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>species_strain</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species strain name for the assembly.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/content/main_static_murine_content.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_content</span><span class="p">(</span><span class="n">info</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">,</span> <span class="n">url_path</span><span class="p">,</span> <span class="n">species_strain</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write assembly and annotation content to HTML files.</span>

<span class="sd">    Args:</span>
<span class="sd">        info (dict): Assembly information to be written.</span>
<span class="sd">        out_dir (Path): Directory path for output files.</span>
<span class="sd">        url_path (str): Name used to title the output files.</span>
<span class="sd">        species_strain (str): Species strain name for the assembly.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span>
        <span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">url_path</span><span class="si">}</span><span class="s2">_assembly.html&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">assembly_out</span><span class="p">:</span>  <span class="c1"># pylint: disable=unspecified-encoding</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;&lt;p&gt;The assembly for </span><span class="si">{</span><span class="n">species_strain</span><span class="si">}</span><span class="s2"> was generated as part of &quot;</span>
            <span class="sa">f</span><span class="s1">&#39;&lt;a href=&quot;https://www.mousegenomes.org/&quot;&gt;The Mouse Genomes Project&#39;</span>
            <span class="sa">f</span><span class="s2">&quot;&lt;/a&gt;, additional species_strain can be found in &quot;</span>
            <span class="sa">f</span><span class="s1">&#39;&lt;a href=&quot;https://www.ensembl.org/Mus_musculus/Info/Strains&quot;&gt;Ensembl&lt;/a&gt;.&lt;/p&gt;&#39;</span>
            <span class="sa">f</span><span class="s2">&quot;&lt;p&gt;The assembly is on the </span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;assembly.level&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;level, consisting of </span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;contig.count&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> contigs assembled &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;into </span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;scaffold.count&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> scaffolds. &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;The N50 size is the length such that 50% of the assembled genome lies in &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;blocks of the N50 size or longer. &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;The N50 length for the contigs is </span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;contig.n50&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> while the &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;scaffold N50 is </span><span class="si">{</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;scaffold.n50&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">.&lt;/p&gt;&quot;</span><span class="p">,</span>
            <span class="n">file</span><span class="o">=</span><span class="n">assembly_out</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span>
        <span class="n">out_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">url_path</span><span class="si">}</span><span class="s2">_annotation.html&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">annotation_out</span><span class="p">:</span>  <span class="c1"># pylint: disable=unspecified-encoding</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;&lt;p&gt;Genome annotation was generated by mapping &quot;</span>  <span class="c1"># pylint: disable=f-string-without-interpolation</span>
            <span class="sa">f</span><span class="s1">&#39;&lt;a href=&quot;https://www.gencodegenes.org/mouse/release_M30.html&quot;&gt;GENCODE M30&lt;/a&gt; &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;genes and transcripts via the &lt;a href=&quot;https://beta.ensembl.org/help/articles/human-genome-automated-annotation&quot;&gt;Ensembl Human automated annotation system&lt;/a&gt;, supplemented by methods from the &#39;</span>  <span class="c1"># pylint: disable=line-too-long</span>
            <span class="sa">f</span><span class="s1">&#39;&lt;a href=&quot;https://beta.ensembl.org/help/articles/vertebrate-genome-annotation&quot;&gt;Ensembl vertebrate annotation pipeline&lt;/a&gt;. &#39;</span>  <span class="c1"># pylint: disable=line-too-long</span>
            <span class="sa">f</span><span class="s2">&quot;Mapped GENCODE structures served as the primary evidence with &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;gaps in the annotations filled using aligned short-read &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;transcriptomic data and full-length transcripts derived from PacBio IsoSeq &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;long-read data.&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;&lt;/p&gt;&lt;p&gt;In accordance with the &quot;</span>
            <span class="sa">f</span><span class="s1">&#39;&lt;a href=&quot;https://en.wikipedia.org/wiki/Fort_Lauderdale_Agreement&quot;&gt;Fort Lauderdale &#39;</span>
            <span class="sa">f</span><span class="s2">&quot;Agreement&lt;/a&gt;, please check the publication &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;status of the genome/assembly before publishing any genome-wide analyses using &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;these data.&lt;/p&gt;&quot;</span><span class="p">,</span>
            <span class="n">file</span><span class="o">=</span><span class="n">annotation_out</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="info-from-registry">Info From Registry<a class="headerlink" href="#info-from-registry" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.assign_species_prefix" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.assign_species_prefix</code>


<a href="#ensembl.genes.info_from_registry.assign_species_prefix" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.assign_species_prefix.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.assign_species_prefix.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_species_prefix.create_prefix" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">create_prefix</span><span class="p">(</span><span class="n">existing_prefix</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_species_prefix.create_prefix" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create a new species prefix for a given taxon ID and insert it into the database.
It will retry up to 10,000 times to ensure uniqueness.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>existing_prefix</code>
            </td>
            <td>
                  <code><span title="list">list</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of existing species prefixes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>taxon_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The taxon ID for the new prefix.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Information about the database server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="RuntimeError">RuntimeError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If a unique prefix cannot be generated after many attempts.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The newly created species prefix.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_species_prefix.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_prefix</span><span class="p">(</span><span class="n">existing_prefix</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a new species prefix for a given taxon ID and insert it into the database.</span>
<span class="sd">    It will retry up to 10,000 times to ensure uniqueness.</span>

<span class="sd">    Args:</span>
<span class="sd">        existing_prefix (list[str]): A list of existing species prefixes.</span>
<span class="sd">        taxon_id (int): The taxon ID for the new prefix.</span>
<span class="sd">        server_info (dict): Information about the database server.</span>

<span class="sd">    Raises:</span>
<span class="sd">        RuntimeError: If a unique prefix cannot be generated after many attempts.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The newly created species prefix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Creating new prefix for taxon ID: </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
        <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user_w&quot;</span><span class="p">],</span>
        <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">password</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
        <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="k">with</span> <span class="n">conn</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>  <span class="c1"># max attempts</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">generate_random_prefix</span><span class="p">(</span><span class="n">existing_prefix</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">insert_prefix_into_db</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">,</span> <span class="n">conn</span><span class="p">):</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Successfully inserted: </span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">prefix</span>
            <span class="n">existing_prefix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span>  <span class="c1"># optimize by adding to &quot;existing prefixes&quot;</span>
    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;Failed to generate unique prefix after many attempts.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_species_prefix.exiting_prefix" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">exiting_prefix</span><span class="p">(</span><span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_species_prefix.exiting_prefix" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Get a list of existing species prefixes from the gb assembly registry and metadata databases.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Information about the database server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="list">list</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list[str]: A list of existing species prefixes.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_species_prefix.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">exiting_prefix</span><span class="p">(</span><span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get a list of existing species prefixes from the gb assembly registry and metadata databases.</span>

<span class="sd">    Args:</span>
<span class="sd">        server_info (dict): Information about the database server.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list[str]: A list of existing species prefixes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Getting existing prefix from registry db. To be removed when the registry is updated.</span>
    <span class="n">prefix_registry_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;SELECT DISTINCT species_prefix FROM assembly ;&quot;</span>
    <span class="n">output_registry</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
        <span class="n">prefix_registry_query</span><span class="p">,</span>
        <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
        <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># Getting existing prefix from metadata db</span>
    <span class="n">prefix_metadata_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;SELECT DISTINCT prefix FROM species_prefix ;&quot;</span>
    <span class="n">output_metadata</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
        <span class="n">prefix_metadata_query</span><span class="p">,</span>
        <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
        <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">prefix_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="nb">list</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">output_registry</span><span class="p">)</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">output_metadata</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="n">existing_prefix</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">prefix_list</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Num Existing prefix: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">existing_prefix</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">existing_prefix</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_species_prefix.generate_random_prefix" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">generate_random_prefix</span><span class="p">(</span><span class="n">existing_prefix</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_species_prefix.generate_random_prefix" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate a random species prefix that does not already exist in the provided list.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>existing_prefix</code>
            </td>
            <td>
                  <code><span title="list">list</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of existing species prefixes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A random three/four letter prefix.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_species_prefix.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_random_prefix</span><span class="p">(</span><span class="n">existing_prefix</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate a random species prefix that does not already exist in the provided list.</span>

<span class="sd">    Args:</span>
<span class="sd">        existing_prefix (list[str]): A list of existing species prefixes.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: A random three/four letter prefix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">letters</span> <span class="o">=</span> <span class="n">string</span><span class="o">.</span><span class="n">ascii_uppercase</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">existing_prefix</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">26</span><span class="o">**</span><span class="mi">3</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating prefix: four random letters&quot;</span><span class="p">)</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating prefix: three random letters&quot;</span><span class="p">)</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">3</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">candidate</span> <span class="o">=</span> <span class="s2">&quot;ENS&quot;</span> <span class="o">+</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">letters</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">length</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">candidate</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">existing_prefix</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">candidate</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_species_prefix.get_special_cases" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_special_cases</span><span class="p">()</span></code>

<a href="#ensembl.genes.info_from_registry.assign_species_prefix.get_special_cases" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Retrieve special cases for species prefixes from a JSON file.
Returns:
    dict: A dictionary mapping taxon IDs to their special species prefixes.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_species_prefix.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_special_cases</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Retrieve special cases for species prefixes from a JSON file.</span>
<span class="sd">    Returns:</span>
<span class="sd">        dict: A dictionary mapping taxon IDs to their special species prefixes.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">json_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">),</span>
        <span class="s2">&quot;ensembl-genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;src&quot;</span><span class="p">,</span>
        <span class="s2">&quot;python&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ensembl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;info_from_registry&quot;</span><span class="p">,</span>
        <span class="s2">&quot;anno_settings.json&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">special_cases</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">special_cases</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_species_prefix.get_species_prefix" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_species_prefix</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_species_prefix.get_species_prefix" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>This function retrieves the species prefix from the assembly registry and metadata databases.
If the prefix is not found, it creates a new one. There are special cases where the prefix is predefined.
- Canis lupus (wolf) -&gt; ENSCAF
- Canis lupus familiaris (Domestic dog) -&gt; ENSCAF
- Heterocephalus glaber (naked mole rat) -&gt; ENSHGL</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>taxon_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>lowest taxon id</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the taxon ID is not found.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>unique species prefix that already exist or a new one when no prefix is found.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_species_prefix.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_species_prefix</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function retrieves the species prefix from the assembly registry and metadata databases.</span>
<span class="sd">    If the prefix is not found, it creates a new one. There are special cases where the prefix is predefined.</span>
<span class="sd">    - Canis lupus (wolf) -&gt; ENSCAF</span>
<span class="sd">    - Canis lupus familiaris (Domestic dog) -&gt; ENSCAF</span>
<span class="sd">    - Heterocephalus glaber (naked mole rat) -&gt; ENSHGL</span>

<span class="sd">    Args:</span>
<span class="sd">        taxon_id (int): lowest taxon id</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the taxon ID is not found.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: unique species prefix that already exist or a new one when no prefix is found.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Special cases</span>
    <span class="n">special_cases</span> <span class="o">=</span> <span class="n">get_special_cases</span><span class="p">()</span>

    <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">)</span> <span class="ow">in</span> <span class="n">special_cases</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The prefix is a special case for taxon ID: </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">species_prefix</span> <span class="o">=</span> <span class="n">special_cases</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Searching for prefix for taxon ID: </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">prefix_registry_query</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;SELECT DISTINCT species_prefix FROM assembly WHERE taxonomy = </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">output_registry</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">prefix_registry_query</span><span class="p">,</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">database</span><span class="o">=</span><span class="s2">&quot;gb_assembly_registry&quot;</span><span class="p">,</span>
            <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">output_registry</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Prefix found in old registry: </span><span class="si">{</span><span class="n">output_registry</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">prefix_metadata_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;SELECT DISTINCT prefix FROM species_prefix WHERE lowest_taxon_id = </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">output_metadata</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">prefix_metadata_query</span><span class="p">,</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
            <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">output_metadata</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Prefix found in new metadata registry: </span><span class="si">{</span><span class="n">output_metadata</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Combine output and get list of unique values</span>
        <span class="n">output</span> <span class="o">=</span> <span class="p">[</span>
            <span class="nb">list</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">output_registry</span><span class="p">)</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">output_metadata</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="n">prefix_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">output</span><span class="p">))</span>

        <span class="c1"># no prefix, create new prefix</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">prefix_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Creating a new prefix for taxon ID: </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">existing_prefix</span> <span class="o">=</span> <span class="n">exiting_prefix</span><span class="p">(</span><span class="n">server_info</span><span class="p">)</span>
            <span class="n">species_prefix</span> <span class="o">=</span> <span class="n">create_prefix</span><span class="p">(</span><span class="n">existing_prefix</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span>

        <span class="c1"># unique prefix detected</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">prefix_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unique prefix detected for taxon ID: </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">species_prefix</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">prefix_list</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saving exting prefix </span><span class="si">{</span><span class="n">species_prefix</span><span class="si">}</span><span class="s2"> in new registry&quot;</span><span class="p">)</span>
            <span class="c1"># Open a new connection</span>
            <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
                <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
                <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user_w&quot;</span><span class="p">],</span>
                <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
                <span class="n">password</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">insert_prefix_into_db</span><span class="p">(</span>
                <span class="n">species_prefix</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">,</span> <span class="n">conn</span><span class="p">,</span> <span class="n">store_new_registry</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">):</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Successfully inserted: </span><span class="si">{</span><span class="n">species_prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;The taxon </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2"> is already registered and multiple prefix were detected: </span><span class="si">{</span><span class="n">prefix_list</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="n">species_prefix</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_species_prefix.insert_prefix_into_db" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">insert_prefix_into_db</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">,</span> <span class="n">conn</span><span class="p">,</span> <span class="n">store_new_registry</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_species_prefix.insert_prefix_into_db" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Insert a new species prefix into the database.
Args:
    prefix (str): The species prefix to insert.
    taxon_id (int): The lowest taxon ID associated with the prefix.
    conn (pymysql.connections.Connection): The database connection object.
    store_new_registry (bool): Whether to allow duplicated exception if the prefix already exists in the database.
Returns:
    bool: True if the prefix was successfully inserted, False if it already exists.
Raises:
    pymysql.err.IntegrityError: If there is a duplicate entry for the prefix.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_species_prefix.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">insert_prefix_into_db</span><span class="p">(</span>
    <span class="n">prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">conn</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span>
    <span class="n">store_new_registry</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Insert a new species prefix into the database.</span>
<span class="sd">    Args:</span>
<span class="sd">        prefix (str): The species prefix to insert.</span>
<span class="sd">        taxon_id (int): The lowest taxon ID associated with the prefix.</span>
<span class="sd">        conn (pymysql.connections.Connection): The database connection object.</span>
<span class="sd">        store_new_registry (bool): Whether to allow duplicated exception if the prefix already exists in the database.</span>
<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the prefix was successfully inserted, False if it already exists.</span>
<span class="sd">    Raises:</span>
<span class="sd">        pymysql.err.IntegrityError: If there is a duplicate entry for the prefix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Inserting prefix </span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2"> for taxon id </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">query</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;INSERT INTO species_prefix (lowest_taxon_id, prefix) &quot;</span>
                <span class="s2">&quot;VALUES (</span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">);&quot;</span>
            <span class="p">)</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">prefix</span><span class="p">))</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">err</span><span class="o">.</span><span class="n">IntegrityError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">e</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1062</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">store_new_registry</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Existing prefix </span><span class="si">{</span><span class="n">prefix</span><span class="si">}</span><span class="s2"> already stored in the new registry. Allow exception&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="kc">True</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>
        <span class="k">raise</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.assign_stable_space" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.assign_stable_space</code>


<a href="#ensembl.genes.info_from_registry.assign_stable_space" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.assign_stable_space.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.assign_stable_space.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_stable_space.assign_stable_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">assign_stable_id</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">gca_accession</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_stable_space.assign_stable_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Assign a stable space ID for a given taxon and GCA accession.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>taxon_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The taxon identifier for which to assign a stable space ID.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gca_accession</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The GCA accession number for the assembly.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The assembly identifier.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The server information for database connection.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="tuple">tuple</span>[<span title="bool">bool</span>, <span title="int">int</span>, <span title="int">int</span>] | <span title="tuple">tuple</span>[<span title="bool">bool</span>, None, None]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>tuple[bool, int|None, int|None]: A tuple containing a success flag and the assigned stable space ID or None if failed.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_stable_space.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">assign_stable_id</span><span class="p">(</span>
    <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">gca_accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">bool</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">bool</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Assign a stable space ID for a given taxon and GCA accession.</span>

<span class="sd">    Args:</span>
<span class="sd">        taxon_id (int): The taxon identifier for which to assign a stable space ID.</span>
<span class="sd">        gca_accession (str): The GCA accession number for the assembly.</span>
<span class="sd">        assembly_id (int): The assembly identifier.</span>
<span class="sd">        server_info (dict): The server information for database connection.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple[bool, int|None, int|None]: A tuple containing a success flag and the assigned stable space ID or None if failed.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">stable_space_id</span> <span class="o">=</span> <span class="n">stable_space_per_taxon</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span>
    <span class="c1"># Check if stable space range exists</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Check if stable space range exists </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">stable_space_start</span> <span class="o">=</span> <span class="n">stable_space_range</span><span class="p">(</span><span class="n">stable_space_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">stable_space_start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">False</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Assigned stable space ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2"> for </span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2"> and taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span>

        <span class="n">insert_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;INSERT INTO stable_space_species_log (stable_space_id, lowest_taxon_id, gca_accession, assembly_id) </span>
<span class="s2">            VALUES (</span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">, &#39;</span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2">&#39;, </span><span class="si">{</span><span class="n">assembly_id</span><span class="si">}</span><span class="s2">);</span>
<span class="s2">            &quot;&quot;&quot;</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user_w&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">password</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
            <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">insert_to_db</span><span class="p">(</span><span class="n">insert_query</span><span class="p">,</span> <span class="n">conn</span><span class="p">,</span> <span class="n">store_new_registry</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Successfully inserted stable space ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2"> for GCA </span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
            <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="k">return</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stable_space_id</span><span class="p">,</span> <span class="n">stable_space_start</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Failed to insert stable space ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2"> for GCA </span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Failed to assign stable space ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2"> for GCA </span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_stable_space.get_stable_space" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_stable_space</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">gca_accession</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_stable_space.get_stable_space" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Get the stable space ID for a given taxon and GCA accession.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>taxon_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The taxon identifier for which to find the stable space ID.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gca_accession</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The GCA accession number for the assembly.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The assembly identifier.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The server information for database connection.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>assigned stable space ID or existing one if already assigned.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_stable_space.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_stable_space</span><span class="p">(</span>
    <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">gca_accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the stable space ID for a given taxon and GCA accession.</span>

<span class="sd">    Args:</span>
<span class="sd">        taxon_id (int): The taxon identifier for which to find the stable space ID.</span>
<span class="sd">        gca_accession (str): The GCA accession number for the assembly.</span>
<span class="sd">        assembly_id (int): The assembly identifier.</span>
<span class="sd">        server_info (dict): The server information for database connection.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: assigned stable space ID or existing one if already assigned.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Check if GCA already has assigned a stable space</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Check if </span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2"> already has assigned a stable space&quot;</span><span class="p">)</span>
    <span class="n">space_gca_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;SELECT stable_space_species_log.stable_space_id, stable_space.stable_space_start FROM stable_space_species_log JOIN stable_space ON stable_space.stable_space_id = stable_space_species_log.stable_space_id  WHERE stable_space_species_log.gca_accession = &#39;</span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2">&#39;;&quot;</span>
    <span class="n">output_query</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
        <span class="n">space_gca_query</span><span class="p">,</span>
        <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
        <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">output_query</span><span class="p">:</span>
        <span class="n">stable_space_id</span> <span class="o">=</span> <span class="n">output_query</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;stable_space_id&quot;</span><span class="p">)</span>
        <span class="n">stable_space_start</span> <span class="o">=</span> <span class="n">output_query</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;stable_space_start&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Stable space </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2"> already assigned for GCA </span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">stable_space_start</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;No stable space found for assembly </span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2">. Checking taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2"> for existing stable space.&quot;</span>
        <span class="p">)</span>

        <span class="n">success</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">while</span> <span class="ow">not</span> <span class="n">success</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Assigning stable space for taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">gca_accession</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
            <span class="n">success</span><span class="p">,</span> <span class="n">stable_space_id</span><span class="p">,</span> <span class="n">stable_space_start</span> <span class="o">=</span> <span class="n">assign_stable_id</span><span class="p">(</span>
                <span class="n">taxon_id</span><span class="p">,</span> <span class="n">gca_accession</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">server_info</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">stable_space_start</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_stable_space.insert_to_db" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">insert_to_db</span><span class="p">(</span><span class="n">insert_query</span><span class="p">,</span> <span class="n">conn</span><span class="p">,</span> <span class="n">store_new_registry</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_stable_space.insert_to_db" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_stable_space.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">insert_to_db</span><span class="p">(</span>
    <span class="n">insert_query</span><span class="p">,</span> <span class="n">conn</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span> <span class="n">store_new_registry</span><span class="p">:</span> <span class="nb">bool</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Inserting data: </span><span class="si">{</span><span class="n">insert_query</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">insert_query</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">err</span><span class="o">.</span><span class="n">IntegrityError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">e</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1062</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">store_new_registry</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Duplicate entry found, but store_new_registry is True. Report as success.&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="kc">True</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>
        <span class="k">raise</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_stable_space.stable_space_per_taxon" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">stable_space_per_taxon</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_stable_space.stable_space_per_taxon" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Get the next stable space ID for a given taxon.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>taxon_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The taxon identifier for which to find the stable space ID.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The server information for database connection.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The next available stable space ID.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_stable_space.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span>
<span class="normal">99</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">stable_space_per_taxon</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the next stable space ID for a given taxon.</span>

<span class="sd">    Args:</span>
<span class="sd">        taxon_id (int): The taxon identifier for which to find the stable space ID.</span>
<span class="sd">        server_info (dict): The server information for database connection.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: The next available stable space ID.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Get the next stable space ID for taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">space_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;SELECT MAX(stable_space_id) as max_stable_id FROM stable_space_species_log WHERE lowest_taxon_id = </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">;&quot;</span>
    <span class="n">output_query</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
        <span class="n">space_query</span><span class="p">,</span>
        <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
        <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">stable_space_tmp</span> <span class="o">=</span> <span class="n">output_query</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;max_stable_id&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

    <span class="c1">####################</span>
    <span class="c1"># remove when all pipelines have moved to the new registry:</span>
    <span class="n">old_stable_space_info</span> <span class="o">=</span> <span class="n">get_old_stable_space_info</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span>
    <span class="c1">####################</span>

    <span class="k">if</span> <span class="n">old_stable_space_info</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">stable_space_tmp</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">stable_space_id</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;No stable space found for taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">. &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;Assigning new stable space ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">stable_space_id</span> <span class="o">=</span> <span class="n">stable_space_tmp</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;No old stable space for taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">. &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;Using tmp value </span><span class="si">{</span><span class="n">stable_space_tmp</span><span class="si">}</span><span class="s2">, assigning new ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>  <span class="c1"># old_stable_space_info is not None</span>
        <span class="k">if</span> <span class="n">stable_space_tmp</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">stable_space_id</span> <span class="o">=</span> <span class="n">old_stable_space_info</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Found old stable space </span><span class="si">{</span><span class="n">old_stable_space_info</span><span class="si">}</span><span class="s2"> for taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">. &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;Assigning new ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">stable_space_tmp</span> <span class="o">==</span> <span class="n">old_stable_space_info</span><span class="p">:</span>
            <span class="n">stable_space_id</span> <span class="o">=</span> <span class="n">stable_space_tmp</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Stable space </span><span class="si">{</span><span class="n">stable_space_tmp</span><span class="si">}</span><span class="s2"> matches old record for taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">. &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;Assigning new ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Conflicting stable space for taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">: &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">old_stable_space_info</span><span class="si">}</span><span class="s2"> != </span><span class="si">{</span><span class="n">stable_space_tmp</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">stable_space_id</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.assign_stable_space.stable_space_range" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">stable_space_range</span><span class="p">(</span><span class="n">stable_space_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.assign_stable_space.stable_space_range" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Check if a stable space range exists for the given stable space ID.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>stable_space_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The stable space ID to check.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The server information for database connection.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the stable space range exists, False otherwise.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/assign_stable_space.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">stable_space_range</span><span class="p">(</span><span class="n">stable_space_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if a stable space range exists for the given stable space ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        stable_space_id (int): The stable space ID to check.</span>
<span class="sd">        server_info (dict): The server information for database connection.</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the stable space range exists, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Get stable space renage for </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;SELECT * FROM stable_space WHERE stable_space_id = &#39;</span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">&#39;;&quot;</span>
    <span class="n">output_query</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
        <span class="n">query</span><span class="p">,</span>
        <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
        <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">output_query</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Stable space ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2"> already exists with range: </span><span class="si">{</span><span class="n">output_query</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;No existing stable space range found for ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">. Creating a new range space.&quot;</span>
        <span class="p">)</span>
        <span class="n">previous_space_id</span> <span class="o">=</span> <span class="n">stable_space_id</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;SELECT * FROM stable_space WHERE stable_space_id = &#39;</span><span class="si">{</span><span class="n">previous_space_id</span><span class="si">}</span><span class="s2">&#39;;&quot;&quot;&quot;</span>
        <span class="n">output_query</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">query</span><span class="p">,</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
            <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Output query for previous stable space ID </span><span class="si">{</span><span class="n">previous_space_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">output_query</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">output_query</span><span class="p">:</span>
            <span class="n">new_start</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">output_query</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;stable_space_end&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="p">)</span>  <span class="c1"># Increment the end of the previous stable space by 1</span>
            <span class="n">new_end</span> <span class="o">=</span> <span class="n">new_start</span> <span class="o">+</span> <span class="mi">4999999</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;New stable space range for ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2"> will be from </span><span class="si">{</span><span class="n">new_start</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">new_end</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>

            <span class="n">insert_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;INSERT INTO stable_space (stable_space_id, stable_space_start, stable_space_end) VALUES (</span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">new_start</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">new_end</span><span class="si">}</span><span class="s2">);&quot;</span>
            <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
                <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
                <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user_w&quot;</span><span class="p">],</span>
                <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
                <span class="n">password</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">insert_to_db</span><span class="p">(</span><span class="n">insert_query</span><span class="p">,</span> <span class="n">conn</span><span class="p">,</span> <span class="n">store_new_registry</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Successfully created stable space range for ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>
                <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
                <span class="k">return</span> <span class="n">new_start</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Failed to insert stable space range for ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="kc">False</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Failed to create stable space range for ID </span><span class="si">{</span><span class="n">stable_space_id</span><span class="si">}</span><span class="s2">. Previous space ID </span><span class="si">{</span><span class="n">previous_space_id</span><span class="si">}</span><span class="s2"> not found.&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.build_anno_commands" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.build_anno_commands</code>


<a href="#ensembl.genes.info_from_registry.build_anno_commands" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.build_anno_commands.build_annotation_commands" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">build_annotation_commands</span><span class="p">(</span><span class="n">core_adaptor</span><span class="p">,</span> <span class="n">output_params</span><span class="p">,</span> <span class="n">anno_settings</span><span class="p">,</span> <span class="n">settings</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.build_anno_commands.build_annotation_commands" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Construct and store command-line strings for genome annotation and repeat analysis.</p>
<p>This function builds two command-line strings based on the provided configuration:
- <code>anno_commandline</code>: full annotation pipeline
- <code>anno_red_commandline</code>: repeat and masking pipeline</p>
<p>The resulting command strings are stored in the <code>output_params</code> dictionary under
the keys "anno_commandline" and "anno_red_commandline".</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>core_adaptor</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Contains database connection info with keys:
'dbname', 'host', 'port', 'user', 'pass'.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_params</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary where input/output file paths and
directories are specified. The final command strings are added here.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>anno_settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Annotation settings such as number of threads and
diamond validation DB path.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>General pipeline settings, including repeat library usage.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>None</code></td>            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Modifies <code>output_params</code> in place by adding two command strings.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/build_anno_commands.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">build_annotation_commands</span><span class="p">(</span>
    <span class="n">core_adaptor</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">output_params</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">anno_settings</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">settings</span><span class="p">:</span> <span class="nb">dict</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Construct and store command-line strings for genome annotation and repeat analysis.</span>

<span class="sd">    This function builds two command-line strings based on the provided configuration:</span>
<span class="sd">    - `anno_commandline`: full annotation pipeline</span>
<span class="sd">    - `anno_red_commandline`: repeat and masking pipeline</span>

<span class="sd">    The resulting command strings are stored in the `output_params` dictionary under</span>
<span class="sd">    the keys &quot;anno_commandline&quot; and &quot;anno_red_commandline&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        core_adaptor (dict): Contains database connection info with keys:</span>
<span class="sd">            &#39;dbname&#39;, &#39;host&#39;, &#39;port&#39;, &#39;user&#39;, &#39;pass&#39;.</span>
<span class="sd">        output_params (dict): Dictionary where input/output file paths and</span>
<span class="sd">            directories are specified. The final command strings are added here.</span>
<span class="sd">        anno_settings (dict): Annotation settings such as number of threads and</span>
<span class="sd">            diamond validation DB path.</span>
<span class="sd">        settings (dict): General pipeline settings, including repeat library usage.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: Modifies `output_params` in place by adding two command strings.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">get</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="n">output_params</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>  <span class="c1"># Short helper</span>

    <span class="n">anno_commandline</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot; --genome_file </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;reheadered_toplevel_genome_file&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --db_details </span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;dbname&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">,</span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;host&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">,</span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;port&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">,</span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;user&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">,</span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;pass&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --output_dir </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;output_path&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --short_read_fastq_dir </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;short_read_dir&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --long_read_fastq_dir </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;long_read_dir&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --max_intron_length </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;max_intron_length&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --protein_file </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;protein_file&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --busco_protein_file </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;busco_protein_file&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --rfam_accessions_file </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;rfam_accessions_file&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --num_threads </span><span class="si">{</span><span class="n">anno_settings</span><span class="p">[</span><span class="s1">&#39;num_threads&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;use_existing_repeatmodeler_library&quot;</span><span class="p">):</span>
        <span class="n">anno_commandline</span> <span class="o">+=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot; --repeatmasker_library </span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;use_existing_repeatmodeler_library&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">anno_settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;diamond_validation_db&quot;</span><span class="p">):</span>
        <span class="n">anno_commandline</span> <span class="o">+=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot; --diamond_validation_db </span><span class="si">{</span><span class="n">anno_settings</span><span class="p">[</span><span class="s1">&#39;diamond_validation_db&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">anno_settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;validation_type&quot;</span><span class="p">):</span>
        <span class="n">anno_commandline</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot; --validation_type </span><span class="si">{</span><span class="n">anno_settings</span><span class="p">[</span><span class="s1">&#39;validation_type&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="n">anno_commandline</span> <span class="o">+=</span> <span class="s2">&quot; --run_full_annotation --load_to_ensembl_db&quot;</span>

    <span class="n">anno_red_commandline</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot; --genome_file </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;reheadered_toplevel_genome_file&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --db_details </span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;dbname&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">,</span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;host&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">,</span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;port&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">,</span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;user&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">,</span><span class="si">{</span><span class="n">core_adaptor</span><span class="p">[</span><span class="s1">&#39;pass&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --output_dir </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;output_path&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --num_threads </span><span class="si">{</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;num_threads&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="s2">&quot; --run_masking --run_repeats --run_simple_features --load_to_ensembl_db&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Add both commands back into the params dict</span>
    <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;anno_commandline&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">anno_commandline</span>
    <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;anno_red_commandline&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">anno_red_commandline</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.check_if_annotated" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.check_if_annotated</code>


<a href="#ensembl.genes.info_from_registry.check_if_annotated" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>check_if_annotated.py</p>
<p>This module defines a function to verify if a given genome assembly accession
is already annotated in the genebuild registry. It logs annotation status and raises
an error if any assembly has already been annotated.</p>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.check_if_annotated.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.check_if_annotated.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">check_if_annotated</span><span class="p">(</span><span class="n">assembly_accession</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Check if a genome assembly is already annotated in the genebuild registry.</p>
<h5 id="ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated--parameters">Parameters:<a class="headerlink" href="#ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated--parameters" title="Permanent link">&para;</a></h5>
<p>assembly_accession : str
    The GCA accession of the assembly to check (e.g., "GCA_000001405.28").</p>


<details class="server_info-" open>
  <summary>dict</summary>
  <p>Dictionary containing server connection details, with a 'registry' key
pointing to another dict with the following keys:
    - db_host : str
    - db_user : str
    - db_port : int
    - db_name : str</p>
</details>        <h5 id="ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated--raises">Raises:<a class="headerlink" href="#ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated--raises" title="Permanent link">&para;</a></h5>
<p>RuntimeError
    If the given assembly accession is already present in the registry.</p>
<h5 id="ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated--returns">Returns:<a class="headerlink" href="#ensembl.genes.info_from_registry.check_if_annotated.check_if_annotated--returns" title="Permanent link">&para;</a></h5>
<p>None</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/check_if_annotated.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_if_annotated</span><span class="p">(</span>
    <span class="n">assembly_accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">server_info</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if a genome assembly is already annotated in the genebuild registry.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    ----------</span>
<span class="sd">    assembly_accession : str</span>
<span class="sd">        The GCA accession of the assembly to check (e.g., &quot;GCA_000001405.28&quot;).</span>

<span class="sd">    server_info : dict</span>
<span class="sd">        Dictionary containing server connection details, with a &#39;registry&#39; key</span>
<span class="sd">        pointing to another dict with the following keys:</span>
<span class="sd">            - db_host : str</span>
<span class="sd">            - db_user : str</span>
<span class="sd">            - db_port : int</span>
<span class="sd">            - db_name : str</span>

<span class="sd">    Raises:</span>
<span class="sd">    ------</span>
<span class="sd">    RuntimeError</span>
<span class="sd">        If the given assembly accession is already present in the registry.</span>

<span class="sd">    Returns:</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">registry_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT </span>
<span class="s2">            gca_accession, </span>
<span class="s2">            gb_status, </span>
<span class="s2">            genebuilder</span>
<span class="s2">        FROM genebuild_status </span>
<span class="s2">        WHERE gca_accession =  %s</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="n">registry_rows</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
        <span class="n">registry_query</span><span class="p">,</span>
        <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
        <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">params</span><span class="o">=</span><span class="p">(</span><span class="n">assembly_accession</span><span class="p">,),</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">registry_rows</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">registry_rows</span><span class="p">:</span>
            <span class="n">gca</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;gca_accession&quot;</span><span class="p">)</span>
            <span class="n">gb_status</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;gb_status&quot;</span><span class="p">)</span>
            <span class="n">genebuilder</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;genebuilder&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"> Annotation already exists for GCA: </span><span class="si">{</span><span class="n">gca</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   Status     : </span><span class="si">{</span><span class="n">gb_status</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;   Genebuilder: </span><span class="si">{</span><span class="n">genebuilder</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
            <span class="s2">&quot;Terminating: One or more assemblies are already annotated. Please set current_genebuild to 1 in user_settings.json if you wish to proceed.&quot;</span>
        <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start check complete: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">)</span>
    <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.check_stable_space_old_registry" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.check_stable_space_old_registry</code>


<a href="#ensembl.genes.info_from_registry.check_stable_space_old_registry" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>check_stable_space_from_old_registry.py</p>
<p>This module gets legacy stable id from old registry. Module will be deprecated when
all pipelines have moved to the new registry. It's main functions is to facilitate
using the new registry to start anno runs.</p>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.check_stable_space_old_registry.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.check_stable_space_old_registry.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.check_stable_space_old_registry.get_old_stable_space_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_old_stable_space_info</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.check_stable_space_old_registry.get_old_stable_space_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Get the stable space that has been assigned to old registry.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>taxon_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The taxon identifier for which to find the stable space ID.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The server information for database connection.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="int">int</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assigned stable space.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/check_stable_space_old_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_old_stable_space_info</span><span class="p">(</span>
    <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">server_info</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the stable space that has been assigned to old registry.</span>

<span class="sd">    Args:</span>
<span class="sd">        taxon_id (int): The taxon identifier for which to find the stable space ID.</span>
<span class="sd">        server_info (dict): The server information for database connection.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: Assigned stable space.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Check if taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2"> has stable space in old registry&quot;</span><span class="p">)</span>
    <span class="n">space_query</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;SELECT MAX(stable_id_space_id) as old_stable_id &quot;</span>
        <span class="sa">f</span><span class="s2">&quot;FROM assembly &quot;</span>
        <span class="sa">f</span><span class="s2">&quot;WHERE taxonomy = </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">;&quot;</span>
    <span class="p">)</span>

    <span class="n">output_query</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
        <span class="n">space_query</span><span class="p">,</span>
        <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">database</span><span class="o">=</span><span class="s2">&quot;gb_assembly_registry&quot;</span><span class="p">,</span>
        <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">old_stable_id</span> <span class="o">=</span> <span class="n">output_query</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;max_stable_id&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">old_stable_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable space not found in old registry for taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Max stable space in the old registry for taxon ID </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2"> is </span><span class="si">{</span><span class="n">old_stable_id</span><span class="si">}</span><span class="s2">.&quot;</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">old_stable_id</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.create_config" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.create_config</code>


<a href="#ensembl.genes.info_from_registry.create_config" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>create_config.py</p>
<p>This module handles copying and editing Ensembl Hive pipeline configuration
files based on runtime settings. It supports substituting key pipeline
parameters like user credentials, pipeline name, and release version.</p>
<p>Functions:
- copy_config: Copies the specified pipeline config file to the working directory.
- edit_config: Edits configuration values in the copied file using the provided settings.</p>


<details class="example-usage" open>
  <summary>Example usage</summary>
  <p>settings = {
    "config": "MyConfig.pm",
    "base_output_dir": "/my/output/path",
    "current_genebuild": "my_genebuild_104",
    "dbowner": "my_db_user",
    "pipeline_name": "my_pipeline",
    "password": "secure_pw",
    "user": "readonly_user",
    "user_r": "reader_user",
    "release_number": "104"
}
edit_config(settings)</p>
</details>









<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.create_config.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.create_config.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.create_config.copy_config" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">copy_config</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">,</span> <span class="n">pipeline</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.create_config.copy_config" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Copies a configuration file from the ENSCODE environment into a local output directory.</p>
<h5 id="ensembl.genes.info_from_registry.create_config.copy_config--parameters">Parameters:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_config.copy_config--parameters" title="Permanent link">&para;</a></h5>
<p>settings : dict
    Dictionary that must contain:
        - 'config': str, filename of the config (e.g., 'MyConfig.pm')
        - 'base_output_dir': str, path where the config will be copied</p>
<h5 id="ensembl.genes.info_from_registry.create_config.copy_config--returns">Returns:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_config.copy_config--returns" title="Permanent link">&para;</a></h5>
<p>str
    Path to the copied local config file.</p>
<h5 id="ensembl.genes.info_from_registry.create_config.copy_config--raises">Raises:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_config.copy_config--raises" title="Permanent link">&para;</a></h5>
<p>RuntimeError
    If the file cannot be copied for any reason (e.g., file not found, permissions).</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/create_config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">copy_config</span><span class="p">(</span>
    <span class="n">settings</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="n">info_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="n">pipeline</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Copies a configuration file from the ENSCODE environment into a local output directory.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    ----------</span>
<span class="sd">    settings : dict</span>
<span class="sd">        Dictionary that must contain:</span>
<span class="sd">            - &#39;config&#39;: str, filename of the config (e.g., &#39;MyConfig.pm&#39;)</span>
<span class="sd">            - &#39;base_output_dir&#39;: str, path where the config will be copied</span>

<span class="sd">    Returns:</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Path to the copied local config file.</span>

<span class="sd">    Raises:</span>
<span class="sd">    ------</span>
<span class="sd">    RuntimeError</span>
<span class="sd">        If the file cannot be copied for any reason (e.g., file not found, permissions).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">original_config</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">),</span>
        <span class="s2">&quot;ensembl-analysis&quot;</span><span class="p">,</span>
        <span class="s2">&quot;modules&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Bio&quot;</span><span class="p">,</span>
        <span class="s2">&quot;EnsEMBL&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Analysis&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Hive&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Config&quot;</span><span class="p">,</span>
        <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;config&quot;</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">pipeline</span> <span class="o">==</span> <span class="s2">&quot;anno&quot;</span><span class="p">:</span>
        <span class="n">anno_parent</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;output_path&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">parent</span><span class="p">)</span>
        <span class="n">local_config</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">anno_parent</span><span class="p">,</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;config&quot;</span><span class="p">])</span>
    <span class="k">elif</span> <span class="n">pipeline</span> <span class="o">==</span> <span class="s2">&quot;main&quot;</span><span class="p">:</span>
        <span class="n">local_config</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;output_path&quot;</span><span class="p">],</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;config&quot;</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Unknown pipeline type: </span><span class="si">{</span><span class="n">pipeline</span><span class="si">}</span><span class="s2">. Can&#39;t copy config. Please check taxon ID.&quot;</span>
        <span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">original_config</span><span class="p">,</span> <span class="n">local_config</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Copied file from </span><span class="si">{</span><span class="n">original_config</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">local_config</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error copying file: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">local_config</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.create_config.edit_config_anno" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">edit_config_anno</span><span class="p">(</span><span class="n">anno_settings</span><span class="p">,</span> <span class="n">settings</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">,</span> <span class="n">pipeline</span><span class="p">,</span> <span class="n">server_settings</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.create_config.edit_config_anno" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Edits specific parameter values in a copied Ensembl Hive config file.</p>
<p>The function performs in-place substitution of key parameters such as:
- current_genebuild
- dbowner
- pipeline_name
- password
- user
- user_r
- release_number</p>
<h5 id="ensembl.genes.info_from_registry.create_config.edit_config_anno--parameters">Parameters:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_config.edit_config_anno--parameters" title="Permanent link">&para;</a></h5>
<p>anno_settings: dict
settings : dict
    Dictionary with required keys:
        - 'config': str, name of the config file (e.g., 'MyConfig.pm')
        - 'base_output_dir': str, destination folder for the config
        - 'current_genebuild': str
        - 'dbowner': str
        - 'pipeline_name': str
        - 'password': str
        - 'user': str
        - 'user_r': str
        - 'release_number': str or int
info_dict : dict
pipeline : str</p>
<h5 id="ensembl.genes.info_from_registry.create_config.edit_config_anno--raises">Raises:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_config.edit_config_anno--raises" title="Permanent link">&para;</a></h5>
<p>FileNotFoundError
    If the copied file cannot be opened.
KeyError
    If required settings keys are missing.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/create_config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">edit_config_anno</span><span class="p">(</span>
    <span class="n">anno_settings</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">settings</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">info_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">pipeline</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">server_settings</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Edits specific parameter values in a copied Ensembl Hive config file.</span>

<span class="sd">    The function performs in-place substitution of key parameters such as:</span>
<span class="sd">    - current_genebuild</span>
<span class="sd">    - dbowner</span>
<span class="sd">    - pipeline_name</span>
<span class="sd">    - password</span>
<span class="sd">    - user</span>
<span class="sd">    - user_r</span>
<span class="sd">    - release_number</span>

<span class="sd">    Parameters:</span>
<span class="sd">    ----------</span>
<span class="sd">    anno_settings: dict</span>
<span class="sd">    settings : dict</span>
<span class="sd">        Dictionary with required keys:</span>
<span class="sd">            - &#39;config&#39;: str, name of the config file (e.g., &#39;MyConfig.pm&#39;)</span>
<span class="sd">            - &#39;base_output_dir&#39;: str, destination folder for the config</span>
<span class="sd">            - &#39;current_genebuild&#39;: str</span>
<span class="sd">            - &#39;dbowner&#39;: str</span>
<span class="sd">            - &#39;pipeline_name&#39;: str</span>
<span class="sd">            - &#39;password&#39;: str</span>
<span class="sd">            - &#39;user&#39;: str</span>
<span class="sd">            - &#39;user_r&#39;: str</span>
<span class="sd">            - &#39;release_number&#39;: str or int</span>
<span class="sd">    info_dict : dict</span>
<span class="sd">    pipeline : str</span>

<span class="sd">    Raises:</span>
<span class="sd">    ------</span>
<span class="sd">    FileNotFoundError</span>
<span class="sd">        If the copied file cannot be opened.</span>
<span class="sd">    KeyError</span>
<span class="sd">        If required settings keys are missing.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">local_config</span> <span class="o">=</span> <span class="n">copy_config</span><span class="p">(</span><span class="n">anno_settings</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">,</span> <span class="n">pipeline</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">local_config</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">content</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

    <span class="c1"># Replace the lines</span>
    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;current_genebuild&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;current_genebuild&#39;            =&gt; &#39;</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;current_genebuild&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;dbowner&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;dbowner&#39;                      =&gt; &#39;</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;dbowner&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;pipeline_name&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;pipeline_name&#39;                =&gt; &#39;</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;pipeline_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;password&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;password&#39;                     =&gt; &#39;</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;password&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;user&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;user&#39;                     =&gt; &#39;</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;user&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;user_r&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;user_r&#39;                     =&gt; &#39;</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;user_r&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;release_number&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;release_number&#39;                     =&gt; &#39;</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;release_number&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;pipe_db_server&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;pipe_db_server&#39;                     =&gt; &#39;</span><span class="si">{</span><span class="n">server_settings</span><span class="p">[</span><span class="s1">&#39;pipeline_db&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;pipe_db_port&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;pipe_db_port&#39;                     =&gt; &#39;</span><span class="si">{</span><span class="n">server_settings</span><span class="p">[</span><span class="s1">&#39;pipeline_db&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;dna_db_server&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;dna_db_server&#39;                     =&gt; &#39;</span><span class="si">{</span><span class="n">server_settings</span><span class="p">[</span><span class="s1">&#39;core_db&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;dna_db_port&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;dna_db_port&#39;                     =&gt; &#39;</span><span class="si">{</span><span class="n">server_settings</span><span class="p">[</span><span class="s1">&#39;core_db&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">content</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;&#39;registry_file&#39;\s*=&gt;\s*[^,]+,&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;&#39;registry_file&#39;                     =&gt; &#39;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;registry_file&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,&quot;</span><span class="p">,</span>
        <span class="n">content</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">local_config</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.create_config.edit_config_main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">edit_config_main</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">,</span> <span class="n">pipeline</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.create_config.edit_config_main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Edits specific parameter values in a copied Ensembl Hive config file safely,
preserving comments and avoiding Perl syntax errors.</p>
<h5 id="ensembl.genes.info_from_registry.create_config.edit_config_main--parameters">Parameters<a class="headerlink" href="#ensembl.genes.info_from_registry.create_config.edit_config_main--parameters" title="Permanent link">&para;</a></h5>
<p>settings : dict
    Must contain 'config' and 'base_output_dir'.
info_dict : dict
    Keys and values to replace in the config. Keys with value None are skipped.
pipeline : str
    Type of pipeline, e.g., "anno" or "main".</p>
<h5 id="ensembl.genes.info_from_registry.create_config.edit_config_main--raises">Raises<a class="headerlink" href="#ensembl.genes.info_from_registry.create_config.edit_config_main--raises" title="Permanent link">&para;</a></h5>
<p>FileNotFoundError
    If the copied config file cannot be found.
ValueError
    If pipeline type is unknown.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/create_config.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">edit_config_main</span><span class="p">(</span>
    <span class="n">settings</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="n">info_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="n">pipeline</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Edits specific parameter values in a copied Ensembl Hive config file safely,</span>
<span class="sd">    preserving comments and avoiding Perl syntax errors.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    settings : dict</span>
<span class="sd">        Must contain &#39;config&#39; and &#39;base_output_dir&#39;.</span>
<span class="sd">    info_dict : dict</span>
<span class="sd">        Keys and values to replace in the config. Keys with value None are skipped.</span>
<span class="sd">    pipeline : str</span>
<span class="sd">        Type of pipeline, e.g., &quot;anno&quot; or &quot;main&quot;.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    FileNotFoundError</span>
<span class="sd">        If the copied config file cannot be found.</span>
<span class="sd">    ValueError</span>
<span class="sd">        If pipeline type is unknown.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">local_config</span> <span class="o">=</span> <span class="n">copy_config</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">,</span> <span class="n">pipeline</span><span class="p">)</span>

    <span class="n">stop_marker</span> <span class="o">=</span> <span class="s2">&quot;# No option below this mark should be modified&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">local_config</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

    <span class="n">new_lines</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">stop_modifying</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
        <span class="c1"># Replace the package declaration at the top</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;package Bio::EnsEMBL::Analysis::Hive::Config::&quot;</span><span class="p">):</span>
            <span class="n">line</span> <span class="o">=</span> <span class="s2">&quot;package Genome_annotation_conf;</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="n">stop_marker</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
            <span class="n">stop_modifying</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">stop_modifying</span><span class="p">:</span>
            <span class="n">updated</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">info_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="c1"># Skip keys with None value</span>
                <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="c1"># Pattern matches key =&gt; old_value, optionally with spaces and comments</span>
                <span class="n">pattern</span> <span class="o">=</span> <span class="sa">rf</span><span class="s2">&quot;^\s*</span><span class="si">{</span><span class="n">re</span><span class="o">.</span><span class="n">escape</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="si">}</span><span class="s2">\s*=&gt;\s*[^,]*,&quot;</span>

                <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">pattern</span><span class="p">,</span> <span class="n">line</span><span class="p">):</span>
                    <span class="c1"># Preserve comment if present</span>
                    <span class="k">if</span> <span class="s2">&quot;#&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
                        <span class="n">parts</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                        <span class="n">line_content</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                        <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span> <span class="o">+</span> <span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">line_content</span> <span class="o">=</span> <span class="n">line</span>
                        <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

                    <span class="c1"># Convert Python value to Perl representation</span>
                    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                        <span class="n">perl_value</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
                        <span class="n">perl_value</span> <span class="o">=</span> <span class="s2">&quot;1&quot;</span> <span class="k">if</span> <span class="n">value</span> <span class="k">else</span> <span class="s2">&quot;0&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">perl_value</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>

                    <span class="c1"># Replace only the value before the comma</span>
                    <span class="n">new_line</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;=&gt;\s*[^,]*,&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;=&gt; </span><span class="si">{</span><span class="n">perl_value</span><span class="si">}</span><span class="s2">,&quot;</span><span class="p">,</span> <span class="n">line_content</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">comment</span><span class="p">:</span>
                        <span class="n">new_line</span> <span class="o">=</span> <span class="n">new_line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">comment</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">new_line</span> <span class="o">=</span> <span class="n">new_line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

                    <span class="n">new_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>
                    <span class="n">updated</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">break</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">updated</span><span class="p">:</span>
                <span class="n">new_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">new_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

    <span class="c1"># Write back the updated config safely</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">local_config</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">new_lines</span><span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Config edited successfully: </span><span class="si">{</span><span class="n">local_config</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">local_config</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.create_pipe_reg" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.create_pipe_reg</code>


<a href="#ensembl.genes.info_from_registry.create_pipe_reg" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>create_pipe_reg.py</p>
<p>This module updates the Ensembl pipeline registry configuration and inserts
the appropriate database adaptor entries for use in a genebuild pipeline.</p>
<p>It performs two key tasks:
1. Updates references to the shared registry file within the pipeline's resource
   description to use a copied, local version.
2. Appends new database connection entries into the copied registry file for use
   by Hive pipeline workers.</p>
<h4 id="ensembl.genes.info_from_registry.create_pipe_reg--functions">Functions:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_pipe_reg--functions" title="Permanent link">&para;</a></h4>
<ul>
<li>update_registry_path_and_create_entry: Updates resource_description in pipeline DB.</li>
<li>create_registry_entry: Adds new DBAdaptor entries into the local registry file.</li>
</ul>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.create_pipe_reg.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.create_pipe_reg.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">create_registry_entry</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="n">server_info</span><span class="p">,</span> <span class="n">core_adaptor</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Updates the local registry file with new DBAdaptor connection details for a
core database, using the provided connection settings.</p>
<h5 id="ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry--parameters">Parameters:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry--parameters" title="Permanent link">&para;</a></h5>
<p>settings : dict
    Must contain:
        - 'base_output_dir': str, where the local Databases.pm file is placed.</p>


<details class="server_info-" open>
  <summary>dict</summary>
  <p>Used to pass to <code>update_registry_path_and_create_entry</code>.</p>
</details>

<details class="core_adaptor-" open>
  <summary>dict</summary>
  <p>Must contain:
    - 'host', 'port', 'dbname', 'user', 'pass', 'species', 'group'</p>
</details>        <h5 id="ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry--returns">Returns:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry--returns" title="Permanent link">&para;</a></h5>
<p>Path to the modified local registry file (Databases.pm)</p>
<h5 id="ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry--raises">Raises:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_pipe_reg.create_registry_entry--raises" title="Permanent link">&para;</a></h5>
<p>FileNotFoundError:
    If the registry file does not exist after copying.</p>


<details class="runtimeerror" open>
  <summary>RuntimeError</summary>
  <p>If there is an error overwriting the original registry file.</p>
</details>

            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/create_pipe_reg.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_registry_entry</span><span class="p">(</span>
    <span class="n">settings</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">server_info</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span>
    <span class="n">core_adaptor</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Path</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Updates the local registry file with new DBAdaptor connection details for a</span>
<span class="sd">    core database, using the provided connection settings.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    ----------</span>
<span class="sd">    settings : dict</span>
<span class="sd">        Must contain:</span>
<span class="sd">            - &#39;base_output_dir&#39;: str, where the local Databases.pm file is placed.</span>

<span class="sd">    server_info : dict</span>
<span class="sd">        Used to pass to `update_registry_path_and_create_entry`.</span>

<span class="sd">    core_adaptor : dict</span>
<span class="sd">        Must contain:</span>
<span class="sd">            - &#39;host&#39;, &#39;port&#39;, &#39;dbname&#39;, &#39;user&#39;, &#39;pass&#39;, &#39;species&#39;, &#39;group&#39;</span>

<span class="sd">    Returns:</span>
<span class="sd">    -------</span>
<span class="sd">    Path to the modified local registry file (Databases.pm)</span>

<span class="sd">    Raises:</span>
<span class="sd">    ------</span>
<span class="sd">    FileNotFoundError:</span>
<span class="sd">        If the registry file does not exist after copying.</span>

<span class="sd">    RuntimeError:</span>
<span class="sd">        If there is an error overwriting the original registry file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">registry_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">settings</span><span class="p">[</span><span class="s2">&quot;base_output_dir&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="s2">&quot;Databases.pm&quot;</span>
    <span class="n">registry_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">),</span>
        <span class="s2">&quot;ensembl-analysis&quot;</span><span class="p">,</span>
        <span class="s2">&quot;scripts&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genebuild&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gbiab&quot;</span><span class="p">,</span>
        <span class="s2">&quot;support_files&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Databases.pm&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">registry_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">registry_file</span><span class="p">,</span> <span class="n">registry_path</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">registry_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;A registry file was not found at: </span><span class="si">{</span><span class="n">registry_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">db_string</span><span class="p">(</span><span class="n">db_details</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Bio::EnsEMBL::DBSQL::DBAdaptor-&gt;new(</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;  -host =&gt; &#39;</span><span class="si">{</span><span class="n">db_details</span><span class="p">[</span><span class="s1">&#39;host&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;  -port =&gt; &#39;</span><span class="si">{</span><span class="n">db_details</span><span class="p">[</span><span class="s1">&#39;port&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;  -dbname =&gt; &#39;</span><span class="si">{</span><span class="n">db_details</span><span class="p">[</span><span class="s1">&#39;dbname&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;  -user =&gt; &#39;</span><span class="si">{</span><span class="n">db_details</span><span class="p">[</span><span class="s1">&#39;user&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;  -pass =&gt; &#39;</span><span class="si">{</span><span class="n">db_details</span><span class="p">[</span><span class="s1">&#39;pass&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;  -species =&gt; &#39;</span><span class="si">{</span><span class="n">db_details</span><span class="p">[</span><span class="s1">&#39;species&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;  -group =&gt; &#39;</span><span class="si">{</span><span class="n">db_details</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39;,</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;);</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Generate registry entries</span>
    <span class="n">core_string</span> <span class="o">=</span> <span class="n">db_string</span><span class="p">(</span><span class="n">core_adaptor</span><span class="p">)</span>

    <span class="c1"># Read the original registry content</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="n">registry_path</span><span class="o">.</span><span class="n">read_text</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(</span><span class="n">keepends</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Insert database connection strings after first `{`</span>
    <span class="n">new_lines</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">inserted</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
        <span class="n">new_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">inserted</span> <span class="ow">and</span> <span class="s2">&quot;{&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
            <span class="n">new_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">core_string</span><span class="p">)</span>
            <span class="n">inserted</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="c1"># Write to a temporary file</span>
    <span class="n">tmp_path</span> <span class="o">=</span> <span class="n">registry_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.pm.tmp&quot;</span><span class="p">)</span>
    <span class="n">tmp_path</span><span class="o">.</span><span class="n">write_text</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">new_lines</span><span class="p">))</span>

    <span class="c1"># Overwrite the original registry file</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">tmp_path</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">registry_path</span><span class="p">))</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Issue overwriting the old registry: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">registry_path</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.create_pipe_reg.update_registry_path_in_pipedb" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">update_registry_path_in_pipedb</span><span class="p">(</span><span class="n">parent_dir</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.create_pipe_reg.update_registry_path_in_pipedb" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Updates the resource_description table in the pipeline database by replacing
the shared registry path with a local copy. This ensures worker nodes access
the correct local registry.</p>
<h5 id="ensembl.genes.info_from_registry.create_pipe_reg.update_registry_path_in_pipedb--parameters">Parameters:<a class="headerlink" href="#ensembl.genes.info_from_registry.create_pipe_reg.update_registry_path_in_pipedb--parameters" title="Permanent link">&para;</a></h5>
<p>settings : dict
    Must contain:
        - 'base_output_dir': Path to where the new registry will be copied.</p>


<details class="server_info-" open>
  <summary>dict</summary>
  <p>Connection info for the pipeline database under:
    - server_info["pipeline_db"]["db_host"]
    - server_info["pipeline_db"]["db_user"]
    - server_info["pipeline_db"]["db_port"]
    - server_info["pipeline_db"]["db_password"]
    - server_info["pipeline_db"]["db_name"]</p>
</details>

            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/create_pipe_reg.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">update_registry_path_in_pipedb</span><span class="p">(</span>
    <span class="n">parent_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">server_info</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Updates the resource_description table in the pipeline database by replacing</span>
<span class="sd">    the shared registry path with a local copy. This ensures worker nodes access</span>
<span class="sd">    the correct local registry.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    ----------</span>
<span class="sd">    settings : dict</span>
<span class="sd">        Must contain:</span>
<span class="sd">            - &#39;base_output_dir&#39;: Path to where the new registry will be copied.</span>

<span class="sd">    server_info : dict</span>
<span class="sd">        Connection info for the pipeline database under:</span>
<span class="sd">            - server_info[&quot;pipeline_db&quot;][&quot;db_host&quot;]</span>
<span class="sd">            - server_info[&quot;pipeline_db&quot;][&quot;db_user&quot;]</span>
<span class="sd">            - server_info[&quot;pipeline_db&quot;][&quot;db_port&quot;]</span>
<span class="sd">            - server_info[&quot;pipeline_db&quot;][&quot;db_password&quot;]</span>
<span class="sd">            - server_info[&quot;pipeline_db&quot;][&quot;db_name&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Updating registry path in pipeline DB&quot;</span><span class="p">)</span>
    <span class="n">registry_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">),</span>
        <span class="s2">&quot;ensembl-analysis&quot;</span><span class="p">,</span>
        <span class="s2">&quot;scripts&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genebuild&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gbiab&quot;</span><span class="p">,</span>
        <span class="s2">&quot;support_files&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Databases.pm&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;parennt_dir : </span><span class="si">{</span><span class="n">parent_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">new_registry_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">parent_dir</span><span class="p">)</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;Databases.pm&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">new_registry_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Registry file doesn&#39;t exist: </span><span class="si">{</span><span class="n">new_registry_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Registry file exist at </span><span class="si">{</span><span class="n">new_registry_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">update_resources_query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">		UPDATE resource_description </span>
<span class="s2">		SET worker_cmd_args = REPLACE(worker_cmd_args, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">);</span>
<span class="s2">	&quot;&quot;&quot;</span>

    <span class="n">success</span> <span class="o">=</span> <span class="n">mysql_update</span><span class="p">(</span>
        <span class="n">query</span><span class="o">=</span><span class="n">update_resources_query</span><span class="p">,</span>
        <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;pipeline_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;pipeline_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;pipeline_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">password</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;pipeline_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_password&quot;</span><span class="p">],</span>
        <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;pipeline_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
        <span class="n">params</span><span class="o">=</span><span class="p">(</span><span class="n">registry_file</span><span class="p">,</span> <span class="n">new_registry_file</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">success</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Registry path updated successfully.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to update registry path.&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Failed to update registry path.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.genebuild_start_pipeline" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.genebuild_start_pipeline</code>


<a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Script to extract metadata, initialise an annotation pipeline using eHive,
and seed the pipeline with jobs based on a generated JSON file.</p>


<details class="modules-used" open>
  <summary>Modules used</summary>
  <ul>
<li>start_pipeline_from_registry.py (main)</li>
<li>seed_nonvert.py (seed_jobs_from_json)</li>
</ul>
</details>

<details class="usage" open>
  <summary>Usage</summary>
  <p>python genebuild_init_pipeline.py --gcas path/to/gcas.txt --settings_file path/to/user_pipeline_settings.json</p>
</details>









<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.genebuild_start_pipeline.args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.genebuild_start_pipeline.ehive_urls" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">ehive_urls</span> <span class="o">=</span> <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gcas</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">settings_file</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">seed_url</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.ehive_urls" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.genebuild_start_pipeline.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.genebuild_start_pipeline.parser" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Extract metadata, initialize and seed annotation pipelines.&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.parser" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.genebuild_start_pipeline.init_pipeline_anno" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">init_pipeline_anno</span><span class="p">(</span><span class="n">config_file</span><span class="p">,</span> <span class="n">hive_force_init</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.init_pipeline_anno" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Initialize an eHive pipeline using a given config file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>config_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the .pm configuration file (e.g. EnsemblAnnoHelixer_conf.pm).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>hive_force_init</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to force initialization (default is 1).</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Extracted eHive database URL if successful.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="subprocess.CalledProcessError">CalledProcessError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the <code>init_pipeline.pl</code> command fails.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="RuntimeError">RuntimeError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the expected MySQL URL is not found in the output.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/genebuild_start_pipeline.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">init_pipeline_anno</span><span class="p">(</span><span class="n">config_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">hive_force_init</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialize an eHive pipeline using a given config file.</span>

<span class="sd">    Args:</span>
<span class="sd">        config_file (str): Path to the .pm configuration file (e.g. EnsemblAnnoHelixer_conf.pm).</span>
<span class="sd">        hive_force_init (int): Whether to force initialization (default is 1).</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Extracted eHive database URL if successful.</span>

<span class="sd">    Raises:</span>
<span class="sd">        subprocess.CalledProcessError: If the `init_pipeline.pl` command fails.</span>
<span class="sd">        RuntimeError: If the expected MySQL URL is not found in the output.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Innitialising anno pipeline&quot;</span><span class="p">)</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;init_pipeline.pl&quot;</span><span class="p">,</span> <span class="n">config_file</span><span class="p">,</span> <span class="s2">&quot;--hive_force_init&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">hive_force_init</span><span class="p">)]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cmd</span><span class="p">))</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
            <span class="n">cmd</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">STDOUT</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">output</span><span class="o">.</span><span class="n">splitlines</span><span class="p">():</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(mysql://[^\s]+)&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># Just the URL part</span>
        <span class="k">return</span>
    <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error running init_pipeline.pl: </span><span class="si">{</span><span class="n">e</span><span class="o">.</span><span class="n">stdout</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">raise</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.genebuild_start_pipeline.init_pipeline_main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">init_pipeline_main</span><span class="p">(</span><span class="n">config_file</span><span class="p">,</span> <span class="n">hive_force_init</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.init_pipeline_main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Initialize an eHive pipeline using a given config file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>config_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the .pm configuration file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>hive_force_init</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to force initialization (default is 1).</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Extracted eHive database URL if successful.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="subprocess.CalledProcessError">CalledProcessError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the <code>init_pipeline.pl</code> command fails.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="RuntimeError">RuntimeError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the expected MySQL URL is not found in the output.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/genebuild_start_pipeline.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">init_pipeline_main</span><span class="p">(</span><span class="n">config_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">hive_force_init</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialize an eHive pipeline using a given config file.</span>

<span class="sd">    Args:</span>
<span class="sd">        config_file (str): Path to the .pm configuration file.</span>
<span class="sd">        hive_force_init (int): Whether to force initialization (default is 1).</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Extracted eHive database URL if successful.</span>

<span class="sd">    Raises:</span>
<span class="sd">        subprocess.CalledProcessError: If the `init_pipeline.pl` command fails.</span>
<span class="sd">        RuntimeError: If the expected MySQL URL is not found in the output.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">config_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">config_file</span><span class="p">)</span>
    <span class="n">config_dir</span> <span class="o">=</span> <span class="n">config_path</span><span class="o">.</span><span class="n">parent</span>
    <span class="n">config_name</span> <span class="o">=</span> <span class="n">config_path</span><span class="o">.</span><span class="n">stem</span>  <span class="c1"># Gets filename without .pm extension</span>

    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;init_pipeline.pl&quot;</span><span class="p">,</span>
        <span class="n">config_name</span><span class="p">,</span>  <span class="c1"># Just the module name, not the full path</span>
        <span class="s2">&quot;--hive_force_init&quot;</span><span class="p">,</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">hive_force_init</span><span class="p">),</span>
    <span class="p">]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cmd</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Working directory: </span><span class="si">{</span><span class="n">config_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
            <span class="n">cmd</span><span class="p">,</span>
            <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
            <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">STDOUT</span><span class="p">,</span>
            <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">cwd</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">config_dir</span><span class="p">),</span>  <span class="c1"># Run from the config directory</span>
        <span class="p">)</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">output</span><span class="o">.</span><span class="n">splitlines</span><span class="p">():</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(mysql://[^\s]+)&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error running init_pipeline.pl: </span><span class="si">{</span><span class="n">e</span><span class="o">.</span><span class="n">stdout</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">raise</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.genebuild_start_pipeline.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">(</span><span class="n">gcas</span><span class="p">,</span> <span class="n">settings_file</span><span class="p">,</span> <span class="n">seed_url</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.genebuild_start_pipeline.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main execution logic: extract metadata, initialize the pipeline, and seed jobs.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gcas</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to file containing GCA accessions (one per line).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>settings_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to settings file (JSON).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed_url</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Existing URL to seed the pipeline. If provided initialisation will be skipped.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If no config file is found.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="RuntimeError">RuntimeError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If multiple config files are found.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/genebuild_start_pipeline.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">gcas</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">settings_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">seed_url</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main execution logic: extract metadata, initialize the pipeline, and seed jobs.</span>

<span class="sd">    Args:</span>
<span class="sd">        gcas (str): Path to file containing GCA accessions (one per line).</span>
<span class="sd">        settings_file (str): Path to settings file (JSON).</span>
<span class="sd">        seed_url (str): Existing URL to seed the pipeline. If provided initialisation will be skipped.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If no config file is found.</span>
<span class="sd">        RuntimeError: If multiple config files are found.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Get info and create input JSON</span>
    <span class="n">server_info</span><span class="p">,</span> <span class="n">all_output_params</span><span class="p">,</span> <span class="n">saved_paths</span> <span class="o">=</span> <span class="n">info</span><span class="p">(</span><span class="n">gcas</span><span class="p">,</span> <span class="n">settings_file</span><span class="p">)</span>

    <span class="c1"># Save EHIVE URLs</span>
    <span class="n">ehive_urls</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Anno</span>
    <span class="k">if</span> <span class="n">saved_paths</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;anno&quot;</span><span class="p">):</span>
        <span class="n">json_file</span> <span class="o">=</span> <span class="n">saved_paths</span><span class="p">[</span><span class="s2">&quot;anno&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">seed_url</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Skipping pipeline initialisation.&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using provided seed URL: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">seed_url</span><span class="p">)</span>
            <span class="n">ehive_urls</span><span class="p">[</span><span class="s2">&quot;anno&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seed_url</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;No seed URL provided. Initialising new pipeline.&quot;</span><span class="p">)</span>
            <span class="n">first_key</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">all_output_params</span><span class="p">))</span>
            <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">all_output_params</span><span class="p">[</span><span class="n">first_key</span><span class="p">][</span><span class="s2">&quot;output_path&quot;</span><span class="p">])</span>
            <span class="n">parent_dir</span> <span class="o">=</span> <span class="n">output_path</span><span class="o">.</span><span class="n">parent</span>
            <span class="n">conf_files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">parent_dir</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;*_conf.pm&quot;</span><span class="p">))</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">conf_files</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No .conf file found in </span><span class="si">{</span><span class="n">parent_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">conf_files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Multiple .conf files found in </span><span class="si">{</span><span class="n">parent_dir</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">conf_files</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

            <span class="n">conf_path</span> <span class="o">=</span> <span class="n">conf_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Config found at </span><span class="si">{</span><span class="n">conf_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">ehive_url</span> <span class="o">=</span> <span class="n">init_pipeline_anno</span><span class="p">(</span><span class="n">conf_path</span><span class="p">)</span>
            <span class="n">ehive_urls</span><span class="p">[</span><span class="s2">&quot;anno&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ehive_url</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Extracted EHIVE_URL: </span><span class="si">{</span><span class="n">ehive_url</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">update_registry_path_in_pipedb</span><span class="p">(</span><span class="n">parent_dir</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span>
            <span class="n">seed_url</span> <span class="o">=</span> <span class="n">ehive_url</span>  <span class="c1"># unify below</span>

        <span class="c1"># Common seeding step</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Seeding non-vertebrate pipeline from </span><span class="si">{</span><span class="n">json_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">params</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

        <span class="n">seed_jobs_from_json</span><span class="p">(</span><span class="n">json_file</span><span class="o">=</span><span class="n">params</span><span class="p">,</span> <span class="n">analysis_id</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ehive_url</span><span class="o">=</span><span class="n">seed_url</span><span class="p">)</span>

    <span class="c1"># --- Handle main (one JSON per GCA) ---</span>
    <span class="k">if</span> <span class="n">saved_paths</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;main&quot;</span><span class="p">):</span>
        <span class="n">ehive_urls</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">gca</span><span class="p">,</span> <span class="n">json_path</span> <span class="ow">in</span> <span class="n">saved_paths</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Initialising vertebrate pipeline for </span><span class="si">{</span><span class="n">gca</span><span class="si">}</span><span class="s2"> from </span><span class="si">{</span><span class="n">json_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">all_output_params</span><span class="p">[</span><span class="n">gca</span><span class="p">][</span><span class="s2">&quot;output_path&quot;</span><span class="p">])</span>
            <span class="n">conf_files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">output_path</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;*_conf.pm&quot;</span><span class="p">))</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">conf_files</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No .conf file found in </span><span class="si">{</span><span class="n">parent_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">conf_files</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Multiple .conf files found in </span><span class="si">{</span><span class="n">parent_dir</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">conf_files</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

            <span class="n">conf_path</span> <span class="o">=</span> <span class="n">conf_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

            <span class="n">ehive_url</span> <span class="o">=</span> <span class="n">init_pipeline_main</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">conf_path</span><span class="p">))</span>
            <span class="n">ehive_urls</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">][</span><span class="n">gca</span><span class="p">]</span> <span class="o">=</span> <span class="n">ehive_url</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Extracted EHIVE_URL for </span><span class="si">{</span><span class="n">gca</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">ehive_url</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Initalised pipelines. EHIVE_URL(s): </span><span class="si">{</span><span class="n">ehive_urls</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">output_dir_current</span> <span class="o">=</span> <span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span>
    <span class="n">ehive_url_file</span> <span class="o">=</span> <span class="n">output_dir_current</span> <span class="o">/</span> <span class="s2">&quot;ehive_urls.json&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">ehive_url_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">ehive_urls</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved EHIVE URLs to </span><span class="si">{</span><span class="n">ehive_url_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ehive_urls</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.mysql_helper" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.mysql_helper</code>


<a href="#ensembl.genes.info_from_registry.mysql_helper" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>MySQL Helper Module</p>
<p>This module provides utility functions for connecting to a MySQL database,
fetching data, and performing updates. Logging is enabled to capture
informative messages and errors during execution.</p>


<p><span class="doc-section-title">Functions:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><span title="ensembl.genes.info_from_registry.mysql_helper.- mysql_fetch_data">- mysql_fetch_data</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Executes a SELECT query and returns the result as a list of dictionaries.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="ensembl.genes.info_from_registry.mysql_helper.- mysql_update">- mysql_update</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Executes an UPDATE/INSERT/DELETE query with optional parameters.</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.mysql_helper.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.mysql_helper.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.mysql_helper.mysql_fetch_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.mysql_helper.mysql_fetch_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Execute a SELECT query on the specified MySQL database and return results.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>query</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>SQL SELECT query to execute.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>database</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the database.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>host</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Host address of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>port</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Port number of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>user</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Username to connect to the database.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>password</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Password for the user.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>params</code>
            </td>
            <td>
                  <code><span title="tuple">tuple</span> or <span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Parameters to safely interpolate into the query.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="list">list</span>[<span title="dict">dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list[dict]: A list of dictionaries representing query results.
        Returns an empty list on failure.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/mysql_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mysql_fetch_data</span><span class="p">(</span>
    <span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">password</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">params</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Execute a SELECT query on the specified MySQL database and return results.</span>

<span class="sd">    Args:</span>
<span class="sd">        query (str): SQL SELECT query to execute.</span>
<span class="sd">        database (str): Name of the database.</span>
<span class="sd">        host (str): Host address of the MySQL server.</span>
<span class="sd">        port (int): Port number of the MySQL server.</span>
<span class="sd">        user (str): Username to connect to the database.</span>
<span class="sd">        password (str): Password for the user.</span>
<span class="sd">        params (tuple or list, optional): Parameters to safely interpolate into the query.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list[dict]: A list of dictionaries representing query results.</span>
<span class="sd">                    Returns an empty list on failure.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span>
            <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span>
            <span class="n">port</span><span class="o">=</span><span class="n">port</span><span class="p">,</span>
            <span class="n">password</span><span class="o">=</span><span class="n">password</span><span class="p">,</span>
            <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
            <span class="n">cursorclass</span><span class="o">=</span><span class="n">pymysql</span><span class="o">.</span><span class="n">cursors</span><span class="o">.</span><span class="n">DictCursor</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span> <span class="ow">or</span> <span class="p">())</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
        <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Query successful&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span>

    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MySQL error during fetch: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.mysql_helper.mysql_get_connection" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_get_connection</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.mysql_helper.mysql_get_connection" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Establish a connection to the MySQL database.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/mysql_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mysql_get_connection</span><span class="p">(</span>
    <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">password</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Establish a connection to the MySQL database.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span>
            <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span>
            <span class="n">port</span><span class="o">=</span><span class="n">port</span><span class="p">,</span>
            <span class="n">password</span><span class="o">=</span><span class="n">password</span><span class="p">,</span>
            <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
            <span class="n">cursorclass</span><span class="o">=</span><span class="n">pymysql</span><span class="o">.</span><span class="n">cursors</span><span class="o">.</span><span class="n">DictCursor</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">conn</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MySQL error: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.mysql_helper.mysql_update" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_update</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.mysql_helper.mysql_update" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Execute an UPDATE, INSERT, or DELETE query on the specified MySQL database.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>query</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>SQL query to execute.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>database</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the database.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>host</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Host address of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>port</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Port number of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>user</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Username to connect to the database.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>password</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Password for the user.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>params</code>
            </td>
            <td>
                  <code><span title="tuple">tuple</span> or <span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Parameters to safely interpolate into the query.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the query was executed successfully, False otherwise.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/mysql_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mysql_update</span><span class="p">(</span>
    <span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">password</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">params</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Execute an UPDATE, INSERT, or DELETE query on the specified MySQL database.</span>

<span class="sd">    Args:</span>
<span class="sd">        query (str): SQL query to execute.</span>
<span class="sd">        database (str): Name of the database.</span>
<span class="sd">        host (str): Host address of the MySQL server.</span>
<span class="sd">        port (int): Port number of the MySQL server.</span>
<span class="sd">        user (str): Username to connect to the database.</span>
<span class="sd">        password (str): Password for the user.</span>
<span class="sd">        params (tuple or list, optional): Parameters to safely interpolate into the query.</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the query was executed successfully, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span>
            <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span>
            <span class="n">port</span><span class="o">=</span><span class="n">port</span><span class="p">,</span>
            <span class="n">password</span><span class="o">=</span><span class="n">password</span><span class="p">,</span>
            <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
            <span class="n">cursorclass</span><span class="o">=</span><span class="n">pymysql</span><span class="o">.</span><span class="n">cursors</span><span class="o">.</span><span class="n">DictCursor</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span> <span class="ow">or</span> <span class="p">())</span>
            <span class="n">conn</span><span class="o">.</span><span class="n">commit</span><span class="p">()</span>
        <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Update successful.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MySQL error during update: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.registry_helper" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.registry_helper</code>


<a href="#ensembl.genes.info_from_registry.registry_helper" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Registry Helper Module</p>
<p>This module provides utility functions for common registry database queries
that are shared across multiple scripts.</p>










<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.registry_helper.fetch_assembly_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">fetch_assembly_id</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.registry_helper.fetch_assembly_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch the assembly ID for a given assembly accession.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>connection</code>
            </td>
            <td>
                  <code><span title="pymysql.connections.Connection">Connection</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL connection object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly Accession (GCA format like GCA_123456789.1)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    int: Assembly ID if found, else None</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/registry_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fetch_assembly_id</span><span class="p">(</span>
    <span class="n">connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch the assembly ID for a given assembly accession.</span>

<span class="sd">    Args:</span>
<span class="sd">        connection: MySQL connection object</span>
<span class="sd">        assembly (str): Assembly Accession (GCA format like GCA_123456789.1)</span>
<span class="sd">    Returns:</span>
<span class="sd">        int: Assembly ID if found, else None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    SELECT assembly_id FROM assembly</span>
<span class="s2">    WHERE CONCAT(gca_chain, &#39;.&#39;, gca_version) = </span><span class="si">%s</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">with</span> <span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="p">(</span><span class="n">assembly</span><span class="p">,))</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchone</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;assembly_id&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">result</span> <span class="k">else</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.registry_helper.fetch_current_genebuild_record" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">fetch_current_genebuild_record</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">genebuilder</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.registry_helper.fetch_current_genebuild_record" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch the full current genebuild record for a given assembly.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>connection</code>
            </td>
            <td>
                  <code><span title="pymysql.connections.Connection">Connection</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL connection object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly Accession (GCA format)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genebuilder</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genebuilder name to filter by. If None, fetches any active record.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    dict: Record with genebuild_status_id, gb_status, genebuilder, annotation_method, genebuild_version, and other fields if found, else None</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/registry_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fetch_current_genebuild_record</span><span class="p">(</span>
    <span class="n">connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span>
    <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genebuilder</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">any</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch the full current genebuild record for a given assembly.</span>

<span class="sd">    Args:</span>
<span class="sd">        connection: MySQL connection object</span>
<span class="sd">        assembly (str): Assembly Accession (GCA format)</span>
<span class="sd">        genebuilder (str, optional): Genebuilder name to filter by. If None, fetches any active record.</span>
<span class="sd">    Returns:</span>
<span class="sd">        dict: Record with genebuild_status_id, gb_status, genebuilder, annotation_method, genebuild_version, and other fields if found, else None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">genebuilder</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT genebuild_status_id, gb_status, genebuilder, annotation_method, genebuild_version</span>
<span class="s2">        FROM genebuild_status</span>
<span class="s2">        WHERE gca_accession = </span><span class="si">%s</span><span class="s2"> AND genebuilder = </span><span class="si">%s</span><span class="s2"> AND last_attempt = 1</span>
<span class="s2">        &quot;&quot;&quot;</span>
        <span class="n">params</span> <span class="o">=</span> <span class="p">(</span><span class="n">assembly</span><span class="p">,</span> <span class="n">genebuilder</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT genebuild_status_id, gb_status, genebuilder, annotation_method, genebuild_version</span>
<span class="s2">        FROM genebuild_status</span>
<span class="s2">        WHERE gca_accession = </span><span class="si">%s</span><span class="s2"> AND last_attempt = 1</span>
<span class="s2">        &quot;&quot;&quot;</span>
        <span class="n">params</span> <span class="o">=</span> <span class="p">(</span><span class="n">assembly</span><span class="p">,)</span>

    <span class="k">with</span> <span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchone</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">result</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.registry_helper.fetch_genebuild_status_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">fetch_genebuild_status_id</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.registry_helper.fetch_genebuild_status_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch the current genebuild status ID for a given assembly.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>connection</code>
            </td>
            <td>
                  <code><span title="pymysql.connections.Connection">Connection</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL connection object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly Accession (GCA format)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    int: genebuild_status_id if found (where last_attempt=1), else None</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/registry_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fetch_genebuild_status_id</span><span class="p">(</span>
    <span class="n">connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch the current genebuild status ID for a given assembly.</span>

<span class="sd">    Args:</span>
<span class="sd">        connection: MySQL connection object</span>
<span class="sd">        assembly (str): Assembly Accession (GCA format)</span>
<span class="sd">    Returns:</span>
<span class="sd">        int: genebuild_status_id if found (where last_attempt=1), else None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">fetch_current_genebuild_record</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;genebuild_status_id&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">record</span> <span class="k">else</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.registry_helper.fetch_registry_ids" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">fetch_registry_ids</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">genebuilder</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.registry_helper.fetch_registry_ids" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch both assembly_id and genebuild_status_id for a given assembly.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>connection</code>
            </td>
            <td>
                  <code><span title="pymysql.connections.Connection">Connection</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL connection object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly Accession (GCA format)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genebuilder</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genebuilder name to filter by. If None and multiple
                         active records exist, returns the first one found.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    tuple: (assembly_id, genebuild_status_id) where genebuild_status_id may be None
Raises:
    ValueError: If assembly is not found in registry</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/registry_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fetch_registry_ids</span><span class="p">(</span>
    <span class="n">connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span>
    <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genebuilder</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch both assembly_id and genebuild_status_id for a given assembly.</span>

<span class="sd">    Args:</span>
<span class="sd">        connection: MySQL connection object</span>
<span class="sd">        assembly (str): Assembly Accession (GCA format)</span>
<span class="sd">        genebuilder (str, optional): Genebuilder name to filter by. If None and multiple</span>
<span class="sd">                                     active records exist, returns the first one found.</span>
<span class="sd">    Returns:</span>
<span class="sd">        tuple: (assembly_id, genebuild_status_id) where genebuild_status_id may be None</span>
<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If assembly is not found in registry</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">assembly_id</span> <span class="o">=</span> <span class="n">fetch_assembly_id</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">assembly_id</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Assembly not found in registry: </span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">record</span> <span class="o">=</span> <span class="n">fetch_current_genebuild_record</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">genebuilder</span><span class="p">)</span>
    <span class="n">genebuild_status_id</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;genebuild_status_id&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">record</span> <span class="k">else</span> <span class="kc">None</span>

    <span class="k">return</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">genebuild_status_id</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.seed_nonvert" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.seed_nonvert</code>


<a href="#ensembl.genes.info_from_registry.seed_nonvert" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>seed_nonvert.py</p>
<p>This script allows you to seed jobs into an eHive pipeline using a JSON file or a dictionary of job parameters.
It converts Python dictionaries to Perl hash syntax and invokes the <code>seed_pipeline.pl</code> script via subprocess.</p>


<details class="typical-usage-example" open>
  <summary>Typical usage example</summary>
  <p>python seed_jobs.py -j jobs.json -a 1 -u mysql://user:pass@host/db</p>
</details>

<p><span class="doc-section-title">Functions:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><span title="ensembl.genes.info_from_registry.seed_nonvert.- dict_to_perl_hash">- dict_to_perl_hash</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Recursively converts a Python dictionary into a Perl-style hash string.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="ensembl.genes.info_from_registry.seed_nonvert.- seed_jobs_from_json">- seed_jobs_from_json</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Loads parameters and invokes eHive seeding commands.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="ensembl.genes.info_from_registry.seed_nonvert.- main">- main</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Parses command-line arguments and calls the seeding function.</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.seed_nonvert.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.seed_nonvert.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.seed_nonvert.dict_to_perl_hash" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">dict_to_perl_hash</span><span class="p">(</span><span class="n">d</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.seed_nonvert.dict_to_perl_hash" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Recursively convert a Python dictionary to a Perl hash string.
Automatically converts database connection parameters to Ensembl format.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>d</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Python dictionary to convert.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A string representing the dictionary in Perl hash format.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/seed_nonvert.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">dict_to_perl_hash</span><span class="p">(</span><span class="n">d</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Recursively convert a Python dictionary to a Perl hash string.</span>
<span class="sd">    Automatically converts database connection parameters to Ensembl format.</span>

<span class="sd">    Args:</span>
<span class="sd">        d (dict): Python dictionary to convert.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: A string representing the dictionary in Perl hash format.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">convert_db_keys</span><span class="p">(</span><span class="n">dictionary</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert db_* keys to Ensembl format if this looks like a DB connection.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dictionary</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">dictionary</span>

        <span class="c1"># Check if this looks like a database connection dictionary</span>
        <span class="n">db_keys</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;db_host&quot;</span><span class="p">,</span> <span class="s2">&quot;db_user&quot;</span><span class="p">,</span> <span class="s2">&quot;db_port&quot;</span><span class="p">,</span> <span class="s2">&quot;db_password&quot;</span><span class="p">,</span> <span class="s2">&quot;db_name&quot;</span><span class="p">}</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">key</span> <span class="ow">in</span> <span class="n">dictionary</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">db_keys</span><span class="p">):</span>
            <span class="n">converted</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">key_mapping</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="s2">&quot;-host&quot;</span><span class="p">,</span>
                <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="s2">&quot;-user&quot;</span><span class="p">,</span>
                <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="s2">&quot;-port&quot;</span><span class="p">,</span>
                <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="s2">&quot;-pass&quot;</span><span class="p">,</span>
                <span class="s2">&quot;db_name&quot;</span><span class="p">:</span> <span class="s2">&quot;-dbname&quot;</span><span class="p">,</span>
            <span class="p">}</span>

            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">dictionary</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">key_mapping</span><span class="p">:</span>
                    <span class="n">converted</span><span class="p">[</span><span class="n">key_mapping</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span> <span class="o">=</span> <span class="n">v</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">converted</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>

            <span class="c1"># Add driver if not present</span>
            <span class="k">if</span> <span class="s2">&quot;-driver&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">converted</span> <span class="ow">and</span> <span class="nb">any</span><span class="p">(</span>
                <span class="n">k</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">converted</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="p">):</span>
                <span class="n">converted</span><span class="p">[</span><span class="s2">&quot;-driver&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;mysql&quot;</span>

            <span class="k">return</span> <span class="n">converted</span>

        <span class="k">return</span> <span class="n">dictionary</span>

    <span class="c1"># Convert database keys if needed</span>
    <span class="n">d</span> <span class="o">=</span> <span class="n">convert_db_keys</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>

    <span class="n">items</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">key_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="c1"># Recursively convert nested dictionaries and check for DB format</span>
            <span class="n">converted_v</span> <span class="o">=</span> <span class="n">convert_db_keys</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
            <span class="n">val_str</span> <span class="o">=</span> <span class="n">dict_to_perl_hash</span><span class="p">(</span><span class="n">converted_v</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">val_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
        <span class="k">elif</span> <span class="n">v</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">val_str</span> <span class="o">=</span> <span class="s2">&quot;undef&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">val_str</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
        <span class="n">items</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">key_str</span><span class="si">}</span><span class="s2"> =&gt; </span><span class="si">{</span><span class="n">val_str</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot;{{</span><span class="si">{}</span><span class="s2">}}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">items</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.seed_nonvert.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.info_from_registry.seed_nonvert.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/seed_nonvert.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Seed eHive pipeline jobs from a JSON file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-j&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--json_file&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to the JSON file with job parameters&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-a&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--analysis_id&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Analysis ID to seed (default: 1)&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-u&quot;</span><span class="p">,</span> <span class="s2">&quot;--url&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;EHIVE URL&quot;</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="n">seed_jobs_from_json</span><span class="p">(</span>
        <span class="n">json_file</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">json_file</span><span class="p">,</span> <span class="n">analysis_id</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">analysis_id</span><span class="p">,</span> <span class="n">ehive_url</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">url</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.seed_nonvert.seed_jobs_from_json" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">seed_jobs_from_json</span><span class="p">(</span><span class="n">json_file</span><span class="p">,</span> <span class="n">analysis_id</span><span class="p">,</span> <span class="n">ehive_url</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.seed_nonvert.seed_jobs_from_json" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Seed jobs in an eHive pipeline using job parameters from a JSON file or dictionary.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>json_file</code>
            </td>
            <td>
                  <code><span title="typing.Union">Union</span>[<span title="str">str</span>, <span title="dict">dict</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to a JSON file or a Python dictionary with job parameters.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>analysis_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>eHive analysis ID to assign jobs to.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ehive_url</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>URL of the eHive database (e.g., mysql://user:pass@host/db).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="subprocess.CalledProcessError">CalledProcessError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the seeding script fails.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/seed_nonvert.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">seed_jobs_from_json</span><span class="p">(</span>
    <span class="n">json_file</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">],</span>
    <span class="n">analysis_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">ehive_url</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Seed jobs in an eHive pipeline using job parameters from a JSON file or dictionary.</span>

<span class="sd">    Args:</span>
<span class="sd">        json_file (Union[str, dict]): Path to a JSON file or a Python dictionary with job parameters.</span>
<span class="sd">        analysis_id (int): eHive analysis ID to assign jobs to.</span>
<span class="sd">        ehive_url (str): URL of the eHive database (e.g., mysql://user:pass@host/db).</span>

<span class="sd">    Raises:</span>
<span class="sd">        subprocess.CalledProcessError: If the seeding script fails.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">json_file</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">params</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">params</span> <span class="o">=</span> <span class="n">json_file</span>  <span class="c1"># assume dict already</span>

    <span class="k">for</span> <span class="n">input_id</span><span class="p">,</span> <span class="n">param_dict</span> <span class="ow">in</span> <span class="n">params</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">perl_hash</span> <span class="o">=</span> <span class="n">dict_to_perl_hash</span><span class="p">(</span><span class="n">param_dict</span><span class="p">)</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;seed_pipeline.pl&quot;</span><span class="p">,</span>
            <span class="s2">&quot;-analysis_id&quot;</span><span class="p">,</span>
            <span class="nb">str</span><span class="p">(</span><span class="n">analysis_id</span><span class="p">),</span>
            <span class="s2">&quot;-input_id&quot;</span><span class="p">,</span>
            <span class="n">perl_hash</span><span class="p">,</span>
            <span class="s2">&quot;-url&quot;</span><span class="p">,</span>
            <span class="n">ehive_url</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Seeding complete&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.start_pipeline_from_registry" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.start_pipeline_from_registry</code>


<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>This script connects to a genomic assembly registry, extracts assembly metadata,
and initializes annotation pipelines based on predefined configurations.</p>
<p>It supports both database-driven metadata retrieval and custom INI file loading,
handles directory setup, and creates necessary configuration and command files
for running genome annotation pipelines.</p>


<details class="typical-usage" open>
  <summary>Typical usage</summary>
  <p>python start_pipeline_from_registry.py --gcas gca_list.txt ---settings_file settings.json</p>
</details>









<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.parser" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Extract metadata and initialize annotation pipelines.&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.parser" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.add_generated_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">add_generated_data</span><span class="p">(</span><span class="n">server_info</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">,</span> <span class="n">settings</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.add_generated_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Enrich registry metadata with clade assignments and derived pipeline variables.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Server connection details.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly_accession</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genome assembly accession.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pipeline settings.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Enriched metadata dictionary with added fields for pipeline usage.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_generated_data</span><span class="p">(</span>
    <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">settings</span><span class="p">:</span> <span class="nb">dict</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Enrich registry metadata with clade assignments and derived pipeline variables.</span>

<span class="sd">    Args:</span>
<span class="sd">        server_info (dict): Server connection details.</span>
<span class="sd">        assembly_accession (str): Genome assembly accession.</span>
<span class="sd">        settings (dict): Pipeline settings.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Enriched metadata dictionary with added fields for pipeline usage.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">registry_info</span> <span class="o">=</span> <span class="n">get_metadata_from_registry</span><span class="p">(</span>
        <span class="n">server_info</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">,</span> <span class="n">settings</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Data collected from registry </span><span class="si">{</span><span class="n">assembly_accession</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">clade</span><span class="p">,</span> <span class="n">genus_id</span><span class="p">,</span> <span class="n">clade_metadata</span> <span class="o">=</span> <span class="n">assign_clade</span><span class="p">(</span><span class="n">server_info</span><span class="p">,</span> <span class="n">registry_info</span><span class="p">)</span>
    <span class="n">registry_info</span><span class="p">[</span><span class="s2">&quot;clade&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">clade</span>
    <span class="n">registry_info</span><span class="p">[</span><span class="s2">&quot;genus_taxon_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genus_id</span>

    <span class="k">if</span> <span class="n">clade_metadata</span><span class="p">:</span>
        <span class="n">registry_info</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">clade_metadata</span><span class="p">)</span>

    <span class="n">info_dict</span> <span class="o">=</span> <span class="n">registry_info</span>
    <span class="c1"># Create variables for pipeline</span>
    <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;strain_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;strain&quot;</span>

    <span class="k">if</span> <span class="s2">&quot;assembly_name&quot;</span> <span class="ow">in</span> <span class="n">registry_info</span> <span class="ow">and</span> <span class="s2">&quot; &quot;</span> <span class="ow">in</span> <span class="n">registry_info</span><span class="p">[</span><span class="s2">&quot;assembly_name&quot;</span><span class="p">]:</span>
        <span class="n">registry_info</span><span class="p">[</span><span class="s2">&quot;assembly_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">registry_info</span><span class="p">[</span><span class="s2">&quot;assembly_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
            <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;alternate_haplotype&quot;</span> <span class="ow">in</span> <span class="n">registry_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_name&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
        <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;common_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;alternate haplotype&quot;</span>
        <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species_strain&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;alternate haplotype&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">registry_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;common_name&quot;</span><span class="p">):</span>
        <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;common_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>

    <span class="n">info_dict</span><span class="p">[</span>
        <span class="s2">&quot;species_display_name&quot;</span>
    <span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">registry_info</span><span class="p">[</span><span class="s1">&#39;species_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">registry_info</span><span class="p">[</span><span class="s1">&#39;common_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">) - </span><span class="si">{</span><span class="n">assembly_accession</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species_strain&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;reference&quot;</span>

    <span class="n">raw_species</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">registry_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;species_name&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">species_name</span> <span class="o">=</span> <span class="n">raw_species</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>  <span class="c1"># Remove trailing underscore</span>

    <span class="c1"># Extract binomial name</span>
    <span class="n">parts</span> <span class="o">=</span> <span class="n">species_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">p1</span><span class="p">,</span> <span class="n">p2</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">max_len</span> <span class="o">=</span> <span class="mi">15</span>
        <span class="n">production_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p1</span><span class="p">[:</span><span class="n">max_len</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">p2</span><span class="p">[:</span><span class="n">max_len</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">production_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="n">production_gca</span> <span class="o">=</span> <span class="n">assembly_accession</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;v&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
    <span class="n">production_name</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">production_gca</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="c1"># Update dictionary</span>
    <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">species_name</span>
    <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;production_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">production_name</span>
    <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species_strain_group&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">production_name</span>
    <span class="n">info_dict</span><span class="p">[</span>
        <span class="s2">&quot;species_url&quot;</span>
    <span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">registry_info</span><span class="p">[</span><span class="s1">&#39;species_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">assembly_accession</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">info_dict</span><span class="p">[</span>
        <span class="s2">&quot;core_dbname&quot;</span>
    <span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;dbowner&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">production_gca</span><span class="si">}</span><span class="s2">_core_</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;release_number&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_1&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Values formatted for </span><span class="si">{</span><span class="n">assembly_accession</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">info_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.copy_general_module" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">copy_general_module</span><span class="p">()</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.copy_general_module" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">copy_general_module</span><span class="p">():</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Copying general module&quot;</span><span class="p">)</span>
    <span class="n">enscode</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">enscode</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span><span class="s2">&quot;Environment variable &#39;ENSCODE&#39; is not set&quot;</span><span class="p">)</span>

    <span class="n">analysis_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">enscode</span><span class="p">,</span> <span class="s2">&quot;ensembl-analysis&quot;</span><span class="p">)</span>
    <span class="n">general_file</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">Path</span><span class="p">(</span><span class="n">analysis_path</span><span class="p">)</span>
        <span class="o">/</span> <span class="s2">&quot;modules&quot;</span>
        <span class="o">/</span> <span class="s2">&quot;Bio&quot;</span>
        <span class="o">/</span> <span class="s2">&quot;EnsEMBL&quot;</span>
        <span class="o">/</span> <span class="s2">&quot;Analysis&quot;</span>
        <span class="o">/</span> <span class="s2">&quot;Config&quot;</span>
        <span class="o">/</span> <span class="s2">&quot;General.pm&quot;</span>
    <span class="p">)</span>
    <span class="n">example_file</span> <span class="o">=</span> <span class="n">general_file</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.pm.example&quot;</span><span class="p">)</span>

    <span class="c1"># Copy if missing</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">general_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">example_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">example_file</span><span class="p">,</span> <span class="n">general_file</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Copied </span><span class="si">{</span><span class="n">example_file</span><span class="si">}</span><span class="s2"> → </span><span class="si">{</span><span class="n">general_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing example file: </span><span class="si">{</span><span class="n">example_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">general_file</span><span class="si">}</span><span class="s2"> already exists, nothing to do.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.create_dir" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">create_dir</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.create_dir" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create a directory and optionally set its permissions.</p>
<p>This function creates the directory at 'path' including any necessary
parent directories. If 'mode' is provided, it changes the directory's permissions.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>path</code>
            </td>
            <td>
                  <code><span title="str">str</span> or <span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory path to create.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mode</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>File system permissions mode (e.g., 0o775). Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="RuntimeError">RuntimeError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the directory creation or permission change fails.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_dir</span><span class="p">(</span><span class="n">path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span><span class="p">,</span> <span class="n">mode</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a directory and optionally set its permissions.</span>

<span class="sd">    This function creates the directory at &#39;path&#39; including any necessary</span>
<span class="sd">    parent directories. If &#39;mode&#39; is provided, it changes the directory&#39;s permissions.</span>

<span class="sd">    Args:</span>
<span class="sd">        path (str or Path): The directory path to create.</span>
<span class="sd">        mode (int, optional): File system permissions mode (e.g., 0o775). Defaults to None.</span>

<span class="sd">    Raises:</span>
<span class="sd">        RuntimeError: If the directory creation or permission change fails.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">chmod</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">mode</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to create dir: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.current_projection_source_db" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">current_projection_source_db</span><span class="p">(</span><span class="n">projection_source_production_name</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.current_projection_source_db" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Find the reference core database and server info for the given projection_source_production_name.
If not found, fall back to homo_sapiens core.
Args:
    projection_source_production_name (str): Name to look for in core name (comes from clade_settings).
Returns a dictionary with keys: db_name, server, port
Raises:
    RuntimeError: If database could not be found.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">current_projection_source_db</span><span class="p">(</span><span class="n">projection_source_production_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find the reference core database and server info for the given projection_source_production_name.</span>
<span class="sd">    If not found, fall back to homo_sapiens core.</span>
<span class="sd">    Args:</span>
<span class="sd">        projection_source_production_name (str): Name to look for in core name (comes from clade_settings).</span>
<span class="sd">    Returns a dictionary with keys: db_name, server, port</span>
<span class="sd">    Raises:</span>
<span class="sd">        RuntimeError: If database could not be found.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">projection_json</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">),</span>
        <span class="s2">&quot;ensembl-genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;src&quot;</span><span class="p">,</span>
        <span class="s2">&quot;python&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ensembl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;info_from_registry&quot;</span><span class="p">,</span>
        <span class="s2">&quot;projection_source.json&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">projection_json</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to read JSON file </span><span class="si">{</span><span class="n">projection_json</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Find a matching nested dictionary</span>
    <span class="n">matched_key</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">subdict</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="nb">isinstance</span><span class="p">(</span><span class="n">subdict</span><span class="p">,</span> <span class="nb">dict</span><span class="p">)</span>
            <span class="ow">and</span> <span class="n">projection_source_production_name</span> <span class="ow">in</span> <span class="n">subdict</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="p">):</span>
            <span class="n">matched_key</span> <span class="o">=</span> <span class="n">key</span>
            <span class="k">break</span>

    <span class="c1"># Fallback logic</span>
    <span class="k">if</span> <span class="n">matched_key</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Match found for &#39;</span><span class="si">{</span><span class="n">projection_source_production_name</span><span class="si">}</span><span class="s2">&#39; in &#39;</span><span class="si">{</span><span class="n">matched_key</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="n">matched_key</span><span class="p">]</span>
    <span class="k">elif</span> <span class="s2">&quot;homo_sapiens&quot;</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;No match for &#39;</span><span class="si">{</span><span class="n">projection_source_production_name</span><span class="si">}</span><span class="s2">&#39;. Falling back to &#39;homo_sapiens&#39;.&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;No match for &#39;</span><span class="si">{</span><span class="n">projection_source_production_name</span><span class="si">}</span><span class="s2">&#39; and &#39;homo_sapiens&#39; not found in </span><span class="si">{</span><span class="n">json_path</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.custom_loading" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">custom_loading</span><span class="p">(</span><span class="n">settings</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.custom_loading" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load custom key-value pairs from an INI-style initialization file specified in settings.</p>
<p>Reads the file line-by-line, extracting key=value pairs into a dictionary.
Lines that do not conform to key=value format or are blank are skipped with a message.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pipeline settings dictionary that must include
             the key 'init_file' pointing to the INI file path.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary containing key-value pairs loaded from the INI file.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="Exception">Exception</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the file cannot be opened or read.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">custom_loading</span><span class="p">(</span><span class="n">settings</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load custom key-value pairs from an INI-style initialization file specified in settings.</span>

<span class="sd">    Reads the file line-by-line, extracting key=value pairs into a dictionary.</span>
<span class="sd">    Lines that do not conform to key=value format or are blank are skipped with a message.</span>

<span class="sd">    Args:</span>
<span class="sd">        settings (dict): Pipeline settings dictionary that must include</span>
<span class="sd">                         the key &#39;init_file&#39; pointing to the INI file path.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary containing key-value pairs loaded from the INI file.</span>

<span class="sd">    Raises:</span>
<span class="sd">        Exception: If the file cannot be opened or read.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Custom loading started&quot;</span><span class="p">)</span>
    <span class="n">init_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">settings</span><span class="p">[</span><span class="s2">&quot;init_file&quot;</span><span class="p">])</span>

    <span class="c1"># Initialize variables</span>
    <span class="n">custom_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">custom_loading</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">if</span> <span class="n">init_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">init_file</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Using custom loading .ini file.&quot;</span><span class="p">)</span>
                <span class="n">custom_loading</span> <span class="o">=</span> <span class="kc">True</span>

                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

                    <span class="k">if</span> <span class="s2">&quot;=&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
                        <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="o">.</span><span class="n">strip</span><span class="p">,</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found key/value pair: </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2"> =&gt; </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="n">custom_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
                    <span class="k">elif</span> <span class="n">line</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Line format not recognised. Skipping line:</span><span class="se">\n</span><span class="si">{</span><span class="n">line</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">OSError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Could not open or read </span><span class="si">{</span><span class="n">init_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>

    <span class="k">return</span> <span class="n">custom_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.get_info_for_pipeline_anno" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_info_for_pipeline_anno</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">,</span> <span class="n">anno_settings</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_info_for_pipeline_anno" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Prepare and add file paths and pipeline-specific parameters needed for running annotation.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pipeline settings.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>info_dict</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Metadata dictionary for the assembly.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly_accession</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly accession string.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>anno_settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Annotation-specific settings.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated info_dict with added file paths and pipeline parameters.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_info_for_pipeline_anno</span><span class="p">(</span>
    <span class="n">settings</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">anno_settings</span><span class="p">:</span> <span class="nb">dict</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare and add file paths and pipeline-specific parameters needed for running annotation.</span>

<span class="sd">    Args:</span>
<span class="sd">        settings (dict): Pipeline settings.</span>
<span class="sd">        info_dict (dict): Metadata dictionary for the assembly.</span>
<span class="sd">        assembly_accession (str): Assembly accession string.</span>
<span class="sd">        anno_settings (dict): Annotation-specific settings.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Updated info_dict with added file paths and pipeline parameters.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting info for pipeline settings for </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">)</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">settings</span><span class="p">[</span><span class="s2">&quot;base_output_dir&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="n">assembly_accession</span>
    <span class="n">genome_files_dir</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;genome_files&quot;</span>
    <span class="n">toplevel_genome_file</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_toplevel.fa&quot;</span>

    <span class="n">short_read_dir</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;short_read_fastq&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;use_existing_short_read_dir&quot;</span> <span class="ow">in</span> <span class="n">anno_settings</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span>
        <span class="n">anno_settings</span><span class="p">[</span><span class="s2">&quot;use_existing_short_read_dir&quot;</span><span class="p">]</span>
    <span class="p">):</span>
        <span class="n">short_read_dir</span> <span class="o">=</span> <span class="n">anno_settings</span><span class="p">[</span><span class="s2">&quot;use_existing_short_read_dir&quot;</span><span class="p">]</span>

    <span class="n">long_read_dir</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;long_read_fastq&quot;</span>
    <span class="n">gst_dir</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;gst&quot;</span>
    <span class="n">rnaseq_summary_file</span> <span class="o">=</span> <span class="n">short_read_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;production_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">.csv&quot;</span>
    <span class="n">rnaseq_summary_file_genus</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">short_read_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;production_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_gen.csv&quot;</span>
    <span class="p">)</span>
    <span class="n">long_read_summary_file</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">long_read_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;production_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_long_read.csv&quot;</span>
    <span class="p">)</span>
    <span class="n">reheadered_toplevel_genome_file</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_reheadered_toplevel.fa&quot;</span>
    <span class="p">)</span>
    <span class="n">diamond_validation_db</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">anno_settings</span><span class="p">[</span><span class="s2">&quot;diamond_validation_db&quot;</span><span class="p">])</span>
    <span class="n">current_genebuild</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;current_genebuild&quot;</span><span class="p">]</span>
    <span class="n">num_threads</span> <span class="o">=</span> <span class="n">anno_settings</span><span class="p">[</span><span class="s2">&quot;num_threads&quot;</span><span class="p">]</span>
    <span class="n">ensembl_release</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;release_number&quot;</span><span class="p">]</span>

    <span class="c1"># Add values back to dictionary</span>
    <span class="n">info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;output_path&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">output_path</span><span class="p">),</span>
            <span class="s2">&quot;genome_files_dir&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">genome_files_dir</span><span class="p">),</span>
            <span class="s2">&quot;toplevel_genome_file&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">toplevel_genome_file</span><span class="p">),</span>
            <span class="s2">&quot;reheadered_toplevel_genome_file&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">reheadered_toplevel_genome_file</span><span class="p">),</span>
            <span class="s2">&quot;short_read_dir&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">short_read_dir</span><span class="p">),</span>
            <span class="s2">&quot;long_read_dir&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">long_read_dir</span><span class="p">),</span>
            <span class="s2">&quot;gst_dir&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">gst_dir</span><span class="p">),</span>
            <span class="s2">&quot;diamond_validation_db&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="n">diamond_validation_db</span><span class="p">),</span>
            <span class="s2">&quot;rnaseq_summary_file&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">rnaseq_summary_file</span><span class="p">),</span>
            <span class="s2">&quot;rnaseq_summary_file_genus&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">rnaseq_summary_file_genus</span><span class="p">),</span>
            <span class="s2">&quot;long_read_summary_file&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">long_read_summary_file</span><span class="p">),</span>
            <span class="s2">&quot;current_genebuild&quot;</span><span class="p">:</span> <span class="n">current_genebuild</span><span class="p">,</span>
            <span class="s2">&quot;assembly_accession&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">assembly_accession</span><span class="p">),</span>
            <span class="s2">&quot;num_threads&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_threads</span><span class="p">),</span>
            <span class="s2">&quot;ensembl_release&quot;</span><span class="p">:</span> <span class="n">ensembl_release</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">info_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.get_info_for_pipeline_main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_info_for_pipeline_main</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_info_for_pipeline_main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Prepare and add file paths and pipeline-specific parameters needed for running annotation.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pipeline settings.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>info_dict</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Metadata dictionary for the assembly.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly_accession</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly accession string.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated info_dict with added file paths and pipeline parameters.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_info_for_pipeline_main</span><span class="p">(</span>
    <span class="n">settings</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare and add file paths and pipeline-specific parameters needed for running annotation.</span>

<span class="sd">    Args:</span>
<span class="sd">        settings (dict): Pipeline settings.</span>
<span class="sd">        info_dict (dict): Metadata dictionary for the assembly.</span>
<span class="sd">        assembly_accession (str): Assembly accession string.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Updated info_dict with added file paths and pipeline parameters.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting info for pipeline settings for GCA </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">)</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">Path</span><span class="p">(</span><span class="n">settings</span><span class="p">[</span><span class="s2">&quot;base_output_dir&quot;</span><span class="p">])</span>
        <span class="o">/</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species_name&quot;</span><span class="p">]</span>
        <span class="o">/</span> <span class="n">assembly_accession</span>
    <span class="p">)</span>
    <span class="n">rnaseq_dir</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;rnaseq&quot;</span>
    <span class="n">long_read_dir</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;long_read&quot;</span>
    <span class="n">gst_dir</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;gst&quot;</span>
    <span class="n">rnaseq_summary_file</span> <span class="o">=</span> <span class="n">rnaseq_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">.csv&quot;</span>
    <span class="n">rnaseq_summary_file_genus</span> <span class="o">=</span> <span class="n">rnaseq_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_gen.csv&quot;</span>
    <span class="n">long_read_fastq_dir</span> <span class="o">=</span> <span class="n">long_read_dir</span> <span class="o">/</span> <span class="s2">&quot;input&quot;</span>
    <span class="n">current_genebuild</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;current_genebuild&quot;</span><span class="p">]</span>
    <span class="n">registry_file</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;Databases.pm&quot;</span>
    <span class="n">release_number</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;release_number&quot;</span><span class="p">]</span>
    <span class="n">dbname_accession</span> <span class="o">=</span> <span class="n">assembly_accession</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;v&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
    <span class="n">email_address</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;email&quot;</span><span class="p">]</span>
    <span class="n">dbowner</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;dbowner&quot;</span><span class="p">]</span>
    <span class="n">user_r</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user_r&quot;</span><span class="p">]</span>
    <span class="n">user</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">]</span>
    <span class="n">password</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">]</span>
    <span class="n">server_set</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;server_set&quot;</span><span class="p">]</span>
    <span class="n">pipeline_name</span> <span class="o">=</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;pipeline_name&quot;</span><span class="p">]</span>
    <span class="n">long_read_summary_file</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">long_read_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_long_read.csv&quot;</span>
    <span class="p">)</span>
    <span class="n">long_read_summary_file_genus</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">long_read_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_long_read_gen.csv&quot;</span>
    <span class="p">)</span>
    <span class="n">pipe_db_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">dbowner</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">dbname_accession</span><span class="si">}</span><span class="s2">_pipe_</span><span class="si">{</span><span class="n">release_number</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="c1"># Add values back to dictionary</span>
    <span class="n">info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;output_path&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">output_path</span><span class="p">),</span>
            <span class="s2">&quot;rnaseq_dir&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">rnaseq_dir</span><span class="p">),</span>
            <span class="s2">&quot;long_read_fastq_dir&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">long_read_fastq_dir</span><span class="p">),</span>
            <span class="s2">&quot;registry_file&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">registry_file</span><span class="p">),</span>
            <span class="s2">&quot;long_read_dir&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">long_read_dir</span><span class="p">),</span>
            <span class="s2">&quot;gst_dir&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">gst_dir</span><span class="p">),</span>
            <span class="s2">&quot;rnaseq_summary_file&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">rnaseq_summary_file</span><span class="p">),</span>
            <span class="s2">&quot;rnaseq_summary_file_genus&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">rnaseq_summary_file_genus</span><span class="p">),</span>
            <span class="s2">&quot;current_genebuild&quot;</span><span class="p">:</span> <span class="n">current_genebuild</span><span class="p">,</span>
            <span class="s2">&quot;assembly_accession&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">assembly_accession</span><span class="p">),</span>
            <span class="s2">&quot;release_number&quot;</span><span class="p">:</span> <span class="n">release_number</span><span class="p">,</span>
            <span class="s2">&quot;dbname_accession&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">dbname_accession</span><span class="p">),</span>
            <span class="s2">&quot;email_address&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">email_address</span><span class="p">),</span>
            <span class="s2">&quot;dbowner&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">dbowner</span><span class="p">),</span>
            <span class="s2">&quot;user_r&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">user_r</span><span class="p">),</span>
            <span class="s2">&quot;user&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">user</span><span class="p">),</span>
            <span class="s2">&quot;password&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">password</span><span class="p">),</span>
            <span class="s2">&quot;server_set&quot;</span><span class="p">:</span> <span class="n">server_set</span><span class="p">,</span>
            <span class="s2">&quot;pipeline_name&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">pipeline_name</span><span class="p">),</span>
            <span class="s2">&quot;long_read_summary_file&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">long_read_summary_file</span><span class="p">),</span>
            <span class="s2">&quot;long_read_summary_file_genus&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">long_read_summary_file_genus</span><span class="p">),</span>
            <span class="s2">&quot;pipe_db_name&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">pipe_db_name</span><span class="p">),</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">info_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.get_metadata_from_registry" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_metadata_from_registry</span><span class="p">(</span><span class="n">server_info</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">,</span> <span class="n">settings</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_metadata_from_registry" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Retrieve registry metadata for a given genome assembly accession or list of accessions.</p>
<p>Supports two modes:
- If 'init_file' is specified in settings, loads registry info via custom INI file parsing.
- Otherwise, queries the MySQL database for metadata matching the accession(s).</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL server connection info with 'registry' key.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly_accession</code>
            </td>
            <td>
                  <code><span title="str">str</span> or <span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Single GCA accession or list of GCAs.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Additional parameters, may include 'init_file'.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict or None: Assembly metadata dictionary if found, None if no record matches,
          or empty dict if an error occurs.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If 'init_file' is specified but not found.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_metadata_from_registry</span><span class="p">(</span>
    <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">,</span> <span class="n">settings</span><span class="p">:</span> <span class="nb">dict</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve registry metadata for a given genome assembly accession or list of accessions.</span>

<span class="sd">    Supports two modes:</span>
<span class="sd">    - If &#39;init_file&#39; is specified in settings, loads registry info via custom INI file parsing.</span>
<span class="sd">    - Otherwise, queries the MySQL database for metadata matching the accession(s).</span>

<span class="sd">    Args:</span>
<span class="sd">        server_info (dict): MySQL server connection info with &#39;registry&#39; key.</span>
<span class="sd">        assembly_accession (str or list): Single GCA accession or list of GCAs.</span>
<span class="sd">        settings (dict): Additional parameters, may include &#39;init_file&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict or None: Assembly metadata dictionary if found, None if no record matches,</span>
<span class="sd">                      or empty dict if an error occurs.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If &#39;init_file&#39; is specified but not found.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Check if using custom init file</span>
    <span class="k">if</span> <span class="s2">&quot;init_file&quot;</span> <span class="ow">in</span> <span class="n">settings</span> <span class="ow">and</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;init_file&quot;</span><span class="p">]:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Initialization file detected&quot;</span><span class="p">)</span>
        <span class="n">init_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">settings</span><span class="p">[</span><span class="s2">&quot;init_file&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">init_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Loading info from init file&quot;</span><span class="p">)</span>
            <span class="n">registry_info</span> <span class="o">=</span> <span class="n">custom_loading</span><span class="p">(</span><span class="n">settings</span><span class="p">)</span>
            <span class="n">registry_info</span> <span class="o">=</span> <span class="n">assign_clade_info_custom_loading</span><span class="p">(</span><span class="n">registry_info</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">registry_info</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;INI file not found: </span><span class="si">{</span><span class="n">init_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">assembly_accession</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="c1"># single accession → wrap in list</span>
            <span class="n">assembly_accessions</span> <span class="o">=</span> <span class="p">[</span><span class="n">assembly_accession</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">assembly_accessions</span> <span class="o">=</span> <span class="n">assembly_accession</span>

        <span class="c1"># Build placeholders for SQL query</span>
        <span class="n">placeholders</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">assembly_accessions</span><span class="p">))</span>

        <span class="n">registry_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">            SELECT </span>
<span class="s2">                s.species_taxon_id, </span>
<span class="s2">                a.lowest_taxon_id AS taxon_id,</span>
<span class="s2">                a.asm_name AS assembly_name,</span>
<span class="s2">                s.common_name, </span>
<span class="s2">                a.refseq_accession AS assembly_refseq_accession,</span>
<span class="s2">                a.release_date AS assembly_date,</span>
<span class="s2">                s.scientific_name AS species_name,</span>
<span class="s2">                a.assembly_id, </span>
<span class="s2">                mb.bioproject_name AS assembly_group</span>
<span class="s2">            FROM assembly a</span>
<span class="s2">            JOIN bioproject b ON a.assembly_id = b.assembly_id</span>
<span class="s2">            JOIN species s ON a.lowest_taxon_id = s.lowest_taxon_id</span>
<span class="s2">            LEFT JOIN main_bioproject mb ON b.bioproject_id = mb.bioproject_id</span>
<span class="s2">            WHERE CONCAT(a.gca_chain, &#39;.&#39;, a.gca_version) IN (</span><span class="si">{</span><span class="n">placeholders</span><span class="si">}</span><span class="s2">)</span>
<span class="s2">        &quot;&quot;&quot;</span>

        <span class="n">registry_info</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">registry_query</span><span class="p">,</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
            <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="n">params</span><span class="o">=</span><span class="n">assembly_accessions</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">registry_info</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;No registry data found for accessions: </span><span class="si">{</span><span class="n">assembly_accessions</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Registry query successful for </span><span class="si">{</span><span class="n">assembly_accessions</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">registry_info</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;MySQL error: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">err</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.get_rna_and_busco_check_threshold" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_rna_and_busco_check_threshold</span><span class="p">(</span><span class="n">anno_settings</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_rna_and_busco_check_threshold" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extract thresholds for RNA-seq and BUSCO checks from the pipeline settings.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>anno_settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pipeline settings dictionary.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary with keys 'busco_threshold', 'busco_lower_threshold',
  'busco_difference_threshold', 'rnaseq_main_file_min_lines',
  and 'rnaseq_genus_file_min_lines'.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_rna_and_busco_check_threshold</span><span class="p">(</span><span class="n">anno_settings</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract thresholds for RNA-seq and BUSCO checks from the pipeline settings.</span>

<span class="sd">    Args:</span>
<span class="sd">        anno_settings (dict): Pipeline settings dictionary.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary with keys &#39;busco_threshold&#39;, &#39;busco_lower_threshold&#39;,</span>
<span class="sd">              &#39;busco_difference_threshold&#39;, &#39;rnaseq_main_file_min_lines&#39;,</span>
<span class="sd">              and &#39;rnaseq_genus_file_min_lines&#39;.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="p">{</span>
        <span class="s2">&quot;busco_threshold&quot;</span><span class="p">:</span> <span class="n">anno_settings</span><span class="p">[</span><span class="s2">&quot;busco_threshold&quot;</span><span class="p">],</span>
        <span class="s2">&quot;busco_lower_threshold&quot;</span><span class="p">:</span> <span class="n">anno_settings</span><span class="p">[</span><span class="s2">&quot;busco_lower_threshold&quot;</span><span class="p">],</span>
        <span class="s2">&quot;busco_difference_threshold&quot;</span><span class="p">:</span> <span class="n">anno_settings</span><span class="p">[</span><span class="s2">&quot;busco_difference_threshold&quot;</span><span class="p">],</span>
        <span class="s2">&quot;rnaseq_main_file_min_lines&quot;</span><span class="p">:</span> <span class="n">anno_settings</span><span class="p">[</span><span class="s2">&quot;rnaseq_main_file_min_lines&quot;</span><span class="p">],</span>
        <span class="s2">&quot;rnaseq_genus_file_min_lines&quot;</span><span class="p">:</span> <span class="n">anno_settings</span><span class="p">[</span><span class="s2">&quot;rnaseq_genus_file_min_lines&quot;</span><span class="p">],</span>
    <span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.get_server_settings_anno" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_server_settings_anno</span><span class="p">(</span><span class="n">settings</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_server_settings_anno" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Determine and return server connection settings for pipeline and core databases.</p>
<p>Uses either custom server settings from the configuration or falls back
to environment-variable-based defaults depending on the 'server_set' parameter.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The pipeline settings dictionary.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Nested dictionary with keys 'pipeline_db' and 'core_db', each containing
  connection parameters like 'db_host', 'db_user', 'db_port', and 'db_password'.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the 'server_set' value is unknown or unsupported.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_server_settings_anno</span><span class="p">(</span><span class="n">settings</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determine and return server connection settings for pipeline and core databases.</span>

<span class="sd">    Uses either custom server settings from the configuration or falls back</span>
<span class="sd">    to environment-variable-based defaults depending on the &#39;server_set&#39; parameter.</span>

<span class="sd">    Args:</span>
<span class="sd">        settings (dict): The pipeline settings dictionary.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Nested dictionary with keys &#39;pipeline_db&#39; and &#39;core_db&#39;, each containing</span>
<span class="sd">              connection parameters like &#39;db_host&#39;, &#39;db_user&#39;, &#39;db_port&#39;, and &#39;db_password&#39;.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the &#39;server_set&#39; value is unknown or unsupported.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting server settings&quot;</span><span class="p">)</span>
    <span class="n">custom</span> <span class="o">=</span> <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;custom_server&quot;</span><span class="p">,</span> <span class="p">{})</span>
    <span class="c1"># Check if all custom_server values are non-empty</span>
    <span class="k">if</span> <span class="nb">all</span><span class="p">(</span>
        <span class="n">custom</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span>
            <span class="s2">&quot;pipeline_db_host&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pipeline_db_port&quot;</span><span class="p">,</span>
            <span class="s2">&quot;core_db_host&quot;</span><span class="p">,</span>
            <span class="s2">&quot;core_db_port&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Custom server settings detected&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;pipeline_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;pipeline_db_host&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;pipeline_db_port&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
            <span class="p">},</span>
            <span class="s2">&quot;core_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;core_db_host&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;core_db_port&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
            <span class="p">},</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Fallback based on server_set</span>
        <span class="n">server_set</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;server_set&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">))</span>  <span class="c1"># default to &quot;1&quot; if missing</span>

        <span class="k">if</span> <span class="n">server_set</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Server set 1 detected&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">{</span>
                <span class="s2">&quot;pipeline_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS4&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP4&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="s2">&quot;core_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS3&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP3&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
            <span class="p">}</span>

        <span class="k">elif</span> <span class="n">server_set</span> <span class="o">==</span> <span class="s2">&quot;2&quot;</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Server set 2 detected&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">{</span>
                <span class="s2">&quot;pipeline_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS7&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP7&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="s2">&quot;core_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS6&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP6&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
            <span class="p">}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown server_set value: </span><span class="si">{</span><span class="n">server_set</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.get_server_settings_main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_server_settings_main</span><span class="p">(</span><span class="n">settings</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.get_server_settings_main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Determine and return server connection settings for pipeline and core databases.</p>
<p>Uses either custom server settings from the configuration or falls back
to environment-variable-based defaults depending on the 'server_set' parameter.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>settings</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The pipeline settings dictionary.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Nested dictionary with keys 'pipeline_db' and 'core_db', each containing
  connection parameters like 'db_host', 'db_user', 'db_port', and 'db_password'.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the 'server_set' value is unknown or unsupported.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_server_settings_main</span><span class="p">(</span><span class="n">settings</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determine and return server connection settings for pipeline and core databases.</span>

<span class="sd">    Uses either custom server settings from the configuration or falls back</span>
<span class="sd">    to environment-variable-based defaults depending on the &#39;server_set&#39; parameter.</span>

<span class="sd">    Args:</span>
<span class="sd">        settings (dict): The pipeline settings dictionary.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Nested dictionary with keys &#39;pipeline_db&#39; and &#39;core_db&#39;, each containing</span>
<span class="sd">              connection parameters like &#39;db_host&#39;, &#39;db_user&#39;, &#39;db_port&#39;, and &#39;db_password&#39;.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the &#39;server_set&#39; value is unknown or unsupported.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting server settings&quot;</span><span class="p">)</span>
    <span class="n">custom</span> <span class="o">=</span> <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;custom_server&quot;</span><span class="p">,</span> <span class="p">{})</span>
    <span class="c1"># Check if all custom_server values are non-empty</span>
    <span class="k">if</span> <span class="nb">all</span><span class="p">(</span>
        <span class="n">custom</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span>
            <span class="s2">&quot;pipeline_db_host&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pipeline_db_port&quot;</span><span class="p">,</span>
            <span class="s2">&quot;core_db_host&quot;</span><span class="p">,</span>
            <span class="s2">&quot;core_db_port&quot;</span><span class="p">,</span>
            <span class="s2">&quot;databases_host&quot;</span><span class="p">,</span>
            <span class="s2">&quot;databases_port&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Custom server settings detected&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;pipeline_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;pipeline_db_host&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;pipeline_db_port&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
            <span class="p">},</span>
            <span class="s2">&quot;core_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;core_db_host&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;core_db_port&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
            <span class="p">},</span>
            <span class="s2">&quot;databases&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;databases_host&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="n">custom</span><span class="p">[</span><span class="s2">&quot;databases_port&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
            <span class="p">},</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Fallback based on server_set</span>
        <span class="n">server_set</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;server_set&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">))</span>  <span class="c1"># default to &quot;1&quot; if missing</span>

        <span class="k">if</span> <span class="n">server_set</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Server set 1 detected&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">{</span>
                <span class="s2">&quot;pipeline_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS4&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP4&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="s2">&quot;core_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS2&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP2&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="s2">&quot;databases&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS3&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP3&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
            <span class="p">}</span>

        <span class="k">elif</span> <span class="n">server_set</span> <span class="o">==</span> <span class="s2">&quot;2&quot;</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Server set 2 detected&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">{</span>
                <span class="s2">&quot;pipeline_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS7&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP7&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="s2">&quot;core_db&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS5&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP5&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="s2">&quot;databases&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS6&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP6&quot;</span><span class="p">)),</span>
                    <span class="s2">&quot;db_password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="p">},</span>
            <span class="p">}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown server_set value: </span><span class="si">{</span><span class="n">server_set</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.load_anno_settings" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">load_anno_settings</span><span class="p">()</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.load_anno_settings" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load the annotation-specific settings from a hardcoded path relative to
the environment variable 'ENSCODE'.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Parsed annotation settings dictionary.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the anno_settings.json file is not found.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="json.JSONDecodeError">JSONDecodeError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the file contents are not valid JSON.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">load_anno_settings</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load the annotation-specific settings from a hardcoded path relative to</span>
<span class="sd">    the environment variable &#39;ENSCODE&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Parsed annotation settings dictionary.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the anno_settings.json file is not found.</span>
<span class="sd">        json.JSONDecodeError: If the file contents are not valid JSON.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Loading anno settings json&quot;</span><span class="p">)</span>
    <span class="n">settings</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">),</span>
        <span class="s2">&quot;ensembl-genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;src&quot;</span><span class="p">,</span>
        <span class="s2">&quot;python&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ensembl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;info_from_registry&quot;</span><span class="p">,</span>
        <span class="s2">&quot;anno_settings.json&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.load_main_settings" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">load_main_settings</span><span class="p">()</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.load_main_settings" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load the annotation-specific settings from a hardcoded path relative to
the environment variable 'ENSCODEf'.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Parsed annotation settings dictionary.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the anno_settings.json file is not found.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="json.JSONDecodeError">JSONDecodeError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the file contents are not valid JSON.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">load_main_settings</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load the annotation-specific settings from a hardcoded path relative to</span>
<span class="sd">    the environment variable &#39;ENSCODEf&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Parsed annotation settings dictionary.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the anno_settings.json file is not found.</span>
<span class="sd">        json.JSONDecodeError: If the file contents are not valid JSON.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Loading main settings json&quot;</span><span class="p">)</span>
    <span class="n">settings</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">),</span>
        <span class="s2">&quot;ensembl-genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;src&quot;</span><span class="p">,</span>
        <span class="s2">&quot;python&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ensembl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;info_from_registry&quot;</span><span class="p">,</span>
        <span class="s2">&quot;main_settings.json&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.load_settings" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">load_settings</span><span class="p">(</span><span class="n">settings_file</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.load_settings" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load JSON-formatted pipeline settings from a given file path.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>settings_file</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the JSON settings file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Parsed settings dictionary.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the settings file does not exist.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="json.JSONDecodeError">JSONDecodeError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the file contents are not valid JSON.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">load_settings</span><span class="p">(</span><span class="n">settings_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load JSON-formatted pipeline settings from a given file path.</span>

<span class="sd">    Args:</span>
<span class="sd">        settings_file (str): Path to the JSON settings file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Parsed settings dictionary.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the settings file does not exist.</span>
<span class="sd">        json.JSONDecodeError: If the file contents are not valid JSON.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading settings from: </span><span class="si">{</span><span class="n">settings_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">settings_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">settings_file</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">settings_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Settings file not found: </span><span class="si">{</span><span class="n">settings_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">settings_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.start_pipeline_from_registry.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">(</span><span class="n">gcas</span><span class="p">,</span> <span class="n">settings_file</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.start_pipeline_from_registry.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create parameters for annotation pipeline.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gcas</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the list of GCAs</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>settings_file </code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pipeline settings JSON</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>all_output_params</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of all GCA output parameters</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>saved_paths</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Paths to saved JSONs, with separate entries for 'anno' and 'main'</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/start_pipeline_from_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 801</span>
<span class="normal"> 802</span>
<span class="normal"> 803</span>
<span class="normal"> 804</span>
<span class="normal"> 805</span>
<span class="normal"> 806</span>
<span class="normal"> 807</span>
<span class="normal"> 808</span>
<span class="normal"> 809</span>
<span class="normal"> 810</span>
<span class="normal"> 811</span>
<span class="normal"> 812</span>
<span class="normal"> 813</span>
<span class="normal"> 814</span>
<span class="normal"> 815</span>
<span class="normal"> 816</span>
<span class="normal"> 817</span>
<span class="normal"> 818</span>
<span class="normal"> 819</span>
<span class="normal"> 820</span>
<span class="normal"> 821</span>
<span class="normal"> 822</span>
<span class="normal"> 823</span>
<span class="normal"> 824</span>
<span class="normal"> 825</span>
<span class="normal"> 826</span>
<span class="normal"> 827</span>
<span class="normal"> 828</span>
<span class="normal"> 829</span>
<span class="normal"> 830</span>
<span class="normal"> 831</span>
<span class="normal"> 832</span>
<span class="normal"> 833</span>
<span class="normal"> 834</span>
<span class="normal"> 835</span>
<span class="normal"> 836</span>
<span class="normal"> 837</span>
<span class="normal"> 838</span>
<span class="normal"> 839</span>
<span class="normal"> 840</span>
<span class="normal"> 841</span>
<span class="normal"> 842</span>
<span class="normal"> 843</span>
<span class="normal"> 844</span>
<span class="normal"> 845</span>
<span class="normal"> 846</span>
<span class="normal"> 847</span>
<span class="normal"> 848</span>
<span class="normal"> 849</span>
<span class="normal"> 850</span>
<span class="normal"> 851</span>
<span class="normal"> 852</span>
<span class="normal"> 853</span>
<span class="normal"> 854</span>
<span class="normal"> 855</span>
<span class="normal"> 856</span>
<span class="normal"> 857</span>
<span class="normal"> 858</span>
<span class="normal"> 859</span>
<span class="normal"> 860</span>
<span class="normal"> 861</span>
<span class="normal"> 862</span>
<span class="normal"> 863</span>
<span class="normal"> 864</span>
<span class="normal"> 865</span>
<span class="normal"> 866</span>
<span class="normal"> 867</span>
<span class="normal"> 868</span>
<span class="normal"> 869</span>
<span class="normal"> 870</span>
<span class="normal"> 871</span>
<span class="normal"> 872</span>
<span class="normal"> 873</span>
<span class="normal"> 874</span>
<span class="normal"> 875</span>
<span class="normal"> 876</span>
<span class="normal"> 877</span>
<span class="normal"> 878</span>
<span class="normal"> 879</span>
<span class="normal"> 880</span>
<span class="normal"> 881</span>
<span class="normal"> 882</span>
<span class="normal"> 883</span>
<span class="normal"> 884</span>
<span class="normal"> 885</span>
<span class="normal"> 886</span>
<span class="normal"> 887</span>
<span class="normal"> 888</span>
<span class="normal"> 889</span>
<span class="normal"> 890</span>
<span class="normal"> 891</span>
<span class="normal"> 892</span>
<span class="normal"> 893</span>
<span class="normal"> 894</span>
<span class="normal"> 895</span>
<span class="normal"> 896</span>
<span class="normal"> 897</span>
<span class="normal"> 898</span>
<span class="normal"> 899</span>
<span class="normal"> 900</span>
<span class="normal"> 901</span>
<span class="normal"> 902</span>
<span class="normal"> 903</span>
<span class="normal"> 904</span>
<span class="normal"> 905</span>
<span class="normal"> 906</span>
<span class="normal"> 907</span>
<span class="normal"> 908</span>
<span class="normal"> 909</span>
<span class="normal"> 910</span>
<span class="normal"> 911</span>
<span class="normal"> 912</span>
<span class="normal"> 913</span>
<span class="normal"> 914</span>
<span class="normal"> 915</span>
<span class="normal"> 916</span>
<span class="normal"> 917</span>
<span class="normal"> 918</span>
<span class="normal"> 919</span>
<span class="normal"> 920</span>
<span class="normal"> 921</span>
<span class="normal"> 922</span>
<span class="normal"> 923</span>
<span class="normal"> 924</span>
<span class="normal"> 925</span>
<span class="normal"> 926</span>
<span class="normal"> 927</span>
<span class="normal"> 928</span>
<span class="normal"> 929</span>
<span class="normal"> 930</span>
<span class="normal"> 931</span>
<span class="normal"> 932</span>
<span class="normal"> 933</span>
<span class="normal"> 934</span>
<span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">gcas</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">settings_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">dict</span><span class="p">,</span> <span class="nb">dict</span><span class="p">,</span> <span class="nb">dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create parameters for annotation pipeline.</span>

<span class="sd">    Args:</span>
<span class="sd">        gcas (str): Path to the list of GCAs</span>
<span class="sd">        settings_file : Pipeline settings JSON</span>

<span class="sd">    Returns:</span>
<span class="sd">        all_output_params (dict): Dictionary of all GCA output parameters</span>
<span class="sd">        saved_paths (dict): Paths to saved JSONs, with separate entries for &#39;anno&#39; and &#39;main&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Load settings</span>
    <span class="n">settings</span> <span class="o">=</span> <span class="n">load_settings</span><span class="p">(</span><span class="n">settings_file</span><span class="p">)</span>

    <span class="c1"># Read in GCAs from file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gcas</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span> <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span>

    <span class="n">gca_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">gca</span><span class="p">:</span> <span class="p">{}</span> <span class="k">for</span> <span class="n">gca</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">}</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">gca_dict</span><span class="p">)</span><span class="si">}</span><span class="s2"> GCAs&quot;</span><span class="p">)</span>

    <span class="c1"># Check if init_file exists</span>
    <span class="n">has_init_file</span> <span class="o">=</span> <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;init_file&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">settings</span><span class="p">[</span><span class="s2">&quot;init_file&quot;</span><span class="p">])</span>

    <span class="n">all_output_params</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Loop through GCAs</span>
    <span class="k">for</span> <span class="n">gca</span> <span class="ow">in</span> <span class="n">gca_dict</span><span class="p">:</span>
        <span class="n">gca_dict</span><span class="p">[</span><span class="n">gca</span><span class="p">][</span><span class="s2">&quot;assembly_accession&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gca</span>

        <span class="c1"># Registry info</span>
        <span class="n">server_info</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;registry&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;db_host&quot;</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBS1&quot;</span><span class="p">),</span>
                <span class="s2">&quot;db_user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user_r&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_user_w&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                <span class="s2">&quot;db_port&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GBP1&quot;</span><span class="p">)),</span>
                <span class="s2">&quot;db_name&quot;</span><span class="p">:</span> <span class="s2">&quot;gb_assembly_metadata&quot;</span><span class="p">,</span>
                <span class="s2">&quot;password&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
            <span class="p">}</span>
        <span class="p">}</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">])</span>

        <span class="c1"># Check if it has annotation history</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">has_init_file</span> <span class="ow">and</span> <span class="nb">int</span><span class="p">(</span><span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;current_genebuild&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">check_if_annotated</span><span class="p">(</span><span class="n">gca</span><span class="p">,</span> <span class="n">server_info</span><span class="p">)</span>

        <span class="c1"># Create output params</span>
        <span class="n">info_dict</span> <span class="o">=</span> <span class="n">add_generated_data</span><span class="p">(</span><span class="n">server_info</span><span class="p">,</span> <span class="n">gca</span><span class="p">,</span> <span class="n">settings</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;repbase_library&quot;</span> <span class="ow">in</span> <span class="n">info_dict</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;repbase_library&quot;</span><span class="p">],</span> <span class="nb">str</span>
        <span class="p">):</span>
            <span class="c1"># Vertebrate (main) pipeline</span>
            <span class="n">pipeline_type</span> <span class="o">=</span> <span class="s2">&quot;main&quot;</span>

            <span class="n">server_settings</span> <span class="o">=</span> <span class="n">get_server_settings_main</span><span class="p">(</span><span class="n">settings</span><span class="p">)</span>
            <span class="n">server_info</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">server_settings</span><span class="p">)</span>
            <span class="n">main_settings</span> <span class="o">=</span> <span class="n">load_main_settings</span><span class="p">()</span>

            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;databases_host&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;databases&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;databases_port&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;databases&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;registry_host&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;registry_port&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;registry_db&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;pipe_db_host&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;pipeline_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;pipe_db_port&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;pipeline_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;dna_db_host&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;core_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;dna_db_port&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;core_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">main_settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;replace_repbase_with_red_to_mask&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;first_choice_repeat&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;repeatdetector&quot;</span>

            <span class="c1"># Protein BLAST DB based on clade</span>
            <span class="n">clade</span> <span class="o">=</span> <span class="n">info_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;clade&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">clade</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;mammalia&quot;</span><span class="p">,</span> <span class="s2">&quot;rodentia&quot;</span><span class="p">,</span> <span class="s2">&quot;primates&quot;</span><span class="p">,</span> <span class="s2">&quot;marsupials&quot;</span><span class="p">]:</span>
                <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;protein_blast_db_file&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;uniprot_mammalia_sp&quot;</span>
            <span class="k">elif</span> <span class="n">clade</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;teleostei&quot;</span><span class="p">,</span> <span class="s2">&quot;sharks&quot;</span><span class="p">]:</span>
                <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;protein_blast_db_file&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;uniprot_vertebrataSP_plus_fishTR&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;protein_blast_db_file&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;uniprot_vertebrata_sp&quot;</span>

            <span class="c1"># RepeatModeler library</span>
            <span class="n">parent_name</span> <span class="o">=</span> <span class="n">get_parent_taxon</span><span class="p">(</span><span class="n">server_info</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species_taxon_id&quot;</span><span class="p">])</span>
            <span class="n">parent_name</span> <span class="o">=</span> <span class="n">parent_name</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;repeatmodeler_library&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;REPEATMODELER_DIR&quot;</span><span class="p">],</span>
                <span class="s2">&quot;species&quot;</span><span class="p">,</span>
                <span class="n">parent_name</span><span class="p">,</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">parent_name</span><span class="si">}</span><span class="s2">.repeatmodeler.fa&quot;</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">output_params</span> <span class="o">=</span> <span class="n">get_info_for_pipeline_main</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">,</span> <span class="n">gca</span><span class="p">)</span>
            <span class="n">create_dir</span><span class="p">(</span><span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;output_path&quot;</span><span class="p">])</span>
            <span class="n">edit_config_main</span><span class="p">(</span><span class="n">main_settings</span><span class="p">,</span> <span class="n">output_params</span><span class="p">,</span> <span class="n">pipeline_type</span><span class="p">)</span>

            <span class="n">projection_source_info</span> <span class="o">=</span> <span class="n">current_projection_source_db</span><span class="p">(</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;projection_source_production_name&quot;</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="n">output_params</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="n">projection_source_info</span>
            <span class="p">)</span>  <span class="c1"># Adds db_name, server, port to output_params</span>

            <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;stable_id_prefix&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_species_prefix</span><span class="p">(</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">],</span> <span class="n">server_info</span>
            <span class="p">)</span>
            <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;stable_id_start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_stable_space</span><span class="p">(</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">],</span>
                <span class="n">gca</span><span class="p">,</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;assembly_id&quot;</span><span class="p">],</span>
                <span class="n">server_info</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">edit_config_main</span><span class="p">(</span><span class="n">main_settings</span><span class="p">,</span> <span class="n">output_params</span><span class="p">,</span> <span class="n">pipeline_type</span><span class="p">)</span>
            <span class="n">copy_general_module</span><span class="p">()</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Non-vertebrate (anno) pipeline</span>
            <span class="n">pipeline_type</span> <span class="o">=</span> <span class="s2">&quot;anno&quot;</span>
            <span class="n">gca_dict</span><span class="p">[</span><span class="n">gca</span><span class="p">][</span>
                <span class="s2">&quot;pipe_db_name&quot;</span>
            <span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;dbowner&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;pipeline_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_pipe_</span><span class="si">{</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;release_number&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">server_settings</span> <span class="o">=</span> <span class="n">get_server_settings_anno</span><span class="p">(</span><span class="n">settings</span><span class="p">)</span>
            <span class="n">server_info</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">server_settings</span><span class="p">)</span>

            <span class="n">server_info</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;pipeline_db&quot;</span><span class="p">,</span> <span class="p">{})[</span><span class="s2">&quot;db_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gca_dict</span><span class="p">[</span><span class="n">gca</span><span class="p">][</span>
                <span class="s2">&quot;pipe_db_name&quot;</span>
            <span class="p">]</span>
            <span class="n">server_info</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;core_db&quot;</span><span class="p">,</span> <span class="p">{})[</span><span class="s2">&quot;db_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;core_dbname&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;core_db&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;core_db&quot;</span><span class="p">]</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;registry_db&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">]</span>

            <span class="c1"># Assign BUSCO lineage</span>
            <span class="n">busco_lineage_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">),</span>
                <span class="s2">&quot;ensembl-genes&quot;</span><span class="p">,</span>
                <span class="s2">&quot;src&quot;</span><span class="p">,</span>
                <span class="s2">&quot;python&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ensembl&quot;</span><span class="p">,</span>
                <span class="s2">&quot;genes&quot;</span><span class="p">,</span>
                <span class="s2">&quot;metrics&quot;</span><span class="p">,</span>
                <span class="s2">&quot;busco_lineage.json&quot;</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">busco_lineage_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">dataset</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

            <span class="n">ncbi_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://api.ncbi.nlm.nih.gov/datasets/v2alpha/taxonomy/taxon/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/dataset_report&quot;</span>

            <span class="n">busco_group_find</span> <span class="o">=</span> <span class="n">get_dataset_match</span><span class="p">(</span><span class="n">ncbi_url</span><span class="p">,</span> <span class="n">dataset</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">busco_group_find</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Closest BUSCO lineage identified as </span><span class="si">{</span><span class="n">busco_group_find</span><span class="si">}</span><span class="s2"> for taxon ID </span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;busco_group&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">busco_group_find</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Falling back on BUSCO lineage from clade settings </span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;busco_group&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

            <span class="c1"># Load anno settings</span>
            <span class="n">anno_settings</span> <span class="o">=</span> <span class="n">load_anno_settings</span><span class="p">()</span>

            <span class="c1"># DB adaptors</span>
            <span class="n">core_adaptor</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;host&quot;</span><span class="p">:</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;core_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
                <span class="s2">&quot;port&quot;</span><span class="p">:</span> <span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;core_db&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
                <span class="s2">&quot;dbname&quot;</span><span class="p">:</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;core_dbname&quot;</span><span class="p">],</span>
                <span class="s2">&quot;user&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;user&quot;</span><span class="p">],</span>
                <span class="s2">&quot;pass&quot;</span><span class="p">:</span> <span class="n">settings</span><span class="p">[</span><span class="s2">&quot;password&quot;</span><span class="p">],</span>
                <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;production_name&quot;</span><span class="p">],</span>
                <span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;core&quot;</span><span class="p">,</span>
            <span class="p">}</span>
            <span class="n">registry_path</span> <span class="o">=</span> <span class="n">create_registry_entry</span><span class="p">(</span><span class="n">settings</span><span class="p">,</span> <span class="n">server_info</span><span class="p">,</span> <span class="n">core_adaptor</span><span class="p">)</span>

            <span class="n">output_params</span> <span class="o">=</span> <span class="n">get_info_for_pipeline_anno</span><span class="p">(</span>
                <span class="n">settings</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">,</span> <span class="n">gca</span><span class="p">,</span> <span class="n">anno_settings</span>
            <span class="p">)</span>
            <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;registry_file&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">registry_path</span><span class="p">)</span>

            <span class="n">build_annotation_commands</span><span class="p">(</span>
                <span class="n">core_adaptor</span><span class="p">,</span> <span class="n">output_params</span><span class="p">,</span> <span class="n">anno_settings</span><span class="p">,</span> <span class="n">settings</span>
            <span class="p">)</span>

            <span class="n">rna_busco_settings</span> <span class="o">=</span> <span class="n">get_rna_and_busco_check_threshold</span><span class="p">(</span><span class="n">anno_settings</span><span class="p">)</span>
            <span class="n">output_params</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">rna_busco_settings</span><span class="p">)</span>
            <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;stable_id_prefix&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_species_prefix</span><span class="p">(</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">],</span> <span class="n">server_info</span>
            <span class="p">)</span>
            <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;stable_id_start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_stable_space</span><span class="p">(</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">],</span>
                <span class="n">gca</span><span class="p">,</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;assembly_id&quot;</span><span class="p">],</span>
                <span class="n">server_info</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="c1"># Create directories</span>
            <span class="n">create_dir</span><span class="p">(</span><span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;output_path&quot;</span><span class="p">])</span>
            <span class="n">dirs_to_create</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;genome_files_dir&quot;</span><span class="p">],</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;short_read_dir&quot;</span><span class="p">],</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;long_read_dir&quot;</span><span class="p">],</span>
                <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;gst_dir&quot;</span><span class="p">],</span>
            <span class="p">]</span>

            <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">dirs_to_create</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to create dir: </span><span class="si">{</span><span class="n">d</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>

        <span class="c1"># Store output params</span>
        <span class="n">output_params</span><span class="p">[</span><span class="s2">&quot;pipeline&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pipeline_type</span>
        <span class="n">all_output_params</span><span class="p">[</span><span class="n">gca</span><span class="p">]</span> <span class="o">=</span> <span class="n">output_params</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Finished with </span><span class="si">{</span><span class="n">gca</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># -------------------------</span>
    <span class="c1"># Save JSONs</span>
    <span class="c1"># -------------------------</span>
    <span class="n">saved_paths</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Save anno GCAs as a single file</span>
    <span class="n">anno_params</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">g</span><span class="p">:</span> <span class="n">p</span> <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_output_params</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">p</span><span class="p">[</span><span class="s2">&quot;pipeline&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;anno&quot;</span>
    <span class="p">}</span>

    <span class="c1"># Edit anno config</span>
    <span class="k">if</span> <span class="n">anno_params</span><span class="p">:</span>
        <span class="c1"># Call edit_config_anno ONCE here with only the generic settings</span>
        <span class="n">first_gca_params</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">anno_params</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
        <span class="n">edit_config_anno</span><span class="p">(</span>
            <span class="n">anno_settings</span><span class="p">,</span>
            <span class="n">settings</span><span class="p">,</span>
            <span class="n">first_gca_params</span><span class="p">,</span>  <span class="c1"># just need a representative dict with registry_file, db names, etc.</span>
            <span class="s2">&quot;anno&quot;</span><span class="p">,</span>
            <span class="n">server_settings</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">anno_json_path</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">settings</span><span class="p">[</span><span class="s2">&quot;base_output_dir&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="s2">&quot;non_vert_pipeline_params.json&quot;</span>
        <span class="p">)</span>
        <span class="n">anno_json_path</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">anno_json_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">anno_params</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved anno parameters to: </span><span class="si">{</span><span class="n">anno_json_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">saved_paths</span><span class="p">[</span><span class="s2">&quot;anno&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">anno_json_path</span>

    <span class="c1"># Save main GCAs as separate files</span>
    <span class="n">main_params</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">g</span><span class="p">:</span> <span class="n">p</span> <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_output_params</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">p</span><span class="p">[</span><span class="s2">&quot;pipeline&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;main&quot;</span>
    <span class="p">}</span>
    <span class="n">main_json_paths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">gca</span><span class="p">,</span> <span class="n">params</span> <span class="ow">in</span> <span class="n">main_params</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">params</span><span class="p">[</span><span class="s2">&quot;output_path&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="s2">&quot;main_pipeline_params.json&quot;</span>
        <span class="n">path</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">({</span><span class="n">gca</span><span class="p">:</span> <span class="n">params</span><span class="p">},</span> <span class="n">f</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved main parameters for </span><span class="si">{</span><span class="n">gca</span><span class="si">}</span><span class="s2"> to: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">main_json_paths</span><span class="p">[</span><span class="n">gca</span><span class="p">]</span> <span class="o">=</span> <span class="n">path</span>

    <span class="k">if</span> <span class="n">main_json_paths</span><span class="p">:</span>
        <span class="n">saved_paths</span><span class="p">[</span><span class="s2">&quot;main&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">main_json_paths</span>

    <span class="c1"># Log if mixed pipelines are present</span>
    <span class="n">pipelines_present</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">p</span><span class="p">[</span><span class="s2">&quot;pipeline&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_output_params</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="k">if</span> <span class="s2">&quot;anno&quot;</span> <span class="ow">in</span> <span class="n">pipelines_present</span> <span class="ow">and</span> <span class="s2">&quot;main&quot;</span> <span class="ow">in</span> <span class="n">pipelines_present</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;ATTENTION: Both &#39;anno&#39; and &#39;main&#39; GCAs detected in input!&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;DONE&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">server_info</span><span class="p">,</span> <span class="n">all_output_params</span><span class="p">,</span> <span class="n">saved_paths</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.taxonomy_helper" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.taxonomy_helper</code>


<a href="#ensembl.genes.info_from_registry.taxonomy_helper" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Module to assign clade information based on taxonomy data retrieved
from a registry MySQL database and static JSON configuration.</p>


<p><span class="doc-section-title">Functions:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><span title="ensembl.genes.info_from_registry.taxonomy_helper.- create_tax_dictionary_from_registry">- create_tax_dictionary_from_registry</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Fetches taxonomy hierarchy from the registry.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="ensembl.genes.info_from_registry.taxonomy_helper.- load_clade_data">- load_clade_data</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Loads clade definitions from a static JSON file.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="ensembl.genes.info_from_registry.taxonomy_helper.- assign_clade">- assign_clade</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Matches a taxon to a clade based on hierarchy.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="ensembl.genes.info_from_registry.taxonomy_helper.- assign_clade_info_custom_loading">- assign_clade_info_custom_loading</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Loads clade details by name from JSON config.</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.taxonomy_helper.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.taxonomy_helper.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.taxonomy_helper.assign_clade" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">assign_clade</span><span class="p">(</span><span class="n">server_info</span><span class="p">,</span> <span class="n">registry_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.taxonomy_helper.assign_clade" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Assign a clade to a given taxon based on clade data and taxonomy hierarchy.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL connection information under 'registry'.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>registry_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary with at least a 'taxon_id' key.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>tuple</code></td>            <td>
                  <code><span title="tuple">tuple</span>[<span title="str">str</span>, <span title="typing.Optional">Optional</span>[<span title="int">int</span>], <span title="typing.Optional">Optional</span>[<span title="dict">dict</span>[<span title="str">str</span>, <span title="any">any</span>]]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <ul>
<li>internal_clade (str): Name of the assigned clade (or 'Unassigned').</li>
<li>genus_taxon_id (int or None): Genus-level taxon ID, if found.</li>
<li>clade_details (dict or None): Dictionary of clade properties excluding taxon_id.</li>
</ul>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/taxonomy_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">assign_clade</span><span class="p">(</span>
    <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">registry_info</span><span class="p">:</span> <span class="nb">dict</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">any</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Assign a clade to a given taxon based on clade data and taxonomy hierarchy.</span>

<span class="sd">    Args:</span>
<span class="sd">        server_info (dict): MySQL connection information under &#39;registry&#39;.</span>
<span class="sd">        registry_info (dict): Dictionary with at least a &#39;taxon_id&#39; key.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple:</span>
<span class="sd">            - internal_clade (str): Name of the assigned clade (or &#39;Unassigned&#39;).</span>
<span class="sd">            - genus_taxon_id (int or None): Genus-level taxon ID, if found.</span>
<span class="sd">            - clade_details (dict or None): Dictionary of clade properties excluding taxon_id.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">clade_data</span> <span class="o">=</span> <span class="n">load_clade_data</span><span class="p">()</span>
    <span class="n">taxonomy_dict</span> <span class="o">=</span> <span class="n">create_tax_dictionary_from_registry</span><span class="p">(</span><span class="n">server_info</span><span class="p">,</span> <span class="n">registry_info</span><span class="p">)</span>

    <span class="c1"># Accept both int and str keys</span>
    <span class="n">lowest_taxon_id</span> <span class="o">=</span> <span class="n">registry_info</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span>
    <span class="n">taxonomy_hierarchy</span> <span class="o">=</span> <span class="n">taxonomy_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">lowest_taxon_id</span><span class="p">))</span> <span class="ow">or</span> <span class="n">taxonomy_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
        <span class="n">lowest_taxon_id</span><span class="p">,</span> <span class="p">[]</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">taxonomy_hierarchy</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Taxonomy hierarchy not found for taxon ID </span><span class="si">{</span><span class="n">registry_info</span><span class="p">[</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="s2">&quot;Unassigned&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>

    <span class="n">internal_clade</span> <span class="o">=</span> <span class="s2">&quot;Unassigned&quot;</span>
    <span class="n">genus_taxon_id</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">clade_details</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Get genus taxon_id</span>
    <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">taxonomy_hierarchy</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">taxon</span><span class="p">[</span><span class="s2">&quot;taxon_class&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;genus&quot;</span><span class="p">:</span>
            <span class="n">genus_taxon_id</span> <span class="o">=</span> <span class="n">taxon</span><span class="p">[</span><span class="s2">&quot;taxon_class_id&quot;</span><span class="p">]</span>

    <span class="c1"># Step 1: try exact match on the lowest_taxon_id</span>
    <span class="k">for</span> <span class="n">clade_name</span><span class="p">,</span> <span class="n">details</span> <span class="ow">in</span> <span class="n">clade_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">clade_taxon_id</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">clade_taxon_id</span> <span class="o">==</span> <span class="n">lowest_taxon_id</span><span class="p">:</span>
            <span class="n">internal_clade</span> <span class="o">=</span> <span class="n">clade_name</span>
            <span class="n">clade_details</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">details</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span> <span class="o">!=</span> <span class="s2">&quot;taxon_id&quot;</span><span class="p">}</span>
            <span class="n">clade_details</span><span class="p">[</span><span class="s2">&quot;helixer_lineage&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">clade_details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;helixer_lineage&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Exact match: Assigned clade &#39;</span><span class="si">{</span><span class="n">internal_clade</span><span class="si">}</span><span class="s2">&#39; for taxon </span><span class="si">{</span><span class="n">registry_info</span><span class="p">[</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="n">internal_clade</span><span class="p">,</span> <span class="n">genus_taxon_id</span><span class="p">,</span> <span class="n">clade_details</span>

    <span class="c1"># Step 2: Walk up the taxonomy hierarchy</span>
    <span class="n">taxon_classes_order</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;species&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genus&quot;</span><span class="p">,</span>
        <span class="s2">&quot;family&quot;</span><span class="p">,</span>
        <span class="s2">&quot;order&quot;</span><span class="p">,</span>
        <span class="s2">&quot;class&quot;</span><span class="p">,</span>
        <span class="s2">&quot;phylum&quot;</span><span class="p">,</span>
        <span class="s2">&quot;kingdom&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">taxon_class</span> <span class="ow">in</span> <span class="n">taxon_classes_order</span><span class="p">:</span>
        <span class="n">matching</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span>
            <span class="p">(</span><span class="n">t</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">taxonomy_hierarchy</span> <span class="k">if</span> <span class="n">t</span><span class="p">[</span><span class="s2">&quot;taxon_class&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">taxon_class</span><span class="p">),</span> <span class="kc">None</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">matching</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">current_taxon_id</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">matching</span><span class="p">[</span><span class="s2">&quot;taxon_class_id&quot;</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">clade_name</span><span class="p">,</span> <span class="n">details</span> <span class="ow">in</span> <span class="n">clade_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">clade_taxon_id</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">clade_taxon_id</span> <span class="o">==</span> <span class="n">current_taxon_id</span><span class="p">:</span>
                <span class="n">internal_clade</span> <span class="o">=</span> <span class="n">clade_name</span>
                <span class="n">clade_details</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">details</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span> <span class="o">!=</span> <span class="s2">&quot;taxon_id&quot;</span><span class="p">}</span>
                <span class="n">clade_details</span><span class="p">[</span><span class="s2">&quot;helixer_lineage&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">clade_details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                    <span class="s2">&quot;helixer_lineage&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
                <span class="p">)</span>

                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Hierarchy match: Assigned clade &#39;</span><span class="si">{</span><span class="n">internal_clade</span><span class="si">}</span><span class="s2">&#39; via </span><span class="si">{</span><span class="n">taxon_class</span><span class="si">}</span><span class="s2"> taxon_id </span><span class="si">{</span><span class="n">current_taxon_id</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="n">internal_clade</span><span class="p">,</span> <span class="n">genus_taxon_id</span><span class="p">,</span> <span class="n">clade_details</span>

    <span class="c1"># No match found</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;No clade found for taxon </span><span class="si">{</span><span class="n">registry_info</span><span class="p">[</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> in full hierarchy.&quot;</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot;Unassigned&quot;</span><span class="p">,</span> <span class="n">genus_taxon_id</span><span class="p">,</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.taxonomy_helper.assign_clade_info_custom_loading" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">assign_clade_info_custom_loading</span><span class="p">(</span><span class="n">registry_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.taxonomy_helper.assign_clade_info_custom_loading" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Look for a specific clade in the JSON data and return all values except for taxon_id.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>registry_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary containing clade information with 'clade' key</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="dict">dict</span>[<span title="str">str</span>, <span title="any">any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict or None: Dictionary containing all clade details except taxon_id,
                 or None if clade not found</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/taxonomy_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">assign_clade_info_custom_loading</span><span class="p">(</span><span class="n">registry_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">any</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Look for a specific clade in the JSON data and return all values except for taxon_id.</span>

<span class="sd">    Args:</span>
<span class="sd">            registry_info (dict): Dictionary containing clade information with &#39;clade&#39; key</span>

<span class="sd">    Returns:</span>
<span class="sd">            dict or None: Dictionary containing all clade details except taxon_id,</span>
<span class="sd">                                     or None if clade not found</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">clade_data</span> <span class="o">=</span> <span class="n">load_clade_data</span><span class="p">()</span>

    <span class="c1"># Extract the clade name from the dictionary</span>
    <span class="n">clade_name</span> <span class="o">=</span> <span class="n">registry_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;clade&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">clade_name</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;No &#39;clade&#39; key found in provided dictionary&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># Look for the specific clade in the dictionary</span>
    <span class="k">if</span> <span class="n">clade_name</span> <span class="ow">in</span> <span class="n">clade_data</span><span class="p">:</span>
        <span class="n">details</span> <span class="o">=</span> <span class="n">clade_data</span><span class="p">[</span><span class="n">clade_name</span><span class="p">]</span>
        <span class="c1"># Return all details except taxon_id</span>
        <span class="n">clade_details</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">details</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span> <span class="o">!=</span> <span class="s2">&quot;taxon_id&quot;</span><span class="p">}</span>
        <span class="n">clade_details</span><span class="p">[</span><span class="s2">&quot;helixer_lineage&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">clade_details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;helixer_lineage&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">clade_details</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Clade &#39;</span><span class="si">{</span><span class="n">clade_name</span><span class="si">}</span><span class="s2">&#39; not found in clade data&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.taxonomy_helper.create_tax_dictionary_from_registry" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">create_tax_dictionary_from_registry</span><span class="p">(</span><span class="n">server_info</span><span class="p">,</span> <span class="n">registry_info</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.taxonomy_helper.create_tax_dictionary_from_registry" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Query the registry MySQL database to construct taxonomy hierarchy for a given taxon ID.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary containing MySQL connection parameters under 'registry' key.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>registry_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary containing the 'taxon_id' key.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <span title="list">list</span>[<span title="dict">dict</span>[<span title="str">str</span>, <span title="any">any</span>]]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary mapping taxon IDs to a list of taxonomy class information,
  or an empty dictionary if query fails.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/taxonomy_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_tax_dictionary_from_registry</span><span class="p">(</span>
    <span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">registry_info</span><span class="p">:</span> <span class="nb">dict</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">any</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Query the registry MySQL database to construct taxonomy hierarchy for a given taxon ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        server_info (dict): Dictionary containing MySQL connection parameters under &#39;registry&#39; key.</span>
<span class="sd">        registry_info (dict): Dictionary containing the &#39;taxon_id&#39; key.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: A dictionary mapping taxon IDs to a list of taxonomy class information,</span>
<span class="sd">              or an empty dictionary if query fails.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">taxon_id</span> <span class="o">=</span> <span class="n">registry_info</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span>

    <span class="n">taxonomy_query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT lowest_taxon_id, taxon_class_id, taxon_class</span>
<span class="s2">        FROM taxonomy</span>
<span class="s2">        WHERE lowest_taxon_id = </span><span class="si">%s</span>
<span class="s2">        ORDER BY FIELD(taxon_class, &#39;species&#39;, &#39;genus&#39;, &#39;family&#39;, &#39;order&#39;, &#39;class&#39;, &#39;phylum&#39;, &#39;kingdom&#39;);</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">taxonomy_info</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">taxonomy_query</span><span class="p">,</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
            <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="n">params</span><span class="o">=</span><span class="n">taxon_id</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">taxonomy_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">taxonomy_info</span><span class="p">:</span>
            <span class="n">lowest_taxon_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;lowest_taxon_id&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">lowest_taxon_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxonomy_dict</span><span class="p">:</span>
                <span class="n">taxonomy_dict</span><span class="p">[</span><span class="n">lowest_taxon_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">taxonomy_dict</span><span class="p">[</span><span class="n">lowest_taxon_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;taxon_class_id&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;taxon_class_id&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;taxon_class&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;taxon_class&quot;</span><span class="p">],</span>
                <span class="p">}</span>
            <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found taxonomy for </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">taxonomy_dict</span>

    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Error while fetching taxonomy info: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">err</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.taxonomy_helper.get_parent_taxon" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_parent_taxon</span><span class="p">(</span><span class="n">server_info</span><span class="p">,</span> <span class="n">species_taxon_id</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.taxonomy_helper.get_parent_taxon" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch the parent taxon name from the taxonomy table for a given taxon ID.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>server_info</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Server connection info with keys 'db_host', 'db_user', 'db_port', 'db_name'.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>species_taxon_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The taxon ID for which to find the parent taxon.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Parent taxon name.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the taxon is not found.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/taxonomy_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_parent_taxon</span><span class="p">(</span><span class="n">server_info</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">species_taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch the parent taxon name from the taxonomy table for a given taxon ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        server_info (dict): Server connection info with keys &#39;db_host&#39;, &#39;db_user&#39;, &#39;db_port&#39;, &#39;db_name&#39;.</span>
<span class="sd">        species_taxon_id (int): The taxon ID for which to find the parent taxon.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Parent taxon name.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the taxon is not found.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parent_query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT taxon_class_name</span>
<span class="s2">        FROM taxonomy_name</span>
<span class="s2">        WHERE taxon_class_id = </span><span class="si">%s</span><span class="s2">;</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">parent_query</span><span class="p">,</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">database</span><span class="o">=</span><span class="n">server_info</span><span class="p">[</span><span class="s2">&quot;registry&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
            <span class="n">password</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="n">params</span><span class="o">=</span><span class="p">(</span><span class="n">species_taxon_id</span><span class="p">,),</span>  <span class="c1"># must be a tuple</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No parent taxon found for taxon ID </span><span class="si">{</span><span class="n">species_taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Return as a single string</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;taxon_class_name&quot;</span><span class="p">])</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error fetching parent taxon: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.taxonomy_helper.load_clade_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">load_clade_data</span><span class="p">()</span></code>

<a href="#ensembl.genes.info_from_registry.taxonomy_helper.load_clade_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Load clade definitions from a static JSON file.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <span title="dict">dict</span>[<span title="str">str</span>, <span title="any">any</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary containing clade configuration data.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="FileNotFoundError">FileNotFoundError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the JSON file cannot be located.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="json.JSONDecodeError">JSONDecodeError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the JSON is malformed.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/taxonomy_helper.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">load_clade_data</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">any</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load clade definitions from a static JSON file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary containing clade configuration data.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the JSON file cannot be located.</span>
<span class="sd">        json.JSONDecodeError: If the JSON is malformed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">json_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ENSCODE&quot;</span><span class="p">),</span>
        <span class="s2">&quot;ensembl-genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;src&quot;</span><span class="p">,</span>
        <span class="s2">&quot;python&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ensembl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;info_from_registry&quot;</span><span class="p">,</span>
        <span class="s2">&quot;clade_settings.json&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Loading clade settings json file.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.update_assembly_registry" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.update_assembly_registry</code>


<a href="#ensembl.genes.info_from_registry.update_assembly_registry" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Utility to update the status of a genebuild in the registry db</p>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.update_assembly_registry.args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.update_assembly_registry.args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.update_assembly_registry.parser" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Update the status of a genebuild in the registry database.&#39;</span><span class="p">,</span> <span class="n">formatter_class</span><span class="o">=</span><span class="p">(</span><span class="n">argparse</span><span class="o">.</span><span class="n">RawDescriptionHelpFormatter</span><span class="p">),</span> <span class="n">epilog</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">            Status values should match the gb_status enum:</span><span class="se">\n</span><span class="s1">            in_progress, insufficient_data, check_busco, completed, </span><span class="se">\n</span><span class="s1">            pre_released, handed_over, archive</span><span class="se">\n\n</span><span class="s1">            Examples:</span><span class="se">\n</span><span class="s1">            </span><span class="si">%(prog)s</span><span class="s1"> --host localhost --user myuser --password mypass --database registry </span><span class="se">\n</span><span class="s1">                    --assembly GCA_123456789.1 --status completed --genebuilder john_doe</span><span class="se">\n\n</span><span class="s1">            </span><span class="si">%(prog)s</span><span class="s1"> --host localhost --user myuser --password mypass --database registry </span><span class="se">\n</span><span class="s1">                    --assembly GCA_123456789.1 --status in_progress --genebuilder jane_smith --dev</span><span class="se">\n</span><span class="s1">        &#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.update_assembly_registry.parser" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.update_assembly_registry.ensure_genebuilder_exists" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">ensure_genebuilder_exists</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">genebuilder</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.update_assembly_registry.ensure_genebuilder_exists" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Ensure genebuilder exists in the genebuilder table.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>connection</code>
            </td>
            <td>
                  <code><span title="pymysql.connections.Connection">Connection</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL connection object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genebuilder</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genebuilder name</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/update_assembly_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">ensure_genebuilder_exists</span><span class="p">(</span>
    <span class="n">connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span> <span class="n">genebuilder</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Ensure genebuilder exists in the genebuilder table.</span>

<span class="sd">    Args:</span>
<span class="sd">        connection: MySQL connection object</span>
<span class="sd">        genebuilder (str): Genebuilder name</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">check_query</span> <span class="o">=</span> <span class="s2">&quot;SELECT genebuilder FROM genebuilder WHERE genebuilder = </span><span class="si">%s</span><span class="s2">&quot;</span>

    <span class="k">with</span> <span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">check_query</span><span class="p">,</span> <span class="p">(</span><span class="n">genebuilder</span><span class="p">,))</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchone</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Genebuilder &#39;</span><span class="si">{</span><span class="n">genebuilder</span><span class="si">}</span><span class="s2">&#39; does not exist in the genebuilder table. Please add it first.&quot;</span>
            <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.update_assembly_registry.insert_new_record" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">insert_new_record</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">genebuilder</span><span class="p">,</span> <span class="n">status</span><span class="p">,</span> <span class="n">annotation_source</span><span class="p">,</span> <span class="n">annotation_method</span><span class="p">,</span> <span class="n">current_date</span><span class="p">,</span> <span class="n">release_type</span><span class="p">,</span> <span class="n">genebuild_version</span><span class="p">,</span> <span class="n">dev</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.update_assembly_registry.insert_new_record" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Insert a new genebuild status record.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>connection</code>
            </td>
            <td>
                  <code><span title="pymysql.connections.Connection">Connection</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL connection object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly ID</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly accession</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genebuilder</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genebuilder name</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>status</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Status to set</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_source</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Annotation source</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_method</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Annotation method</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>current_date</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Current date</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Release type (default: "not_available")</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dev</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, only print SQL without executing</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/update_assembly_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span>
<span class="normal">99</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">insert_new_record</span><span class="p">(</span>
    <span class="n">connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span>
    <span class="n">assembly_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genebuilder</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">status</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">annotation_source</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">annotation_method</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">current_date</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">release_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genebuild_version</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">dev</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Insert a new genebuild status record.</span>

<span class="sd">    Args:</span>
<span class="sd">        connection: MySQL connection object</span>
<span class="sd">        assembly_id (int): Assembly ID</span>
<span class="sd">        assembly (str): Assembly accession</span>
<span class="sd">        genebuilder (str): Genebuilder name</span>
<span class="sd">        status (str): Status to set</span>
<span class="sd">        annotation_source (str): Annotation source</span>
<span class="sd">        annotation_method (str): Annotation method</span>
<span class="sd">        current_date (str): Current date</span>
<span class="sd">        release_type (str): Release type (default: &quot;not_available&quot;)</span>
<span class="sd">        dev (bool): If True, only print SQL without executing</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># date completed is being used to track the last time this record was updated</span>
    <span class="c1"># so we set it to the current date</span>
    <span class="n">date_status_update</span> <span class="o">=</span> <span class="n">current_date</span>

    <span class="c1"># Add not avaialable for release type</span>
    <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    INSERT INTO genebuild_status (</span>
<span class="s2">        assembly_id, gca_accession, gb_status, last_attempt, </span>
<span class="s2">        genebuilder, annotation_source, annotation_method,</span>
<span class="s2">        date_started, date_status_update, release_type,</span>
<span class="s2">        genebuild_version</span>
<span class="s2">    )</span>
<span class="s2">    VALUES (</span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, 1, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">)</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="n">params</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">assembly_id</span><span class="p">,</span>
        <span class="n">assembly</span><span class="p">,</span>
        <span class="n">status</span><span class="p">,</span>
        <span class="n">genebuilder</span><span class="p">,</span>
        <span class="n">annotation_source</span><span class="p">,</span>
        <span class="n">annotation_method</span><span class="p">,</span>
        <span class="n">current_date</span><span class="p">,</span>
        <span class="n">date_status_update</span><span class="p">,</span>
        <span class="n">release_type</span><span class="p">,</span>
        <span class="n">genebuild_version</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">dev</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Would execute:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">query</span> <span class="o">%</span> <span class="n">params</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Inserted new record for GCA </span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2"> with status &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.update_assembly_registry.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">(</span><span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">status</span><span class="p">,</span> <span class="n">genebuilder</span><span class="p">,</span> <span class="n">annotation_source</span><span class="p">,</span> <span class="n">annotation_method</span><span class="p">,</span> <span class="n">release_type</span><span class="p">,</span> <span class="n">genebuild_version</span><span class="p">,</span> <span class="n">dev</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.update_assembly_registry.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main function to update the genebuild status in the registry database.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>host</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL host</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>port</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL port</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>user</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL user</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>password</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL password</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>database</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL database name</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly Accession (GCA format)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>status</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Status to update to</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genebuilder</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genebuilder name</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_source</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Annotation source</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_method</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Annotation method</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Release type (default: "not_available")</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genebuild_version</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genebuild version</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dev</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, only print SQL statements without executing</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/update_assembly_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">(</span>
    <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">password</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">status</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genebuilder</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">annotation_source</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">annotation_method</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">release_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genebuild_version</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">dev</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main function to update the genebuild status in the registry database.</span>

<span class="sd">    Args:</span>
<span class="sd">        host (str): MySQL host</span>
<span class="sd">        port (int): MySQL port</span>
<span class="sd">        user (str): MySQL user</span>
<span class="sd">        password (str): MySQL password</span>
<span class="sd">        database (str): MySQL database name</span>
<span class="sd">        assembly (str): Assembly Accession (GCA format)</span>
<span class="sd">        status (str): Status to update to</span>
<span class="sd">        genebuilder (str): Genebuilder name</span>
<span class="sd">        annotation_source (str): Annotation source</span>
<span class="sd">        annotation_method (str): Annotation method</span>
<span class="sd">        release_type (str): Release type (default: &quot;not_available&quot;)</span>
<span class="sd">        genebuild_version (str): Genebuild version</span>
<span class="sd">        dev (bool): If True, only print SQL statements without executing</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">connection</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Connect to database</span>
        <span class="n">connection</span> <span class="o">=</span> <span class="n">mysql_get_connection</span><span class="p">(</span>
            <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="o">=</span><span class="n">password</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">connection</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Failed to connect to the database&quot;</span><span class="p">)</span>

        <span class="c1"># Start transaction</span>
        <span class="n">connection</span><span class="o">.</span><span class="n">begin</span><span class="p">()</span>

        <span class="n">ensure_genebuilder_exists</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">genebuilder</span><span class="p">)</span>

        <span class="n">assembly_id</span> <span class="o">=</span> <span class="n">fetch_assembly_id</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">assembly_id</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Assembly not found for GCA: </span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">current_date</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">existing_record</span> <span class="o">=</span> <span class="n">fetch_current_genebuild_record</span><span class="p">(</span>
            <span class="n">connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">genebuilder</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">existing_record</span><span class="p">:</span>
            <span class="c1"># No existing record for this genebuilder - INSERT new</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No existing record found for </span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2"> by </span><span class="si">{</span><span class="n">genebuilder</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Creating new record with status: </span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Use defaults for new records if not specified</span>
            <span class="n">method_to_insert</span> <span class="o">=</span> <span class="n">annotation_method</span> <span class="k">if</span> <span class="n">annotation_method</span> <span class="k">else</span> <span class="s2">&quot;pending&quot;</span>
            <span class="n">source_to_insert</span> <span class="o">=</span> <span class="n">annotation_source</span> <span class="k">if</span> <span class="n">annotation_source</span> <span class="k">else</span> <span class="s2">&quot;ensembl&quot;</span>
            <span class="n">version_to_insert</span> <span class="o">=</span> <span class="n">genebuild_version</span> <span class="k">if</span> <span class="n">genebuild_version</span> <span class="k">else</span> <span class="s2">&quot;ENS01&quot;</span>

            <span class="n">insert_new_record</span><span class="p">(</span>
                <span class="n">connection</span><span class="p">,</span>
                <span class="n">assembly_id</span><span class="p">,</span>
                <span class="n">assembly</span><span class="p">,</span>
                <span class="n">genebuilder</span><span class="p">,</span>
                <span class="n">status</span><span class="p">,</span>
                <span class="n">source_to_insert</span><span class="p">,</span>
                <span class="n">method_to_insert</span><span class="p">,</span>
                <span class="n">current_date</span><span class="p">,</span>
                <span class="n">release_type</span><span class="p">,</span>
                <span class="n">version_to_insert</span><span class="p">,</span>
                <span class="n">dev</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Existing record found for this genebuilder</span>
            <span class="n">current_status</span> <span class="o">=</span> <span class="n">existing_record</span><span class="p">[</span><span class="s2">&quot;gb_status&quot;</span><span class="p">]</span>
            <span class="n">current_method</span> <span class="o">=</span> <span class="n">existing_record</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;annotation_method&quot;</span><span class="p">)</span>
            <span class="n">current_source</span> <span class="o">=</span> <span class="n">existing_record</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;annotation_source&quot;</span><span class="p">)</span>
            <span class="n">current_version</span> <span class="o">=</span> <span class="n">existing_record</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;genebuild_version&quot;</span><span class="p">)</span>
            <span class="n">record_id</span> <span class="o">=</span> <span class="n">existing_record</span><span class="p">[</span><span class="s2">&quot;genebuild_status_id&quot;</span><span class="p">]</span>

            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found existing record for </span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2"> by </span><span class="si">{</span><span class="n">genebuilder</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Current status: &#39;</span><span class="si">{</span><span class="n">current_status</span><span class="si">}</span><span class="s2">&#39;, requested status: &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>

            <span class="c1"># Status categories</span>
            <span class="n">terminal_statuses</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;live&quot;</span><span class="p">,</span> <span class="s2">&quot;pre_released&quot;</span><span class="p">,</span> <span class="s2">&quot;handed_over&quot;</span><span class="p">,</span> <span class="s2">&quot;archive&quot;</span><span class="p">]</span>
            <span class="n">active_statuses</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;insufficient_data&quot;</span><span class="p">,</span> <span class="s2">&quot;in_progress&quot;</span><span class="p">,</span> <span class="s2">&quot;check_busco&quot;</span><span class="p">]</span>
            <span class="n">completed_status</span> <span class="o">=</span> <span class="s2">&quot;completed&quot;</span>  <span class="c1"># terminal-like status</span>

            <span class="c1"># Determine if method/source/version should be updated (preserve existing if not specified)</span>
            <span class="n">method_to_update</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">annotation_method</span> <span class="k">if</span> <span class="n">annotation_method</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="n">source_to_update</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">annotation_source</span> <span class="k">if</span> <span class="n">annotation_source</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="n">version_to_update</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">genebuild_version</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="n">genebuild_version</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">genebuild_version</span> <span class="o">!=</span> <span class="n">current_version</span>
                <span class="p">)</span>
                <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>

            <span class="c1"># Same status - check for method/source/version changes</span>
            <span class="k">if</span> <span class="n">current_status</span> <span class="o">==</span> <span class="n">status</span><span class="p">:</span>
                <span class="c1"># Check if annotation_method, annotation_source, or genebuild_version is provided and different</span>
                <span class="n">method_changed</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">annotation_method</span> <span class="ow">and</span> <span class="n">annotation_method</span> <span class="o">!=</span> <span class="n">current_method</span>
                <span class="p">)</span>
                <span class="n">source_changed</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">annotation_source</span> <span class="ow">and</span> <span class="n">annotation_source</span> <span class="o">!=</span> <span class="n">current_source</span>
                <span class="p">)</span>
                <span class="n">version_changed</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">genebuild_version</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">genebuild_version</span> <span class="o">!=</span> <span class="n">current_version</span>
                <span class="p">)</span>

                <span class="k">if</span> <span class="n">method_changed</span> <span class="ow">or</span> <span class="n">source_changed</span> <span class="ow">or</span> <span class="n">version_changed</span><span class="p">:</span>
                    <span class="n">changes</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">if</span> <span class="n">method_changed</span><span class="p">:</span>
                        <span class="n">changes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;method from &#39;</span><span class="si">{</span><span class="n">current_method</span><span class="si">}</span><span class="s2">&#39; to &#39;</span><span class="si">{</span><span class="n">annotation_method</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                        <span class="p">)</span>
                    <span class="k">if</span> <span class="n">source_changed</span><span class="p">:</span>
                        <span class="n">changes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;source from &#39;</span><span class="si">{</span><span class="n">current_source</span><span class="si">}</span><span class="s2">&#39; to &#39;</span><span class="si">{</span><span class="n">annotation_source</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                        <span class="p">)</span>
                    <span class="k">if</span> <span class="n">version_changed</span><span class="p">:</span>
                        <span class="n">changes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;version from &#39;</span><span class="si">{</span><span class="n">current_version</span><span class="si">}</span><span class="s2">&#39; to &#39;</span><span class="si">{</span><span class="n">genebuild_version</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                        <span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Status is already &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;, but updating </span><span class="si">{</span><span class="s1">&#39; and &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">changes</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                    <span class="n">update_existing_record</span><span class="p">(</span>
                        <span class="n">connection</span><span class="p">,</span>
                        <span class="n">record_id</span><span class="p">,</span>
                        <span class="n">status</span><span class="p">,</span>
                        <span class="n">current_date</span><span class="p">,</span>
                        <span class="n">dev</span><span class="p">,</span>
                        <span class="n">annotation_method</span> <span class="k">if</span> <span class="n">method_changed</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                        <span class="n">annotation_source</span> <span class="k">if</span> <span class="n">source_changed</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                        <span class="n">genebuild_version</span> <span class="k">if</span> <span class="n">version_changed</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Status is already &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;. No changes needed.&quot;</span><span class="p">)</span>
                    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

            <span class="c1"># Block backwards transitions from completed</span>
            <span class="k">elif</span> <span class="n">current_status</span> <span class="o">==</span> <span class="n">completed_status</span> <span class="ow">and</span> <span class="n">status</span> <span class="ow">in</span> <span class="n">active_statuses</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ERROR: Cannot move backwards from &#39;completed&#39; to &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Completed is a terminal-like status. To restart work, use a new genebuild_version.&quot;</span>
                <span class="p">)</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="c1"># Block backwards transitions from terminal statuses</span>
            <span class="k">elif</span> <span class="n">current_status</span> <span class="ow">in</span> <span class="n">terminal_statuses</span> <span class="ow">and</span> <span class="n">status</span> <span class="ow">in</span> <span class="p">(</span>
                <span class="n">active_statuses</span> <span class="o">+</span> <span class="p">[</span><span class="n">completed_status</span><span class="p">]</span>
            <span class="p">):</span>
                <span class="c1"># Check if version changed - if so, allow new attempt</span>
                <span class="c1"># If genebuild_version not provided, use current version (no version change)</span>
                <span class="n">effective_version</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">genebuild_version</span> <span class="k">if</span> <span class="n">genebuild_version</span> <span class="k">else</span> <span class="n">current_version</span>
                <span class="p">)</span>

                <span class="k">if</span> <span class="n">effective_version</span> <span class="o">!=</span> <span class="n">current_version</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Moving from terminal status &#39;</span><span class="si">{</span><span class="n">current_status</span><span class="si">}</span><span class="s2">&#39; to &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39; with new version </span><span class="si">{</span><span class="n">effective_version</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Creating new attempt.&quot;</span><span class="p">)</span>

                    <span class="n">method_to_insert</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">annotation_method</span> <span class="k">if</span> <span class="n">annotation_method</span> <span class="k">else</span> <span class="s2">&quot;pending&quot;</span>
                    <span class="p">)</span>
                    <span class="n">source_to_insert</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">annotation_source</span> <span class="k">if</span> <span class="n">annotation_source</span> <span class="k">else</span> <span class="n">current_source</span>
                    <span class="p">)</span>

                    <span class="n">set_old_record_historical</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">record_id</span><span class="p">,</span> <span class="n">dev</span><span class="p">)</span>
                    <span class="n">insert_new_record</span><span class="p">(</span>
                        <span class="n">connection</span><span class="p">,</span>
                        <span class="n">assembly_id</span><span class="p">,</span>
                        <span class="n">assembly</span><span class="p">,</span>
                        <span class="n">genebuilder</span><span class="p">,</span>
                        <span class="n">status</span><span class="p">,</span>
                        <span class="n">source_to_insert</span><span class="p">,</span>
                        <span class="n">method_to_insert</span><span class="p">,</span>
                        <span class="n">current_date</span><span class="p">,</span>
                        <span class="n">release_type</span><span class="p">,</span>
                        <span class="n">effective_version</span><span class="p">,</span>
                        <span class="n">dev</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;ERROR: Cannot move from terminal status &#39;</span><span class="si">{</span><span class="n">current_status</span><span class="si">}</span><span class="s2">&#39; to &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39; with same genebuild_version &#39;</span><span class="si">{</span><span class="n">current_version</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                    <span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;To restart work, you must provide a new --genebuild_version (e.g., ENS02, ENS03, etc.)&quot;</span>
                    <span class="p">)</span>
                    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="c1"># Terminal to terminal transition - UPDATE same record</span>
            <span class="k">elif</span> <span class="n">current_status</span> <span class="ow">in</span> <span class="n">terminal_statuses</span> <span class="ow">and</span> <span class="n">status</span> <span class="ow">in</span> <span class="n">terminal_statuses</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Moving from terminal status &#39;</span><span class="si">{</span><span class="n">current_status</span><span class="si">}</span><span class="s2">&#39; to terminal status &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                <span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Updating existing record.&quot;</span><span class="p">)</span>
                <span class="n">update_existing_record</span><span class="p">(</span>
                    <span class="n">connection</span><span class="p">,</span>
                    <span class="n">record_id</span><span class="p">,</span>
                    <span class="n">status</span><span class="p">,</span>
                    <span class="n">current_date</span><span class="p">,</span>
                    <span class="n">dev</span><span class="p">,</span>
                    <span class="n">method_to_update</span><span class="p">,</span>
                    <span class="n">source_to_update</span><span class="p">,</span>
                    <span class="n">version_to_update</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="c1"># Completed to terminal transition - UPDATE same record</span>
            <span class="k">elif</span> <span class="n">current_status</span> <span class="o">==</span> <span class="n">completed_status</span> <span class="ow">and</span> <span class="n">status</span> <span class="ow">in</span> <span class="n">terminal_statuses</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Moving from &#39;completed&#39; to terminal status &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Updating existing record.&quot;</span><span class="p">)</span>
                <span class="n">update_existing_record</span><span class="p">(</span>
                    <span class="n">connection</span><span class="p">,</span>
                    <span class="n">record_id</span><span class="p">,</span>
                    <span class="n">status</span><span class="p">,</span>
                    <span class="n">current_date</span><span class="p">,</span>
                    <span class="n">dev</span><span class="p">,</span>
                    <span class="n">method_to_update</span><span class="p">,</span>
                    <span class="n">source_to_update</span><span class="p">,</span>
                    <span class="n">version_to_update</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="c1"># Active to active or active to completed - UPDATE same record</span>
            <span class="k">elif</span> <span class="n">current_status</span> <span class="ow">in</span> <span class="n">active_statuses</span> <span class="ow">and</span> <span class="p">(</span>
                <span class="n">status</span> <span class="ow">in</span> <span class="n">active_statuses</span> <span class="ow">or</span> <span class="n">status</span> <span class="o">==</span> <span class="n">completed_status</span>
            <span class="p">):</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Current status &#39;</span><span class="si">{</span><span class="n">current_status</span><span class="si">}</span><span class="s2">&#39; is active. Updating to &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;.&quot;</span>
                <span class="p">)</span>
                <span class="n">update_existing_record</span><span class="p">(</span>
                    <span class="n">connection</span><span class="p">,</span>
                    <span class="n">record_id</span><span class="p">,</span>
                    <span class="n">status</span><span class="p">,</span>
                    <span class="n">current_date</span><span class="p">,</span>
                    <span class="n">dev</span><span class="p">,</span>
                    <span class="n">method_to_update</span><span class="p">,</span>
                    <span class="n">source_to_update</span><span class="p">,</span>
                    <span class="n">version_to_update</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="c1"># Active to terminal - UPDATE same record</span>
            <span class="k">elif</span> <span class="n">current_status</span> <span class="ow">in</span> <span class="n">active_statuses</span> <span class="ow">and</span> <span class="n">status</span> <span class="ow">in</span> <span class="n">terminal_statuses</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Moving from active status &#39;</span><span class="si">{</span><span class="n">current_status</span><span class="si">}</span><span class="s2">&#39; to terminal status &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                <span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Updating existing record.&quot;</span><span class="p">)</span>
                <span class="n">update_existing_record</span><span class="p">(</span>
                    <span class="n">connection</span><span class="p">,</span>
                    <span class="n">record_id</span><span class="p">,</span>
                    <span class="n">status</span><span class="p">,</span>
                    <span class="n">current_date</span><span class="p">,</span>
                    <span class="n">dev</span><span class="p">,</span>
                    <span class="n">method_to_update</span><span class="p">,</span>
                    <span class="n">source_to_update</span><span class="p">,</span>
                    <span class="n">version_to_update</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unexpected status transition: &#39;</span><span class="si">{</span><span class="n">current_status</span><span class="si">}</span><span class="s2">&#39; to &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">dev</span><span class="p">:</span>
            <span class="c1"># Commit transaction</span>
            <span class="n">connection</span><span class="o">.</span><span class="n">commit</span><span class="p">()</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Transaction completed successfully&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;DEV MODE: No changes were made to the database&quot;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">connection</span><span class="p">:</span>
            <span class="n">connection</span><span class="o">.</span><span class="n">rollback</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ERROR: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">connection</span><span class="p">:</span>
            <span class="n">connection</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.update_assembly_registry.set_old_record_historical" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">set_old_record_historical</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">record_id</span><span class="p">,</span> <span class="n">dev</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.update_assembly_registry.set_old_record_historical" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Set an existing record to historical (last_attempt = 0).</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>connection</code>
            </td>
            <td>
                  <code><span title="pymysql.connections.Connection">Connection</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL connection object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>record_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>genebuild_status_id to update</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dev</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, only print SQL without executing</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/update_assembly_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">set_old_record_historical</span><span class="p">(</span>
    <span class="n">connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span> <span class="n">record_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">dev</span><span class="p">:</span> <span class="nb">bool</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Set an existing record to historical (last_attempt = 0).</span>

<span class="sd">    Args:</span>
<span class="sd">        connection: MySQL connection object</span>
<span class="sd">        record_id (int): genebuild_status_id to update</span>
<span class="sd">        dev (bool): If True, only print SQL without executing</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">query</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;UPDATE genebuild_status SET last_attempt = 0 WHERE genebuild_status_id = </span><span class="si">%s</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">dev</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Would execute:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">query</span> <span class="o">%</span> <span class="p">(</span><span class="n">record_id</span><span class="p">,))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="p">(</span><span class="n">record_id</span><span class="p">,))</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Set record </span><span class="si">{</span><span class="n">record_id</span><span class="si">}</span><span class="s2"> to historical&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.update_assembly_registry.update_existing_record" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">update_existing_record</span><span class="p">(</span><span class="n">connection</span><span class="p">,</span> <span class="n">record_id</span><span class="p">,</span> <span class="n">status</span><span class="p">,</span> <span class="n">current_date</span><span class="p">,</span> <span class="n">dev</span><span class="p">,</span> <span class="n">annotation_method</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">annotation_source</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">genebuild_version</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.update_assembly_registry.update_existing_record" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/update_assembly_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">update_existing_record</span><span class="p">(</span>
    <span class="n">connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span>
    <span class="n">record_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">status</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">current_date</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">dev</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">annotation_method</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">annotation_source</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">genebuild_version</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">query_parts</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gb_status = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;date_status_update = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">]</span>
    <span class="n">params</span> <span class="o">=</span> <span class="p">[</span><span class="n">status</span><span class="p">,</span> <span class="n">current_date</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">annotation_method</span><span class="p">:</span>
        <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;annotation_method = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">params</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">annotation_method</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">annotation_source</span><span class="p">:</span>
        <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;annotation_source = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">params</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">annotation_source</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">genebuild_version</span><span class="p">:</span>
        <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;genebuild_version = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">params</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genebuild_version</span><span class="p">)</span>
    <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">UPDATE genebuild_status</span>
<span class="s2">SET </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">query_parts</span><span class="p">)</span><span class="si">}</span>
<span class="s2">WHERE genebuild_status_id = %s</span>
<span class="s2">&quot;&quot;&quot;</span>
    <span class="n">params</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record_id</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">dev</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Would execute:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Updated record </span><span class="si">{</span><span class="n">record_id</span><span class="si">}</span><span class="s2"> to status &#39;</span><span class="si">{</span><span class="n">status</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.info_from_registry.write_metrics_to_registry" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.info_from_registry.write_metrics_to_registry</code>


<a href="#ensembl.genes.info_from_registry.write_metrics_to_registry" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Write metrics from core database to registry database.</p>


<details class="reads-metrics-from-core-db-meta-table-and-writes-them-to-registry-tables" open>
  <summary>Reads metrics from core DB meta table and writes them to registry tables</summary>
  <ul>
<li>assembly.* keys -&gt; assembly_metrics</li>
<li>genebuild.* keys -&gt; annotation_metrics</li>
</ul>
</details>        <p>Uses DELETE then INSERT pattern to avoid stale/duplicated metrics.</p>










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.write_metrics_to_registry.args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.info_from_registry.write_metrics_to_registry.parser" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Write metrics from core database to registry database.&#39;</span><span class="p">,</span> <span class="n">formatter_class</span><span class="o">=</span><span class="p">(</span><span class="n">argparse</span><span class="o">.</span><span class="n">RawDescriptionHelpFormatter</span><span class="p">),</span> <span class="n">epilog</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">            Fetches metrics from core meta table and writes to registry tables.</span><span class="se">\n\n</span><span class="s1">            Examples:</span><span class="se">\n</span><span class="s1">            </span><span class="si">%(prog)s</span><span class="s1"> --registry_host localhost --registry_user myuser --registry_password mypass </span><span class="se">\\\n</span><span class="s1">                     --registry_db registry --core_host corehost --core_user myuser </span><span class="se">\\\n</span><span class="s1">                     --core_password mypass --core_db my_core_db --assembly GCA_123456789.1 </span><span class="se">\\\n</span><span class="s1">                     --genebuilder john_doe</span><span class="se">\n\n</span><span class="s1">            </span><span class="si">%(prog)s</span><span class="s1"> --registry_host localhost --registry_user myuser --registry_password mypass </span><span class="se">\\\n</span><span class="s1">                     --registry_db registry --core_host corehost --core_user myuser </span><span class="se">\\\n</span><span class="s1">                     --core_password mypass --core_db my_core_db --assembly GCA_123456789.1 </span><span class="se">\\\n</span><span class="s1">                     --genebuilder john_doe --dev</span><span class="se">\n</span><span class="s1">        &#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.parser" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.write_metrics_to_registry.fetch_core_metrics" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">fetch_core_metrics</span><span class="p">(</span><span class="n">core_connection</span><span class="p">,</span> <span class="n">species_id</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.fetch_core_metrics" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch metrics from core database meta table.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>core_connection</code>
            </td>
            <td>
                  <code><span title="pymysql.connections.Connection">Connection</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL connection object for core DB</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>species_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species ID in core meta table</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    list: List of dictionaries with meta_key and meta_value</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/write_metrics_to_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fetch_core_metrics</span><span class="p">(</span>
    <span class="n">core_connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span> <span class="n">species_id</span><span class="p">:</span> <span class="nb">int</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch metrics from core database meta table.</span>

<span class="sd">    Args:</span>
<span class="sd">        core_connection: MySQL connection object for core DB</span>
<span class="sd">        species_id (int): Species ID in core meta table</span>
<span class="sd">    Returns:</span>
<span class="sd">        list: List of dictionaries with meta_key and meta_value</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    SELECT meta_key, meta_value</span>
<span class="s2">    FROM meta</span>
<span class="s2">    WHERE species_id=</span><span class="si">%s</span><span class="s2"> AND (</span>
<span class="s2">    meta_key LIKE &#39;assembly.busco</span><span class="si">%%</span><span class="s2">&#39; OR</span>
<span class="s2">    meta_key LIKE &#39;assembly.stats.</span><span class="si">%%</span><span class="s2">&#39; OR</span>
<span class="s2">    meta_key LIKE &#39;genebuild.busco</span><span class="si">%%</span><span class="s2">&#39; OR</span>
<span class="s2">    meta_key LIKE &#39;genebuild.stats.</span><span class="si">%%</span><span class="s2">&#39; OR</span>
<span class="s2">        meta_key = &#39;genebuild.last_geneset_update&#39;</span>
<span class="s2">    )</span>
<span class="s2">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">core_connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="p">(</span><span class="n">species_id</span><span class="p">,))</span>
        <span class="k">return</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.write_metrics_to_registry.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">(</span><span class="n">registry_host</span><span class="p">,</span> <span class="n">registry_port</span><span class="p">,</span> <span class="n">registry_user</span><span class="p">,</span> <span class="n">registry_password</span><span class="p">,</span> <span class="n">registry_db</span><span class="p">,</span> <span class="n">core_host</span><span class="p">,</span> <span class="n">core_port</span><span class="p">,</span> <span class="n">core_user</span><span class="p">,</span> <span class="n">core_password</span><span class="p">,</span> <span class="n">core_db</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">species_id</span><span class="p">,</span> <span class="n">genebuilder</span><span class="p">,</span> <span class="n">dev</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main function to write metrics from core DB to registry DB.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>registry_host, registry_port, registry_user, registry_password, registry_db</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Registry DB connection</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>core_host, core_port, core_user, core_password, core_db</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Core DB connection</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assembly</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Assembly Accession (GCA format)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>species_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species ID in core meta table</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genebuilder</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genebuilder name to identify which genebuild record to update</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dev</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, only print SQL statements without executing</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/write_metrics_to_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">(</span>
    <span class="n">registry_host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">registry_port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">registry_user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">registry_password</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">registry_db</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">core_host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">core_port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">core_user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">core_password</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">core_db</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">species_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">genebuilder</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">dev</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main function to write metrics from core DB to registry DB.</span>

<span class="sd">    Args:</span>
<span class="sd">        registry_host, registry_port, registry_user, registry_password, registry_db: Registry DB connection</span>
<span class="sd">        core_host, core_port, core_user, core_password, core_db: Core DB connection</span>
<span class="sd">        assembly (str): Assembly Accession (GCA format)</span>
<span class="sd">        species_id (int): Species ID in core meta table</span>
<span class="sd">        genebuilder (str): Genebuilder name to identify which genebuild record to update</span>
<span class="sd">        dev (bool): If True, only print SQL statements without executing</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">registry_connection</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">core_connection</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">registry_connection</span> <span class="o">=</span> <span class="n">mysql_get_connection</span><span class="p">(</span>
            <span class="n">database</span><span class="o">=</span><span class="n">registry_db</span><span class="p">,</span>
            <span class="n">host</span><span class="o">=</span><span class="n">registry_host</span><span class="p">,</span>
            <span class="n">port</span><span class="o">=</span><span class="n">registry_port</span><span class="p">,</span>
            <span class="n">user</span><span class="o">=</span><span class="n">registry_user</span><span class="p">,</span>
            <span class="n">password</span><span class="o">=</span><span class="n">registry_password</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">registry_connection</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Failed to connect to the registry database&quot;</span><span class="p">)</span>

        <span class="n">core_connection</span> <span class="o">=</span> <span class="n">mysql_get_connection</span><span class="p">(</span>
            <span class="n">database</span><span class="o">=</span><span class="n">core_db</span><span class="p">,</span>
            <span class="n">host</span><span class="o">=</span><span class="n">core_host</span><span class="p">,</span>
            <span class="n">port</span><span class="o">=</span><span class="n">core_port</span><span class="p">,</span>
            <span class="n">user</span><span class="o">=</span><span class="n">core_user</span><span class="p">,</span>
            <span class="n">password</span><span class="o">=</span><span class="n">core_password</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">core_connection</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Failed to connect to the core database&quot;</span><span class="p">)</span>

        <span class="c1"># Start transaction on registry</span>
        <span class="n">registry_connection</span><span class="o">.</span><span class="n">begin</span><span class="p">()</span>

        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Fetching registry IDs for assembly </span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2"> and genebuilder </span><span class="si">{</span><span class="n">genebuilder</span><span class="si">}</span><span class="s2">...&quot;</span>
        <span class="p">)</span>
        <span class="n">assembly_id</span><span class="p">,</span> <span class="n">genebuild_status_id</span> <span class="o">=</span> <span class="n">fetch_registry_ids</span><span class="p">(</span>
            <span class="n">registry_connection</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">genebuilder</span>
        <span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Found assembly_id: </span><span class="si">{</span><span class="n">assembly_id</span><span class="si">}</span><span class="s2">, genebuild_status_id: </span><span class="si">{</span><span class="n">genebuild_status_id</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Fetching metrics from core database...&quot;</span><span class="p">)</span>
        <span class="n">meta_rows</span> <span class="o">=</span> <span class="n">fetch_core_metrics</span><span class="p">(</span><span class="n">core_connection</span><span class="p">,</span> <span class="n">species_id</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">meta_rows</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No metrics found in core database&quot;</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># Partition metrics</span>
        <span class="n">assembly_rows</span><span class="p">,</span> <span class="n">genebuild_rows</span> <span class="o">=</span> <span class="n">partition_metrics</span><span class="p">(</span><span class="n">meta_rows</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">assembly_rows</span><span class="p">)</span><span class="si">}</span><span class="s2"> assembly metrics and </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">genebuild_rows</span><span class="p">)</span><span class="si">}</span><span class="s2"> genebuild metrics&quot;</span>
        <span class="p">)</span>

        <span class="c1"># Write metrics to registry</span>
        <span class="n">write_assembly_metrics</span><span class="p">(</span><span class="n">registry_connection</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">assembly_rows</span><span class="p">,</span> <span class="n">dev</span><span class="p">)</span>
        <span class="n">write_genebuild_metrics</span><span class="p">(</span>
            <span class="n">registry_connection</span><span class="p">,</span> <span class="n">genebuild_status_id</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">genebuild_rows</span><span class="p">,</span> <span class="n">dev</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">dev</span><span class="p">:</span>
            <span class="c1"># Commit transaction</span>
            <span class="n">registry_connection</span><span class="o">.</span><span class="n">commit</span><span class="p">()</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Metrics written successfully&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;DEV MODE: No changes were made to the registry database&quot;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">registry_connection</span><span class="p">:</span>
            <span class="n">registry_connection</span><span class="o">.</span><span class="n">rollback</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ERROR: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">registry_connection</span><span class="p">:</span>
            <span class="n">registry_connection</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">core_connection</span><span class="p">:</span>
            <span class="n">core_connection</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.write_metrics_to_registry.partition_metrics" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">partition_metrics</span><span class="p">(</span><span class="n">rows</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.partition_metrics" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Partition metrics into assembly and genebuild categories.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>rows</code>
            </td>
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of dicts with meta_key and meta_value</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    tuple: (assembly_rows, genebuild_rows) as lists of (key, value) tuples</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/write_metrics_to_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">partition_metrics</span><span class="p">(</span>
    <span class="n">rows</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Partition metrics into assembly and genebuild categories.</span>

<span class="sd">    Args:</span>
<span class="sd">        rows (list): List of dicts with meta_key and meta_value</span>
<span class="sd">    Returns:</span>
<span class="sd">        tuple: (assembly_rows, genebuild_rows) as lists of (key, value) tuples</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">assembly_rows</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">genebuild_rows</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
        <span class="n">key</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;meta_key&quot;</span><span class="p">]</span>
        <span class="n">value</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;meta_value&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">key</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;assembly.busco&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">key</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;assembly.stats.&quot;</span><span class="p">):</span>
            <span class="n">assembly_rows</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">))</span>
        <span class="k">elif</span> <span class="p">(</span>
            <span class="n">key</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;genebuild.busco&quot;</span><span class="p">)</span>
            <span class="ow">or</span> <span class="n">key</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;genebuild.stats.&quot;</span><span class="p">)</span>
            <span class="ow">or</span> <span class="n">key</span> <span class="o">==</span> <span class="s2">&quot;genebuild.last_geneset_update&quot;</span>
        <span class="p">):</span>
            <span class="n">genebuild_rows</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">assembly_rows</span><span class="p">,</span> <span class="n">genebuild_rows</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.write_metrics_to_registry.write_assembly_metrics" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">write_assembly_metrics</span><span class="p">(</span><span class="n">registry_connection</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">rows</span><span class="p">,</span> <span class="n">dev</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.write_assembly_metrics" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Write assembly metrics to registry using DELETE then INSERT pattern.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/write_metrics_to_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_assembly_metrics</span><span class="p">(</span>
    <span class="n">registry_connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span>
    <span class="n">assembly_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">rows</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="n">dev</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write assembly metrics to registry using DELETE then INSERT pattern.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">rows</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No assembly metrics to write&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="c1"># Extract metric names for deletion</span>
    <span class="n">metric_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">name</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">]</span>

    <span class="c1"># Single DELETE for all metrics at once</span>
    <span class="n">placeholders</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">metric_names</span><span class="p">))</span>
    <span class="n">delete_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    DELETE FROM assembly_metrics</span>
<span class="s2">    WHERE assembly_id=%s AND metrics_name IN (</span><span class="si">{</span><span class="n">placeholders</span><span class="si">}</span><span class="s2">)</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="n">insert_query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    INSERT INTO assembly_metrics (assembly_id, metrics_name, metrics_value)</span>
<span class="s2">    VALUES (</span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">)</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">dev</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Would execute:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">delete_query</span><span class="p">,</span> <span class="p">(</span><span class="n">assembly_id</span><span class="p">,</span> <span class="o">*</span><span class="n">metric_names</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">insert_query</span><span class="p">,</span> <span class="p">(</span><span class="n">assembly_id</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">registry_connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">delete_query</span><span class="p">,</span> <span class="p">(</span><span class="n">assembly_id</span><span class="p">,</span> <span class="o">*</span><span class="n">metric_names</span><span class="p">))</span>

            <span class="c1"># Single DELETE for all metrics</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">delete_query</span><span class="p">,</span> <span class="p">(</span><span class="n">assembly_id</span><span class="p">,</span> <span class="o">*</span><span class="n">metric_names</span><span class="p">))</span>
            <span class="n">deleted</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">rowcount</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Deleted </span><span class="si">{</span><span class="n">deleted</span><span class="si">}</span><span class="s2"> existing assembly metrics for assembly_id </span><span class="si">{</span><span class="n">assembly_id</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

            <span class="c1"># Bulk INSERT</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">executemany</span><span class="p">(</span>
                <span class="n">insert_query</span><span class="p">,</span> <span class="p">[(</span><span class="n">assembly_id</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">rows</span><span class="p">)</span><span class="si">}</span><span class="s2"> assembly metrics for assembly_id </span><span class="si">{</span><span class="n">assembly_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.info_from_registry.write_metrics_to_registry.write_genebuild_metrics" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">write_genebuild_metrics</span><span class="p">(</span><span class="n">registry_connection</span><span class="p">,</span> <span class="n">genebuild_status_id</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">rows</span><span class="p">,</span> <span class="n">dev</span><span class="p">)</span></code>

<a href="#ensembl.genes.info_from_registry.write_metrics_to_registry.write_genebuild_metrics" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Write genebuild metrics to registry using DELETE then INSERT pattern.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/info_from_registry/write_metrics_to_registry.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_genebuild_metrics</span><span class="p">(</span>
    <span class="n">registry_connection</span><span class="p">:</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">,</span>
    <span class="n">genebuild_status_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span>
    <span class="n">assembly_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">rows</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="n">dev</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write genebuild metrics to registry using DELETE then INSERT pattern.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">rows</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No genebuild metrics to write&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">genebuild_status_id</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No genebuild_status_id available; skipping genebuild metrics&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="n">metric_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">name</span> <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">]</span>

    <span class="n">placeholders</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">metric_names</span><span class="p">))</span>
    <span class="n">delete_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    DELETE FROM annotation_metrics</span>
<span class="s2">    WHERE genebuild_status_id=%s AND metrics_name IN (</span><span class="si">{</span><span class="n">placeholders</span><span class="si">}</span><span class="s2">)</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="n">insert_query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    INSERT INTO annotation_metrics (genebuild_status_id, assembly_id, metrics_name, metrics_value)</span>
<span class="s2">    VALUES (</span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">)</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">dev</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Would execute:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">delete_query</span><span class="p">,</span> <span class="p">(</span><span class="n">genebuild_status_id</span><span class="p">,</span> <span class="o">*</span><span class="n">metric_names</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">insert_query</span><span class="p">,</span> <span class="p">(</span><span class="n">genebuild_status_id</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">registry_connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">delete_query</span><span class="p">,</span> <span class="p">(</span><span class="n">genebuild_status_id</span><span class="p">,</span> <span class="o">*</span><span class="n">metric_names</span><span class="p">))</span>
            <span class="n">deleted</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">rowcount</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Deleted </span><span class="si">{</span><span class="n">deleted</span><span class="si">}</span><span class="s2"> existing genebuild metrics for genebuild_status_id </span><span class="si">{</span><span class="n">genebuild_status_id</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

            <span class="n">cursor</span><span class="o">.</span><span class="n">executemany</span><span class="p">(</span>
                <span class="n">insert_query</span><span class="p">,</span>
                <span class="p">[</span>
                    <span class="p">(</span><span class="n">genebuild_status_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">assembly_id</span><span class="p">),</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">rows</span>
                <span class="p">],</span>
            <span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Wrote </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">rows</span><span class="p">)</span><span class="si">}</span><span class="s2"> genebuild metrics for genebuild_status_id </span><span class="si">{</span><span class="n">genebuild_status_id</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="metadata">Metadata<a class="headerlink" href="#metadata" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.metadata.beta_patcher" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.metadata.beta_patcher</code>


<a href="#ensembl.genes.metadata.beta_patcher" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Beta Metadata Patcher Script</p>
<p>Generates standardized SQL patch files for fixing metadata issues in:
1. Production metadata database (ensembl_genome_metadata)
2. Core database (meta table)</p>
<p>This ensures metadata consistency between both databases and provides
a standardized workflow for creating and applying patches.</p>


<details class="usage" open>
  <summary>Usage</summary>
  <p>python beta_patcher.py patches.csv --jira-ticket EBD-1111 --output-dir ./patches/
python beta_patcher.py patches.csv --jira-ticket EBD-1111 --core-suffix _core_115_1</p>
</details>









<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.beta_patcher.METADATA_API_AVAILABLE" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">METADATA_API_AVAILABLE</span> <span class="o">=</span> <span class="kc">True</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.beta_patcher.METADATA_API_AVAILABLE" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.generate_all_patches" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">generate_all_patches</span><span class="p">(</span><span class="n">grouped_patches</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">jira_ticket</span><span class="p">,</span> <span class="n">logger</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.generate_all_patches" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Generate consolidated patch files for all genomes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>grouped_patches</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of genome patches grouped by UUID</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_dir</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Output directory for patch files</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>jira_ticket</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Jira ticket reference</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>logger</code>
            </td>
            <td>
                  <code><span title="logging.Logger">Logger</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Logger instance</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if successful, False otherwise</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_all_patches</span><span class="p">(</span>
    <span class="n">grouped_patches</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">],</span>
    <span class="n">output_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">jira_ticket</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">logger</span><span class="p">:</span> <span class="n">logging</span><span class="o">.</span><span class="n">Logger</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate consolidated patch files for all genomes.</span>

<span class="sd">    Args:</span>
<span class="sd">        grouped_patches: Dictionary of genome patches grouped by UUID</span>
<span class="sd">        output_dir: Output directory for patch files</span>
<span class="sd">        jira_ticket: Jira ticket reference</span>
<span class="sd">        logger: Logger instance</span>

<span class="sd">    Returns:</span>
<span class="sd">        True if successful, False otherwise</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Metadata DB files</span>
    <span class="n">metadata_patch_file</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;patch_metadata_</span><span class="si">{</span><span class="n">jira_ticket</span><span class="si">}</span><span class="s2">.sql&quot;</span>
    <span class="n">metadata_validate_file</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;validate_metadata_</span><span class="si">{</span><span class="n">jira_ticket</span><span class="si">}</span><span class="s2">.sql&quot;</span>

    <span class="c1"># Core DB files</span>
    <span class="n">core_patch_file</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;patch_core_</span><span class="si">{</span><span class="n">jira_ticket</span><span class="si">}</span><span class="s2">.sql&quot;</span>
    <span class="n">core_validate_file</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;validate_core_</span><span class="si">{</span><span class="n">jira_ticket</span><span class="si">}</span><span class="s2">.sql&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Write metadata patches</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">metadata_validate_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">val_f</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span>
            <span class="n">metadata_patch_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span>
        <span class="p">)</span> <span class="k">as</span> <span class="n">patch_f</span><span class="p">:</span>
            <span class="n">val_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;-- Validation: Metadata DB | </span><span class="si">{</span><span class="n">jira_ticket</span><span class="si">}</span><span class="s2"> | </span><span class="si">{</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">isoformat</span><span class="p">()</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">val_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;USE ensembl_genome_metadata;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">patch_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;-- Patch: Metadata DB | </span><span class="si">{</span><span class="n">jira_ticket</span><span class="si">}</span><span class="s2"> | </span><span class="si">{</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">isoformat</span><span class="p">()</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">patch_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;-- Validate first: </span><span class="si">{</span><span class="n">metadata_validate_file</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;USE ensembl_genome_metadata;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">genome_uuid</span><span class="p">,</span> <span class="n">genome_data</span> <span class="ow">in</span> <span class="n">grouped_patches</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">production_name</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;production_name&quot;</span><span class="p">]</span>
                <span class="n">patches</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;patches&quot;</span><span class="p">]</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;  Metadata: </span><span class="si">{</span><span class="n">production_name</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2">): </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">patches</span><span class="p">)</span><span class="si">}</span><span class="s2"> patches&quot;</span>
                <span class="p">)</span>

                <span class="n">write_metadata_patch_for_genome</span><span class="p">(</span>
                    <span class="n">val_f</span><span class="p">,</span> <span class="n">patch_f</span><span class="p">,</span> <span class="n">genome_uuid</span><span class="p">,</span> <span class="n">patches</span><span class="p">,</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;dataset_type&quot;</span><span class="p">]</span>
                <span class="p">)</span>

        <span class="c1"># Write core patches</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">core_validate_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">val_f</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span>
            <span class="n">core_patch_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span>
        <span class="p">)</span> <span class="k">as</span> <span class="n">patch_f</span><span class="p">:</span>
            <span class="n">val_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;-- Validation: Core DBs | </span><span class="si">{</span><span class="n">jira_ticket</span><span class="si">}</span><span class="s2"> | </span><span class="si">{</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">isoformat</span><span class="p">()</span><span class="si">}</span><span class="se">\n\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>

            <span class="n">patch_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;-- Patch: Core DBs | </span><span class="si">{</span><span class="n">jira_ticket</span><span class="si">}</span><span class="s2"> | </span><span class="si">{</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">isoformat</span><span class="p">()</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">patch_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;-- Validate first: </span><span class="si">{</span><span class="n">core_validate_file</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">genome_uuid</span><span class="p">,</span> <span class="n">genome_data</span> <span class="ow">in</span> <span class="n">grouped_patches</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">production_name</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;production_name&quot;</span><span class="p">]</span>
                <span class="n">core_db_name</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;core_db_name&quot;</span><span class="p">]</span>
                <span class="n">patches</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;patches&quot;</span><span class="p">]</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Core: </span><span class="si">{</span><span class="n">core_db_name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">patches</span><span class="p">)</span><span class="si">}</span><span class="s2"> patches&quot;</span><span class="p">)</span>

                <span class="n">write_core_patch_for_genome</span><span class="p">(</span>
                    <span class="n">val_f</span><span class="p">,</span> <span class="n">patch_f</span><span class="p">,</span> <span class="n">core_db_name</span><span class="p">,</span> <span class="n">patches</span><span class="p">,</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;species_id&quot;</span><span class="p">]</span>
                <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Generated files:&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">metadata_validate_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">metadata_patch_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">core_validate_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">core_patch_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="kc">True</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to generate patches: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.get_genome_adaptor" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_genome_adaptor</span><span class="p">()</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.get_genome_adaptor" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create a GenomeAdaptor instance using environment variables.</p>


<details class="requires" open>
  <summary>Requires</summary>
  <p>METADATA_URI: Connection string for metadata database
TAXONOMY_URI: Connection string for taxonomy database</p>
</details>

    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="ensembl.production.metadata.api.adaptors.genome.GenomeAdaptor">GenomeAdaptor</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>GenomeAdaptor instance or None if not available</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_genome_adaptor</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">GenomeAdaptor</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a GenomeAdaptor instance using environment variables.</span>

<span class="sd">    Requires:</span>
<span class="sd">        METADATA_URI: Connection string for metadata database</span>
<span class="sd">        TAXONOMY_URI: Connection string for taxonomy database</span>

<span class="sd">    Returns:</span>
<span class="sd">        GenomeAdaptor instance or None if not available</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">METADATA_API_AVAILABLE</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;ensembl-metadata-api package not available&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">metadata_uri</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getenv</span><span class="p">(</span><span class="s2">&quot;METADATA_URI&quot;</span><span class="p">)</span>
    <span class="n">taxonomy_uri</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getenv</span><span class="p">(</span><span class="s2">&quot;TAXONOMY_URI&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">metadata_uri</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">taxonomy_uri</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="s2">&quot;METADATA_URI and TAXONOMY_URI environment variables required. &quot;</span>
            <span class="s2">&quot;Format: mysql+pymysql://user:pass@host:port/database&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">GenomeAdaptor</span><span class="p">(</span><span class="n">metadata_uri</span><span class="o">=</span><span class="n">metadata_uri</span><span class="p">,</span> <span class="n">taxonomy_uri</span><span class="o">=</span><span class="n">taxonomy_uri</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to create GenomeAdaptor: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.get_genome_by_production_name" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_genome_by_production_name</span><span class="p">(</span><span class="n">production_name</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.get_genome_by_production_name" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch genome UUID by production name.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>production_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species production name (e.g., homo_sapiens)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict containing genome information or None if not found</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_genome_by_production_name</span><span class="p">(</span><span class="n">production_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch genome UUID by production name.</span>

<span class="sd">    Args:</span>
<span class="sd">        production_name: Species production name (e.g., homo_sapiens)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict containing genome information or None if not found</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">adaptor</span> <span class="o">=</span> <span class="n">get_genome_adaptor</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">adaptor</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">adaptor</span><span class="o">.</span><span class="n">fetch_genomes</span><span class="p">(</span><span class="n">production_name</span><span class="o">=</span><span class="n">production_name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">genome</span><span class="p">,</span> <span class="n">organism</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">release</span><span class="p">,</span> <span class="n">_site</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">return</span> <span class="p">{</span>
                <span class="s2">&quot;genome_uuid&quot;</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">genome_uuid</span><span class="p">,</span>
                <span class="s2">&quot;production_name&quot;</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">production_name</span><span class="p">,</span>
                <span class="s2">&quot;assembly_accession&quot;</span><span class="p">:</span> <span class="n">assembly</span><span class="o">.</span><span class="n">accession</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Failed to fetch genome by production name </span><span class="si">{</span><span class="n">production_name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.get_genome_by_uuid" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_genome_by_uuid</span><span class="p">(</span><span class="n">genome_uuid</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.get_genome_by_uuid" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch genome details by UUID using ensembl-metadata-api.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>genome_uuid</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genome UUID to query</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict containing genome information or None if not found</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_genome_by_uuid</span><span class="p">(</span><span class="n">genome_uuid</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch genome details by UUID using ensembl-metadata-api.</span>

<span class="sd">    Args:</span>
<span class="sd">        genome_uuid: Genome UUID to query</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict containing genome information or None if not found</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">adaptor</span> <span class="o">=</span> <span class="n">get_genome_adaptor</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">adaptor</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">adaptor</span><span class="o">.</span><span class="n">fetch_genomes_by_genome_uuid</span><span class="p">(</span><span class="n">genome_uuid</span><span class="o">=</span><span class="n">genome_uuid</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">genome</span><span class="p">,</span> <span class="n">organism</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">release</span><span class="p">,</span> <span class="n">site</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">return</span> <span class="p">{</span>
                <span class="s2">&quot;genome_uuid&quot;</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">genome_uuid</span><span class="p">,</span>
                <span class="s2">&quot;production_name&quot;</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">production_name</span><span class="p">,</span>
                <span class="s2">&quot;genebuild_version&quot;</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">genebuild_version</span><span class="p">,</span>
                <span class="s2">&quot;genebuild_date&quot;</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">genebuild_date</span><span class="p">,</span>
                <span class="s2">&quot;organism&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;organism_uuid&quot;</span><span class="p">:</span> <span class="n">organism</span><span class="o">.</span><span class="n">organism_uuid</span><span class="p">,</span>
                    <span class="s2">&quot;taxonomy_id&quot;</span><span class="p">:</span> <span class="n">organism</span><span class="o">.</span><span class="n">taxonomy_id</span><span class="p">,</span>
                    <span class="s2">&quot;scientific_name&quot;</span><span class="p">:</span> <span class="n">organism</span><span class="o">.</span><span class="n">scientific_name</span><span class="p">,</span>
                    <span class="s2">&quot;strain&quot;</span><span class="p">:</span> <span class="n">organism</span><span class="o">.</span><span class="n">strain</span><span class="p">,</span>
                    <span class="s2">&quot;biosample_id&quot;</span><span class="p">:</span> <span class="n">organism</span><span class="o">.</span><span class="n">biosample_id</span><span class="p">,</span>
                <span class="p">},</span>
                <span class="s2">&quot;assembly&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;assembly_uuid&quot;</span><span class="p">:</span> <span class="n">assembly</span><span class="o">.</span><span class="n">assembly_uuid</span><span class="p">,</span>
                    <span class="s2">&quot;accession&quot;</span><span class="p">:</span> <span class="n">assembly</span><span class="o">.</span><span class="n">accession</span><span class="p">,</span>
                    <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">assembly</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                    <span class="s2">&quot;level&quot;</span><span class="p">:</span> <span class="n">assembly</span><span class="o">.</span><span class="n">level</span><span class="p">,</span>
                <span class="p">},</span>
            <span class="p">}</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to fetch genome </span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.group_patches_by_genome" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">group_patches_by_genome</span><span class="p">(</span><span class="n">patches</span><span class="p">,</span> <span class="n">logger</span><span class="p">,</span> <span class="n">jira_ticket</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.group_patches_by_genome" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Group patches by genome UUID.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>patches</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span>[<span title="typing.Dict">Dict</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of patch dictionaries from CSV</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>logger</code>
            </td>
            <td>
                  <code><span title="logging.Logger">Logger</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Logger instance</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>jira_ticket</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Jira ticket reference</p>
              </div>
            </td>
            <td>
                  <code>&#39;&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Dict">Dict</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict mapping genome_uuid to genome info and list of patches</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">group_patches_by_genome</span><span class="p">(</span>
    <span class="n">patches</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Dict</span><span class="p">],</span> <span class="n">logger</span><span class="p">:</span> <span class="n">logging</span><span class="o">.</span><span class="n">Logger</span><span class="p">,</span> <span class="n">jira_ticket</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Group patches by genome UUID.</span>

<span class="sd">    Args:</span>
<span class="sd">        patches: List of patch dictionaries from CSV</span>
<span class="sd">        logger: Logger instance</span>
<span class="sd">        jira_ticket: Jira ticket reference</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict mapping genome_uuid to genome info and list of patches</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">grouped</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">patch</span> <span class="ow">in</span> <span class="n">patches</span><span class="p">:</span>
        <span class="c1"># Resolve genome information</span>
        <span class="n">genome_info</span> <span class="o">=</span> <span class="n">resolve_genome_info</span><span class="p">(</span><span class="n">patch</span><span class="p">,</span> <span class="n">logger</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">genome_info</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">patch</span><span class="p">[</span><span class="s1">&#39;row_num&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">: Skipping due to resolution failure&quot;</span>
            <span class="p">)</span>
            <span class="k">continue</span>

        <span class="n">genome_uuid</span> <span class="o">=</span> <span class="n">genome_info</span><span class="p">[</span><span class="s2">&quot;genome_uuid&quot;</span><span class="p">]</span>

        <span class="c1"># Initialize genome entry if not exists</span>
        <span class="k">if</span> <span class="n">genome_uuid</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">grouped</span><span class="p">:</span>
            <span class="n">grouped</span><span class="p">[</span><span class="n">genome_uuid</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;genome_uuid&quot;</span><span class="p">:</span> <span class="n">genome_uuid</span><span class="p">,</span>
                <span class="s2">&quot;production_name&quot;</span><span class="p">:</span> <span class="n">genome_info</span><span class="p">[</span><span class="s2">&quot;production_name&quot;</span><span class="p">],</span>
                <span class="s2">&quot;core_db_name&quot;</span><span class="p">:</span> <span class="n">genome_info</span><span class="p">[</span><span class="s2">&quot;core_db_name&quot;</span><span class="p">],</span>
                <span class="s2">&quot;dataset_type&quot;</span><span class="p">:</span> <span class="n">patch</span><span class="p">[</span><span class="s2">&quot;dataset_type&quot;</span><span class="p">],</span>
                <span class="s2">&quot;species_id&quot;</span><span class="p">:</span> <span class="n">patch</span><span class="p">[</span><span class="s2">&quot;species_id&quot;</span><span class="p">],</span>
                <span class="s2">&quot;jira_ticket&quot;</span><span class="p">:</span> <span class="n">jira_ticket</span><span class="p">,</span>
                <span class="s2">&quot;patches&quot;</span><span class="p">:</span> <span class="p">[],</span>
                <span class="s2">&quot;row_nums&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="p">}</span>

        <span class="c1"># Add patch to genome&#39;s list (include table_location)</span>
        <span class="n">grouped</span><span class="p">[</span><span class="n">genome_uuid</span><span class="p">][</span><span class="s2">&quot;patches&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">(</span><span class="n">patch</span><span class="p">[</span><span class="s2">&quot;meta_key&quot;</span><span class="p">],</span> <span class="n">patch</span><span class="p">[</span><span class="s2">&quot;desired_meta_value&quot;</span><span class="p">],</span> <span class="n">patch</span><span class="p">[</span><span class="s2">&quot;table_location&quot;</span><span class="p">])</span>
        <span class="p">)</span>
        <span class="n">grouped</span><span class="p">[</span><span class="n">genome_uuid</span><span class="p">][</span><span class="s2">&quot;row_nums&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">patch</span><span class="p">[</span><span class="s2">&quot;row_num&quot;</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">grouped</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main entry point for the beta patcher script.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main entry point for the beta patcher script.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Generate standardized SQL patches for Ensembl beta metadata issues from CSV input&quot;</span><span class="p">,</span>
        <span class="n">formatter_class</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">RawDescriptionHelpFormatter</span><span class="p">,</span>
        <span class="n">epilog</span><span class="o">=</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">Examples:</span>
<span class="s2">  # Process patches from CSV file</span>
<span class="s2">  python beta_patcher.py patches.csv --jira-ticket EBD-1111 --output-dir ./patches/</span>

<span class="s2">  # Process CSV with custom core suffix</span>
<span class="s2">  python beta_patcher.py patches.csv --jira-ticket EBD-1111 --core-suffix _core_115_1</span>

<span class="s2">  # Validation files are automatically generated</span>
<span class="s2">  # Run validate_*.sql files before applying patch_*.sql files</span>

<span class="s2">CSV Format:</span>
<span class="s2">  Required columns: meta_key, desired_meta_value</span>
<span class="s2">  Identifier (one required): production_name OR genome_uuid</span>
<span class="s2">  Optional columns: dataset_type, species_id, table_location</span>

<span class="s2">  Example:</span>
<span class="s2">    production_name,genome_uuid,meta_key,desired_meta_value,dataset_type,species_id,table_location</span>
<span class="s2">    homo_sapiens,,assembly.name,GRCh38.p14,genebuild,1,dataset_attribute</span>
<span class="s2">    ,a7335667-93e7-11ec-a39d-005056b38ce3,organism.strain,Reference,genebuild,1,genome</span>

<span class="s2">See patches_template.csv for a complete example.</span>
<span class="s2">        &quot;&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Positional argument</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;csv_file&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="n">Path</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;CSV file with patches (see patches_template.csv for format)&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Jira ticket (required)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--jira-ticket&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Jira ticket reference (e.g., EBD-1111)&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Core database suffix</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--core-suffix&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;_core_114_1&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Suffix to append to production_name for core DB name (default: _core_114_1)&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Database connection (uses environment variables METADATA_URI and TAXONOMY_URI)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--metadata-uri&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Metadata database URI (overrides METADATA_URI env var)&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--taxonomy-uri&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxonomy database URI (overrides TAXONOMY_URI env var)&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Output options</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--output-dir&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="n">Path</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">(),</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output directory for patch files (default: current directory)&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="c1"># Validate CSV file exists</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">csv_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CSV file not found: </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">csv_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Override environment variables if provided</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">metadata_uri</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;METADATA_URI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">metadata_uri</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">taxonomy_uri</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;TAXONOMY_URI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">taxonomy_uri</span>

    <span class="c1"># Create output directory</span>
    <span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Set up logging</span>
    <span class="n">log_file</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">args</span><span class="o">.</span><span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;patch_csv_</span><span class="si">{</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y%m</span><span class="si">%d</span><span class="s1">_%H%M%S&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">.log&quot;</span>
    <span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
        <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span>
        <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> - </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">handlers</span><span class="o">=</span><span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">FileHandler</span><span class="p">(</span><span class="n">log_file</span><span class="p">),</span> <span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)],</span>
    <span class="p">)</span>
    <span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">patches</span> <span class="o">=</span> <span class="n">read_csv_patches</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">csv_file</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">core_suffix</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">patches</span><span class="p">)</span><span class="si">}</span><span class="s2"> patches from </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">csv_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to read CSV: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">1</span>

    <span class="n">grouped_patches</span> <span class="o">=</span> <span class="n">group_patches_by_genome</span><span class="p">(</span><span class="n">patches</span><span class="p">,</span> <span class="n">logger</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">jira_ticket</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Grouped into </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">grouped_patches</span><span class="p">)</span><span class="si">}</span><span class="s2"> genome(s)&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">grouped_patches</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;No valid patches to process after grouping&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">1</span>

    <span class="c1"># Generate consolidated patch files</span>
    <span class="k">if</span> <span class="n">generate_all_patches</span><span class="p">(</span><span class="n">grouped_patches</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">jira_ticket</span><span class="p">,</span> <span class="n">logger</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Success! Log: </span><span class="si">{</span><span class="n">log_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed. Log: </span><span class="si">{</span><span class="n">log_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">1</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.read_csv_patches" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">read_csv_patches</span><span class="p">(</span><span class="n">csv_file</span><span class="p">,</span> <span class="n">core_suffix</span><span class="o">=</span><span class="s1">&#39;_core_114_1&#39;</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.read_csv_patches" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Read patches from CSV file.</p>
<p>CSV columns:
- production_name OR genome_uuid (one required)
- meta_key (required)
- desired_meta_value (required)
- dataset_type (optional, default: genebuild)
- species_id (optional, default: 1)
- table_location (optional, default: dataset_attribute)</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>csv_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to CSV file</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>core_suffix</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Suffix to append to production_name for core DB name</p>
              </div>
            </td>
            <td>
                  <code>&#39;_core_114_1&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[<span title="typing.Dict">Dict</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of patch dictionaries</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">read_csv_patches</span><span class="p">(</span><span class="n">csv_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">core_suffix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;_core_114_1&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read patches from CSV file.</span>

<span class="sd">    CSV columns:</span>
<span class="sd">    - production_name OR genome_uuid (one required)</span>
<span class="sd">    - meta_key (required)</span>
<span class="sd">    - desired_meta_value (required)</span>
<span class="sd">    - dataset_type (optional, default: genebuild)</span>
<span class="sd">    - species_id (optional, default: 1)</span>
<span class="sd">    - table_location (optional, default: dataset_attribute)</span>

<span class="sd">    Args:</span>
<span class="sd">        csv_file: Path to CSV file</span>
<span class="sd">        core_suffix: Suffix to append to production_name for core DB name</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of patch dictionaries</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError if CSV format is invalid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">patches</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">csv_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

        <span class="c1"># Validate required columns</span>
        <span class="n">required_cols</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;meta_key&quot;</span><span class="p">,</span> <span class="s2">&quot;desired_meta_value&quot;</span><span class="p">}</span>
        <span class="n">identifier_cols</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;production_name&quot;</span><span class="p">,</span> <span class="s2">&quot;genome_uuid&quot;</span><span class="p">}</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">required_cols</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">reader</span><span class="o">.</span><span class="n">fieldnames</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CSV must contain columns: </span><span class="si">{</span><span class="n">required_cols</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">identifier_cols</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">reader</span><span class="o">.</span><span class="n">fieldnames</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CSV must contain at least one of: </span><span class="si">{</span><span class="n">identifier_cols</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">row_num</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
            <span class="n">reader</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">2</span>
        <span class="p">):</span>  <span class="c1"># Start at 2 for header + 1-indexed</span>
            <span class="c1"># Check we have at least one identifier</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;production_name&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;genome_uuid&quot;</span><span class="p">):</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row_num</span><span class="si">}</span><span class="s2">: Must provide either production_name or genome_uuid&quot;</span>
                <span class="p">)</span>
                <span class="k">continue</span>

            <span class="c1"># Check required fields</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;meta_key&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;desired_meta_value&quot;</span><span class="p">):</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row_num</span><span class="si">}</span><span class="s2">: Missing required fields meta_key or desired_meta_value&quot;</span>
                <span class="p">)</span>
                <span class="k">continue</span>

            <span class="n">patch</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;production_name&quot;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;production_name&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                <span class="s2">&quot;genome_uuid&quot;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;genome_uuid&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                <span class="s2">&quot;meta_key&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;meta_key&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                <span class="s2">&quot;desired_meta_value&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;desired_meta_value&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                <span class="s2">&quot;dataset_type&quot;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dataset_type&quot;</span><span class="p">,</span> <span class="s2">&quot;genebuild&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                <span class="s2">&quot;species_id&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;species_id&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)),</span>
                <span class="s2">&quot;table_location&quot;</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                    <span class="s2">&quot;table_location&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_attribute&quot;</span>
                <span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                <span class="s2">&quot;core_suffix&quot;</span><span class="p">:</span> <span class="n">core_suffix</span><span class="p">,</span>
                <span class="s2">&quot;row_num&quot;</span><span class="p">:</span> <span class="n">row_num</span><span class="p">,</span>
            <span class="p">}</span>

            <span class="n">patches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">patch</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">patches</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.resolve_genome_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">resolve_genome_info</span><span class="p">(</span><span class="n">patch</span><span class="p">,</span> <span class="n">logger</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.resolve_genome_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Resolve genome UUID and production name from patch data.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>patch</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Patch dictionary from CSV</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>logger</code>
            </td>
            <td>
                  <code><span title="logging.Logger">Logger</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Logger instance</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict with genome_uuid, production_name, core_db_name, or None if failed</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">resolve_genome_info</span><span class="p">(</span><span class="n">patch</span><span class="p">:</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">logger</span><span class="p">:</span> <span class="n">logging</span><span class="o">.</span><span class="n">Logger</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Resolve genome UUID and production name from patch data.</span>

<span class="sd">    Args:</span>
<span class="sd">        patch: Patch dictionary from CSV</span>
<span class="sd">        logger: Logger instance</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict with genome_uuid, production_name, core_db_name, or None if failed</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">row_num</span> <span class="o">=</span> <span class="n">patch</span><span class="p">[</span><span class="s2">&quot;row_num&quot;</span><span class="p">]</span>
    <span class="n">genome_uuid</span> <span class="o">=</span> <span class="n">patch</span><span class="p">[</span><span class="s2">&quot;genome_uuid&quot;</span><span class="p">]</span>
    <span class="n">production_name</span> <span class="o">=</span> <span class="n">patch</span><span class="p">[</span><span class="s2">&quot;production_name&quot;</span><span class="p">]</span>

    <span class="c1"># If production_name provided, resolve to genome_uuid</span>
    <span class="k">if</span> <span class="n">production_name</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">genome_uuid</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row_num</span><span class="si">}</span><span class="s2">: Resolving genome_uuid from production_name: </span><span class="si">{</span><span class="n">production_name</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">genomes</span> <span class="o">=</span> <span class="n">suggest_genomes_by_production_name</span><span class="p">(</span><span class="n">production_name</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">genomes</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row_num</span><span class="si">}</span><span class="s2">: No genomes found for production_name: </span><span class="si">{</span><span class="n">production_name</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genomes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row_num</span><span class="si">}</span><span class="s2">: Ambiguous production_name &#39;</span><span class="si">{</span><span class="n">production_name</span><span class="si">}</span><span class="s2">&#39; matches </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">genomes</span><span class="p">)</span><span class="si">}</span><span class="s2"> genomes. &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;Please specify genome_uuid instead. Found UUIDs: </span><span class="si">{</span><span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="s1">&#39;genome_uuid&#39;</span><span class="p">]</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">g</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">genomes</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">genome_uuid</span> <span class="o">=</span> <span class="n">genomes</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;genome_uuid&quot;</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row_num</span><span class="si">}</span><span class="s2">: Resolved to genome_uuid: </span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># If genome_uuid provided, get production_name if not already set</span>
    <span class="k">if</span> <span class="n">genome_uuid</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">production_name</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row_num</span><span class="si">}</span><span class="s2">: Fetching production_name for genome_uuid: </span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">genome_info</span> <span class="o">=</span> <span class="n">get_genome_by_uuid</span><span class="p">(</span><span class="n">genome_uuid</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">genome_info</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row_num</span><span class="si">}</span><span class="s2">: Could not fetch genome information for UUID: </span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">production_name</span> <span class="o">=</span> <span class="n">genome_info</span><span class="p">[</span><span class="s2">&quot;production_name&quot;</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Row </span><span class="si">{</span><span class="n">row_num</span><span class="si">}</span><span class="s2">: Found production_name: </span><span class="si">{</span><span class="n">production_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Build core database name</span>
    <span class="n">core_db_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">production_name</span><span class="si">}{</span><span class="n">patch</span><span class="p">[</span><span class="s1">&#39;core_suffix&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="k">return</span> <span class="p">{</span>
        <span class="s2">&quot;genome_uuid&quot;</span><span class="p">:</span> <span class="n">genome_uuid</span><span class="p">,</span>
        <span class="s2">&quot;production_name&quot;</span><span class="p">:</span> <span class="n">production_name</span><span class="p">,</span>
        <span class="s2">&quot;core_db_name&quot;</span><span class="p">:</span> <span class="n">core_db_name</span><span class="p">,</span>
    <span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.setup_logging" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">setup_logging</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">genome_uuid</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.setup_logging" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Set up logging configuration.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">setup_logging</span><span class="p">(</span><span class="n">output_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">genome_uuid</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">logging</span><span class="o">.</span><span class="n">Logger</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Set up logging configuration.&quot;&quot;&quot;</span>
    <span class="n">log_file</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">output_dir</span>
        <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;patch_</span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s1">&#39;%Y%m</span><span class="si">%d</span><span class="s1">_%H%M%S&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">.log&quot;</span>
    <span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
        <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span>
        <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> - </span><span class="si">%(levelname)s</span><span class="s2"> - </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">handlers</span><span class="o">=</span><span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">FileHandler</span><span class="p">(</span><span class="n">log_file</span><span class="p">),</span> <span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)],</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.suggest_genomes_by_production_name" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">suggest_genomes_by_production_name</span><span class="p">(</span><span class="n">production_name</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.suggest_genomes_by_production_name" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Suggest genome UUIDs based on production name.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>production_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species production name (e.g., homo_sapiens)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[<span title="typing.Dict">Dict</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of matching genomes with their UUIDs</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">suggest_genomes_by_production_name</span><span class="p">(</span><span class="n">production_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Suggest genome UUIDs based on production name.</span>

<span class="sd">    Args:</span>
<span class="sd">        production_name: Species production name (e.g., homo_sapiens)</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of matching genomes with their UUIDs</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">adaptor</span> <span class="o">=</span> <span class="n">get_genome_adaptor</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">adaptor</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">adaptor</span><span class="o">.</span><span class="n">fetch_genomes</span><span class="p">(</span><span class="n">production_name</span><span class="o">=</span><span class="n">production_name</span><span class="p">)</span>
        <span class="n">genomes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">genome</span><span class="p">,</span> <span class="n">organism</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">release</span><span class="p">,</span> <span class="n">_site</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">genomes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;genome_uuid&quot;</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">genome_uuid</span><span class="p">,</span>
                    <span class="s2">&quot;production_name&quot;</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">production_name</span><span class="p">,</span>
                    <span class="s2">&quot;assembly_accession&quot;</span><span class="p">:</span> <span class="n">assembly</span><span class="o">.</span><span class="n">accession</span><span class="p">,</span>
                    <span class="s2">&quot;assembly_name&quot;</span><span class="p">:</span> <span class="n">assembly</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                    <span class="s2">&quot;genebuild_version&quot;</span><span class="p">:</span> <span class="n">genome</span><span class="o">.</span><span class="n">genebuild_version</span><span class="p">,</span>
                    <span class="s2">&quot;release_version&quot;</span><span class="p">:</span> <span class="n">release</span><span class="o">.</span><span class="n">version</span> <span class="k">if</span> <span class="n">release</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="p">}</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">genomes</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Failed to fetch genomes by production name </span><span class="si">{</span><span class="n">production_name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="p">[]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.write_core_patch_for_genome" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">write_core_patch_for_genome</span><span class="p">(</span><span class="n">validate_file</span><span class="p">,</span> <span class="n">patch_file</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">patches</span><span class="p">,</span> <span class="n">species_id</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.write_core_patch_for_genome" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Write validation and patch SQL for a single core database to open file handles.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>validate_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Open file handle for validation SQL</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>patch_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Open file handle for patch SQL</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>database</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Core database name</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>patches</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[<span title="str">str</span>, <span title="str">str</span>, <span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of (meta_key, new_value, table_location) tuples (table_location ignored for core DB)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>species_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Species ID</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_core_patch_for_genome</span><span class="p">(</span>
    <span class="n">validate_file</span><span class="p">,</span>
    <span class="n">patch_file</span><span class="p">,</span>
    <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">patches</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="n">species_id</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write validation and patch SQL for a single core database to open file handles.</span>

<span class="sd">    Args:</span>
<span class="sd">        validate_file: Open file handle for validation SQL</span>
<span class="sd">        patch_file: Open file handle for patch SQL</span>
<span class="sd">        database: Core database name</span>
<span class="sd">        patches: List of (meta_key, new_value, table_location) tuples (table_location ignored for core DB)</span>
<span class="sd">        species_id: Species ID</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">meta_key</span><span class="p">,</span> <span class="n">new_value</span><span class="p">,</span> <span class="n">_table_location</span> <span class="ow">in</span> <span class="n">patches</span><span class="p">:</span>
        <span class="n">escaped_value</span> <span class="o">=</span> <span class="n">new_value</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;&#39;&#39;&quot;</span><span class="p">)</span>

        <span class="c1"># Write validation SQL</span>
        <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;-- </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> | </span><span class="si">{</span><span class="n">meta_key</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;USE </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2">;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;SELECT &#39;</span><span class="si">{</span><span class="n">meta_key</span><span class="si">}</span><span class="s2">&#39; AS meta_key, meta_value AS current_value, &quot;</span>
        <span class="p">)</span>
        <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">escaped_value</span><span class="si">}</span><span class="s2">&#39; AS proposed_value</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;FROM meta</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">_core_db_where</span><span class="p">(</span><span class="n">meta_key</span><span class="p">,</span> <span class="n">species_id</span><span class="p">))</span>
        <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Write patch SQL</span>
        <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;-- </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> | </span><span class="si">{</span><span class="n">meta_key</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;USE </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2">;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;UPDATE meta</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;SET meta_value = &#39;</span><span class="si">{</span><span class="n">escaped_value</span><span class="si">}</span><span class="s2">&#39;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">_core_db_where</span><span class="p">(</span><span class="n">meta_key</span><span class="p">,</span> <span class="n">species_id</span><span class="p">))</span>
        <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;INSERT IGNORE INTO meta (species_id, meta_key, meta_value)</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;VALUES (</span><span class="si">{</span><span class="n">species_id</span><span class="si">}</span><span class="s2">, &#39;</span><span class="si">{</span><span class="n">meta_key</span><span class="si">}</span><span class="s2">&#39;, &#39;</span><span class="si">{</span><span class="n">escaped_value</span><span class="si">}</span><span class="s2">&#39;);</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.beta_patcher.write_metadata_patch_for_genome" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">write_metadata_patch_for_genome</span><span class="p">(</span><span class="n">validate_file</span><span class="p">,</span> <span class="n">patch_file</span><span class="p">,</span> <span class="n">genome_uuid</span><span class="p">,</span> <span class="n">patches</span><span class="p">,</span> <span class="n">dataset_type</span><span class="o">=</span><span class="s1">&#39;genebuild&#39;</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.beta_patcher.write_metadata_patch_for_genome" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Write validation and patch SQL for a single genome to open file handles.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>validate_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Open file handle for validation SQL</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>patch_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Open file handle for patch SQL</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genome_uuid</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Genome UUID</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>patches</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[<span title="str">str</span>, <span title="str">str</span>, <span title="str">str</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of (attribute_name, new_value, table_location) tuples</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dataset_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dataset type (genebuild, assembly, etc.)</p>
              </div>
            </td>
            <td>
                  <code>&#39;genebuild&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/beta_patcher.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_metadata_patch_for_genome</span><span class="p">(</span>
    <span class="n">validate_file</span><span class="p">,</span>
    <span class="n">patch_file</span><span class="p">,</span>
    <span class="n">genome_uuid</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">patches</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="n">dataset_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;genebuild&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write validation and patch SQL for a single genome to open file handles.</span>

<span class="sd">    Args:</span>
<span class="sd">        validate_file: Open file handle for validation SQL</span>
<span class="sd">        patch_file: Open file handle for patch SQL</span>
<span class="sd">        genome_uuid: Genome UUID</span>
<span class="sd">        patches: List of (attribute_name, new_value, table_location) tuples</span>
<span class="sd">        dataset_type: Dataset type (genebuild, assembly, etc.)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">attribute_name</span><span class="p">,</span> <span class="n">new_value</span><span class="p">,</span> <span class="n">table_location</span> <span class="ow">in</span> <span class="n">patches</span><span class="p">:</span>
        <span class="n">escaped_value</span> <span class="o">=</span> <span class="n">new_value</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;&#39;&#39;&quot;</span><span class="p">)</span>

        <span class="c1"># Write validation SQL</span>
        <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;-- </span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2"> | </span><span class="si">{</span><span class="n">attribute_name</span><span class="si">}</span><span class="s2"> (table: </span><span class="si">{</span><span class="n">table_location</span><span class="si">}</span><span class="s2">)</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">table_location</span> <span class="o">==</span> <span class="s2">&quot;dataset_attribute&quot;</span><span class="p">:</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;SELECT genome.genome_uuid, dataset_attribute.value AS current_value, &quot;</span>
            <span class="p">)</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">escaped_value</span><span class="si">}</span><span class="s2">&#39; AS proposed_value</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;FROM dataset_attribute</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">_metadata_db_joins</span><span class="p">())</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="n">_metadata_db_where</span><span class="p">(</span><span class="n">genome_uuid</span><span class="p">,</span> <span class="n">dataset_type</span><span class="p">,</span> <span class="n">attribute_name</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">table_location</span> <span class="o">==</span> <span class="s2">&quot;genome&quot;</span><span class="p">:</span>
            <span class="n">column_name</span> <span class="o">=</span> <span class="n">attribute_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;SELECT genome.genome_uuid, genome.</span><span class="si">{</span><span class="n">column_name</span><span class="si">}</span><span class="s2"> AS current_value, &quot;</span>
            <span class="p">)</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">escaped_value</span><span class="si">}</span><span class="s2">&#39; AS proposed_value</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;FROM genome</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;WHERE genome.genome_uuid = &#39;</span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2">&#39;;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;-- WARNING: Unknown table_location &#39;</span><span class="si">{</span><span class="n">table_location</span><span class="si">}</span><span class="s2">&#39;</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">validate_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;-- Manual validation required for </span><span class="si">{</span><span class="n">attribute_name</span><span class="si">}</span><span class="se">\n\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Write patch SQL</span>
        <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;-- </span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2"> | </span><span class="si">{</span><span class="n">attribute_name</span><span class="si">}</span><span class="s2"> (table: </span><span class="si">{</span><span class="n">table_location</span><span class="si">}</span><span class="s2">)</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">table_location</span> <span class="o">==</span> <span class="s2">&quot;dataset_attribute&quot;</span><span class="p">:</span>
            <span class="c1"># DELETE existing attribute value (if exists)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;DELETE dataset_attribute FROM dataset_attribute</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">_metadata_db_joins</span><span class="p">())</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="n">_metadata_db_where</span><span class="p">(</span><span class="n">genome_uuid</span><span class="p">,</span> <span class="n">dataset_type</span><span class="p">,</span> <span class="n">attribute_name</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># INSERT new attribute value</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;INSERT INTO dataset_attribute (dataset_id, attribute_id, value)</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;SELECT dataset.dataset_id, attribute.attribute_id, &quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">escaped_value</span><span class="si">}</span><span class="s2">&#39;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;FROM dataset</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;JOIN genome_dataset ON dataset.dataset_id = genome_dataset.dataset_id</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;JOIN genome ON genome_dataset.genome_id = genome.genome_id</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;JOIN attribute ON attribute.name = &quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">attribute_name</span><span class="si">}</span><span class="s2">&#39;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;WHERE genome.genome_uuid = &#39;</span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2">&#39;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;AND dataset.name = &#39;</span><span class="si">{</span><span class="n">dataset_type</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">elif</span> <span class="n">table_location</span> <span class="o">==</span> <span class="s2">&quot;genome&quot;</span><span class="p">:</span>
            <span class="n">column_name</span> <span class="o">=</span> <span class="n">attribute_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;UPDATE genome</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;SET </span><span class="si">{</span><span class="n">column_name</span><span class="si">}</span><span class="s2"> = &#39;</span><span class="si">{</span><span class="n">escaped_value</span><span class="si">}</span><span class="s2">&#39;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;WHERE genome_uuid = &#39;</span><span class="si">{</span><span class="n">genome_uuid</span><span class="si">}</span><span class="s2">&#39;;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;-- ERROR: Unknown table_location &#39;</span><span class="si">{</span><span class="n">table_location</span><span class="si">}</span><span class="s2">&#39;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">patch_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;-- Manual SQL required for </span><span class="si">{</span><span class="n">attribute_name</span><span class="si">}</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.metadata.core_meta_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.metadata.core_meta_data</code>


<a href="#ensembl.genes.metadata.core_meta_data" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.core_dict" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">core_dict</span> <span class="o">=</span> <span class="p">{}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.core_dict" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.core_meta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">core_meta</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">core_query</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;staging&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]),</span> <span class="n">user</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;staging&#39;</span><span class="p">][</span><span class="s1">&#39;db_user&#39;</span><span class="p">]),</span> <span class="n">port</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;staging&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]),</span> <span class="n">database</span><span class="o">=</span><span class="n">db</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.core_meta" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.core_query" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">core_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;SELECT meta_key,meta_value FROM meta WHERE species_id = </span><span class="si">{</span><span class="n">species_id</span><span class="si">}</span><span class="s1"> AND (meta_key LIKE &#39;</span><span class="n">assembly</span><span class="o">%</span><span class="s1">&#39; OR meta_key LIKE &#39;</span><span class="n">species</span><span class="o">%</span><span class="s1">&#39; OR meta_key LIKE &#39;</span><span class="n">genebuild</span><span class="o">%</span><span class="s1">&#39; OR meta_key LIKE &#39;</span><span class="n">organism</span><span class="o">%</span><span class="s1">&#39; OR meta_key LIKE &#39;</span><span class="n">sample</span><span class="o">%</span><span class="s1">&#39; OR meta_key LIKE &#39;</span><span class="n">annotation</span><span class="o">%</span><span class="s1">&#39; OR meta_key LIKE &#39;</span><span class="n">gencode</span><span class="o">%</span><span class="s1">&#39;);&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.core_query" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.current_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">current_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">rank_dict</span><span class="p">[</span><span class="s1">&#39;parent_id&#39;</span><span class="p">])</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.current_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.db" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">db</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">db_name</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.db" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.gca_accession" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">gca_accession</span> <span class="o">=</span> <span class="n">core_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.gca_accession" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.log_file_path" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">log_file_path</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">production_name</span><span class="si">}</span><span class="s1">_metadata.log&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.log_file_path" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.log_ini_path" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">log_ini_path</span> <span class="o">=</span> <span class="n">metadata_dir</span> <span class="o">/</span> <span class="s1">&#39;logging.conf&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.log_ini_path" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.logger" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.logger" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.meta_keys_to_remove" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">meta_keys_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;species.strain&#39;</span><span class="p">,</span> <span class="s1">&#39;strain.type&#39;</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.meta_keys_to_remove" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.meta_value" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">meta_value</span> <span class="o">=</span> <span class="n">truth_dict</span><span class="p">[</span><span class="n">meta_key</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;&#39;&#39;&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="n">truth_dict</span><span class="p">[</span><span class="n">meta_key</span><span class="p">]</span> <span class="k">else</span> <span class="kc">None</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.meta_value" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.metadata_dir" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">metadata_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.metadata_dir" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.name_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">name_info</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">name_query</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]),</span> <span class="n">user</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_user&#39;</span><span class="p">]),</span> <span class="n">port</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]),</span> <span class="n">database</span><span class="o">=</span><span class="s1">&#39;ncbi_taxonomy&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.name_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.name_query" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">name_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;SELECT name FROM ncbi_taxa_name WHERE taxon_id=</span><span class="si">{</span><span class="n">truth_dict</span><span class="p">[</span><span class="s1">&#39;organism.taxonomy_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s1"> AND name_class=&#39;</span><span class="n">genbank</span> <span class="n">common</span> <span class="n">name</span><span class="s1">&#39;;&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.name_query" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.output_dir" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">output_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.output_dir" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.parent_rank_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parent_rank_info</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">parent_taxonomy_query</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]),</span> <span class="n">user</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_user&#39;</span><span class="p">]),</span> <span class="n">port</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]),</span> <span class="n">database</span><span class="o">=</span><span class="s1">&#39;ncbi_taxonomy&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.parent_rank_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.parent_taxonomy_query" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parent_taxonomy_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;SELECT rank, parent_id FROM ncbi_taxa_node WHERE taxon_id=</span><span class="si">{</span><span class="n">rank_dict</span><span class="p">[</span><span class="s1">&#39;parent_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s1">;&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.parent_taxonomy_query" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.parser" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Prepare SQL updates for core dbs&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.parser" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.parts" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">parts</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.parts" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.provider_dict" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">provider_dict</span> <span class="o">=</span> <span class="p">{}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.provider_dict" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.provider_static_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">provider_static_file</span> <span class="o">=</span> <span class="n">metadata_dir</span> <span class="o">/</span> <span class="s1">&#39;provider_static.txt&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.provider_static_file" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.rank_dict" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">rank_dict</span> <span class="o">=</span> <span class="p">{}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.rank_dict" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.rank_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">rank_info</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">taxonomy_query</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]),</span> <span class="n">user</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_user&#39;</span><span class="p">]),</span> <span class="n">port</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]),</span> <span class="n">database</span><span class="o">=</span><span class="s1">&#39;ncbi_taxonomy&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.rank_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.ref_accession" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">ref_accession</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.ref_accession" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.ref_list" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">ref_list</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">ref_static_file</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.ref_list" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.ref_static_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">ref_static_file</span> <span class="o">=</span> <span class="n">metadata_dir</span> <span class="o">/</span> <span class="s1">&#39;ref_static.txt&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.ref_static_file" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.required_meta_keys" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">required_meta_keys</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;organism.production_name&#39;</span><span class="p">,</span> <span class="s1">&#39;organism.taxonomy_id&#39;</span><span class="p">,</span> <span class="s1">&#39;organism.scientific_name&#39;</span><span class="p">,</span> <span class="s1">&#39;assembly.accession&#39;</span><span class="p">,</span> <span class="s1">&#39;assembly.name&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.version&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.method&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.method_display&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.start_date&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.annotation_source&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.provider_name&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.provider_url&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.sample_gene&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.sample_location&#39;</span><span class="p">,</span> <span class="s1">&#39;genebuild.last_geneset_update&#39;</span><span class="p">,</span> <span class="s1">&#39;organism.biosample_id&#39;</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.required_meta_keys" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.s_name_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">s_name_info</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">s_name_query</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]),</span> <span class="n">user</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_user&#39;</span><span class="p">]),</span> <span class="n">port</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]),</span> <span class="n">database</span><span class="o">=</span><span class="s1">&#39;ncbi_taxonomy&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.s_name_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.s_name_query" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">s_name_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;SELECT name FROM ncbi_taxa_name WHERE taxon_id=</span><span class="si">{</span><span class="n">truth_dict</span><span class="p">[</span><span class="s1">&#39;organism.taxonomy_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s1"> AND name_class=&#39;</span><span class="n">scientific</span> <span class="n">name</span><span class="s1">&#39;;&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.s_name_query" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.server_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">server_info</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;staging&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;db_host&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">host</span><span class="p">,</span> <span class="s1">&#39;db_port&#39;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">port</span><span class="p">),</span> <span class="s1">&#39;db_user&#39;</span><span class="p">:</span> <span class="s1">&#39;ensro&#39;</span><span class="p">},</span> <span class="s1">&#39;meta&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;db_host&#39;</span><span class="p">:</span> <span class="s1">&#39;mysql-ens-meta-prod-1.ebi.ac.uk&#39;</span><span class="p">,</span> <span class="s1">&#39;db_port&#39;</span><span class="p">:</span> <span class="mi">4483</span><span class="p">,</span> <span class="s1">&#39;db_user&#39;</span><span class="p">:</span> <span class="s1">&#39;ensro&#39;</span><span class="p">,</span> <span class="s1">&#39;db_pass&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">}}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.server_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.snp" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">snp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.snp" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.snp_list" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">snp_list</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">snp_static_file</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.snp_list" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.snp_static_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">snp_static_file</span> <span class="o">=</span> <span class="n">metadata_dir</span> <span class="o">/</span> <span class="s1">&#39;snp_static.txt&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.snp_static_file" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.species_id" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">species_id</span> <span class="o">=</span> <span class="n">species_meta</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.species_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.species_meta" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">species_meta</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">species_query</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;staging&#39;</span><span class="p">][</span><span class="s1">&#39;db_host&#39;</span><span class="p">]),</span> <span class="n">user</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;staging&#39;</span><span class="p">][</span><span class="s1">&#39;db_user&#39;</span><span class="p">]),</span> <span class="n">port</span><span class="o">=</span><span class="p">(</span><span class="n">server_info</span><span class="p">[</span><span class="s1">&#39;staging&#39;</span><span class="p">][</span><span class="s1">&#39;db_port&#39;</span><span class="p">]),</span> <span class="n">database</span><span class="o">=</span><span class="n">db</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.species_meta" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.species_query" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">species_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;SELECT species_id FROM meta WHERE meta_key=&#39;</span><span class="n">species</span><span class="o">.</span><span class="n">production_name</span><span class="s1">&#39; AND meta_value=&#39;</span><span class="p">{</span><span class="n">core_dict</span><span class="p">[</span><span class="s1">&#39;species.production_name&#39;</span><span class="p">]}</span><span class="s1">&#39;;&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.species_query" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.sql_out" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">sql_out</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">production_name</span><span class="si">}</span><span class="s1">.sql&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.sql_out" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.taxonomy_query" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">taxonomy_query</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;SELECT rank, parent_id FROM ncbi_taxa_node WHERE taxon_id=</span><span class="si">{</span><span class="n">truth_dict</span><span class="p">[</span><span class="s1">&#39;organism.taxonomy_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s1">;&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.taxonomy_query" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.truth_dict" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">truth_dict</span> <span class="o">=</span> <span class="p">{}</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.truth_dict" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.url" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">url</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.url" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.url_list" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">url_list</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">url_static_file</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.url_list" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.metadata.core_meta_data.url_static_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">url_static_file</span> <span class="o">=</span> <span class="n">metadata_dir</span> <span class="o">/</span> <span class="s1">&#39;url_static.txt&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.metadata.core_meta_data.url_static_file" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.core_meta_data.get_biosample_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_biosample_metadata</span><span class="p">(</span><span class="n">biosample_id</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">,</span> <span class="n">assembly_name</span><span class="p">,</span> <span class="n">scientific_name</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.core_meta_data.get_biosample_metadata" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/core_meta_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_biosample_metadata</span><span class="p">(</span>
    <span class="n">biosample_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">assembly_accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">assembly_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">scientific_name</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
    <span class="n">biosample_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/biosamples/samples/</span><span class="si">{</span><span class="n">biosample_id</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">biosample_return</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">biosample_url</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
    <span class="n">return_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;organism.strain&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;organism.strain_type&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">}</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">biosample_data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">biosample_return</span><span class="p">)</span>

        <span class="n">strain_types</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;population&quot;</span><span class="p">,</span> <span class="s2">&quot;race&quot;</span><span class="p">,</span> <span class="s2">&quot;ecotype&quot;</span><span class="p">,</span> <span class="s2">&quot;breed&quot;</span><span class="p">,</span> <span class="s2">&quot;strain&quot;</span><span class="p">,</span> <span class="s2">&quot;cultivar&quot;</span><span class="p">}</span>

        <span class="k">for</span> <span class="n">strain_type</span> <span class="ow">in</span> <span class="n">strain_types</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.strain&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">biosample_data</span><span class="p">[</span><span class="s2">&quot;characteristics&quot;</span><span class="p">][</span>
                    <span class="n">strain_type</span>
                <span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;text&quot;</span><span class="p">]</span>
                <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.strain_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">strain_type</span>
                <span class="k">break</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">continue</span>

        <span class="k">if</span> <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.strain&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Caucasian&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.strain&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;European&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.tol_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">biosample_data</span><span class="p">[</span><span class="s2">&quot;characteristics&quot;</span><span class="p">][</span><span class="s2">&quot;tolid&quot;</span><span class="p">][</span>
                <span class="mi">0</span>
            <span class="p">][</span><span class="s2">&quot;text&quot;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.tol_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="k">except</span> <span class="n">json</span><span class="o">.</span><span class="n">decoder</span><span class="o">.</span><span class="n">JSONDecodeError</span><span class="p">:</span>
        <span class="c1"># fallback to NCBI</span>
        <span class="n">biosample_dict</span> <span class="o">=</span> <span class="n">get_ncbi_metadata</span><span class="p">(</span>
            <span class="n">assembly_accession</span><span class="p">,</span> <span class="n">assembly_name</span><span class="p">,</span> <span class="n">scientific_name</span><span class="p">,</span> <span class="s2">&quot;biosample&quot;</span>
        <span class="p">)</span>
        <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.strain&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">biosample_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;organism.strain&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.strain_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">biosample_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="s2">&quot;organism.strain_type&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
        <span class="p">)</span>
        <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.tol_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="k">return</span> <span class="n">return_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.core_meta_data.get_ena_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_ena_metadata</span><span class="p">(</span><span class="n">accession</span><span class="p">,</span> <span class="n">truth_dict</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.core_meta_data.get_ena_metadata" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/core_meta_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_ena_metadata</span><span class="p">(</span><span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">truth_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
    <span class="n">assembly_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/browser/api/xml/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">assembly_xml</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">assembly_url</span><span class="p">)</span>
    <span class="n">assembly_dict</span> <span class="o">=</span> <span class="n">xmltodict</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">assembly_xml</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>

    <span class="n">attribs</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">assembly_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">,</span> <span class="p">{})</span>
        <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">,</span> <span class="p">{})</span>
        <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY_ATTRIBUTES&quot;</span><span class="p">,</span> <span class="p">{})</span>
        <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY_ATTRIBUTE&quot;</span><span class="p">,</span> <span class="p">[])</span>
    <span class="p">)</span>
    <span class="c1"># normalize to a list</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">attribs</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
        <span class="n">assembly_attribs</span> <span class="o">=</span> <span class="p">[</span><span class="n">attribs</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">assembly_attribs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">attribs</span><span class="p">)</span>

    <span class="n">return_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># assembly metadata</span>
    <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="p">(</span>
            <span class="n">assembly_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;NAME&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">assembly_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;NAME&quot;</span><span class="p">)</span>
        <span class="k">else</span> <span class="s2">&quot;&quot;</span>
    <span class="p">)</span>

    <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">assembly_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">,</span> <span class="p">{})</span>
        <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">,</span> <span class="p">{})</span>
        <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY_LEVEL&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">attrib_set</span> <span class="ow">in</span> <span class="n">assembly_attribs</span><span class="p">:</span>
        <span class="c1"># assembly date</span>
        <span class="k">if</span> <span class="n">attrib_set</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;TAG&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;ENA-LAST-UPDATED&quot;</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.date&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrib_set</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;VALUE&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="c1"># organism metadata: sample id</span>
    <span class="k">if</span> <span class="s2">&quot;organism.biosample_id&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">truth_dict</span><span class="p">:</span>
        <span class="n">biosample_id</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">assembly_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">,</span> <span class="p">{})</span>
            <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">,</span> <span class="p">{})</span>
            <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;SAMPLE_REF&quot;</span><span class="p">,</span> <span class="p">{})</span>
            <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;IDENTIFIERS&quot;</span><span class="p">,</span> <span class="p">{})</span>
            <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;PRIMARY_ID&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">biosample_id</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.biosample_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">biosample_id</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span>
                <span class="s2">&quot; | BIOSAMPLE_ID | organism.biosample_id could not be found in the ENA metadata, this is a required key!&quot;</span>
            <span class="p">)</span>

    <span class="c1"># taxonomy id (workaround for bad records)</span>
    <span class="k">if</span> <span class="n">accession</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;GCA_944452655.1&quot;</span><span class="p">,</span> <span class="s2">&quot;GCA_944452715.1&quot;</span><span class="p">):</span>
        <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.taxonomy_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;1539398&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">tax_id</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">assembly_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY_SET&quot;</span><span class="p">,</span> <span class="p">{})</span>
            <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ASSEMBLY&quot;</span><span class="p">,</span> <span class="p">{})</span>
            <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;TAXON&quot;</span><span class="p">,</span> <span class="p">{})</span>
            <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;TAXON_ID&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">tax_id</span><span class="p">:</span>
            <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.taxonomy_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tax_id</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot; | TAXONOMY_ID | organism.taxonomy_id could not be found in the ENA metadata&quot;</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="n">return_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.core_meta_data.get_ncbi_metadata" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_ncbi_metadata</span><span class="p">(</span><span class="n">accession</span><span class="p">,</span> <span class="n">assembly_name</span><span class="p">,</span> <span class="n">scientific_name</span><span class="p">,</span> <span class="n">search</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.core_meta_data.get_ncbi_metadata" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/core_meta_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_ncbi_metadata</span><span class="p">(</span>
    <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">assembly_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">scientific_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">search</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
    <span class="n">organism</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">assembly_name</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">ncbi_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://ftp.ncbi.nlm.nih.gov/genomes/all/</span><span class="si">{</span><span class="n">accession</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">accession</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">7</span><span class="p">]</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">accession</span><span class="p">[</span><span class="mi">7</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">accession</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">13</span><span class="p">]</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">organism</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">organism</span><span class="si">}</span><span class="s2">_assembly_report.txt&quot;</span>
    <span class="n">ncbi_return</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">ncbi_url</span><span class="p">)</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>

    <span class="n">return_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;assembly.ucsc_alias&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;organism.strain&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;organism.strain_type&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="k">if</span> <span class="n">search</span> <span class="o">==</span> <span class="s2">&quot;ucsc&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ncbi_return</span><span class="p">:</span>
            <span class="n">ucsc_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;# Synonyms:\s*([A-Za-z0-9]+)&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ucsc_match</span><span class="p">:</span>
                <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;assembly.ucsc_alias&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ucsc_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">search</span> <span class="o">==</span> <span class="s2">&quot;biosample&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ncbi_return</span><span class="p">:</span>
            <span class="n">strain_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
                <span class="sa">r</span><span class="s2">&quot;# Infraspecific name:\s*([A-Za-z]+)=([A-Za-z0-9 \-\./]+)&quot;</span><span class="p">,</span> <span class="n">line</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">strain_match</span><span class="p">:</span>
                <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.strain_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">strain_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">return_dict</span><span class="p">[</span><span class="s2">&quot;organism.strain&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">strain_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">return_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metadata.core_meta_data.mysql_fetch_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">)</span></code>

<a href="#ensembl.genes.metadata.core_meta_data.mysql_fetch_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Run a simple SELECT query and return fetched rows.</p>
<p>Returns an empty list on error.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metadata/core_meta_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mysql_fetch_data</span><span class="p">(</span>
    <span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">user</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="o">...</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Run a simple SELECT query and return fetched rows.</span>

<span class="sd">    Returns an empty list on error.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">conn</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">cursor</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">info</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="o">...</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="n">port</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">cursor</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
        <span class="n">info</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">())</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">exception</span><span class="p">(</span><span class="s2">&quot;MySQL error while executing query: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
    <span class="k">finally</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">cursor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">cursor</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">exception</span><span class="p">(</span><span class="s2">&quot;Error closing cursor&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">conn</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">exception</span><span class="p">(</span><span class="s2">&quot;Error closing connection&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">info</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="metrics">Metrics<a class="headerlink" href="#metrics" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.metrics.busco_lineage_selector" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.metrics.busco_lineage_selector</code>


<a href="#ensembl.genes.metrics.busco_lineage_selector" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.busco_lineage_selector.get_dataset_match" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_dataset_match</span><span class="p">(</span><span class="n">ncbi_url</span><span class="p">,</span> <span class="n">dataset</span><span class="p">)</span></code>

<a href="#ensembl.genes.metrics.busco_lineage_selector.get_dataset_match" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Get taxonomy tree from ncbi taxonomy datasets and find the closest match with the input list</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>ncbi_url</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Ncbi dataset url</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dataset</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of data to match</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.Any">Any</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>closest match in the dataset list</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="requests.HTTPError">HTTPError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If an HTTP error occurs during the API request.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="Exception">Exception</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If any other error occurs during the function's operation.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/busco_lineage_selector.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_dataset_match</span><span class="p">(</span><span class="n">ncbi_url</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">dataset</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get taxonomy tree from ncbi taxonomy datasets and find the closest match with the input list</span>


<span class="sd">    Args:</span>
<span class="sd">        ncbi_url (str): Ncbi dataset url</span>
<span class="sd">        dataset (dict): list of data to match</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: closest match in the dataset list</span>

<span class="sd">    Raises:</span>
<span class="sd">        requests.HTTPError: If an HTTP error occurs during the API request.</span>
<span class="sd">        Exception: If any other error occurs during the function&#39;s operation.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Fetch data from the URL</span>
        <span class="k">with</span> <span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="n">ncbi_url</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span> <span class="k">as</span> <span class="n">response</span><span class="p">:</span>
            <span class="c1"># Read the response and decode it</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span>
            <span class="c1"># Parse the JSON data</span>
            <span class="n">json_data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

            <span class="c1"># Extract classification names</span>
            <span class="n">parents</span> <span class="o">=</span> <span class="n">json_data</span><span class="p">[</span><span class="s2">&quot;reports&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;taxonomy&quot;</span><span class="p">][</span><span class="s2">&quot;parents&quot;</span><span class="p">]</span>
            <span class="c1"># Variable to store the matched result</span>
            <span class="n">matched_value</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="c1"># Match parents against the dictionary</span>
            <span class="k">for</span> <span class="n">parent_id</span> <span class="ow">in</span> <span class="nb">reversed</span><span class="p">(</span><span class="n">parents</span><span class="p">):</span>
                <span class="n">parent_id_str</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">parent_id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">parent_id_str</span> <span class="ow">in</span> <span class="n">dataset</span><span class="p">:</span>
                    <span class="n">matched_value</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">[</span><span class="n">parent_id_str</span><span class="p">]</span>
                    <span class="k">break</span>
    <span class="k">except</span> <span class="n">urllib</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">URLError</span> <span class="k">as</span> <span class="n">url_err</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;URL error occurred: </span><span class="si">{</span><span class="n">url_err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">json</span><span class="o">.</span><span class="n">JSONDecodeError</span> <span class="k">as</span> <span class="n">json_err</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error decoding JSON: </span><span class="si">{</span><span class="n">json_err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># print (matched_value)</span>
    <span class="k">return</span> <span class="n">matched_value</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.busco_lineage_selector.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.metrics.busco_lineage_selector.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Entry-point.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/busco_lineage_selector.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Entry-point.&quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parse_args</span><span class="p">()</span>

    <span class="n">ncbi_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">ncbi_url</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">/dataset_report&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">datasets</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>  <span class="c1"># pylint:disable=unspecified-encoding</span>
        <span class="c1"># datasets = [line[: max(line.find(&quot; &quot;), 0) or None] for line in file]</span>
        <span class="n">datasets</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
    <span class="n">clade_match</span> <span class="o">=</span> <span class="n">get_dataset_match</span><span class="p">(</span><span class="n">ncbi_url</span><span class="p">,</span> <span class="n">datasets</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">clade_match</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No match found&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">output</span> <span class="o">==</span> <span class="s2">&quot;stdout&quot;</span><span class="p">:</span>  <span class="c1"># pylint:disable=no-else-return</span>
        <span class="c1"># print(clade_match[0].strip(&quot;\n&quot;))</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">clade_match</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s2">&quot;w+&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output</span><span class="p">:</span>  <span class="c1"># pylint:disable=unspecified-encoding</span>
            <span class="k">if</span> <span class="n">clade_match</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">args</span><span class="o">.</span><span class="n">species</span><span class="p">:</span>
                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">clade_match</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">clade_match</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.busco_lineage_selector.parse_args" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">parse_args</span><span class="p">()</span></code>

<a href="#ensembl.genes.metrics.busco_lineage_selector.parse_args" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parse command line arguments.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/busco_lineage_selector.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">parse_args</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse command line arguments.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s2">&quot;Clade selector arguments&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-d&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--datasets&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to file containing list of datasets (one per line)&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-t&quot;</span><span class="p">,</span> <span class="s2">&quot;--taxon_id&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxon id &quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--output&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output file&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;stdout&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--ncbi_url&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;NCBI dataset url&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;https://api.ncbi.nlm.nih.gov/datasets/v2alpha/taxonomy/taxon/&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.metrics.busco_metakeys_patch" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.metrics.busco_metakeys_patch</code>


<a href="#ensembl.genes.metrics.busco_metakeys_patch" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.busco_metakeys_patch.execute_sql_patches" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">execute_sql_patches</span><span class="p">(</span><span class="n">db_name</span><span class="p">,</span> <span class="n">sql_statements</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">port</span><span class="p">)</span></code>

<a href="#ensembl.genes.metrics.busco_metakeys_patch.execute_sql_patches" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create SQL patch for database and execute it</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>db_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Database name</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sql_statements</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Union">Union</span>[<span title="str">str</span>, <span title="float">float</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of queryes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>host</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL server host.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>user</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL user.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>password</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL password.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>port</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MySQL port.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/busco_metakeys_patch.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">execute_sql_patches</span><span class="p">(</span>  <span class="c1"># pylint: disable=too-many-arguments, too-many-locals</span>
    <span class="n">db_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">sql_statements</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">password</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># pylint: disable=line-too-long</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create SQL patch for database and execute it</span>

<span class="sd">    Args:</span>
<span class="sd">        db_name (str): Database name</span>
<span class="sd">        sql_statements (Dict[str, Union[str, float]]): List of queryes</span>
<span class="sd">        host (str, optional): MySQL server host.</span>
<span class="sd">        user (str, optional): MySQL user.</span>
<span class="sd">        password (str, optional): MySQL password.</span>
<span class="sd">        port (int, optional): MySQL port.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">statements</span> <span class="o">=</span> <span class="n">sql_statements</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">connection</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># Initialize connection variable</span>
    <span class="c1"># Connect to the database and execute the SQL statements</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">connection</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="o">=</span><span class="n">password</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">db_name</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">port</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">statement</span> <span class="ow">in</span> <span class="n">statements</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">statement</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
                <span class="n">statement</span> <span class="o">=</span> <span class="n">statement</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">statement</span><span class="p">)</span>  <span class="c1"># Execute each SQL statement</span>
            <span class="n">connection</span><span class="o">.</span><span class="n">commit</span><span class="p">()</span>  <span class="c1"># Commit the changes</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">MySQLError</span> <span class="k">as</span> <span class="n">err_msg</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error while executing SQL: </span><span class="si">{</span><span class="n">err_msg</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">finally</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">connection</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">connection</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c1"># Close the database connection</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.busco_metakeys_patch.generate_sql_patches" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">generate_sql_patches</span><span class="p">(</span><span class="n">db_name</span><span class="p">,</span> <span class="n">json_data</span><span class="p">,</span> <span class="n">species_id</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="s1">&#39;meta&#39;</span><span class="p">)</span></code>

<a href="#ensembl.genes.metrics.busco_metakeys_patch.generate_sql_patches" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Creat Sql patch for database</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>db_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>db name</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>json_data</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Union">Union</span>[<span title="str">str</span>, <span title="float">float</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict of metakeys</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>species_id</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>species_id Defaults to 1.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>table_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Table name where to store data Defaults to "meta".</p>
              </div>
            </td>
            <td>
                  <code>&#39;meta&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of Mysql patches</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/busco_metakeys_patch.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_sql_patches</span><span class="p">(</span>
    <span class="n">db_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">json_data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]],</span>
    <span class="n">species_id</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">table_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;meta&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>  <span class="c1"># pylint: disable=line-too-long</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creat Sql patch for database</span>

<span class="sd">    Args:</span>
<span class="sd">        db_name (str): db name</span>
<span class="sd">        json_data (Dict[str, Union[str, float]]): Dict of metakeys</span>
<span class="sd">        species_id (int, optional): species_id Defaults to 1.</span>
<span class="sd">        table_name (str, optional): Table name where to store data Defaults to &quot;meta&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: list of Mysql patches</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sql_statements</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">sql_statements</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;USE </span><span class="si">{</span><span class="n">db_name</span><span class="si">}</span><span class="s2">;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># Replace with your actual DB name</span>

    <span class="c1"># Iterate through the JSON key-value pairs</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">json_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Skip if the value is None (or can handle it differently if needed)</span>
            <span class="k">continue</span>
        <span class="c1"># Convert value to string and escape single quotes if necessary</span>
        <span class="n">value_str</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;&#39;&#39;&quot;</span><span class="p">)</span>
        <span class="c1"># Create the SQL INSERT statement</span>
        <span class="n">sql_statements</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;INSERT IGNORE INTO </span><span class="si">{</span><span class="n">table_name</span><span class="si">}</span><span class="s2"> (species_id, meta_key, meta_value) VALUES (</span><span class="si">{</span><span class="n">species_id</span><span class="si">}</span><span class="s2">, &#39;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">&#39;, &#39;</span><span class="si">{</span><span class="n">value_str</span><span class="si">}</span><span class="s2">&#39;);&quot;</span>  <span class="c1"># pylint: disable=line-too-long</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sql_statements</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.busco_metakeys_patch.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.metrics.busco_metakeys_patch.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main function to parse a BUSCO result file and output the parsed data in JSON format.</p>
<p>It expects the file path to the BUSCO result as a command-line argument.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/busco_metakeys_patch.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main function to parse a BUSCO result file and output the parsed data in JSON format.</span>

<span class="sd">    It expects the file path to the BUSCO result as a command-line argument.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Set up argument parser</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Parse a BUSCO result file and generate JSON output.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-file&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to the BUSCO result file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-db&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Core db&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-input_dir&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path for directory containing the busco output files&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-output_dir&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path for output directory&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-run_query&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;true&quot;</span><span class="p">,</span> <span class="s2">&quot;false&quot;</span><span class="p">],</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Add busco metakeys to the db&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-host&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Server host&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-port&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Server port&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-user&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Db user with writable permission&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-password&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Server password&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-assembly_id&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Registry assembly id&quot;</span><span class="p">)</span>
    <span class="c1"># Parse arguments</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">file</span><span class="p">:</span>
        <span class="c1"># Process the single file and write to JSON and SQL</span>
        <span class="n">sql_patches</span> <span class="o">=</span> <span class="n">process_busco_file</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">file</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">assembly_id</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">db</span><span class="si">}</span><span class="s2">.sql&quot;</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span>
        <span class="p">)</span> <span class="k">as</span> <span class="n">f_in</span><span class="p">:</span>  <span class="c1"># pylint: disable =unspecified-encoding</span>
            <span class="n">f_in</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sql_patches</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">input_dir</span><span class="p">:</span>
        <span class="c1"># Process all files that end with &#39;busco_short_summary&#39;</span>
        <span class="n">busco_files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">input_dir</span><span class="p">)</span><span class="o">.</span><span class="n">rglob</span><span class="p">(</span><span class="s2">&quot;*busco_short_summary.txt&quot;</span><span class="p">))</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span>
            <span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">db</span><span class="si">}</span><span class="s2">.sql&quot;</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span>
        <span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>  <span class="c1"># pylint: disable =unspecified-encoding</span>
            <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">busco_files</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing file: </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">sql_patches</span> <span class="o">=</span> <span class="n">process_busco_file</span><span class="p">(</span>
                    <span class="n">file</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">assembly_id</span>
                <span class="p">)</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sql_patches</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">run_query</span> <span class="o">==</span> <span class="s2">&quot;true&quot;</span><span class="p">:</span>
        <span class="n">execute_sql_patches</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="n">sql_patches</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">host</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">user</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">password</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">port</span><span class="p">)</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.busco_metakeys_patch.parse_busco_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">parse_busco_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span></code>

<a href="#ensembl.genes.metrics.busco_metakeys_patch.parse_busco_file" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parses a BUSCO result file and extracts relevant data into a dictionary.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>file_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the BUSCO result file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>db</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Core db name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Union">Union</span>[<span title="str">str</span>, <span title="int">int</span>, <span title="float">float</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict[str, Union[str, int]]: A dictionary containing parsed BUSCO data,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Union">Union</span>[<span title="str">str</span>, <span title="int">int</span>, <span title="float">float</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>including the dataset, completeness values, and mode (proteins or genome).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/busco_metakeys_patch.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">parse_busco_file</span><span class="p">(</span>
    <span class="n">file_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span>
    <span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span>
<span class="p">]:</span>  <span class="c1"># pylint: disable =too-many-locals, too-many-statements</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parses a BUSCO result file and extracts relevant data into a dictionary.</span>

<span class="sd">    Args:</span>
<span class="sd">        file_path (str): The path to the BUSCO result file.</span>
<span class="sd">        db(str): Core db name.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Union[str, int]]: A dictionary containing parsed BUSCO data,</span>
<span class="sd">        including the dataset, completeness values, and mode (proteins or genome).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Declare the dictionary to accept str as keys and str or float as values</span>
    <span class="n">data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># data[&quot;core_db&quot;] = db</span>
    <span class="c1"># Open and read the file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>  <span class="c1"># pylint: disable =unspecified-encoding</span>
        <span class="n">content</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

    <span class="c1"># Define regular expressions to match the relevant numbers</span>
    <span class="n">version_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;BUSCO version is: ((\d+\.\d+.\d+))&quot;</span><span class="p">,</span> <span class="n">content</span>
    <span class="p">)</span>
    <span class="n">dataset_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;The lineage dataset is: ([\w_]+)&quot;</span><span class="p">,</span> <span class="n">content</span>
    <span class="p">)</span>
    <span class="n">mode_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;BUSCO was run in mode: ([\w_]+)&quot;</span><span class="p">,</span> <span class="n">content</span>
    <span class="p">)</span>
    <span class="n">completeness_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;(\d+)\s+Complete BUSCOs \(C\)&quot;</span><span class="p">,</span> <span class="n">content</span>
    <span class="p">)</span>
    <span class="n">single_copy_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;(\d+)\s+Complete and single-copy BUSCOs \(S\)&quot;</span><span class="p">,</span> <span class="n">content</span>
    <span class="p">)</span>
    <span class="n">duplicates_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;(\d+)\s+Complete and duplicated BUSCOs \(D\)&quot;</span><span class="p">,</span> <span class="n">content</span>
    <span class="p">)</span>
    <span class="n">fragmented_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;(\d+)\s+Fragmented BUSCOs \(F\)&quot;</span><span class="p">,</span> <span class="n">content</span>
    <span class="p">)</span>
    <span class="n">missing_pattern</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">Match</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;(\d+)\s+Missing BUSCOs \(M\)&quot;</span><span class="p">,</span> <span class="n">content</span>
    <span class="p">)</span>

    <span class="c1"># Initialize mode_match as None or str</span>
    <span class="n">mode_match</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># If match is not None, extract the group and assign it to mode_match</span>
    <span class="k">if</span> <span class="n">mode_pattern</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mode_match</span> <span class="o">=</span> <span class="n">mode_pattern</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mode_match</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;genome&quot;</span><span class="p">,</span> <span class="s2">&quot;euk_genome_met&quot;</span><span class="p">,</span> <span class="s2">&quot;euk_genome_min&quot;</span><span class="p">):</span>
        <span class="n">busco_mode</span> <span class="o">=</span> <span class="s2">&quot;genome&quot;</span>
    <span class="k">elif</span> <span class="n">mode_match</span> <span class="o">==</span> <span class="s2">&quot;proteins&quot;</span><span class="p">:</span>
        <span class="n">busco_mode</span> <span class="o">=</span> <span class="s2">&quot;protein&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mode_match</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">version</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">version_pattern</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">version_pattern</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">dataset</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">dataset_pattern</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">dataset_pattern</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">completeness</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">completeness_pattern</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">completeness_pattern</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">single_copy</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">single_copy_pattern</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">single_copy_pattern</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">duplicated</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">duplicates_pattern</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">duplicates_pattern</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">fragmented</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fragmented_pattern</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">fragmented_pattern</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">missing</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">missing_pattern</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="k">if</span> <span class="n">missing_pattern</span> <span class="k">else</span> <span class="kc">None</span>

    <span class="c1"># Extract the BUSCO summary line with completeness values</span>
    <span class="k">if</span> <span class="n">mode_match</span> <span class="o">==</span> <span class="s2">&quot;euk_genome_min&quot;</span><span class="p">:</span>
        <span class="n">score_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
            <span class="sa">r</span><span class="s2">&quot;C:(\d+\.\d+)%\[S:(\d+\.\d+)%.*,D:(\d+\.\d+)%\],F:(\d+\.\d+)%.*,M:(\d+\.\d+)%,n:(\d+),E:(\d+\.\d+)%&quot;</span><span class="p">,</span>  <span class="c1"># pylint: disable=line-too-long</span>
            <span class="n">content</span><span class="p">,</span>  <span class="c1"># pylint: disable=line-too-long</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">score_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
            <span class="sa">r</span><span class="s2">&quot;C:(\d+\.\d+)%\[S:(\d+\.\d+)%.*,D:(\d+\.\d+)%\],F:(\d+\.\d+)%.*,M:(\d+\.\d+)%,n:(\d+)&quot;</span><span class="p">,</span>
            <span class="n">content</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">score_match</span><span class="p">:</span>
        <span class="n">score</span> <span class="o">=</span> <span class="n">score_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">total_buscos</span> <span class="o">=</span> <span class="n">score_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mode_match</span> <span class="o">==</span> <span class="s2">&quot;euk_genome_min&quot;</span><span class="p">:</span>
            <span class="n">erroneus</span> <span class="o">=</span> <span class="n">score_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mode_match</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;genome&quot;</span><span class="p">,</span> <span class="s2">&quot;euk_genome_met&quot;</span><span class="p">,</span> <span class="s2">&quot;euk_genome_min&quot;</span><span class="p">):</span>
            <span class="c1"># Extract the BUSCO version</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">version</span><span class="p">)</span>
            <span class="c1"># Extract the BUSCO dataset</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_dataset&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
            <span class="c1"># Store the BUSCO completeness summary with erroneous</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_mode&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">busco_mode</span>
            <span class="c1"># Store the BUSCO values into individual fields</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_completeness&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">completeness</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_single_copy&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">single_copy</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_duplicated&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">duplicated</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_fragmented&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">fragmented</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_missing&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">missing</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_total&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">total_buscos</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">mode_match</span> <span class="o">==</span> <span class="s2">&quot;euk_genome_min&quot;</span><span class="p">:</span>
                <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco_erroneus&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">erroneus</span><span class="p">)</span>

            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;assembly.busco&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>  <span class="c1"># pylint: disable=line-too-long</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Extract the BUSCO version</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco_version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">version</span><span class="p">)</span>
            <span class="c1"># Extract the BUSCO dataset</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco_dataset&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
            <span class="c1"># Store the BUSCO completeness summary</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco_mode&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">busco_mode</span>
            <span class="c1"># Store the BUSCO values into individual fields</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco_completeness&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">completeness</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco_single_copy&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">single_copy</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco_duplicated&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">duplicated</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco_fragmented&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">fragmented</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco_missing&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">missing</span><span class="p">)</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;genebuild.busco_total&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">total_buscos</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.busco_metakeys_patch.process_busco_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">process_busco_file</span><span class="p">(</span><span class="n">busco_file</span><span class="p">,</span> <span class="n">dbname</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">assembly_id</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span></code>

<a href="#ensembl.genes.metrics.busco_metakeys_patch.process_busco_file" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Parses the BUSCO file, generates a JSON, writes it to an output file,
and generates SQL patches.
Args:
    busco_file (str): Busco file to process
    dbname (str) : db name
    output_dir (str): Output directory path
    assembly_id (str, optional): Metadata used in downstream analyses.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of Mysql patches</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/busco_metakeys_patch.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">process_busco_file</span><span class="p">(</span>
    <span class="n">busco_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">dbname</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">assembly_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parses the BUSCO file, generates a JSON, writes it to an output file,</span>
<span class="sd">    and generates SQL patches.</span>
<span class="sd">    Args:</span>
<span class="sd">        busco_file (str): Busco file to process</span>
<span class="sd">        dbname (str) : db name</span>
<span class="sd">        output_dir (str): Output directory path</span>
<span class="sd">        assembly_id (str, optional): Metadata used in downstream analyses.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: list of Mysql patches</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Parse the BUSCO file and generate the JSON</span>
    <span class="n">busco_data</span> <span class="o">=</span> <span class="n">parse_busco_file</span><span class="p">(</span><span class="n">busco_file</span><span class="p">)</span>

    <span class="c1"># Determine the file name based on the mode (protein or genome)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">busco_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">key</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.busco_mode&quot;</span><span class="p">):</span>
            <span class="n">busco_mode</span> <span class="o">=</span> <span class="n">value</span>
            <span class="k">break</span>  <span class="c1"># Exit the loop once we find the first match</span>

    <span class="n">output_file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">dbname</span><span class="si">}</span><span class="s2">_busco_</span><span class="si">{</span><span class="n">busco_mode</span><span class="si">}</span><span class="s2">_metakey.json&quot;</span>
    <span class="n">busco_data_json</span> <span class="o">=</span> <span class="n">busco_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">busco_data_json</span><span class="p">[</span><span class="s2">&quot;core_db&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dbname</span>
    <span class="n">busco_data_json</span><span class="p">[</span><span class="s2">&quot;assembly_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">assembly_id</span>
    <span class="c1"># Convert the dictionary to a JSON object</span>
    <span class="n">busco_json</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">busco_data_json</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

    <span class="c1"># Write the JSON output to the dynamically named file</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span> <span class="o">/</span> <span class="n">output_file_name</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>  <span class="c1"># pylint: disable =unspecified-encoding</span>
        <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">busco_json</span><span class="p">)</span>

    <span class="c1"># Output the JSON</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">busco_json</span><span class="p">)</span>

    <span class="c1"># Generate SQL patches from the JSON</span>
    <span class="n">sql_patches</span> <span class="o">=</span> <span class="n">generate_sql_patches</span><span class="p">(</span><span class="n">dbname</span><span class="p">,</span> <span class="n">busco_data</span><span class="p">)</span>

    <span class="c1"># Return SQL patches to write them to an SQL file later</span>
    <span class="k">return</span> <span class="n">sql_patches</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.metrics.check_busco_score" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.metrics.check_busco_score</code>


<a href="#ensembl.genes.metrics.check_busco_score" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.check_busco_score.evaluate_busco" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">evaluate_busco</span><span class="p">(</span><span class="n">genome_json_path</span><span class="p">,</span> <span class="n">protein_json_path</span><span class="p">,</span> <span class="n">min_range_protein_score</span><span class="p">,</span> <span class="n">max_range_protein_score</span><span class="p">,</span> <span class="n">diff_prot_gen_mode</span><span class="p">)</span></code>

<a href="#ensembl.genes.metrics.check_busco_score.evaluate_busco" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Evaluate BUSCO scores from genome and protein JSON files.
This function loads the BUSCO JSON files for genome and protein,
extracts the completeness scores, and compares them to determine
if the protein BUSCO score is significantly higher than the genome BUSCO score.
The criteria for significance are:
- If the protein BUSCO score is &gt;= 70%, it is considered significant.
- If the protein BUSCO score is between 50% and 70%, it is considered significant
  if the difference between the protein and genome BUSCO scores is &gt;= 10%.
If the genome or protein BUSCO JSON files cannot be loaded, an error message is printed
and the function returns False.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>genome_json_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>path to the genome BUSCO JSON file</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>protein_json_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>path to the protein BUSCO JSON file</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_range_protein_score</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Lowest threshold to analyse busco score in protein mode</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>max_range_protein_score</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Highest threshold to analyse busco score in protein mode</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>diff_prot_gen_mode</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Max difference between Busco in protein and in genome mode</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="KeyError">KeyError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the expected '.busco' key is not found in either JSON file.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>exit 0 if the protein BUSCO score soddisfy criteria, otherwise exit 42</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/check_busco_score.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">evaluate_busco</span><span class="p">(</span>
    <span class="n">genome_json_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">protein_json_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">min_range_protein_score</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">max_range_protein_score</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">diff_prot_gen_mode</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Evaluate BUSCO scores from genome and protein JSON files.</span>
<span class="sd">    This function loads the BUSCO JSON files for genome and protein,</span>
<span class="sd">    extracts the completeness scores, and compares them to determine</span>
<span class="sd">    if the protein BUSCO score is significantly higher than the genome BUSCO score.</span>
<span class="sd">    The criteria for significance are:</span>
<span class="sd">    - If the protein BUSCO score is &gt;= 70%, it is considered significant.</span>
<span class="sd">    - If the protein BUSCO score is between 50% and 70%, it is considered significant</span>
<span class="sd">      if the difference between the protein and genome BUSCO scores is &gt;= 10%.</span>
<span class="sd">    If the genome or protein BUSCO JSON files cannot be loaded, an error message is printed</span>
<span class="sd">    and the function returns False.</span>

<span class="sd">    Args:</span>
<span class="sd">        genome_json_path (str): _path to the genome BUSCO JSON file_</span>
<span class="sd">        protein_json_path (str): _path to the protein BUSCO JSON file_</span>
<span class="sd">        min_range_protein_score (int): Lowest threshold to analyse busco score in protein mode</span>
<span class="sd">        max_range_protein_score (int): Highest threshold to analyse busco score in protein mode</span>
<span class="sd">        diff_prot_gen_mode (int): Max difference between Busco in protein and in genome mode</span>

<span class="sd">    Raises:</span>
<span class="sd">        KeyError: If the expected &#39;.busco&#39; key is not found in either JSON file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        exit 0 if the protein BUSCO score soddisfy criteria, otherwise exit 42</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">genome_json_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">g_file</span><span class="p">:</span>  <span class="c1"># pylint:disable=unspecified-encoding</span>
            <span class="n">genome_data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">g_file</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err_msg</span><span class="p">:</span>  <span class="c1"># pylint: disable=broad-except</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to load genome BUSCO JSON: </span><span class="si">{</span><span class="n">err_msg</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">protein_json_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">p_file</span><span class="p">:</span>  <span class="c1"># pylint:disable=unspecified-encoding</span>
            <span class="n">protein_data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">p_file</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err_msg</span><span class="p">:</span>  <span class="c1"># pylint: disable=broad-except</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to load protein BUSCO JSON: </span><span class="si">{</span><span class="n">err_msg</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>

    <span class="n">genome_busco_key</span> <span class="o">=</span> <span class="nb">next</span><span class="p">((</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">genome_data</span> <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.busco&quot;</span><span class="p">)),</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">protein_busco_key</span> <span class="o">=</span> <span class="nb">next</span><span class="p">((</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">protein_data</span> <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.busco&quot;</span><span class="p">)),</span> <span class="kc">None</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">genome_busco_key</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">protein_busco_key</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="s2">&quot;Missing &#39;.busco&#39; key in one of the JSON files&quot;</span><span class="p">)</span>

    <span class="n">genome_busco_score</span> <span class="o">=</span> <span class="n">extract_completeness</span><span class="p">(</span><span class="n">genome_data</span><span class="p">[</span><span class="n">genome_busco_key</span><span class="p">])</span>
    <span class="n">protein_busco_score</span> <span class="o">=</span> <span class="n">extract_completeness</span><span class="p">(</span><span class="n">protein_data</span><span class="p">[</span><span class="n">protein_busco_key</span><span class="p">])</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Genome BUSCO completeness: </span><span class="si">{</span><span class="n">genome_busco_score</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Protein BUSCO completeness: </span><span class="si">{</span><span class="n">protein_busco_score</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">protein_busco_score</span> <span class="o">&gt;=</span> <span class="n">max_range_protein_score</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Protein BUSCO completeness </span><span class="si">{</span><span class="n">protein_busco_score</span><span class="si">}</span><span class="s2">% is above threshold (70%)&quot;</span>
        <span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">min_range_protein_score</span> <span class="o">&lt;</span> <span class="n">protein_busco_score</span> <span class="o">&lt;</span> <span class="n">max_range_protein_score</span><span class="p">:</span>
        <span class="n">difference</span> <span class="o">=</span> <span class="n">protein_busco_score</span> <span class="o">-</span> <span class="n">genome_busco_score</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Difference (protein - genome): </span><span class="si">{</span><span class="n">difference</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">difference</span> <span class="o">&lt;=</span> <span class="n">diff_prot_gen_mode</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Protein BUSCO completeness </span><span class="si">{</span><span class="n">protein_busco_score</span><span class="si">}</span><span class="s2">% is too low&quot;</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.check_busco_score.extract_completeness" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">extract_completeness</span><span class="p">(</span><span class="n">busco_str</span><span class="p">)</span></code>

<a href="#ensembl.genes.metrics.check_busco_score.extract_completeness" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extract the BUSCO completeness value from a string like:
"C:99.2%[S:98.4%,D:0.8%],F:0.6%,M:0.2%,n:3640"
and return it as a float.
If the format is invalid, raise a ValueError.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>busco_str</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The BUSCO string to parse.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Return: The completeness percentage as a float.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/check_busco_score.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">extract_completeness</span><span class="p">(</span><span class="n">busco_str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract the BUSCO completeness value from a string like:</span>
<span class="sd">    &quot;C:99.2%[S:98.4%,D:0.8%],F:0.6%,M:0.2%,n:3640&quot;</span>
<span class="sd">    and return it as a float.</span>
<span class="sd">    If the format is invalid, raise a ValueError.</span>

<span class="sd">    Args:</span>
<span class="sd">        busco_str (str): The BUSCO string to parse.</span>
<span class="sd">    Return: The completeness percentage as a float.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;C:([\d.]+)%&quot;</span><span class="p">,</span> <span class="n">busco_str</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid BUSCO format: </span><span class="si">{</span><span class="n">busco_str</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.metrics.check_busco_score.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.metrics.check_busco_score.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main function to parse command line arguments and evaluate BUSCO scores.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/metrics/check_busco_score.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main function to parse command line arguments and evaluate BUSCO scores.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Evaluate BUSCO scores from genome and protein JSON files.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--genome&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to genome BUSCO JSON file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--protein&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to protein BUSCO JSON file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--min_range_protein_score&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Lowest threshold to analyse busco score in protein mode&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--max_range_protein_score&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">70</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Highest threshold to analyse busco score in protein mode&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--diff_prot_gen_mode&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Max difference between Busco in protein and in genome mode&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="k">try</span><span class="p">:</span>  <span class="c1"># pylint: disable=broad-except</span>
        <span class="n">evaluate_busco</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">protein</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">min_range_protein_score</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">max_range_protein_score</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">diff_prot_gen_mode</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err_msg</span><span class="p">:</span>  <span class="c1"># pylint: disable=broad-except</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ERROR: </span><span class="si">{</span><span class="n">err_msg</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="projects">Projects<a class="headerlink" href="#projects" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.projects.write_yaml" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.projects.write_yaml</code>


<a href="#ensembl.genes.projects.write_yaml" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">









<div class="doc doc-object doc-class">



<h3 id="ensembl.genes.projects.write_yaml.EnsemblFTP" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>            <code>EnsemblFTP</code>


<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">



        <p>Robust FTP client for Ensembl/EBI with reconnect + tolerant shutdown.</p>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">EnsemblFTP</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Robust FTP client for Ensembl/EBI with reconnect + tolerant shutdown.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">timeout</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">max_retries</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">retry_sleep</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">3.0</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">timeout</span> <span class="o">=</span> <span class="n">timeout</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_retries</span> <span class="o">=</span> <span class="n">max_retries</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">retry_sleep</span> <span class="o">=</span> <span class="n">retry_sleep</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_connect_ensembl</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_connect_ebi</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp_path</span> <span class="o">=</span> <span class="s2">&quot;https://ftp.ensembl.org/&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp_path</span> <span class="o">=</span> <span class="s2">&quot;https://ftp.ebi.ac.uk/&quot;</span>

    <span class="c1"># ---- connections ----</span>
    <span class="k">def</span> <span class="nf">_connect_ensembl</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span> <span class="o">=</span> <span class="n">FTP</span><span class="p">(</span><span class="s2">&quot;ftp.ensembl.org&quot;</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">timeout</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span><span class="o">.</span><span class="n">set_pasv</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span><span class="o">.</span><span class="n">login</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">raise</span>

    <span class="k">def</span> <span class="nf">_connect_ebi</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span> <span class="o">=</span> <span class="n">FTP</span><span class="p">(</span><span class="s2">&quot;ftp.ebi.ac.uk&quot;</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">timeout</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span><span class="o">.</span><span class="n">set_pasv</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span><span class="o">.</span><span class="n">login</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">raise</span>

    <span class="k">def</span> <span class="nf">_retry</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fn</span><span class="p">,</span> <span class="n">which</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retry FTP operation; on failure, reconnect that endpoint and try again.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">last_exc</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_retries</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">fn</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
            <span class="k">except</span> <span class="p">(</span>
                <span class="ne">ConnectionResetError</span><span class="p">,</span>
                <span class="ne">EOFError</span><span class="p">,</span>
                <span class="ne">OSError</span><span class="p">,</span>
                <span class="n">error_temp</span><span class="p">,</span>
                <span class="n">socket</span><span class="o">.</span><span class="n">timeout</span><span class="p">,</span>
            <span class="p">)</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">last_exc</span> <span class="o">=</span> <span class="n">e</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">which</span> <span class="o">==</span> <span class="s2">&quot;ensembl&quot;</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_connect_ensembl</span><span class="p">()</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_connect_ebi</span><span class="p">()</span>
                <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">retry_sleep</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">last_exc</span>

    <span class="c1"># ---- utils ----</span>
    <span class="k">def</span> <span class="nf">return_to_root</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ftp_connection</span><span class="p">:</span> <span class="n">FTP</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">which</span> <span class="o">=</span> <span class="s2">&quot;ensembl&quot;</span> <span class="k">if</span> <span class="n">ftp_connection</span> <span class="ow">is</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span> <span class="k">else</span> <span class="s2">&quot;ebi&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp_connection</span><span class="o">.</span><span class="n">cwd</span><span class="p">,</span> <span class="n">which</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">)</span>

    <span class="c1"># ---- lookups ----</span>
    <span class="k">def</span> <span class="nf">check_for_file</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">species_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">prod_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">source</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">file_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s2">&quot;repeatmodeler&quot;</span><span class="p">:</span>
            <span class="n">ftp_connection</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span>
            <span class="n">which</span> <span class="o">=</span> <span class="s2">&quot;ebi&quot;</span>
            <span class="n">ftp_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp_path</span>
            <span class="n">path</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;pub/databases/ensembl/repeats/unfiltered_repeatmodeler/species/&quot;</span>
                <span class="o">+</span> <span class="n">species_name</span>
                <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
            <span class="p">)</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="n">accession</span> <span class="o">+</span> <span class="s2">&quot;.repeatmodeler.fa&quot;</span>
        <span class="k">elif</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s2">&quot;busco&quot;</span><span class="p">:</span>
            <span class="n">ftp_connection</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span>
            <span class="n">which</span> <span class="o">=</span> <span class="s2">&quot;ensembl&quot;</span>
            <span class="n">ftp_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp_path</span>
            <span class="n">path</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;pub/rapid-release/species/&quot;</span>
                <span class="o">+</span> <span class="n">species_name</span>
                <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
                <span class="o">+</span> <span class="n">accession</span>
                <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
                <span class="o">+</span> <span class="n">source</span>
                <span class="o">+</span> <span class="s2">&quot;/statistics/&quot;</span>
            <span class="p">)</span>
            <span class="n">file_name_protein</span> <span class="o">=</span> <span class="n">prod_name</span> <span class="o">+</span> <span class="s2">&quot;_protein_busco_short_summary.txt&quot;</span>
            <span class="n">file_name_alternative</span> <span class="o">=</span> <span class="n">prod_name</span> <span class="o">+</span> <span class="s2">&quot;_busco_short_summary.txt&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">return_to_root</span><span class="p">(</span><span class="n">ftp_connection</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp_connection</span><span class="o">.</span><span class="n">cwd</span><span class="p">,</span> <span class="n">which</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
            <span class="n">files_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp_connection</span><span class="o">.</span><span class="n">nlst</span><span class="p">,</span> <span class="n">which</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s2">&quot;busco&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">file_name_protein</span> <span class="ow">in</span> <span class="n">files_list</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">ftp_path</span> <span class="o">+</span> <span class="n">path</span> <span class="o">+</span> <span class="n">file_name_protein</span>
                <span class="k">if</span> <span class="n">file_name_alternative</span> <span class="ow">in</span> <span class="n">files_list</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">ftp_path</span> <span class="o">+</span> <span class="n">path</span> <span class="o">+</span> <span class="n">file_name_alternative</span>
                <span class="k">return</span> <span class="s2">&quot;&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">ftp_path</span> <span class="o">+</span> <span class="n">path</span> <span class="o">+</span> <span class="n">file_name</span> <span class="k">if</span> <span class="n">file_name</span> <span class="ow">in</span> <span class="n">files_list</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
        <span class="k">except</span> <span class="n">error_perm</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;550&quot;</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">):</span>
                <span class="k">return</span> <span class="s2">&quot;&quot;</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;FTP permission error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
            <span class="k">return</span> <span class="s2">&quot;&quot;</span>
        <span class="k">except</span> <span class="n">error_temp</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;FTP temporary error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
            <span class="k">return</span> <span class="s2">&quot;&quot;</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error while checking FTP file: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
            <span class="k">return</span> <span class="s2">&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">check_pre_release_file</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">species_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">extension</span><span class="p">:</span> <span class="nb">str</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">ftp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span>
        <span class="n">which</span> <span class="o">=</span> <span class="s2">&quot;ebi&quot;</span>
        <span class="n">base</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp_path</span>
        <span class="n">path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;pub/databases/ensembl/pre-release/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">/&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">return_to_root</span><span class="p">(</span><span class="n">ftp</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp</span><span class="o">.</span><span class="n">cwd</span><span class="p">,</span> <span class="n">which</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">fname</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp</span><span class="o">.</span><span class="n">nlst</span><span class="p">,</span> <span class="n">which</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">fname</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">extension</span><span class="p">):</span>
                    <span class="k">return</span> <span class="n">base</span> <span class="o">+</span> <span class="n">path</span> <span class="o">+</span> <span class="n">fname</span>
        <span class="k">except</span> <span class="n">error_perm</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;&quot;</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error while checking pre-release file: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">close_connections</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Best-effort shutdown: if QUIT fails because the server dropped us,</span>
<span class="sd">        fall back to close() and never raise.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">conn</span> <span class="ow">in</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">conn</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">conn</span><span class="o">.</span><span class="n">quit</span><span class="p">()</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
                <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                    <span class="k">pass</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.ebi_ftp" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">ebi_ftp</span> <span class="o">=</span> <span class="kc">None</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.ebi_ftp" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.ebi_ftp_path" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">ebi_ftp_path</span> <span class="o">=</span> <span class="s1">&#39;https://ftp.ebi.ac.uk/&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.ebi_ftp_path" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.ensembl_ftp" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">ensembl_ftp</span> <span class="o">=</span> <span class="kc">None</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.ensembl_ftp" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.ensembl_ftp_path" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">ensembl_ftp_path</span> <span class="o">=</span> <span class="s1">&#39;https://ftp.ensembl.org/&#39;</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.ensembl_ftp_path" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.max_retries" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">max_retries</span> <span class="o">=</span> <span class="n">max_retries</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.max_retries" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.retry_sleep" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">retry_sleep</span> <span class="o">=</span> <span class="n">retry_sleep</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.retry_sleep" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.timeout" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">timeout</span> <span class="o">=</span> <span class="n">timeout</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-instance-attribute"><code>instance-attribute</code></small>
  </span>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.timeout" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.check_for_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">check_for_file</span><span class="p">(</span><span class="n">species_name</span><span class="p">,</span> <span class="n">prod_name</span><span class="p">,</span> <span class="n">accession</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="n">file_type</span><span class="p">)</span></code>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.check_for_file" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_for_file</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">species_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">prod_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">source</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">file_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s2">&quot;repeatmodeler&quot;</span><span class="p">:</span>
        <span class="n">ftp_connection</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span>
        <span class="n">which</span> <span class="o">=</span> <span class="s2">&quot;ebi&quot;</span>
        <span class="n">ftp_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp_path</span>
        <span class="n">path</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;pub/databases/ensembl/repeats/unfiltered_repeatmodeler/species/&quot;</span>
            <span class="o">+</span> <span class="n">species_name</span>
            <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
        <span class="p">)</span>
        <span class="n">file_name</span> <span class="o">=</span> <span class="n">accession</span> <span class="o">+</span> <span class="s2">&quot;.repeatmodeler.fa&quot;</span>
    <span class="k">elif</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s2">&quot;busco&quot;</span><span class="p">:</span>
        <span class="n">ftp_connection</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span>
        <span class="n">which</span> <span class="o">=</span> <span class="s2">&quot;ensembl&quot;</span>
        <span class="n">ftp_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp_path</span>
        <span class="n">path</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;pub/rapid-release/species/&quot;</span>
            <span class="o">+</span> <span class="n">species_name</span>
            <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
            <span class="o">+</span> <span class="n">accession</span>
            <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
            <span class="o">+</span> <span class="n">source</span>
            <span class="o">+</span> <span class="s2">&quot;/statistics/&quot;</span>
        <span class="p">)</span>
        <span class="n">file_name_protein</span> <span class="o">=</span> <span class="n">prod_name</span> <span class="o">+</span> <span class="s2">&quot;_protein_busco_short_summary.txt&quot;</span>
        <span class="n">file_name_alternative</span> <span class="o">=</span> <span class="n">prod_name</span> <span class="o">+</span> <span class="s2">&quot;_busco_short_summary.txt&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">return_to_root</span><span class="p">(</span><span class="n">ftp_connection</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp_connection</span><span class="o">.</span><span class="n">cwd</span><span class="p">,</span> <span class="n">which</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
        <span class="n">files_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp_connection</span><span class="o">.</span><span class="n">nlst</span><span class="p">,</span> <span class="n">which</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s2">&quot;busco&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file_name_protein</span> <span class="ow">in</span> <span class="n">files_list</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">ftp_path</span> <span class="o">+</span> <span class="n">path</span> <span class="o">+</span> <span class="n">file_name_protein</span>
            <span class="k">if</span> <span class="n">file_name_alternative</span> <span class="ow">in</span> <span class="n">files_list</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">ftp_path</span> <span class="o">+</span> <span class="n">path</span> <span class="o">+</span> <span class="n">file_name_alternative</span>
            <span class="k">return</span> <span class="s2">&quot;&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ftp_path</span> <span class="o">+</span> <span class="n">path</span> <span class="o">+</span> <span class="n">file_name</span> <span class="k">if</span> <span class="n">file_name</span> <span class="ow">in</span> <span class="n">files_list</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
    <span class="k">except</span> <span class="n">error_perm</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;550&quot;</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">):</span>
            <span class="k">return</span> <span class="s2">&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;FTP permission error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>
    <span class="k">except</span> <span class="n">error_temp</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;FTP temporary error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error while checking FTP file: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.check_pre_release_file" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">check_pre_release_file</span><span class="p">(</span><span class="n">species_name</span><span class="p">,</span> <span class="n">accession</span><span class="p">,</span> <span class="n">extension</span><span class="p">)</span></code>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.check_pre_release_file" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_pre_release_file</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">species_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">extension</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
    <span class="n">ftp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span>
    <span class="n">which</span> <span class="o">=</span> <span class="s2">&quot;ebi&quot;</span>
    <span class="n">base</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp_path</span>
    <span class="n">path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;pub/databases/ensembl/pre-release/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">/&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">return_to_root</span><span class="p">(</span><span class="n">ftp</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp</span><span class="o">.</span><span class="n">cwd</span><span class="p">,</span> <span class="n">which</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">fname</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp</span><span class="o">.</span><span class="n">nlst</span><span class="p">,</span> <span class="n">which</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">fname</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">extension</span><span class="p">):</span>
                <span class="k">return</span> <span class="n">base</span> <span class="o">+</span> <span class="n">path</span> <span class="o">+</span> <span class="n">fname</span>
    <span class="k">except</span> <span class="n">error_perm</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error while checking pre-release file: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.close_connections" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">close_connections</span><span class="p">()</span></code>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.close_connections" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Best-effort shutdown: if QUIT fails because the server dropped us,
fall back to close() and never raise.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">close_connections</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Best-effort shutdown: if QUIT fails because the server dropped us,</span>
<span class="sd">    fall back to close() and never raise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">conn</span> <span class="ow">in</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ebi_ftp</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">conn</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">conn</span><span class="o">.</span><span class="n">quit</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">pass</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="ensembl.genes.projects.write_yaml.EnsemblFTP.return_to_root" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-method"></code>            <code class="highlight language-python"><span class="n">return_to_root</span><span class="p">(</span><span class="n">ftp_connection</span><span class="p">)</span></code>

<a href="#ensembl.genes.projects.write_yaml.EnsemblFTP.return_to_root" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">return_to_root</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ftp_connection</span><span class="p">:</span> <span class="n">FTP</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">which</span> <span class="o">=</span> <span class="s2">&quot;ensembl&quot;</span> <span class="k">if</span> <span class="n">ftp_connection</span> <span class="ow">is</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_ftp</span> <span class="k">else</span> <span class="s2">&quot;ebi&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_retry</span><span class="p">(</span><span class="n">ftp_connection</span><span class="o">.</span><span class="n">cwd</span><span class="p">,</span> <span class="n">which</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.projects.write_yaml.check_database_on_server" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">check_database_on_server</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">server_key</span><span class="p">,</span> <span class="n">server_dict</span><span class="p">)</span></code>

<a href="#ensembl.genes.projects.write_yaml.check_database_on_server" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Checks if a database exists on a given server.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>db</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the database to check.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_key</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The key of the server in the config.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_dict</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary containing server connection details.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the database exists, False otherwise.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_database_on_server</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">server_key</span><span class="p">,</span> <span class="n">server_dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Checks if a database exists on a given server.</span>

<span class="sd">    Args:</span>
<span class="sd">        db (str): The name of the database to check.</span>
<span class="sd">        server_key (str): The key of the server in the config.</span>
<span class="sd">        server_dict (dict): Dictionary containing server connection details.</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the database exists, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_dict</span><span class="p">[</span><span class="n">server_key</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_dict</span><span class="p">[</span><span class="n">server_key</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">passwd</span><span class="o">=</span><span class="n">server_dict</span><span class="p">[</span><span class="n">server_key</span><span class="p">][</span><span class="s2">&quot;db_pass&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_dict</span><span class="p">[</span><span class="n">server_key</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cur</span><span class="p">:</span>
            <span class="c1"># print(f&quot;Checking for DB &#39;{db}&#39; on server &#39;{server_key}&#39;&quot;, file=sys.stderr)</span>
            <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                <span class="s2">&quot;SELECT SCHEMA_NAME FROM information_schema.SCHEMATA WHERE SCHEMA_NAME = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">(</span><span class="n">db</span><span class="p">,),</span>
            <span class="p">)</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">cur</span><span class="o">.</span><span class="n">fetchone</span><span class="p">()</span>
            <span class="c1"># print(f&quot;Query result on {server_key}: {result}&quot;, file=sys.stderr)</span>
            <span class="k">return</span> <span class="n">result</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>

    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">MySQLError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error connecting to </span><span class="si">{</span><span class="n">server_key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">finally</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;conn&quot;</span> <span class="ow">in</span> <span class="nb">locals</span><span class="p">()</span> <span class="ow">and</span> <span class="n">conn</span><span class="p">:</span>
            <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.projects.write_yaml.check_url_status" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">check_url_status</span><span class="p">(</span><span class="n">url</span><span class="p">)</span></code>

<a href="#ensembl.genes.projects.write_yaml.check_url_status" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Checks if a given URL is reachable.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>url</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The URL to check.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the URL returns a 200 status code, False otherwise.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_url_status</span><span class="p">(</span><span class="n">url</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Checks if a given URL is reachable.</span>

<span class="sd">    Args:</span>
<span class="sd">        url (str): The URL to check.</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the URL returns a 200 status code, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">head</span><span class="p">(</span>
            <span class="n">url</span><span class="p">,</span> <span class="n">allow_redirects</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">5</span>
        <span class="p">)</span>  <span class="c1"># Use HEAD for efficiency</span>
        <span class="k">return</span> <span class="n">response</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span>
    <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error checking URL </span><span class="si">{</span><span class="n">url</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.projects.write_yaml.find_database_server" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">find_database_server</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">server_dict</span><span class="p">)</span></code>

<a href="#ensembl.genes.projects.write_yaml.find_database_server" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Determines which server the given database exists on.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>db</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the database to check.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>server_dict</code>
            </td>
            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary containing server connection details from the config file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the server where the database is found.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="Exception">Exception</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the database is not found on any of the servers.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_database_server</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">server_dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determines which server the given database exists on.</span>

<span class="sd">    Args:</span>
<span class="sd">        db (str): The name of the database to check.</span>
<span class="sd">        server_dict (dict): Dictionary containing server connection details from the config file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The name of the server where the database is found.</span>

<span class="sd">    Raises:</span>
<span class="sd">        Exception: If the database is not found on any of the servers.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check servers in order: st5 -&gt; st6 -&gt; main</span>
    <span class="k">for</span> <span class="n">server_key</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;st5&quot;</span><span class="p">,</span> <span class="s2">&quot;st6&quot;</span><span class="p">,</span> <span class="s2">&quot;main&quot;</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">check_database_on_server</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">server_key</span><span class="p">,</span> <span class="n">server_dict</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">server_key</span>

    <span class="c1"># If no server found, raise an error</span>
    <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unable to find database </span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s2"> on any configured servers!&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.projects.write_yaml.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.projects.write_yaml.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main function that parses arguments, reads server config from JSON,
queries MySQL databases, and writes YAML content.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main function that parses arguments, reads server config from JSON,</span>
<span class="sd">    queries MySQL databases, and writes YAML content.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Create a species.yaml file for a given project page.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-f&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--db_file&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Name of the file containing list of databases on Beta or Main servers and their Genome UUIDs&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-p&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--project&quot;</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span>
            <span class="s2">&quot;aquafaang&quot;</span><span class="p">,</span>
            <span class="s2">&quot;asg&quot;</span><span class="p">,</span>
            <span class="s2">&quot;bge&quot;</span><span class="p">,</span>
            <span class="s2">&quot;bovreg&quot;</span><span class="p">,</span>
            <span class="s2">&quot;cbp&quot;</span><span class="p">,</span>
            <span class="s2">&quot;dtol&quot;</span><span class="p">,</span>
            <span class="s2">&quot;erga&quot;</span><span class="p">,</span>
            <span class="s2">&quot;geneswitch&quot;</span><span class="p">,</span>
            <span class="s2">&quot;vgp&quot;</span><span class="p">,</span>
        <span class="p">],</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Name of the project this set of databases belongs to&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-c&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--config_file&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to the JSON configuration file containing server connection details&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">project</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">project</span>

    <span class="c1"># Load server connection details from JSON</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">config_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">config_f</span><span class="p">:</span>
        <span class="n">server_dict</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">config_f</span><span class="p">)</span>

    <span class="c1"># Read icons</span>
    <span class="n">icon_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;icons.txt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">icon_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">icon_file</span><span class="p">:</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">icon_dict</span><span class="p">[</span><span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Read database list</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">db_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">db_file</span><span class="p">:</span>
        <span class="n">db_list</span> <span class="o">=</span> <span class="n">db_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">sorted_db_list</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">db_list</span><span class="p">)</span>

    <span class="c1"># Special rearrangements for certain projects</span>
    <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;aquafaang&quot;</span><span class="p">:</span>
        <span class="c1"># Move danio rerio reference DBs to the end</span>
        <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">sorted_db_list</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;danio_rerio_core&quot;</span> <span class="ow">in</span> <span class="n">db</span><span class="p">:</span>
                <span class="n">sorted_db_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sorted_db_list</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">sorted_db_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">db</span><span class="p">)))</span>

    <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;bovreg&quot;</span><span class="p">:</span>
        <span class="c1"># Move bos taurus reference DBs to the top</span>
        <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">sorted_db_list</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;bos_taurus_core&quot;</span> <span class="ow">in</span> <span class="n">db</span><span class="p">:</span>
                <span class="n">sorted_db_list</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">sorted_db_list</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">sorted_db_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">db</span><span class="p">)))</span>

    <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;geneswitch&quot;</span><span class="p">:</span>
        <span class="c1"># Move sus scrofa or gallus gallus reference DBs to top</span>
        <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">sorted_db_list</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;sus_scrofa_core&quot;</span> <span class="ow">in</span> <span class="n">db</span> <span class="ow">or</span> <span class="s2">&quot;gallus_gallus_core&quot;</span> <span class="ow">in</span> <span class="n">db</span><span class="p">:</span>
                <span class="n">sorted_db_list</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">sorted_db_list</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">sorted_db_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">db</span><span class="p">)))</span>

    <span class="c1"># Initialize FTP connection</span>
    <span class="n">ftp_client</span> <span class="o">=</span> <span class="n">EnsemblFTP</span><span class="p">(</span><span class="n">timeout</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">max_retries</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">retry_sleep</span><span class="o">=</span><span class="mf">3.0</span><span class="p">)</span>

    <span class="c1"># Open the output YAML file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">_species.yaml&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">yaml_out</span><span class="p">:</span>
        <span class="c1"># will replace with guiid eventually</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">sorted_db_list</span><span class="p">:</span>
            <span class="n">db</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="n">guuid</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">guuid</span> <span class="o">==</span> <span class="s2">&quot;unknown&quot;</span><span class="p">:</span>
                <span class="n">use_server</span> <span class="o">=</span> <span class="s2">&quot;gb1&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">use_server</span> <span class="o">=</span> <span class="n">find_database_server</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">server_dict</span><span class="p">)</span>

            <span class="c1"># Retrieve metadata from the chosen server</span>
            <span class="c1"># update this with query from the metadata db</span>
            <span class="n">info_query</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;SELECT meta_key, meta_value FROM meta &quot;</span>
                <span class="s2">&quot;WHERE meta_key IN &quot;</span>
                <span class="s2">&quot;(&#39;species.scientific_name&#39;,&#39;assembly.accession&#39;,&#39;assembly.name&#39;,&quot;</span>
                <span class="s2">&quot;&#39;species.production_name&#39;,&#39;species.strain&#39;,&#39;schema_version&#39;,&quot;</span>
                <span class="s2">&quot;&#39;genebuild.last_geneset_update&#39;,&#39;species.annotation_source&#39;,&#39;genebuild.busco&#39;,&#39;genebuild.busco_dataset&#39;) &quot;</span>
                <span class="s2">&quot;OR meta_key LIKE &#39;genebuild.method%&#39;&quot;</span>
            <span class="p">)</span>
            <span class="n">info</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
                <span class="n">info_query</span><span class="p">,</span>
                <span class="n">db</span><span class="p">,</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="n">use_server</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="n">use_server</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="n">use_server</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="n">use_server</span><span class="p">][</span><span class="s2">&quot;db_pass&quot;</span><span class="p">],</span>
            <span class="p">)</span>

            <span class="n">info_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">info</span><span class="p">}</span>

            <span class="c1"># Check if assembly has an alternate</span>
            <span class="c1"># update this with query from the metadata db</span>
            <span class="n">alternate_assembly_name</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.name&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_alternate_haplotype&quot;</span>
            <span class="p">)</span>
            <span class="n">alternate_query</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;SELECT organism.url_name &quot;</span>
                <span class="s2">&quot;FROM assembly &quot;</span>
                <span class="s2">&quot;JOIN genome USING (assembly_id) &quot;</span>
                <span class="s2">&quot;JOIN organism USING (organism_id) &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;WHERE assembly.assembly_name=&#39;</span><span class="si">{</span><span class="n">alternate_assembly_name</span><span class="si">}</span><span class="s2">&#39; LIMIT 1;&quot;</span>
            <span class="p">)</span>
            <span class="n">alternate_fetch</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
                <span class="n">alternate_query</span><span class="p">,</span>
                <span class="s2">&quot;ensembl_metadata_qrp&quot;</span><span class="p">,</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;db_pass&quot;</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="n">alternate</span> <span class="o">=</span> <span class="n">alternate_fetch</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">alternate_fetch</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>

            <span class="c1"># Retrieve classification info</span>
            <span class="c1"># replace this with call to datasets</span>
            <span class="n">class_query</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;SELECT meta_value FROM meta WHERE meta_key=&#39;species.classification&#39;&quot;</span>
            <span class="p">)</span>
            <span class="n">classifications</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
                <span class="n">class_query</span><span class="p">,</span>
                <span class="n">db</span><span class="p">,</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="n">use_server</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="n">use_server</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="n">use_server</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
                <span class="n">server_dict</span><span class="p">[</span><span class="n">use_server</span><span class="p">][</span><span class="s2">&quot;db_pass&quot;</span><span class="p">],</span>
            <span class="p">)</span>

            <span class="n">class_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">classifications</span><span class="p">]</span>

            <span class="n">icon</span> <span class="o">=</span> <span class="s2">&quot;Metazoa.png&quot;</span>
            <span class="n">chordate</span> <span class="o">=</span> <span class="s2">&quot;Chordata&quot;</span> <span class="ow">in</span> <span class="n">class_list</span>
            <span class="k">for</span> <span class="n">classification</span> <span class="ow">in</span> <span class="n">class_list</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">classification</span> <span class="ow">in</span> <span class="n">icon_dict</span><span class="p">:</span>
                    <span class="n">icon</span> <span class="o">=</span> <span class="n">icon_dict</span><span class="p">[</span><span class="n">classification</span><span class="p">]</span>
                    <span class="k">break</span>

            <span class="k">if</span> <span class="n">chordate</span> <span class="ow">and</span> <span class="n">icon</span> <span class="o">==</span> <span class="s2">&quot;Metazoa.png&quot;</span><span class="p">:</span>
                <span class="n">icon</span> <span class="o">=</span> <span class="s2">&quot;Chordates.png&quot;</span>

            <span class="n">write_yaml</span><span class="p">(</span>
                <span class="n">info_dict</span><span class="p">,</span>
                <span class="n">icon</span><span class="p">,</span>
                <span class="n">yaml_out</span><span class="p">,</span>
                <span class="n">project</span><span class="p">,</span>
                <span class="n">use_server</span><span class="p">,</span>
                <span class="n">alternate</span><span class="p">,</span>
                <span class="n">guuid</span><span class="p">,</span>
                <span class="n">ftp_client</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="n">ftp_client</span><span class="o">.</span><span class="n">close_connections</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.projects.write_yaml.mysql_fetch_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">)</span></code>

<a href="#ensembl.genes.projects.write_yaml.mysql_fetch_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetch data from a MySQL database using a given query.</p>
<h5 id="ensembl.genes.projects.write_yaml.mysql_fetch_data--parameters">Parameters<a class="headerlink" href="#ensembl.genes.projects.write_yaml.mysql_fetch_data--parameters" title="Permanent link">&para;</a></h5>
<p>query : str
    The SQL query to execute.
database : str
    The database name.
host : str
    The database host.
port : int
    The database port.
user : str
    The username for the database.
password : str
    The password for the database.</p>
<h5 id="ensembl.genes.projects.write_yaml.mysql_fetch_data--returns">Returns<a class="headerlink" href="#ensembl.genes.projects.write_yaml.mysql_fetch_data--returns" title="Permanent link">&para;</a></h5>
<p>Tuple
    A tuple of results fetched from the database query.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mysql_fetch_data</span><span class="p">(</span>
    <span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">password</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch data from a MySQL database using a given query.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    query : str</span>
<span class="sd">        The SQL query to execute.</span>
<span class="sd">    database : str</span>
<span class="sd">        The database name.</span>
<span class="sd">    host : str</span>
<span class="sd">        The database host.</span>
<span class="sd">    port : int</span>
<span class="sd">        The database port.</span>
<span class="sd">    user : str</span>
<span class="sd">        The username for the database.</span>
<span class="sd">    password : str</span>
<span class="sd">        The password for the database.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Tuple</span>
<span class="sd">        A tuple of results fetched from the database query.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
        <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span> <span class="n">passwd</span><span class="o">=</span><span class="n">password</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="n">port</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">cursor</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
    <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="n">info</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
    <span class="n">cursor</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">info</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.projects.write_yaml.write_yaml" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">write_yaml</span><span class="p">(</span><span class="n">info_dict</span><span class="p">,</span> <span class="n">icon</span><span class="p">,</span> <span class="n">yaml_out</span><span class="p">,</span> <span class="n">project</span><span class="p">,</span> <span class="n">use_server</span><span class="p">,</span> <span class="n">alternate</span><span class="p">,</span> <span class="n">guuid</span><span class="p">,</span> <span class="n">ftp_client</span><span class="p">)</span></code>

<a href="#ensembl.genes.projects.write_yaml.write_yaml" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Write YAML content for a species entry based on the provided metadata and project type.</p>
<h5 id="ensembl.genes.projects.write_yaml.write_yaml--parameters">Parameters<a class="headerlink" href="#ensembl.genes.projects.write_yaml.write_yaml--parameters" title="Permanent link">&para;</a></h5>
<p>info_dict : Dict[str, str]
    Dictionary containing species metadata.
icon : str
    Icon filename for the species.
yaml_out : file-like object
    The output file handle for writing the YAML.
project : str
    The project name (e.g., 'aquafaang', 'asg', 'bge', ...).
use_server : str
    Which server the data was fetched from ('st5', 'st6' or 'main').
alternate : str
    Alternate assembly name, if available.
ftp_client : EnsemblFTP
    An initialized FTP client instance for checking extra files.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/projects/write_yaml.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_yaml</span><span class="p">(</span>
    <span class="n">info_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="n">icon</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">yaml_out</span><span class="p">,</span>
    <span class="n">project</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">use_server</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">alternate</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">guuid</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ftp_client</span><span class="p">:</span> <span class="n">EnsemblFTP</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write YAML content for a species entry based on the provided metadata and project type.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    info_dict : Dict[str, str]</span>
<span class="sd">        Dictionary containing species metadata.</span>
<span class="sd">    icon : str</span>
<span class="sd">        Icon filename for the species.</span>
<span class="sd">    yaml_out : file-like object</span>
<span class="sd">        The output file handle for writing the YAML.</span>
<span class="sd">    project : str</span>
<span class="sd">        The project name (e.g., &#39;aquafaang&#39;, &#39;asg&#39;, &#39;bge&#39;, ...).</span>
<span class="sd">    use_server : str</span>
<span class="sd">        Which server the data was fetched from (&#39;st5&#39;, &#39;st6&#39; or &#39;main&#39;).</span>
<span class="sd">    alternate : str</span>
<span class="sd">        Alternate assembly name, if available.</span>
<span class="sd">    ftp_client : EnsemblFTP</span>
<span class="sd">        An initialized FTP client instance for checking extra files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">prod_url_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;bos_taurus_hybrid&quot;</span><span class="p">,</span> <span class="s2">&quot;bos_indicus_hybrid&quot;</span><span class="p">]</span>

    <span class="n">assembly_name</span> <span class="o">=</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">date</span> <span class="o">=</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;genebuild.last_geneset_update&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="n">date</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">species_name</span> <span class="o">=</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species.scientific_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="n">species_name</span> <span class="o">=</span> <span class="n">species_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">lc_species_name</span> <span class="o">=</span> <span class="n">species_name</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>

    <span class="k">if</span> <span class="s2">&quot;species.strain&quot;</span> <span class="ow">in</span> <span class="n">info_dict</span> <span class="ow">and</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species.strain&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;reference&quot;</span><span class="p">:</span>
        <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species.scientific_name&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species.strain&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>

    <span class="n">uc_prod_name</span> <span class="o">=</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species.production_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>

    <span class="c1"># Get submitter from assembly report</span>
    <span class="n">assembly_report_url</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;https://www.ncbi.nlm.nih.gov/assembly/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="n">assembly_report_response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">assembly_report_url</span><span class="p">)</span>
    <span class="n">submitter_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;Submitter: &lt;/dt&gt;&lt;dd&gt;([^&lt;]*)&lt;/dd&gt;&lt;dt&gt;&quot;</span><span class="p">,</span> <span class="n">assembly_report_response</span><span class="o">.</span><span class="n">text</span>
    <span class="p">)</span>
    <span class="n">submitter</span> <span class="o">=</span> <span class="n">submitter_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="k">if</span> <span class="n">submitter_match</span> <span class="k">else</span> <span class="s2">&quot;unknown&quot;</span>
    <span class="k">if</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GCA_000002315.5&quot;</span><span class="p">:</span>
        <span class="n">submitter</span> <span class="o">=</span> <span class="s2">&quot;Genome Reference Consortium&quot;</span>

    <span class="n">source</span> <span class="o">=</span> <span class="n">info_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;species.annotation_source&quot;</span><span class="p">,</span> <span class="s2">&quot;ensembl&quot;</span><span class="p">)</span>

    <span class="c1"># Start building YAML content</span>
    <span class="n">yaml</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;- species: </span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species.scientific_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="k">if</span> <span class="n">use_server</span> <span class="o">==</span> <span class="s2">&quot;st5&quot;</span> <span class="ow">or</span> <span class="n">use_server</span> <span class="o">==</span> <span class="s2">&quot;st6&quot;</span> <span class="ow">or</span> <span class="n">use_server</span> <span class="o">==</span> <span class="s2">&quot;gb1&quot;</span><span class="p">:</span>
        <span class="n">ftp_base</span> <span class="o">=</span> <span class="s2">&quot;https://ftp.ebi.ac.uk/pub/ensemblorganisms&quot;</span>

        <span class="k">if</span> <span class="n">project</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;vgp&quot;</span><span class="p">,</span> <span class="s2">&quot;dtol&quot;</span><span class="p">,</span> <span class="s2">&quot;erga&quot;</span><span class="p">,</span> <span class="s2">&quot;cbp&quot;</span><span class="p">,</span> <span class="s2">&quot;bge&quot;</span><span class="p">,</span> <span class="s2">&quot;asg&quot;</span><span class="p">):</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  image: </span><span class="si">{</span><span class="n">icon</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  submitted_by: </span><span class="si">{</span><span class="n">submitter</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  accession: </span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="n">project</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;dtol&quot;</span><span class="p">,</span> <span class="s2">&quot;erga&quot;</span><span class="p">,</span> <span class="s2">&quot;cbp&quot;</span><span class="p">,</span> <span class="s2">&quot;bge&quot;</span><span class="p">,</span> <span class="s2">&quot;asg&quot;</span><span class="p">):</span>
            <span class="c1"># Add annotation method if available, else default to BRAKER2</span>
            <span class="n">method</span> <span class="o">=</span> <span class="n">info_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;genebuild.method_display&quot;</span><span class="p">,</span> <span class="s2">&quot;BRAKER2&quot;</span><span class="p">)</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  annotation_method: </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="c1"># ==== annotation_gtf ====</span>
        <span class="n">gtf</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s2">/geneset/</span><span class="si">{</span><span class="n">date</span><span class="si">}</span><span class="s2">/genes.gtf.gz&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">check_url_status</span><span class="p">(</span><span class="n">gtf</span><span class="p">):</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  annotation_gtf: </span><span class="si">{</span><span class="n">gtf</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">fb</span> <span class="o">=</span> <span class="n">ftp_client</span><span class="o">.</span><span class="n">check_pre_release_file</span><span class="p">(</span>
                <span class="n">species_name</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">],</span> <span class="s2">&quot;.gtf.gz&quot;</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">fb</span><span class="p">:</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  annotation_gtf: </span><span class="si">{</span><span class="n">fb</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="c1"># ==== annotation_gff3: try gzipped, then uncompressed, then pre‑release ====</span>
        <span class="n">gff_gz</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s2">/geneset/</span><span class="si">{</span><span class="n">date</span><span class="si">}</span><span class="s2">/genes.gff3.gz&quot;</span>
        <span class="p">)</span>
        <span class="n">gff</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s2">/geneset/</span><span class="si">{</span><span class="n">date</span><span class="si">}</span><span class="s2">/genes.gff3&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">check_url_status</span><span class="p">(</span><span class="n">gff_gz</span><span class="p">):</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  annotation_gff3: </span><span class="si">{</span><span class="n">gff_gz</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">elif</span> <span class="n">check_url_status</span><span class="p">(</span><span class="n">gff</span><span class="p">):</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  annotation_gff3: </span><span class="si">{</span><span class="n">gff</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># fallback: look in pre‑release for gzipped first, then uncompressed</span>
            <span class="n">fb</span> <span class="o">=</span> <span class="n">ftp_client</span><span class="o">.</span><span class="n">check_pre_release_file</span><span class="p">(</span>
                <span class="n">species_name</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">],</span> <span class="s2">&quot;.gff3.gz&quot;</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">fb</span><span class="p">:</span>
                <span class="n">fb</span> <span class="o">=</span> <span class="n">ftp_client</span><span class="o">.</span><span class="n">check_pre_release_file</span><span class="p">(</span>
                    <span class="n">species_name</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">],</span> <span class="s2">&quot;.gff3&quot;</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="n">fb</span><span class="p">:</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  annotation_gff3: </span><span class="si">{</span><span class="n">fb</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="c1"># ==== proteins ====</span>
        <span class="n">pep</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s2">/geneset/</span><span class="si">{</span><span class="n">date</span><span class="si">}</span><span class="s2">/pep.fa.gz&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">check_url_status</span><span class="p">(</span><span class="n">pep</span><span class="p">):</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  proteins: </span><span class="si">{</span><span class="n">pep</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="c1"># ==== transcripts ====</span>
        <span class="n">cdna</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s2">/geneset/</span><span class="si">{</span><span class="n">date</span><span class="si">}</span><span class="s2">/cdna.fa.gz&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">check_url_status</span><span class="p">(</span><span class="n">cdna</span><span class="p">):</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  transcripts: </span><span class="si">{</span><span class="n">cdna</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="c1"># ==== softmasked genome ====</span>
        <span class="n">soft</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/genome/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;softmasked.fa.gz&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">check_url_status</span><span class="p">(</span><span class="n">soft</span><span class="p">):</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  softmasked_genome: </span><span class="si">{</span><span class="n">soft</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">fb</span> <span class="o">=</span> <span class="n">ftp_client</span><span class="o">.</span><span class="n">check_pre_release_file</span><span class="p">(</span>
                <span class="n">species_name</span><span class="p">,</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">],</span> <span class="s2">&quot;.dna.softmasked.fa.gz&quot;</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">fb</span><span class="p">:</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  softmasked_genome: </span><span class="si">{</span><span class="n">fb</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">rm_file</span> <span class="o">=</span> <span class="n">ftp_client</span><span class="o">.</span><span class="n">check_for_file</span><span class="p">(</span>
            <span class="n">lc_species_name</span><span class="p">,</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species.production_name&quot;</span><span class="p">],</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">],</span>
            <span class="n">source</span><span class="p">,</span>
            <span class="s2">&quot;repeatmodeler&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">rm_file</span><span class="p">:</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  repeat_library: </span><span class="si">{</span><span class="n">rm_file</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="c1"># ==== ftp dumps ====</span>
        <span class="n">dumps</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="n">check_url_status</span><span class="p">(</span><span class="n">dumps</span><span class="p">):</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  ftp_dumps: </span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  ftp_dumps: https://ftp.ebi.ac.uk/pub/databases/ensembl/pre-release/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">main_species_url</span> <span class="o">=</span> <span class="s2">&quot;http://www.ensembl.org/info/about/species.html&quot;</span>
        <span class="n">main_species_response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">main_species_url</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">main_species_response</span><span class="o">.</span><span class="n">text</span><span class="p">:</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;  ensembl_link: https://www.ensembl.org/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/Info/Index</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">beta_link</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://beta.ensembl.org/species/</span><span class="si">{</span><span class="n">guuid</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">check_url_status</span><span class="p">(</span><span class="n">beta_link</span><span class="p">):</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  beta_link: https://beta.ensembl.org/species/</span><span class="si">{</span><span class="n">guuid</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  beta_link: Coming soon!</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="n">project</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;dtol&quot;</span><span class="p">,</span> <span class="s2">&quot;erga&quot;</span><span class="p">,</span> <span class="s2">&quot;cbp&quot;</span><span class="p">,</span> <span class="s2">&quot;bge&quot;</span><span class="p">,</span> <span class="s2">&quot;asg&quot;</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  busco_score: </span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;genebuild.busco&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">busco_file</span> <span class="o">=</span> <span class="n">ftp_client</span><span class="o">.</span><span class="n">check_for_file</span><span class="p">(</span>
                    <span class="n">species_name</span><span class="p">,</span>
                    <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species.production_name&quot;</span><span class="p">],</span>
                    <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">],</span>
                    <span class="n">source</span><span class="p">,</span>
                    <span class="s2">&quot;busco&quot;</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">busco_file</span><span class="p">:</span>
                    <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  busco_score: </span><span class="si">{</span><span class="n">busco_file</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  busco_lineage: </span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;genebuild.busco_dataset&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">busco_file</span><span class="p">:</span>
                    <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  busco_lineage: </span><span class="si">{</span><span class="n">busco_file</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">alternate</span><span class="p">:</span>
                <span class="n">alternate_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://rapid.ensembl.org/</span><span class="si">{</span><span class="n">alternate</span><span class="si">}</span><span class="s2">/Info/Index&quot;</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  alternate: </span><span class="si">{</span><span class="n">alternate_url</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="k">elif</span> <span class="n">use_server</span> <span class="o">==</span> <span class="s2">&quot;main&quot;</span><span class="p">:</span>
        <span class="c1"># Handle &quot;geneswitch&quot; project special case with release 102</span>
        <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;geneswitch&quot;</span><span class="p">:</span>
            <span class="n">release</span> <span class="o">=</span> <span class="s2">&quot;release-102&quot;</span>
            <span class="n">release_number</span> <span class="o">=</span> <span class="s2">&quot;102&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">release</span> <span class="o">=</span> <span class="s2">&quot;release-&quot;</span> <span class="o">+</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;schema_version&quot;</span><span class="p">]</span>
            <span class="n">release_number</span> <span class="o">=</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;schema_version&quot;</span><span class="p">]</span>

        <span class="n">ftp_base</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://ftp.ensembl.org/pub/</span><span class="si">{</span><span class="n">release</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="n">project</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;vgp&quot;</span><span class="p">,</span> <span class="s2">&quot;dtol&quot;</span><span class="p">,</span> <span class="s2">&quot;erga&quot;</span><span class="p">,</span> <span class="s2">&quot;cbp&quot;</span><span class="p">,</span> <span class="s2">&quot;bge&quot;</span><span class="p">,</span> <span class="s2">&quot;asg&quot;</span><span class="p">):</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  image: </span><span class="si">{</span><span class="n">icon</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  submitted_by: </span><span class="si">{</span><span class="n">submitter</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  accession: </span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;assembly.accession&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="n">project</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;dtol&quot;</span><span class="p">,</span> <span class="s2">&quot;erga&quot;</span><span class="p">,</span> <span class="s2">&quot;cbp&quot;</span><span class="p">,</span> <span class="s2">&quot;bge&quot;</span><span class="p">,</span> <span class="s2">&quot;asg&quot;</span><span class="p">):</span>
            <span class="c1"># Annotation method is always Ensembl genebuild on main</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="s2">&quot;  annotation_method: Ensembl genebuild</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">source</span> <span class="o">=</span> <span class="s2">&quot;ensembl&quot;</span>  <span class="c1"># for consistency with busco retrieval logic</span>

        <span class="n">yaml</span> <span class="o">+=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;  annotation_gtf: </span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/gtf/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species.production_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">uc_prod_name</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">assembly_name</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">release_number</span><span class="si">}</span><span class="s2">.gtf.gz</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">yaml</span> <span class="o">+=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;  annotation_gff3: </span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/gff3/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species.production_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">uc_prod_name</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">assembly_name</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">release_number</span><span class="si">}</span><span class="s2">.gff3.gz</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">yaml</span> <span class="o">+=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;  proteins: </span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/fasta/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species.production_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/pep/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">uc_prod_name</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">assembly_name</span><span class="si">}</span><span class="s2">.pep.all.fa.gz</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">yaml</span> <span class="o">+=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;  transcripts: </span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/fasta/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species.production_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/cdna/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">uc_prod_name</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">assembly_name</span><span class="si">}</span><span class="s2">.cdna.all.fa.gz</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">yaml</span> <span class="o">+=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;  softmasked_genome: </span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="s2">/fasta/</span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;species.production_name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">/dna/&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">uc_prod_name</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">assembly_name</span><span class="si">}</span><span class="s2">.dna_sm.toplevel.fa.gz</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">rm_file</span> <span class="o">=</span> <span class="n">ftp_client</span><span class="o">.</span><span class="n">check_for_file</span><span class="p">(</span>
            <span class="n">lc_species_name</span><span class="p">,</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species.production_name&quot;</span><span class="p">],</span>
            <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">],</span>
            <span class="n">source</span><span class="p">,</span>
            <span class="s2">&quot;repeatmodeler&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">rm_file</span><span class="p">:</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  repeat_library: </span><span class="si">{</span><span class="n">rm_file</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  ftp_dumps: </span><span class="si">{</span><span class="n">ftp_base</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species.production_name&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">prod_url_list</span><span class="p">:</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;  ensembl_link: https://www.ensembl.org/</span><span class="si">{</span><span class="n">uc_prod_name</span><span class="si">}</span><span class="s2">/Info/Index</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;geneswitch&quot;</span><span class="p">:</span>
            <span class="c1"># geneswitch project frozen at e102</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;  ensembl_link: https://e102.ensembl.org/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/Info/Index</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">yaml</span> <span class="o">+=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;  ensembl_link: https://www.ensembl.org/</span><span class="si">{</span><span class="n">species_name</span><span class="si">}</span><span class="s2">/Info/Index</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">project</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;dtol&quot;</span><span class="p">,</span> <span class="s2">&quot;erga&quot;</span><span class="p">,</span> <span class="s2">&quot;cbp&quot;</span><span class="p">,</span> <span class="s2">&quot;bge&quot;</span><span class="p">,</span> <span class="s2">&quot;asg&quot;</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  busco_score: </span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;genebuild.busco&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">busco_file</span> <span class="o">=</span> <span class="n">ftp_client</span><span class="o">.</span><span class="n">check_for_file</span><span class="p">(</span>
                    <span class="n">lc_species_name</span><span class="p">,</span>
                    <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;species.production_name&quot;</span><span class="p">],</span>
                    <span class="n">info_dict</span><span class="p">[</span><span class="s2">&quot;assembly.accession&quot;</span><span class="p">],</span>
                    <span class="n">source</span><span class="p">,</span>
                    <span class="s2">&quot;busco&quot;</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">busco_file</span><span class="p">:</span>
                    <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  busco_score: </span><span class="si">{</span><span class="n">busco_file</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  busco_lineage: </span><span class="si">{</span><span class="n">info_dict</span><span class="p">[</span><span class="s1">&#39;genebuild.busco_dataset&#39;</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">busco_file</span><span class="p">:</span>
                    <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  busco_lineage: </span><span class="si">{</span><span class="n">busco_file</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">if</span> <span class="n">alternate</span><span class="p">:</span>
                <span class="n">alternate_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://rapid.ensembl.org/</span><span class="si">{</span><span class="n">alternate</span><span class="si">}</span><span class="s2">/Info/Index&quot;</span>
                <span class="n">yaml</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  alternate: </span><span class="si">{</span><span class="n">alternate_url</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">yaml</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">yaml_out</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="stable-id">Stable ID<a class="headerlink" href="#stable-id" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.stable_id.pid_overlap" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.stable_id.pid_overlap</code>


<a href="#ensembl.genes.stable_id.pid_overlap" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.stable_id.pid_overlap.gff3_file_new" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">gff3_file_new</span> <span class="o">=</span> <span class="nb">input</span><span class="p">(</span><span class="s1">&#39;Enter the path to the newer GFF3 file: &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.stable_id.pid_overlap.gff3_file_new" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.stable_id.pid_overlap.gff3_file_old" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">gff3_file_old</span> <span class="o">=</span> <span class="nb">input</span><span class="p">(</span><span class="s1">&#39;Enter the path to the older GFF3 file: &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.stable_id.pid_overlap.gff3_file_old" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.stable_id.pid_overlap.output_report" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">output_report</span> <span class="o">=</span> <span class="nb">input</span><span class="p">(</span><span class="s1">&#39;Enter the path for the output report file: &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.stable_id.pid_overlap.output_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.stable_id.pid_overlap.compare_protein_sets" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">compare_protein_sets</span><span class="p">(</span><span class="n">gff3_file_old</span><span class="p">,</span> <span class="n">gff3_file_new</span><span class="p">,</span> <span class="n">output_report</span><span class="p">)</span></code>

<a href="#ensembl.genes.stable_id.pid_overlap.compare_protein_sets" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/stable_id/pid_overlap.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compare_protein_sets</span><span class="p">(</span><span class="n">gff3_file_old</span><span class="p">,</span> <span class="n">gff3_file_new</span><span class="p">,</span> <span class="n">output_report</span><span class="p">):</span>
    <span class="c1"># Extract protein information from both files</span>
    <span class="n">proteins_old</span> <span class="o">=</span> <span class="n">extract_proteins_from_gff3</span><span class="p">(</span><span class="n">gff3_file_old</span><span class="p">)</span>
    <span class="n">proteins_new</span> <span class="o">=</span> <span class="n">extract_proteins_from_gff3</span><span class="p">(</span><span class="n">gff3_file_new</span><span class="p">)</span>

    <span class="c1"># Find shared protein IDs</span>
    <span class="n">shared_proteins</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">proteins_old</span><span class="p">,</span> <span class="n">proteins_new</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;ProteinID&quot;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;_old&quot;</span><span class="p">,</span> <span class="s2">&quot;_new&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Compute proportions</span>
    <span class="n">proportion_old</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">shared_proteins</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">proteins_old</span><span class="p">)</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">proteins_old</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="p">)</span>
    <span class="n">proportion_new</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">shared_proteins</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">proteins_new</span><span class="p">)</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">proteins_new</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Proportion of proteins in the old GFF3 that have shared IDs: </span><span class="si">{</span><span class="n">proportion_old</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Proportion of proteins in the new GFF3 that have shared IDs: </span><span class="si">{</span><span class="n">proportion_new</span><span class="si">:</span><span class="s2">.2%</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Add columns for genomic locations in the report</span>
    <span class="n">shared_proteins</span> <span class="o">=</span> <span class="n">shared_proteins</span><span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;ProteinID&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Chromosome_old&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Start_old&quot;</span><span class="p">,</span>
            <span class="s2">&quot;End_old&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Strand_old&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Chromosome_new&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Start_new&quot;</span><span class="p">,</span>
            <span class="s2">&quot;End_new&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Strand_new&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>

    <span class="c1"># Save report</span>
    <span class="n">shared_proteins</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="n">output_report</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_MINIMAL</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Report saved to </span><span class="si">{</span><span class="n">output_report</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.stable_id.pid_overlap.extract_proteins_from_gff3" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">extract_proteins_from_gff3</span><span class="p">(</span><span class="n">gff3_file</span><span class="p">)</span></code>

<a href="#ensembl.genes.stable_id.pid_overlap.extract_proteins_from_gff3" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/stable_id/pid_overlap.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">extract_proteins_from_gff3</span><span class="p">(</span><span class="n">gff3_file</span><span class="p">):</span>
    <span class="n">protein_data</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gff3_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="n">columns</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">attributes</span> <span class="o">=</span> <span class="n">columns</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;ID=CDS:([^;]+)&quot;</span><span class="p">,</span> <span class="n">attributes</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">protein_id</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">protein_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">protein_data</span><span class="p">:</span>
                        <span class="n">protein_data</span><span class="p">[</span><span class="n">protein_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                            <span class="s2">&quot;ProteinID&quot;</span><span class="p">:</span> <span class="n">protein_id</span><span class="p">,</span>
                            <span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">columns</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span>
                            <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">]),</span>
                            <span class="s2">&quot;Strand&quot;</span><span class="p">:</span> <span class="n">columns</span><span class="p">[</span><span class="mi">6</span><span class="p">],</span>
                            <span class="s2">&quot;Attributes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">attributes</span><span class="p">],</span>
                        <span class="p">}</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Update the start and end to encompass all CDS regions for the protein</span>
                        <span class="n">protein_data</span><span class="p">[</span><span class="n">protein_id</span><span class="p">][</span><span class="s2">&quot;Start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span>
                            <span class="n">protein_data</span><span class="p">[</span><span class="n">protein_id</span><span class="p">][</span><span class="s2">&quot;Start&quot;</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">columns</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                        <span class="p">)</span>
                        <span class="n">protein_data</span><span class="p">[</span><span class="n">protein_id</span><span class="p">][</span><span class="s2">&quot;End&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span>
                            <span class="n">protein_data</span><span class="p">[</span><span class="n">protein_id</span><span class="p">][</span><span class="s2">&quot;End&quot;</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
                        <span class="p">)</span>
                        <span class="n">protein_data</span><span class="p">[</span><span class="n">protein_id</span><span class="p">][</span><span class="s2">&quot;Attributes&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">attributes</span><span class="p">)</span>
    <span class="c1"># Flatten attributes into a single string for each protein</span>
    <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">protein_data</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">protein</span><span class="p">[</span><span class="s2">&quot;Attributes&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein</span><span class="p">[</span><span class="s2">&quot;Attributes&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">protein_data</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="tracking">Tracking<a class="headerlink" href="#tracking" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.tracking.bioproject_tracking" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.tracking.bioproject_tracking</code>


<a href="#ensembl.genes.tracking.bioproject_tracking" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.tracking.bioproject_tracking.config" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">config</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.tracking.bioproject_tracking.config" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.add_ftp" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">add_ftp</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">release_type</span><span class="o">=</span><span class="s1">&#39;live&#39;</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.add_ftp" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_ftp</span><span class="p">(</span>
    <span class="n">annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span> <span class="n">release_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;live&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]:</span>
    <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">annotation</span> <span class="ow">in</span> <span class="n">annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;matches&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">release_type</span> <span class="o">==</span> <span class="s2">&quot;pre&quot;</span><span class="p">:</span>
            <span class="n">dbnames</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">):</span>
                <span class="n">dbnames</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">])</span>

            <span class="k">for</span> <span class="n">dbname</span> <span class="ow">in</span> <span class="n">dbnames</span><span class="p">:</span>
                <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                SELECT meta_value</span>
<span class="s2">                FROM meta</span>
<span class="s2">                WHERE meta_key = &#39;organism.scientific_name&#39;</span>
<span class="s2">                &quot;&quot;&quot;</span>
                <span class="n">result</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
                    <span class="n">query</span><span class="p">,</span>
                    <span class="p">(),</span>
                    <span class="n">server_group</span><span class="o">=</span><span class="s2">&quot;pre-release&quot;</span><span class="p">,</span>
                    <span class="n">server_name</span><span class="o">=</span><span class="s2">&quot;gb1&quot;</span><span class="p">,</span>
                    <span class="n">db_name</span><span class="o">=</span><span class="n">dbname</span><span class="p">,</span>
                <span class="p">)</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span><span class="p">:</span>
                    <span class="k">continue</span>  <span class="c1"># skip if no scientific name found</span>

                <span class="n">scientific_name</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ftp_link</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://ftp.ebi.ac.uk/pub/databases/ensembl/pre-release/</span><span class="si">{</span><span class="n">scientific_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">&quot;</span>

                <span class="c1"># attach to matching entries</span>
                <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">matches</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">dbname</span><span class="p">:</span>
                        <span class="n">m</span><span class="p">[</span><span class="s2">&quot;ftp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ftp_link</span>
                <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">dbname</span><span class="p">:</span>
                    <span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;ftp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ftp_link</span>  <span class="c1"># top-level</span>

        <span class="k">elif</span> <span class="n">release_type</span> <span class="o">==</span> <span class="s2">&quot;live&quot;</span><span class="p">:</span>
            <span class="c1"># Build map guuid-&gt;match for easy updates</span>
            <span class="n">by_guuid</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">):</span> <span class="n">m</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">matches</span> <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">)}</span>
            <span class="n">guuids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">by_guuid</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
            <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">):</span>
                <span class="n">guuids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;guuid&quot;</span><span class="p">])</span>

            <span class="k">for</span> <span class="n">guuid</span> <span class="ow">in</span> <span class="n">guuids</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">guuid</span> <span class="ow">or</span> <span class="n">guuid</span> <span class="o">==</span> <span class="s2">&quot;unknown&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">                SELECT genome.genebuild_date, organism.scientific_name, dataset_attribute.value</span>
<span class="s2">                FROM genome</span>
<span class="s2">                JOIN organism USING(organism_id)</span>
<span class="s2">                JOIN genome_dataset USING(genome_id) </span>
<span class="s2">                JOIN dataset_attribute USING(dataset_id)</span>
<span class="s2">                WHERE genome.genome_uuid = </span><span class="si">%s</span>
<span class="s2">                AND attribute_id=169</span>
<span class="s2">                &quot;&quot;&quot;</span>
                <span class="n">data_fetch</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="p">(</span><span class="n">guuid</span><span class="p">,))</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">data_fetch</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="n">date</span><span class="p">,</span> <span class="n">scientific_name</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="n">data_fetch</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">date</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">date</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
                <span class="n">scientific_name</span> <span class="o">=</span> <span class="n">scientific_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
                <span class="n">ftp_link</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://ftp.ebi.ac.uk/pub/ensemblorganisms/</span><span class="si">{</span><span class="n">scientific_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s2">/geneset/</span><span class="si">{</span><span class="n">date</span><span class="si">}</span><span class="s2">/&quot;</span>

                <span class="k">if</span> <span class="n">guuid</span> <span class="ow">in</span> <span class="n">by_guuid</span><span class="p">:</span>
                    <span class="n">by_guuid</span><span class="p">[</span><span class="n">guuid</span><span class="p">][</span><span class="s2">&quot;ftp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ftp_link</span>
                    <span class="n">by_guuid</span><span class="p">[</span><span class="n">guuid</span><span class="p">][</span><span class="s2">&quot;date&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">date</span>
                <span class="k">if</span> <span class="n">annotation</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">guuid</span><span class="p">:</span>
                    <span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;ftp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ftp_link</span>
                    <span class="n">annotation</span><span class="p">[</span><span class="s2">&quot;date&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">date</span>

    <span class="k">return</span> <span class="n">annotations</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.get_assembly_accessions" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_assembly_accessions</span><span class="p">(</span><span class="n">query_id</span><span class="p">,</span> <span class="n">query_type</span><span class="p">,</span> <span class="n">only_haploid</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.get_assembly_accessions" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Prefer NCBI Datasets CLI (all versions), fall back to Datasets API (latest only).
Supports query_type in {"bioproject", "taxon"}.
Always returns a dict (possibly empty): {accession: {"taxon_id": int}}</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_assembly_accessions</span><span class="p">(</span>
    <span class="n">query_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">query_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">only_haploid</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prefer NCBI Datasets CLI (all versions), fall back to Datasets API (latest only).</span>
<span class="sd">    Supports query_type in {&quot;bioproject&quot;, &quot;taxon&quot;}.</span>
<span class="sd">    Always returns a dict (possibly empty): {accession: {&quot;taxon_id&quot;: int}}</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">_parse_reports</span><span class="p">(</span><span class="n">lines</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
        <span class="n">accs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">obj</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
            <span class="k">except</span> <span class="n">json</span><span class="o">.</span><span class="n">JSONDecodeError</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">report</span> <span class="o">=</span> <span class="n">obj</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;report&quot;</span><span class="p">,</span> <span class="n">obj</span><span class="p">)</span>
            <span class="n">assembly_info</span> <span class="o">=</span> <span class="n">report</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_info&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">{}</span>
            <span class="k">if</span> <span class="n">only_haploid</span> <span class="ow">and</span> <span class="n">assembly_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_type&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;haploid&quot;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">accession</span> <span class="o">=</span> <span class="n">report</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;accession&quot;</span><span class="p">)</span>
            <span class="n">taxon_id</span> <span class="o">=</span> <span class="p">(</span><span class="n">report</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;organism&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tax_id&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">accession</span><span class="p">:</span>
                <span class="n">accs</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">:</span> <span class="n">taxon_id</span><span class="p">}</span>
        <span class="k">return</span> <span class="n">accs</span>

    <span class="c1"># --- 1) Try CLI (gets ALL versions) ---</span>
    <span class="n">datasets_cli</span> <span class="o">=</span> <span class="n">shutil</span><span class="o">.</span><span class="n">which</span><span class="p">(</span><span class="s2">&quot;datasets&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">datasets_cli</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">query_type</span> <span class="o">==</span> <span class="s2">&quot;bioproject&quot;</span><span class="p">:</span>
                <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">datasets_cli</span><span class="p">,</span>
                    <span class="s2">&quot;summary&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;genome&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;accession&quot;</span><span class="p">,</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">query_id</span><span class="p">),</span>
                    <span class="s2">&quot;--assembly-version&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;all&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;--as-json-lines&quot;</span><span class="p">,</span>
                <span class="p">]</span>
            <span class="k">elif</span> <span class="n">query_type</span> <span class="o">==</span> <span class="s2">&quot;taxon&quot;</span><span class="p">:</span>
                <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">datasets_cli</span><span class="p">,</span>
                    <span class="s2">&quot;summary&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;genome&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;taxon&quot;</span><span class="p">,</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">query_id</span><span class="p">),</span>
                    <span class="s2">&quot;--assembly-version&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;all&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;--as-json-lines&quot;</span><span class="p">,</span>
                <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid query_type &#39;</span><span class="si">{</span><span class="n">query_type</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">{}</span>

            <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">parsed</span> <span class="o">=</span> <span class="n">_parse_reports</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">splitlines</span><span class="p">())</span>
            <span class="k">if</span> <span class="n">parsed</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;[Datasets CLI] Retrieved </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">parsed</span><span class="p">)</span><span class="si">}</span><span class="s2"> assemblies (all versions).&quot;</span>
                <span class="p">)</span>
                <span class="k">return</span> <span class="n">parsed</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;[Datasets CLI] Parsed zero rows from stdout; falling back to API.&quot;</span>
                <span class="p">)</span>
        <span class="k">except</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">CalledProcessError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;[Datasets CLI] Exit </span><span class="si">{</span><span class="n">e</span><span class="o">.</span><span class="n">returncode</span><span class="si">}</span><span class="s2">. STDERR:</span><span class="se">\n</span><span class="si">{</span><span class="n">e</span><span class="o">.</span><span class="n">stderr</span><span class="si">}</span><span class="se">\n</span><span class="s2">Falling back to API.&quot;</span>
            <span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;[Datasets CLI] Unexpected error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">. Falling back to API.&quot;</span>
            <span class="p">)</span>

    <span class="c1"># --- 2) Fallback: API (latest only) ---</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">base_url</span> <span class="o">=</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;urls&quot;</span><span class="p">][</span><span class="s2">&quot;datasets&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">query_type</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
        <span class="n">base_url</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">base_url</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="s2">&quot;[Datasets API] Base URL not found in config; returning empty set.&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="p">{}</span>

    <span class="n">assembly_accessions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">page_size</span> <span class="o">=</span> <span class="mi">5000</span>
    <span class="n">next_page_token</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_url</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">query_id</span><span class="si">}</span><span class="s2">/dataset_report?page_size=</span><span class="si">{</span><span class="n">page_size</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="n">next_page_token</span><span class="p">:</span>
            <span class="n">url</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;&amp;page_token=</span><span class="si">{</span><span class="n">next_page_token</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
            <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">assembly</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;reports&quot;</span><span class="p">,</span> <span class="p">[]):</span>
                <span class="n">assembly_info</span> <span class="o">=</span> <span class="n">assembly</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_info&quot;</span><span class="p">,</span> <span class="p">{})</span> <span class="ow">or</span> <span class="p">{}</span>
                <span class="k">if</span> <span class="n">only_haploid</span> <span class="ow">and</span> <span class="n">assembly_info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;assembly_type&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;haploid&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">accession</span> <span class="o">=</span> <span class="n">assembly</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;accession&quot;</span><span class="p">)</span>
                <span class="n">taxon_id</span> <span class="o">=</span> <span class="p">(</span><span class="n">assembly</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;organism&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tax_id&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">accession</span><span class="p">:</span>
                    <span class="n">assembly_accessions</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">:</span> <span class="n">taxon_id</span><span class="p">}</span>
            <span class="n">next_page_token</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;next_page_token&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">next_page_token</span><span class="p">:</span>
                <span class="k">break</span>
        <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">RequestException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Datasets API] Error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">break</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;[Datasets API] Retrieved </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">assembly_accessions</span><span class="p">)</span><span class="si">}</span><span class="s2"> assemblies (latest only).&quot;</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">assembly_accessions</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.get_ensembl_live" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_ensembl_live</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.get_ensembl_live" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Returns (annotated_dict, missing_list).
annotated_dict[accession] contains:
  - taxon_id (from input)
  - matches: List[{"guuid","dbname", optional "ftp","date"}]
  - guuid, dbname: best match (latest core; prefer non-cm)</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_ensembl_live</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns (annotated_dict, missing_list).</span>
<span class="sd">    annotated_dict[accession] contains:</span>
<span class="sd">      - taxon_id (from input)</span>
<span class="sd">      - matches: List[{&quot;guuid&quot;,&quot;dbname&quot;, optional &quot;ftp&quot;,&quot;date&quot;}]</span>
<span class="sd">      - guuid, dbname: best match (latest core; prefer non-cm)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">accessions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">accessions</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">{},</span> <span class="p">[]</span>

    <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    SELECT DISTINCT</span>
<span class="s2">        assembly.accession,</span>
<span class="s2">        genome.genome_uuid,</span>
<span class="s2">        dataset_source.name AS database_name</span>
<span class="s2">    FROM genome</span>
<span class="s2">    JOIN assembly ON genome.assembly_id = assembly.assembly_id</span>
<span class="s2">    JOIN genome_dataset ON genome.genome_id = genome_dataset.genome_id</span>
<span class="s2">    JOIN dataset ON genome_dataset.dataset_id = dataset.dataset_id</span>
<span class="s2">    JOIN dataset_source ON dataset.dataset_source_id = dataset_source.dataset_source_id</span>
<span class="s2">    JOIN genome_release ON genome.genome_id = genome_release.genome_id</span>
<span class="s2">    JOIN ensembl_release ON genome_release.release_id = ensembl_release.release_id</span>
<span class="s2">    WHERE dataset.name = &#39;genebuild&#39;</span>
<span class="s2">      AND assembly.accession IN (</span><span class="si">{}</span><span class="s2">)</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">accessions</span><span class="p">))</span>
    <span class="p">)</span>

    <span class="n">rows</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">accessions</span><span class="p">))</span>

    <span class="c1"># Seed with taxon info so missing ones can still be tracked</span>
    <span class="n">live_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">guuid</span><span class="p">,</span> <span class="n">dbname</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
        <span class="n">entry</span> <span class="o">=</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span>
            <span class="n">accession</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">accession</span><span class="p">,</span> <span class="p">{}))</span>
        <span class="p">)</span>
        <span class="n">entry</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;matches&quot;</span><span class="p">,</span> <span class="p">[])</span>
        <span class="c1"># Avoid duplicates across joins if any</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">any</span><span class="p">(</span>
            <span class="n">m</span><span class="p">[</span><span class="s2">&quot;guuid&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">guuid</span> <span class="ow">and</span> <span class="n">m</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">dbname</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">entry</span><span class="p">[</span><span class="s2">&quot;matches&quot;</span><span class="p">]</span>
        <span class="p">):</span>
            <span class="n">entry</span><span class="p">[</span><span class="s2">&quot;matches&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;guuid&quot;</span><span class="p">:</span> <span class="n">guuid</span><span class="p">,</span> <span class="s2">&quot;dbname&quot;</span><span class="p">:</span> <span class="n">dbname</span><span class="p">})</span>

    <span class="c1"># Choose best match for backward compatibility fields</span>
    <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="n">entry</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;matches&quot;</span><span class="p">,</span> <span class="p">[])</span>
        <span class="k">if</span> <span class="n">matches</span><span class="p">:</span>
            <span class="n">best</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">matches</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">m</span><span class="p">:</span> <span class="n">_score_dbname</span><span class="p">(</span><span class="n">m</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">]))</span>
            <span class="n">entry</span><span class="p">[</span><span class="s2">&quot;guuid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">best</span><span class="p">[</span><span class="s2">&quot;guuid&quot;</span><span class="p">]</span>
            <span class="n">entry</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">best</span><span class="p">[</span><span class="s2">&quot;dbname&quot;</span><span class="p">]</span>

    <span class="c1"># Compute missing (accessions with no rows at all)</span>
    <span class="n">matched_accessions</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">live_annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="n">missing_annotations</span> <span class="o">=</span> <span class="p">[</span><span class="n">acc</span> <span class="k">for</span> <span class="n">acc</span> <span class="ow">in</span> <span class="n">accessions</span> <span class="k">if</span> <span class="n">acc</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">matched_accessions</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">live_annotations</span><span class="p">,</span> <span class="n">missing_annotations</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.get_pre_release" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_pre_release</span><span class="p">(</span><span class="n">missing_annotations</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.get_pre_release" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Checks for the existence of pre-release Ensembl database schemas for a list of missing accessions.
This function takes a list of GenBank/RefSeq assembly accessions that were not found in the
Ensembl live database lookup (e.g., from <code>get_ensembl_live</code>). It reformats each accession
to match the expected pre-release Ensembl database schema naming convention (e.g.,
'GCA_000001405.39' → 'gca000001405v39') and queries the MySQL server's
<code>information_schema.schemata</code> table to check if a database with that name exists.
If such a schema is found, the function records the corresponding accession and
database name in the returned dictionary.
Args:
    missing_annotations (List[str]): A list of assembly accessions (e.g., 'GCA_000001405.39')
        for which no live annotation database was found.
Returns:
    Dict[str, Dict[str, str]]: A dictionary mapping each accession with a found pre-release
    database to a nested dictionary containing:
        - "dbname": The name of the pre-release schema (e.g., 'gca000001405v39').
    Example:
        {
            "GCA_000001405.39": {
                "dbname": "gca000001405v39"
            },
            ...
        }
Notes:
    - Only accessions matching the pattern 'GCA/GCF_XXXXXXXXX.XX' are processed.
    - If no matching schema is found for an accession, it is omitted from the result.
    - The function uses <code>mysql_fetch_data</code> to query the MySQL server's schema metadata.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_pre_release</span><span class="p">(</span><span class="n">missing_annotations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Checks for the existence of pre-release Ensembl database schemas for a list of missing accessions.</span>
<span class="sd">    This function takes a list of GenBank/RefSeq assembly accessions that were not found in the</span>
<span class="sd">    Ensembl live database lookup (e.g., from `get_ensembl_live`). It reformats each accession</span>
<span class="sd">    to match the expected pre-release Ensembl database schema naming convention (e.g.,</span>
<span class="sd">    &#39;GCA_000001405.39&#39; → &#39;gca000001405v39&#39;) and queries the MySQL server&#39;s</span>
<span class="sd">    `information_schema.schemata` table to check if a database with that name exists.</span>
<span class="sd">    If such a schema is found, the function records the corresponding accession and</span>
<span class="sd">    database name in the returned dictionary.</span>
<span class="sd">    Args:</span>
<span class="sd">        missing_annotations (List[str]): A list of assembly accessions (e.g., &#39;GCA_000001405.39&#39;)</span>
<span class="sd">            for which no live annotation database was found.</span>
<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Dict[str, str]]: A dictionary mapping each accession with a found pre-release</span>
<span class="sd">        database to a nested dictionary containing:</span>
<span class="sd">            - &quot;dbname&quot;: The name of the pre-release schema (e.g., &#39;gca000001405v39&#39;).</span>
<span class="sd">        Example:</span>
<span class="sd">            {</span>
<span class="sd">                &quot;GCA_000001405.39&quot;: {</span>
<span class="sd">                    &quot;dbname&quot;: &quot;gca000001405v39&quot;</span>
<span class="sd">                },</span>
<span class="sd">                ...</span>
<span class="sd">            }</span>
<span class="sd">    Notes:</span>
<span class="sd">        - Only accessions matching the pattern &#39;GCA/GCF_XXXXXXXXX.XX&#39; are processed.</span>
<span class="sd">        - If no matching schema is found for an accession, it is omitted from the result.</span>
<span class="sd">        - The function uses `mysql_fetch_data` to query the MySQL server&#39;s schema metadata.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pre_release_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">missing_annotations</span><span class="p">:</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(GCA|GCF)_(\d+)\.(\d+)&quot;</span><span class="p">,</span> <span class="n">accession</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">match</span><span class="p">:</span>
            <span class="k">continue</span>  <span class="c1"># skip malformed accession</span>

        <span class="n">prefix</span><span class="p">,</span> <span class="n">number</span><span class="p">,</span> <span class="n">version</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
        <span class="n">lookup_string</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;%</span><span class="si">{</span><span class="n">prefix</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="si">}{</span><span class="n">number</span><span class="si">}</span><span class="s2">v</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">%&quot;</span>

        <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT SCHEMA_NAME</span>
<span class="s2">        FROM information_schema.schemata</span>
<span class="s2">        WHERE SCHEMA_NAME like </span><span class="si">%s</span>
<span class="s2">        &quot;&quot;&quot;</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">query</span><span class="p">,</span> <span class="p">(</span><span class="n">lookup_string</span><span class="p">,),</span> <span class="n">server_group</span><span class="o">=</span><span class="s2">&quot;pre-release&quot;</span><span class="p">,</span> <span class="n">server_name</span><span class="o">=</span><span class="s2">&quot;gb1&quot;</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">result</span><span class="p">:</span>
            <span class="n">pre_release_annotations</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;guuid&quot;</span><span class="p">:</span> <span class="s2">&quot;unknown&quot;</span><span class="p">,</span>  <span class="c1"># typo fixed from guiid -&gt; guuid</span>
                <span class="s2">&quot;dbname&quot;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
            <span class="p">}</span>
    <span class="k">return</span> <span class="n">pre_release_annotations</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.get_taxonomy_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_taxonomy_info</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">,</span> <span class="n">accessions_taxon</span><span class="p">,</span> <span class="n">rank</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.get_taxonomy_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetches taxonomy information for a given rank from the NCBI Datasets API.
For each accession in <code>live_annotations</code>, this function retrieves its <code>taxon_id</code>
and queries the NCBI taxonomy endpoint to get the classification at the specified rank.
The retrieved rank name is then added to the corresponding annotation data.
Args:
    live_annotations (Dict[str, Dict[str, str]]): A dictionary where each key is
        an assembly accession and its value contains annotation details (including
        at least "taxon_id").
    accessions_taxon (Dict[str, Dict[str, int]]): A dictionary mapping accessions
        to basic taxon information, such as {"GCA_000001405.39": {"taxon_id": 9606}}.
    rank (str): The taxonomic rank to retrieve (e.g., "order", "class", "phylum").
Returns:
    Dict[str, Dict[str, str]]: The updated <code>live_annotations</code> dictionary with an
    additional key for the requested rank. For example:
        {
            "GCA_000001405.39": {
                "taxon_id": 9606,
                "guuid": "some-uuid",
                "dbname": "ensembl_core",
                "order": "Primates"
            },
            ...
        }
    If the API query fails or the rank is not found, the value for that rank is not added
    (or set to None/unknown).</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_taxonomy_info</span><span class="p">(</span>
    <span class="n">live_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span>
    <span class="n">accessions_taxon</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
    <span class="n">rank</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches taxonomy information for a given rank from the NCBI Datasets API.</span>
<span class="sd">    For each accession in `live_annotations`, this function retrieves its `taxon_id`</span>
<span class="sd">    and queries the NCBI taxonomy endpoint to get the classification at the specified rank.</span>
<span class="sd">    The retrieved rank name is then added to the corresponding annotation data.</span>
<span class="sd">    Args:</span>
<span class="sd">        live_annotations (Dict[str, Dict[str, str]]): A dictionary where each key is</span>
<span class="sd">            an assembly accession and its value contains annotation details (including</span>
<span class="sd">            at least &quot;taxon_id&quot;).</span>
<span class="sd">        accessions_taxon (Dict[str, Dict[str, int]]): A dictionary mapping accessions</span>
<span class="sd">            to basic taxon information, such as {&quot;GCA_000001405.39&quot;: {&quot;taxon_id&quot;: 9606}}.</span>
<span class="sd">        rank (str): The taxonomic rank to retrieve (e.g., &quot;order&quot;, &quot;class&quot;, &quot;phylum&quot;).</span>
<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Dict[str, str]]: The updated `live_annotations` dictionary with an</span>
<span class="sd">        additional key for the requested rank. For example:</span>
<span class="sd">            {</span>
<span class="sd">                &quot;GCA_000001405.39&quot;: {</span>
<span class="sd">                    &quot;taxon_id&quot;: 9606,</span>
<span class="sd">                    &quot;guuid&quot;: &quot;some-uuid&quot;,</span>
<span class="sd">                    &quot;dbname&quot;: &quot;ensembl_core&quot;,</span>
<span class="sd">                    &quot;order&quot;: &quot;Primates&quot;</span>
<span class="sd">                },</span>
<span class="sd">                ...</span>
<span class="sd">            }</span>
<span class="sd">        If the API query fails or the rank is not found, the value for that rank is not added</span>
<span class="sd">        (or set to None/unknown).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="p">:</span>
        <span class="n">taxon_id</span> <span class="o">=</span> <span class="n">accessions_taxon</span><span class="p">[</span><span class="n">accession</span><span class="p">][</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span>
        <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://api.ncbi.nlm.nih.gov/datasets/v2alpha/taxonomy/taxon/</span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">/dataset_report&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
            <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>

            <span class="n">taxonomies</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;reports&quot;</span><span class="p">,</span> <span class="p">[])</span>
            <span class="k">for</span> <span class="n">taxonomy</span> <span class="ow">in</span> <span class="n">taxonomies</span><span class="p">:</span>
                <span class="n">rank_name</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">taxonomy</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;taxonomy&quot;</span><span class="p">,</span> <span class="p">{})</span>
                    <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;classification&quot;</span><span class="p">,</span> <span class="p">{})</span>
                    <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rank</span><span class="p">,</span> <span class="p">{})</span>
                    <span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="n">taxonomy_info</span> <span class="o">=</span> <span class="p">{</span><span class="n">rank</span><span class="p">:</span> <span class="n">rank_name</span><span class="p">}</span>
                <span class="n">live_annotations</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">taxonomy_info</span><span class="p">)</span>

        <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">HTTPError</span> <span class="k">as</span> <span class="n">http_err</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HTTP error occurred: </span><span class="si">{</span><span class="n">http_err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;An error occurred: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">live_annotations</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Main entry point of the script.
Parses command-line arguments to determine if a BioProject ID or Taxon ID is provided,
whether to filter only haploid assemblies, the desired output file path, the taxonomic
rank to retrieve, and whether to include FTP links. It then:
  1. Fetches assembly accessions from the NCBI API based on the provided ID.
  2. Fetches corresponding Ensembl live database information from the local MySQL database.
  3. Retrieves taxonomy information for the specified rank from the NCBI API.
  4. Optionally adds FTP links if requested.
  5. Writes all collected data into a tab-separated report file.
Usage Example:
    python script_name.py --bioproject_id PRJNA12345 --haploid --rank order --ftp
Returns:
    None</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main entry point of the script.</span>
<span class="sd">    Parses command-line arguments to determine if a BioProject ID or Taxon ID is provided,</span>
<span class="sd">    whether to filter only haploid assemblies, the desired output file path, the taxonomic</span>
<span class="sd">    rank to retrieve, and whether to include FTP links. It then:</span>
<span class="sd">      1. Fetches assembly accessions from the NCBI API based on the provided ID.</span>
<span class="sd">      2. Fetches corresponding Ensembl live database information from the local MySQL database.</span>
<span class="sd">      3. Retrieves taxonomy information for the specified rank from the NCBI API.</span>
<span class="sd">      4. Optionally adds FTP links if requested.</span>
<span class="sd">      5. Writes all collected data into a tab-separated report file.</span>
<span class="sd">    Usage Example:</span>
<span class="sd">        python script_name.py --bioproject_id PRJNA12345 --haploid --rank order --ftp</span>
<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Fetch assembly data and report annotations.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--bioproject_id&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;NCBI BioProject ID&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--taxon_id&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxonomy ID&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--haploid&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Fetch only haploid assemblies&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--report_file&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;./report_file.tsv&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--classification&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Provide breakdown of taxonomic classification&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--rank&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;order&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--ftp&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Include FTP links in the report&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--pre_release&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Include list of pre-release databases in the report&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="c1"># Determine which ID is provided and which query type is relevant</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="p">:</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;You must provide either --bioproject_id or --taxon_id&quot;</span><span class="p">)</span>

    <span class="n">accessions_taxon</span> <span class="o">=</span> <span class="n">get_assembly_accessions</span><span class="p">(</span>
        <span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="p">,</span>
        <span class="s2">&quot;bioproject&quot;</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span> <span class="k">else</span> <span class="s2">&quot;taxon&quot;</span><span class="p">,</span>
        <span class="n">args</span><span class="o">.</span><span class="n">haploid</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Fetch annotations (live)</span>
    <span class="n">live_annotations</span><span class="p">,</span> <span class="n">missing_annotations</span> <span class="o">=</span> <span class="n">get_ensembl_live</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="p">)</span>

    <span class="c1"># Taxonomy classification (top-level only; rows inherit from details)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">classification</span><span class="p">:</span>
        <span class="n">get_taxonomy_info</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">,</span> <span class="n">accessions_taxon</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">rank</span><span class="p">)</span>

    <span class="c1"># Optionally add FTP links</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">ftp</span><span class="p">:</span>
        <span class="n">live_annotations</span> <span class="o">=</span> <span class="n">add_ftp</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">,</span> <span class="s2">&quot;live&quot;</span><span class="p">)</span>

    <span class="c1"># Unique species count from live annotations</span>
    <span class="n">unique_taxon_ids</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">details</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">details</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="k">if</span> <span class="s2">&quot;taxon_id&quot;</span> <span class="ow">in</span> <span class="n">details</span>
    <span class="p">}</span>

    <span class="c1"># Optionally add Pre-release data</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">pre_release</span> <span class="ow">and</span> <span class="n">missing_annotations</span><span class="p">:</span>
        <span class="n">pre_release_annotations</span> <span class="o">=</span> <span class="n">get_pre_release</span><span class="p">(</span><span class="n">missing_annotations</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">pre_release_annotations</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">accession</span> <span class="ow">in</span> <span class="n">accessions_taxon</span><span class="p">:</span>
                <span class="n">pre_release_annotations</span><span class="p">[</span><span class="n">accession</span><span class="p">][</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">accessions_taxon</span><span class="p">[</span>
                    <span class="n">accession</span>
                <span class="p">][</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">classification</span> <span class="ow">and</span> <span class="n">pre_release_annotations</span><span class="p">:</span>
            <span class="n">get_taxonomy_info</span><span class="p">(</span><span class="n">pre_release_annotations</span><span class="p">,</span> <span class="n">accessions_taxon</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">rank</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">ftp</span> <span class="ow">and</span> <span class="n">pre_release_annotations</span><span class="p">:</span>
            <span class="n">pre_release_annotations</span> <span class="o">=</span> <span class="n">add_ftp</span><span class="p">(</span><span class="n">pre_release_annotations</span><span class="p">,</span> <span class="s2">&quot;pre&quot;</span><span class="p">)</span>

        <span class="n">all_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="o">**</span><span class="n">live_annotations</span><span class="p">,</span>
            <span class="o">**</span><span class="n">pre_release_annotations</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">all_annotations</span> <span class="o">=</span> <span class="n">live_annotations</span>

    <span class="c1"># Console summary</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="p">)</span><span class="si">}</span><span class="s2"> assemblies under BioProject ID </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">bioproject_id</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">accessions_taxon</span><span class="p">)</span><span class="si">}</span><span class="s2"> assemblies for taxon ID </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">)</span><span class="si">}</span><span class="s2"> annotations in beta.ensembl.org for </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_taxon_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> unique species&quot;</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">classification</span><span class="p">:</span>
        <span class="n">rank_values</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">rank</span><span class="p">,</span> <span class="s2">&quot;unknown&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">details</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="p">]</span>
        <span class="n">rank_counts</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">rank_values</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Breakdown:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">rank_counts</span><span class="p">)</span>

    <span class="c1"># Write final report</span>
    <span class="n">write_report</span><span class="p">(</span>
        <span class="n">all_annotations</span><span class="p">,</span>
        <span class="n">args</span><span class="o">.</span><span class="n">report_file</span><span class="p">,</span>
        <span class="n">args</span><span class="o">.</span><span class="n">rank</span><span class="p">,</span>
        <span class="n">include_class</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">classification</span><span class="p">,</span>
        <span class="n">include_ftp</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">ftp</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.mysql_fetch_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="p">(),</span> <span class="n">server_group</span><span class="o">=</span><span class="s1">&#39;meta&#39;</span><span class="p">,</span> <span class="n">server_name</span><span class="o">=</span><span class="s1">&#39;beta&#39;</span><span class="p">,</span> <span class="n">db_name</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.mysql_fetch_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Executes a SQL query with optional parameters to fetch results from a specified MySQL server.</p>
<p>Returns empty list on error or no data.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mysql_fetch_data</span><span class="p">(</span>
    <span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">params</span><span class="p">:</span> <span class="n">Tuple</span> <span class="o">=</span> <span class="p">(),</span>
    <span class="n">server_group</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;meta&quot;</span><span class="p">,</span>
    <span class="n">server_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;beta&quot;</span><span class="p">,</span>
    <span class="n">db_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="o">...</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Executes a SQL query with optional parameters to fetch results from a specified MySQL server.</span>

<span class="sd">    Returns empty list on error or no data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">server_config</span> <span class="o">=</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="n">server_group</span><span class="p">][</span><span class="n">server_name</span><span class="p">]</span>
        <span class="n">connection</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">server_config</span><span class="p">[</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">server_config</span><span class="p">[</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">server_config</span><span class="p">[</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">database</span><span class="o">=</span><span class="n">db_name</span> <span class="ow">or</span> <span class="n">server_config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;db_name&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cursor</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">params</span><span class="p">)</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
        <span class="n">connection</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">result</span>

    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">key_err</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid server group or name in config: </span><span class="si">{</span><span class="n">key_err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">sql_err</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MySQL Error: </span><span class="si">{</span><span class="n">sql_err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">[]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.bioproject_tracking.write_report" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">write_report</span><span class="p">(</span><span class="n">live_annotations</span><span class="p">,</span> <span class="n">report_file</span><span class="p">,</span> <span class="n">rank</span><span class="p">,</span> <span class="n">include_class</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">include_ftp</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.bioproject_tracking.write_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Writes a tab-separated report. If an accession has multiple matches,
produces one line per match; otherwise one line.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/bioproject_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_report</span><span class="p">(</span>
    <span class="n">live_annotations</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span>
    <span class="n">report_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">rank</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">include_class</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">include_ftp</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Writes a tab-separated report. If an accession has multiple matches,</span>
<span class="sd">    produces one line per match; otherwise one line.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">report_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">report_file</span><span class="p">)</span>
    <span class="n">lines_written</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">report_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">details</span> <span class="ow">in</span> <span class="n">live_annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">matches</span> <span class="o">=</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;matches&quot;</span><span class="p">)</span>

            <span class="c1"># If we have multiple, write one row per match</span>
            <span class="n">rows</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">matches</span>
                <span class="k">if</span> <span class="n">matches</span>
                <span class="k">else</span> <span class="p">[</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;guuid&quot;</span><span class="p">:</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">),</span>
                        <span class="s2">&quot;dbname&quot;</span><span class="p">:</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">),</span>
                        <span class="s2">&quot;ftp&quot;</span><span class="p">:</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ftp&quot;</span><span class="p">),</span>
                        <span class="s2">&quot;date&quot;</span><span class="p">:</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;date&quot;</span><span class="p">),</span>
                    <span class="p">}</span>
                <span class="p">]</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">accession</span><span class="p">),</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;guuid&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;unknown&quot;</span><span class="p">),</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;dbname&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;unknown&quot;</span><span class="p">),</span>
                <span class="p">]</span>
                <span class="k">if</span> <span class="n">include_class</span><span class="p">:</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rank</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;unknown&quot;</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">include_ftp</span><span class="p">:</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;date&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;date&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
                    <span class="n">row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ftp&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">details</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ftp&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="s2">&quot;N/A&quot;</span><span class="p">))</span>

                <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">lines_written</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Report written to </span><span class="si">{</span><span class="n">report_path</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span><span class="si">}</span><span class="s2"> with </span><span class="si">{</span><span class="n">lines_written</span><span class="si">}</span><span class="s2"> lines.&quot;</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.tracking.live_tracking" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.tracking.live_tracking</code>


<a href="#ensembl.genes.tracking.live_tracking" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










<div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h3 id="ensembl.genes.tracking.live_tracking.config" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-attribute"></code>            <code class="highlight language-python"><span class="n">config</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#ensembl.genes.tracking.live_tracking.config" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

    </div>

</div>




<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.live_tracking.check_database_on_server" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">check_database_on_server</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">server_key</span><span class="p">,</span> <span class="n">config</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.live_tracking.check_database_on_server" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Checks if a database exists on a given server.
Args:
    db (str): The name of the database to check.
    server_key (str): The key of the server in the config.
    server_dict (dict): Dictionary containing server connection details.
Returns:
    bool: True if the database exists, False otherwise.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/live_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_database_on_server</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">server_key</span><span class="p">,</span> <span class="n">config</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Checks if a database exists on a given server.</span>
<span class="sd">    Args:</span>
<span class="sd">        db (str): The name of the database to check.</span>
<span class="sd">        server_key (str): The key of the server in the config.</span>
<span class="sd">        server_dict (dict): Dictionary containing server connection details.</span>
<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the database exists, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;staging&quot;</span><span class="p">][</span><span class="n">server_key</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">user</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;staging&quot;</span><span class="p">][</span><span class="n">server_key</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
            <span class="n">passwd</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;staging&quot;</span><span class="p">][</span><span class="n">server_key</span><span class="p">][</span><span class="s2">&quot;db_pass&quot;</span><span class="p">],</span>
            <span class="n">port</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;staging&quot;</span><span class="p">][</span><span class="n">server_key</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span> <span class="k">as</span> <span class="n">cur</span><span class="p">:</span>
            <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                <span class="s2">&quot;SELECT SCHEMA_NAME FROM information_schema.SCHEMATA WHERE SCHEMA_NAME = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">(</span><span class="n">db</span><span class="p">,),</span>
            <span class="p">)</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">cur</span><span class="o">.</span><span class="n">fetchone</span><span class="p">()</span>
            <span class="k">return</span> <span class="n">result</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>

    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">MySQLError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error connecting to </span><span class="si">{</span><span class="n">server_key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">finally</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;conn&quot;</span> <span class="ow">in</span> <span class="nb">locals</span><span class="p">()</span> <span class="ow">and</span> <span class="n">conn</span><span class="p">:</span>
            <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.live_tracking.clean_server" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">clean_server</span><span class="p">(</span><span class="n">config</span><span class="p">,</span> <span class="n">mode</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.live_tracking.clean_server" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Fetches a list of core databases that are both current in the rapid database and present on the MySQL server.</p>
<p>This function performs the following steps:
1. Executes a query on the rapid database to fetch all current core databases.
2. Executes a query on the MySQL server to fetch all core databases.
3. Compares the results from both queries and compiles a list of core databases that are present in both.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of core databases that are current in the rapid database and present on the MySQL server.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/live_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">clean_server</span><span class="p">(</span><span class="n">config</span><span class="p">,</span> <span class="n">mode</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches a list of core databases that are both current in the rapid database and present on the MySQL server.</span>

<span class="sd">    This function performs the following steps:</span>
<span class="sd">    1. Executes a query on the rapid database to fetch all current core databases.</span>
<span class="sd">    2. Executes a query on the MySQL server to fetch all core databases.</span>
<span class="sd">    3. Compares the results from both queries and compiles a list of core databases that are present in both.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: A list of core databases that are current in the rapid database and present on the MySQL server.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">rapid_live_databases</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">beta_live_databases</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">cores_query</span> <span class="o">=</span> <span class="s2">&quot;SHOW DATABASES like &#39;</span><span class="si">%c</span><span class="s2">ore%&#39;;&quot;</span>
    <span class="n">cores_fetch</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
        <span class="n">cores_query</span><span class="p">,</span>
        <span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;genebuild&quot;</span><span class="p">][</span><span class="s2">&quot;prod_1&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
        <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;genebuild&quot;</span><span class="p">][</span><span class="s2">&quot;prod_1&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
        <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;genebuild&quot;</span><span class="p">][</span><span class="s2">&quot;prod_1&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">core_dbs</span> <span class="o">=</span> <span class="p">[</span><span class="n">db</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">cores_fetch</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;rapid&quot;</span><span class="p">:</span>
        <span class="n">rapid_query</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;SELECT dbname FROM genome_database JOIN genome USING(genome_id) &quot;</span>
            <span class="s2">&quot;JOIN data_release USING(data_release_id) WHERE is_current = 1 and dbname like &#39;</span><span class="si">%c</span><span class="s2">ore%&#39;;&quot;</span>
        <span class="p">)</span>
        <span class="n">rapid_fetch</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">rapid_query</span><span class="p">,</span>
            <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;rapid&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
            <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;rapid&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;rapid&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;rapid&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="n">rapid_databases</span> <span class="o">=</span> <span class="p">[</span><span class="n">rrdb</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">rrdb</span> <span class="ow">in</span> <span class="n">rapid_fetch</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">core_db</span> <span class="ow">in</span> <span class="n">core_dbs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">core_db</span> <span class="ow">in</span> <span class="n">rapid_databases</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">server_key</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;st5&quot;</span><span class="p">,</span> <span class="s2">&quot;st6&quot;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">check_database_on_server</span><span class="p">(</span><span class="n">core_db</span><span class="p">,</span> <span class="n">server_key</span><span class="p">,</span> <span class="n">config</span><span class="p">):</span>
                        <span class="n">rapid_live_databases</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">core_db</span><span class="p">)</span>
                        <span class="k">break</span>
        <span class="k">return</span> <span class="n">rapid_live_databases</span>

    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;beta&quot;</span><span class="p">:</span>
        <span class="n">beta_query</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;SELECT DISTINCT(dataset_source.name) AS dbname FROM genome &quot;</span>
            <span class="s2">&quot;JOIN assembly ON genome.assembly_id = assembly.assembly_id &quot;</span>
            <span class="s2">&quot;JOIN genome_dataset ON genome.genome_id = genome_dataset.genome_id &quot;</span>
            <span class="s2">&quot;JOIN dataset ON genome_dataset.dataset_id = dataset.dataset_id &quot;</span>
            <span class="s2">&quot;JOIN dataset_source ON dataset.dataset_source_id = dataset_source.dataset_source_id &quot;</span>
            <span class="s2">&quot;JOIN genome_release ON genome.genome_id = genome_release.genome_id &quot;</span>
            <span class="s2">&quot;JOIN ensembl_release ON genome_release.release_id = ensembl_release.release_id &quot;</span>
            <span class="s2">&quot;WHERE dataset.name=&#39;assembly&#39; AND dataset.status=&#39;Released&#39;;&quot;</span>
        <span class="p">)</span>
        <span class="n">beta_fetch</span> <span class="o">=</span> <span class="n">mysql_fetch_data</span><span class="p">(</span>
            <span class="n">beta_query</span><span class="p">,</span>
            <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;beta&quot;</span><span class="p">][</span><span class="s2">&quot;db_name&quot;</span><span class="p">],</span>
            <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;beta&quot;</span><span class="p">][</span><span class="s2">&quot;db_host&quot;</span><span class="p">],</span>
            <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;beta&quot;</span><span class="p">][</span><span class="s2">&quot;db_port&quot;</span><span class="p">],</span>
            <span class="n">config</span><span class="p">[</span><span class="s2">&quot;server_details&quot;</span><span class="p">][</span><span class="s2">&quot;meta&quot;</span><span class="p">][</span><span class="s2">&quot;beta&quot;</span><span class="p">][</span><span class="s2">&quot;db_user&quot;</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="n">beta_databases</span> <span class="o">=</span> <span class="p">[</span><span class="n">rrdb</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">rrdb</span> <span class="ow">in</span> <span class="n">beta_fetch</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">core_db</span> <span class="ow">in</span> <span class="n">core_dbs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">core_db</span> <span class="ow">in</span> <span class="n">beta_databases</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">server_key</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;st5&quot;</span><span class="p">,</span> <span class="s2">&quot;st6&quot;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">check_database_on_server</span><span class="p">(</span><span class="n">core_db</span><span class="p">,</span> <span class="n">server_key</span><span class="p">,</span> <span class="n">config</span><span class="p">):</span>
                        <span class="n">beta_live_databases</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">core_db</span><span class="p">)</span>
                        <span class="k">break</span>
        <span class="k">return</span> <span class="n">beta_live_databases</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Mode must be either &#39;rapid&#39; or &#39;beta&#39;&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.live_tracking.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.tracking.live_tracking.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/live_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Check rapid or beta metadata databases.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--mode&quot;</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;rapid&quot;</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">],</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Choose either &#39;rapid&#39; or &#39;beta&#39;&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="n">live_databases</span> <span class="o">=</span> <span class="n">clean_server</span><span class="p">(</span><span class="n">config</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">mode</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Here is a list of databases on genebuild-prod-1 that can also be found in the </span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">mode</span><span class="si">}</span><span class="s2"> metadata:&quot;</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">live_databases</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.tracking.live_tracking.mysql_fetch_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">mysql_fetch_data</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">port</span><span class="p">,</span> <span class="n">user</span><span class="p">)</span></code>

<a href="#ensembl.genes.tracking.live_tracking.mysql_fetch_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Executes a given SQL query on a MySQL database and fetches the result.</p>
<p>This function establishes a connection to the MySQL database using provided connection details.
It then executes the given query using a cursor obtained from the connection. After executing the query,
it fetches all the rows of the query result and returns them. The function handles any errors that might
occur during the process and ensures that the database connection is closed before returning the result.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>query</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The SQL query to be executed.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>database</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the database to connect to.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>host</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The host name or IP address of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>port</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The port number to use for the connection.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>user</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The username to use for the database connection.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>tuple</code></td>            <td>
                  <code><span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[<span title="typing.Any">Any</span>, ...]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A tuple of tuples containing the rows returned by the query execution.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Note</summary>
  <p>This function does not handle database password authentication. Ensure that the provided user
has the necessary permissions and that the database is configured to allow password-less connections
from the given host.</p>
</details>

            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/tracking/live_tracking.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mysql_fetch_data</span><span class="p">(</span>
    <span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">port</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">user</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="o">...</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Executes a given SQL query on a MySQL database and fetches the result.</span>

<span class="sd">    This function establishes a connection to the MySQL database using provided connection details.</span>
<span class="sd">    It then executes the given query using a cursor obtained from the connection. After executing the query,</span>
<span class="sd">    it fetches all the rows of the query result and returns them. The function handles any errors that might</span>
<span class="sd">    occur during the process and ensures that the database connection is closed before returning the result.</span>

<span class="sd">    Args:</span>
<span class="sd">        query (str): The SQL query to be executed.</span>
<span class="sd">        database (str): The name of the database to connect to.</span>
<span class="sd">        host (str): The host name or IP address of the MySQL server.</span>
<span class="sd">        port (int): The port number to use for the connection.</span>
<span class="sd">        user (str): The username to use for the database connection.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple of tuples containing the rows returned by the query execution.</span>

<span class="sd">    Note:</span>
<span class="sd">        This function does not handle database password authentication. Ensure that the provided user</span>
<span class="sd">        has the necessary permissions and that the database is configured to allow password-less connections</span>
<span class="sd">        from the given host.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="n">port</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="n">cursor</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
        <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>

        <span class="n">cursor</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">result</span>

    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">Error</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error: </span><span class="si">{</span><span class="n">err</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">cursor</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">return</span> <span class="p">[]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="transcriptomic-data">Transcriptomic Data<a class="headerlink" href="#transcriptomic-data" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.transcriptomic_data.get_transcriptomic_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.transcriptomic_data.get_transcriptomic_data</code>


<a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Download short and long read data from ENA website</p>










<div class="doc doc-children">









<div class="doc doc-object doc-class">



<h3 id="ensembl.genes.transcriptomic_data.get_transcriptomic_data.InputSchema" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>            <code>InputSchema</code>


<a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data.InputSchema" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="argparse.ArgumentParser">ArgumentParser</span></code></p>



        <p>Input arguments</p>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/get_transcriptomic_data.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">InputSchema</span><span class="p">(</span><span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Input arguments&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-t&quot;</span><span class="p">,</span> <span class="s2">&quot;--taxon_id&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxon id&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--tree&quot;</span><span class="p">,</span>
            <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
            <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Turn on the &#39;Include subordinate taxa&#39; option in your query to ENA&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;-f&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--csv_file&quot;</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
            <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output file path (csv format)&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;-r&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--read_type&quot;</span><span class="p">,</span>
            <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;short&quot;</span><span class="p">,</span> <span class="s2">&quot;long&quot;</span><span class="p">],</span>
            <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Specify the type of transcriptomic data to download [&#39;short&#39;, &#39;long&#39;]&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;-l&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--limit&quot;</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
            <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The number of runs to be included in your csv file - consider that 1 run = 2 files, so setting &#39;-l 50&#39; will result in 100 lines in your csv (WARNING: this limit does not consider quality of data, it is a simple subsampling)&quot;</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">












  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.get_transcriptomic_data.get_data_from_ena" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_data_from_ena</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">read_type</span><span class="p">,</span> <span class="n">tree</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data.get_data_from_ena" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Query ENA API to get short or long read data</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/get_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_data_from_ena</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">read_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">tree</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Query ENA API to get short or long read data&quot;&quot;&quot;</span>
    <span class="n">csv_data</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="n">tree</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;tax_tree(</span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">)&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;tax_eq(</span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">)&quot;</span>

    <span class="k">if</span> <span class="n">read_type</span> <span class="o">==</span> <span class="s2">&quot;short&quot;</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">+=</span> <span class="s2">&quot; AND instrument_platform=ILLUMINA AND library_layout=PAIRED&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">+=</span> <span class="s2">&quot; AND (instrument_platform=OXFORD_NANOPORE OR instrument_platform=PACBIO_SMRT)&quot;</span>

    <span class="n">query</span> <span class="o">+=</span> <span class="s2">&quot; AND library_source=TRANSCRIPTOMIC&quot;</span>

    <span class="n">search_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/portal/api/search?display=report&amp;query=</span><span class="si">{</span><span class="n">query</span><span class="si">}</span><span class="s2">&amp;domain=read&amp;result=read_run&amp;fields=sample_accession,run_accession,fastq_ftp,read_count,instrument_platform,fastq_md5&quot;</span>
    <span class="n">search_result</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">search_url</span><span class="p">)</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">search_result</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="n">is_paired</span> <span class="o">=</span> <span class="s2">&quot;1&quot;</span>
    <span class="n">is_mate_1</span> <span class="o">=</span> <span class="s2">&quot;-1&quot;</span>
    <span class="n">read_length</span> <span class="o">=</span> <span class="s2">&quot;1&quot;</span>
    <span class="n">is_plus_13</span> <span class="o">=</span> <span class="s2">&quot;0&quot;</span>
    <span class="n">centre</span> <span class="o">=</span> <span class="s2">&quot;ENA&quot;</span>

    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span> <span class="o">==</span> <span class="mi">6</span><span class="p">:</span>
            <span class="n">row_data</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">sample_accession</span> <span class="o">=</span> <span class="n">row_data</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">run_accession</span> <span class="o">=</span> <span class="n">row_data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

            <span class="n">multi_samples</span> <span class="o">=</span> <span class="n">sample_accession</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">multi_samples</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">sample</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;multiple&quot;</span><span class="p">,</span> <span class="s2">&quot;multiple&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">sample</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="n">get_sample_info</span><span class="p">(</span><span class="n">sample_accession</span><span class="p">)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="n">read_count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row_data</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">instrument_platform</span> <span class="o">=</span> <span class="n">row_data</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="n">read_count</span> <span class="o">=</span> <span class="s2">&quot;0&quot;</span>
                <span class="n">instrument_platform</span> <span class="o">=</span> <span class="n">row_data</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>

            <span class="n">file_entries</span> <span class="o">=</span> <span class="n">row_data</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
            <span class="n">md5_entries</span> <span class="o">=</span> <span class="n">row_data</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_entries</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">instrument_platform</span> <span class="o">==</span> <span class="s2">&quot;ILLUMINA&quot;</span><span class="p">:</span>
                <span class="k">pass</span>  <span class="c1"># Nothing special to do, continue as normal</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_entries</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">instrument_platform</span> <span class="o">==</span> <span class="s2">&quot;PACBIO_SMRT&quot;</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_entries</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span> <span class="ow">and</span> <span class="n">instrument_platform</span> <span class="o">==</span> <span class="s2">&quot;ILLUMINA&quot;</span><span class="p">:</span>
                <span class="c1"># Identify the unwanted entry and remove it</span>
                <span class="n">file_entries</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">file_entries</span> <span class="k">if</span> <span class="s2">&quot;_1.fastq.gz&quot;</span> <span class="ow">in</span> <span class="n">f</span> <span class="ow">or</span> <span class="s2">&quot;_2.fastq.gz&quot;</span> <span class="ow">in</span> <span class="n">f</span>
                <span class="p">]</span>
                <span class="c1"># Assuming the order of md5 corresponds to files and the unwanted file is in the middle</span>
                <span class="n">md5_entries</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">md5_entries</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">row_data</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">))</span>
                    <span class="k">if</span> <span class="s2">&quot;_1.fastq.gz&quot;</span> <span class="ow">in</span> <span class="n">f</span> <span class="ow">or</span> <span class="s2">&quot;_2.fastq.gz&quot;</span> <span class="ow">in</span> <span class="n">f</span>
                <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="s2">&quot;Warning: Unexpected number of file entries, skipping &quot;</span>
                    <span class="o">+</span> <span class="n">run_accession</span>
                <span class="p">)</span>
                <span class="nb">next</span>  <span class="c1"># Skip further processing for this row</span>

            <span class="c1"># Only proceed if we have exactly 2 entries after any necessary filtering</span>
            <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">file_entries</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">instrument_platform</span> <span class="o">==</span> <span class="s2">&quot;PACBIO_SMRT&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">file_entries</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">instrument_platform</span> <span class="o">==</span> <span class="s2">&quot;ILLUMINA&quot;</span>
            <span class="p">):</span>
                <span class="c1"># if &quot;ftp&quot; in row_data[2] and &quot;;&quot; in row_data[5]:</span>
                <span class="k">for</span> <span class="n">file</span><span class="p">,</span> <span class="n">md5_file</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">file_entries</span><span class="p">,</span> <span class="n">md5_entries</span><span class="p">):</span>
                    <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                    <span class="n">md5_file_value</span> <span class="o">=</span> <span class="n">md5_file</span>
                    <span class="n">csv_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="p">(</span>
                            <span class="n">sample</span><span class="p">,</span>
                            <span class="n">run_accession</span><span class="p">,</span>
                            <span class="n">is_paired</span><span class="p">,</span>
                            <span class="n">file_path</span><span class="p">,</span>
                            <span class="n">is_mate_1</span><span class="p">,</span>
                            <span class="n">read_length</span><span class="p">,</span>
                            <span class="n">is_plus_13</span><span class="p">,</span>
                            <span class="n">centre</span><span class="p">,</span>
                            <span class="n">instrument_platform</span><span class="p">,</span>
                            <span class="n">description</span><span class="p">,</span>
                            <span class="n">file</span><span class="p">,</span>
                            <span class="n">md5_file_value</span><span class="p">,</span>
                        <span class="p">)</span>
                    <span class="p">)</span>

    <span class="k">return</span> <span class="n">csv_data</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.get_transcriptomic_data.get_sample_info" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">get_sample_info</span><span class="p">(</span><span class="n">accession</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data.get_sample_info" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Get info about sample name and description for the run accession</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/get_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_sample_info</span><span class="p">(</span><span class="n">accession</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get info about sample name and description for the run accession&quot;&quot;&quot;</span>
    <span class="n">biosample_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/biosamples/samples/</span><span class="si">{</span><span class="n">accession</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">biosample_url</span><span class="p">)</span>
        <span class="n">response</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>  <span class="c1"># Raise an HTTPError if the request was not successful</span>

        <span class="n">biosample_data</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>

        <span class="k">if</span> <span class="p">(</span>
            <span class="s2">&quot;characteristics&quot;</span> <span class="ow">in</span> <span class="n">biosample_data</span>
            <span class="ow">and</span> <span class="s2">&quot;tissue&quot;</span> <span class="ow">in</span> <span class="n">biosample_data</span><span class="p">[</span><span class="s2">&quot;characteristics&quot;</span><span class="p">]</span>
        <span class="p">):</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="n">biosample_data</span><span class="p">[</span><span class="s2">&quot;characteristics&quot;</span><span class="p">][</span><span class="s2">&quot;tissue&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;text&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="k">elif</span> <span class="p">(</span>
            <span class="s2">&quot;characteristics&quot;</span> <span class="ow">in</span> <span class="n">biosample_data</span>
            <span class="ow">and</span> <span class="s2">&quot;organism_part&quot;</span> <span class="ow">in</span> <span class="n">biosample_data</span><span class="p">[</span><span class="s2">&quot;characteristics&quot;</span><span class="p">]</span>
        <span class="p">):</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="n">biosample_data</span><span class="p">[</span><span class="s2">&quot;characteristics&quot;</span><span class="p">][</span><span class="s2">&quot;organism_part&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span>
                <span class="s2">&quot;text&quot;</span>
            <span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="n">accession</span>

        <span class="k">if</span> <span class="p">(</span>
            <span class="s2">&quot;characteristics&quot;</span> <span class="ow">in</span> <span class="n">biosample_data</span>
            <span class="ow">and</span> <span class="s2">&quot;description&quot;</span> <span class="ow">in</span> <span class="n">biosample_data</span><span class="p">[</span><span class="s2">&quot;characteristics&quot;</span><span class="p">]</span>
        <span class="p">):</span>
            <span class="n">description</span> <span class="o">=</span> <span class="n">biosample_data</span><span class="p">[</span><span class="s2">&quot;characteristics&quot;</span><span class="p">][</span><span class="s2">&quot;description&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;text&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">description</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">250</span><span class="p">:</span>
                <span class="n">description</span> <span class="o">=</span> <span class="n">accession</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">description</span> <span class="o">=</span> <span class="n">accession</span>

        <span class="n">sample</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[ ;\(\)\/</span><span class="se">\\</span><span class="s2">]&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="p">)</span>
        <span class="c1"># remove punctuation</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[!</span><span class="se">\&quot;</span><span class="s2">#$%&amp;()*\+,\-\&#39;.\/:;&lt;=&gt;?@\[\]^`{|}~]&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="p">)</span>
        <span class="n">multi_tissuesREGEX</span> <span class="o">=</span> <span class="s2">&quot;([a-zA-Z]+\,)+&quot;</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">multi_tissuesREGEX</span><span class="p">,</span> <span class="n">sample</span><span class="p">):</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="s2">&quot;mixed_tissues&quot;</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">description</span><span class="p">)</span>

    <span class="k">except</span> <span class="p">(</span><span class="n">RequestException</span><span class="p">,</span> <span class="n">HTTPError</span><span class="p">,</span> <span class="ne">ConnectionError</span><span class="p">,</span> <span class="n">Timeout</span><span class="p">)</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;An error occurred while fetching data from </span><span class="si">{</span><span class="n">biosample_url</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># Handle the error here, you can log it or take other appropriate actions.</span>
        <span class="k">return</span> <span class="p">(</span><span class="s2">&quot;unknown&quot;</span><span class="p">,</span> <span class="s2">&quot;unknown&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.get_transcriptomic_data.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.transcriptomic_data.get_transcriptomic_data.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Entrypoint</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/get_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Entrypoint&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">InputSchema</span><span class="p">()</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">csv_file</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">csv_file</span><span class="p">)</span><span class="o">.</span><span class="n">st_size</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="s2">&quot;File &quot;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">csv_file</span> <span class="o">+</span> <span class="s2">&quot; exists, and is not empty, will not overwrite!&quot;</span>
        <span class="p">)</span>
        <span class="n">exit</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">csv_data</span> <span class="o">=</span> <span class="n">get_data_from_ena</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">read_type</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">tree</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">limit</span><span class="p">:</span>
                <span class="n">csv_data</span> <span class="o">=</span> <span class="n">csv_data</span><span class="p">[:</span> <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">limit</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)]</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">csv_file</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">csv_file</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_data</span><span class="p">:</span>
                    <span class="n">csv_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="p">(</span><span class="n">RequestException</span><span class="p">,</span> <span class="n">HTTPError</span><span class="p">,</span> <span class="ne">ConnectionError</span><span class="p">,</span> <span class="n">Timeout</span><span class="p">)</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;An error occurred during the data retrieval process: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.transcriptomic_data.check_for_transcriptomic" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.transcriptomic_data.check_for_transcriptomic</code>


<a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Check the availability for short and long read data from ENA website given a taxon id</p>










<div class="doc doc-children">









<div class="doc doc-object doc-class">



<h3 id="ensembl.genes.transcriptomic_data.check_for_transcriptomic.InputSchema" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-class"></code>            <code>InputSchema</code>


<a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic.InputSchema" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="argparse.ArgumentParser">ArgumentParser</span></code></p>



        <p>Input arguments</p>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/check_for_transcriptomic.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">InputSchema</span><span class="p">(</span><span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Input arguments&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-t&quot;</span><span class="p">,</span> <span class="s2">&quot;--taxon_id&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxon id&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--tree&quot;</span><span class="p">,</span>
            <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
            <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Turn on the &#39;Include subordinate taxa&#39; option in your query to ENA&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
            <span class="s2">&quot;--file&quot;</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
            <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to the file containing a list of taxon ids&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--output_dir&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output directory path&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">












  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.check_for_transcriptomic.check_data_from_ena" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">check_data_from_ena</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">,</span> <span class="n">tree</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic.check_data_from_ena" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Query ENA API to get short or long read data</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/check_for_transcriptomic.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_data_from_ena</span><span class="p">(</span>  <span class="c1"># pylint: disable=too-many-locals</span>
    <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">tree</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Query ENA API to get short or long read data&quot;&quot;&quot;</span>

    <span class="n">TEXT_FORMAT</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;BOLD&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="se">\033</span><span class="s2">[1m&quot;</span><span class="p">,</span> <span class="s2">&quot;UNDERLINE&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="se">\033</span><span class="s2">[4m&quot;</span><span class="p">,</span> <span class="s2">&quot;END&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="se">\033</span><span class="s2">[0m&quot;</span><span class="p">}</span>
    <span class="k">if</span> <span class="n">tree</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;tax_tree(</span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">)&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;tax_eq(</span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">)&quot;</span>

    <span class="n">query_short_paired</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">query</span>
        <span class="o">+</span> <span class="s2">&quot; AND instrument_platform=ILLUMINA AND library_layout=PAIRED&quot;</span>
        <span class="o">+</span> <span class="s2">&quot; AND library_source=TRANSCRIPTOMIC&quot;</span>
    <span class="p">)</span>
    <span class="n">query_short_single</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">query</span>
        <span class="o">+</span> <span class="s2">&quot; AND instrument_platform=ILLUMINA AND library_layout=SINGLE&quot;</span>
        <span class="o">+</span> <span class="s2">&quot; AND library_source=TRANSCRIPTOMIC&quot;</span>
    <span class="p">)</span>
    <span class="n">query_pacbio</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">query</span> <span class="o">+</span> <span class="s2">&quot; AND instrument_platform=PACBIO_SMRT AND library_source=TRANSCRIPTOMIC&quot;</span>
    <span class="p">)</span>
    <span class="n">query_onp</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">query</span>
        <span class="o">+</span> <span class="s2">&quot; AND instrument_platform=OXFORD_NANOPORE AND library_source=TRANSCRIPTOMIC&quot;</span>
    <span class="p">)</span>

    <span class="n">short_paired_runs</span> <span class="o">=</span> <span class="n">ena_rest_api</span><span class="p">(</span><span class="n">query_short_paired</span><span class="p">)</span>
    <span class="n">short_single_runs</span> <span class="o">=</span> <span class="n">ena_rest_api</span><span class="p">(</span><span class="n">query_short_single</span><span class="p">)</span>
    <span class="n">pacbio_read_runs</span> <span class="o">=</span> <span class="n">ena_rest_api</span><span class="p">(</span><span class="n">query_pacbio</span><span class="p">)</span>
    <span class="n">onp_read_runs</span> <span class="o">=</span> <span class="n">ena_rest_api</span><span class="p">(</span><span class="n">query_onp</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span>
        <span class="n">TEXT_FORMAT</span><span class="p">[</span><span class="s2">&quot;BOLD&quot;</span><span class="p">]</span>
        <span class="o">+</span> <span class="s2">&quot;Short-read paired-end illumina data available! &quot;</span>
        <span class="o">+</span> <span class="n">TEXT_FORMAT</span><span class="p">[</span><span class="s2">&quot;END&quot;</span><span class="p">]</span>
        <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">short_paired_runs</span><span class="si">}</span><span class="s2"> runs.&quot;</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="n">TEXT_FORMAT</span><span class="p">[</span><span class="s2">&quot;BOLD&quot;</span><span class="p">]</span>
        <span class="o">+</span> <span class="s2">&quot;Short-read single-end illumina data available! &quot;</span>
        <span class="o">+</span> <span class="n">TEXT_FORMAT</span><span class="p">[</span><span class="s2">&quot;END&quot;</span><span class="p">]</span>
        <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">short_single_runs</span><span class="si">}</span><span class="s2"> runs.&quot;</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="n">TEXT_FORMAT</span><span class="p">[</span><span class="s2">&quot;BOLD&quot;</span><span class="p">]</span>
        <span class="o">+</span> <span class="s2">&quot;Long-read PacBio data available! &quot;</span>
        <span class="o">+</span> <span class="n">TEXT_FORMAT</span><span class="p">[</span><span class="s2">&quot;END&quot;</span><span class="p">]</span>
        <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">pacbio_read_runs</span><span class="si">}</span><span class="s2"> runs.&quot;</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="n">TEXT_FORMAT</span><span class="p">[</span><span class="s2">&quot;BOLD&quot;</span><span class="p">]</span>
        <span class="o">+</span> <span class="s2">&quot;Long_read ONP data available! &quot;</span>
        <span class="o">+</span> <span class="n">TEXT_FORMAT</span><span class="p">[</span><span class="s2">&quot;END&quot;</span><span class="p">]</span>
        <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">onp_read_runs</span><span class="si">}</span><span class="s2"> runs.&quot;</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="p">{</span>
        <span class="s2">&quot;Short-read paired-end illumina&quot;</span><span class="p">:</span> <span class="n">short_paired_runs</span><span class="p">,</span>
        <span class="s2">&quot;Short-read single-end illumina&quot;</span><span class="p">:</span> <span class="n">short_single_runs</span><span class="p">,</span>
        <span class="s2">&quot;Long-read PacBio&quot;</span><span class="p">:</span> <span class="n">pacbio_read_runs</span><span class="p">,</span>
        <span class="s2">&quot;Long-read ONP&quot;</span><span class="p">:</span> <span class="n">onp_read_runs</span><span class="p">,</span>
    <span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.check_for_transcriptomic.ena_rest_api" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">ena_rest_api</span><span class="p">(</span><span class="n">query</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic.ena_rest_api" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Call to ENA API</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/check_for_transcriptomic.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">ena_rest_api</span><span class="p">(</span><span class="n">query</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Call to ENA API&quot;&quot;&quot;</span>
    <span class="n">search_url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/portal/api/search?display=report&amp;query=</span><span class="si">{</span><span class="n">query</span><span class="si">}</span><span class="s2">&amp;domain=read&amp;result=read_run&amp;fields=sample_accession,run_accession,fastq_ftp,read_count,instrument_platform&quot;</span>  <span class="c1"># pylint: disable=line-too-long</span>
    <span class="n">search_result</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">search_url</span><span class="p">)</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">search_result</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.check_for_transcriptomic.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.transcriptomic_data.check_for_transcriptomic.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Entrypoint</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/check_for_transcriptomic.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Entrypoint&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">InputSchema</span><span class="p">()</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">output_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span>
    <span class="n">output_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">taxon_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="p">]</span>  <span class="c1"># Start with the single taxon_id</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">file</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">input_file</span><span class="p">:</span>
            <span class="n">taxon_ids</span> <span class="o">=</span> <span class="n">input_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>

    <span class="c1"># Prepare CSV file</span>
    <span class="n">csv_file</span> <span class="o">=</span> <span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;taxon_summary.csv&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">csv_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

        <span class="c1"># Write the header row</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="s2">&quot;Taxon ID&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Short-read paired-end illumina&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Short-read single-end illumina&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Long-read PacBio&quot;</span><span class="p">,</span>
                <span class="s2">&quot;Long-read ONP&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">taxon_id</span> <span class="ow">in</span> <span class="n">taxon_ids</span><span class="p">:</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">check_data_from_ena</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">),</span> <span class="n">args</span><span class="o">.</span><span class="n">tree</span><span class="p">)</span>
            <span class="n">writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="n">taxon_id</span><span class="p">,</span>
                    <span class="n">results</span><span class="p">[</span><span class="s2">&quot;Short-read paired-end illumina&quot;</span><span class="p">],</span>
                    <span class="n">results</span><span class="p">[</span><span class="s2">&quot;Short-read single-end illumina&quot;</span><span class="p">],</span>
                    <span class="n">results</span><span class="p">[</span><span class="s2">&quot;Long-read PacBio&quot;</span><span class="p">],</span>
                    <span class="n">results</span><span class="p">[</span><span class="s2">&quot;Long-read ONP&quot;</span><span class="p">],</span>
                <span class="p">]</span>
            <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-module"></code>            <code>ensembl.genes.transcriptomic_data.select_transcriptomic_data</code>


<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Select assessed data from the database and create a report.
This script connects to a MySQL database, retrieves data related to sequencing runs,
applies quality checks using FastQC and STAR, and generates a report summarizing the results.</p>










<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.check_fastqc_star_quality" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">check_fastqc_star_quality</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.check_fastqc_star_quality" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Apply FastQC and STAR quality to each row
Args:
    df (pd.DataFrame): DataFrame containing the data to be processed.
Returns:
    pd.DataFrame: DataFrame with additional columns for FastQC and STAR quality.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_fastqc_star_quality</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Apply FastQC and STAR quality to each row</span>
<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): DataFrame containing the data to be processed.</span>
<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: DataFrame with additional columns for FastQC and STAR quality.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Apply FastQC and STAR quality to each row</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;fastqc_pass&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">fastqc_quality</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;star_quality&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">star_quality</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Assign final quality labels</span>

    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">)[</span><span class="s2">&quot;fastqc_pass&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;Failed&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;Passed&quot;</span>
    <span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">)[</span><span class="s2">&quot;star_quality&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;Failed&quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;Passed&quot;</span>
    <span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span>
    <span class="p">)</span>
    <span class="c1"># print(df.head(10))</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.clean_repeated_words" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">clean_repeated_words</span><span class="p">(</span><span class="n">text</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.clean_repeated_words" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Remove consecutive duplicate words from a string."</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>text</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The input string from which to remove</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A string with consecutive duplicate words removed,</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>preserving the first occurrence of each word.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">clean_repeated_words</span><span class="p">(</span><span class="n">text</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove consecutive duplicate words from a string.&quot;</span>

<span class="sd">    Args:</span>
<span class="sd">        text (str): The input string from which to remove</span>
<span class="sd">        consecutive duplicate words.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: A string with consecutive duplicate words removed,</span>
<span class="sd">        preserving the first occurrence of each word.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">words</span> <span class="o">=</span> <span class="n">text</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">words</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span>
    <span class="c1"># Keep only the first occurrence, skip consecutive duplicates</span>
    <span class="n">cleaned_words</span> <span class="o">=</span> <span class="p">[</span><span class="n">words</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="k">for</span> <span class="n">w</span> <span class="ow">in</span> <span class="n">words</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="k">if</span> <span class="n">w</span> <span class="o">!=</span> <span class="n">cleaned_words</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">cleaned_words</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">w</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cleaned_words</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.connect_to_db" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">connect_to_db</span><span class="p">(</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="mi">3306</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.connect_to_db" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Establishes a connection to the MySQL database.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>host</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The hostname or IP address of the MySQL server.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>user</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The username to connect to the database.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>password</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The password for the user.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>db</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the database to connect to.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>port</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The port of the MySQL server (default is 3306).</p>
              </div>
            </td>
            <td>
                  <code>3306</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    A pymysql connection object.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">connect_to_db</span><span class="p">(</span>
    <span class="n">host</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">user</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">password</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">database</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="mi">3306</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connections</span><span class="o">.</span><span class="n">Connection</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Establishes a connection to the MySQL database.</span>

<span class="sd">    Args:</span>
<span class="sd">        host: The hostname or IP address of the MySQL server.</span>
<span class="sd">        user: The username to connect to the database.</span>
<span class="sd">        password: The password for the user.</span>
<span class="sd">        db: The name of the database to connect to.</span>
<span class="sd">        port: The port of the MySQL server (default is 3306).</span>
<span class="sd">    Returns:</span>
<span class="sd">        A pymysql connection object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">connection</span> <span class="o">=</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span>
            <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">user</span><span class="o">=</span><span class="n">user</span><span class="p">,</span> <span class="n">password</span><span class="o">=</span><span class="n">password</span><span class="p">,</span> <span class="n">database</span><span class="o">=</span><span class="n">database</span><span class="p">,</span> <span class="n">port</span><span class="o">=</span><span class="n">port</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">connection</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">MySQLError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error connecting to MySQL: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.create_report" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">create_report</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">tissue_report_file</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.create_report" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Create a report based on the DataFrame.
This function summarizes the FastQC and STAR quality results for each taxon_id and run_accession.
It calculates the number of runs that passed both quality checks and the percentage of runs that passed.
The report is printed to the console.
1. Group the DataFrame by taxon_id and run_accession.
2. Aggregate the final_fastqc_status and final_star_status columns.
3. Create a new column passed_both to indicate if both quality checks passed.
4. Group the DataFrame by taxon_id and aggregate the results.
5. Calculate the percentage of runs that passed both quality checks.
6. Print the report to the console.
7. Return the summarized DataFrame.
8. The report includes the following columns:
    - taxon_id: The taxon ID.
    - passed_fastqc: The number of runs that passed FastQC.
    - passed_star: The number of runs that passed STAR.
    - passed_both: The number of runs that passed both quality checks.
    - total_runs: The total number of runs for the taxon_id.
    - percentage_passed_both: The percentage of runs that passed both quality checks.
Args:
    df (pd.DataFrame): DataFrame containing the data to be processed.
Returns:
    pd.DataFrame: DataFrame with the summarized report.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_report</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">tissue_report_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a report based on the DataFrame.</span>
<span class="sd">    This function summarizes the FastQC and STAR quality results for each taxon_id and run_accession.</span>
<span class="sd">    It calculates the number of runs that passed both quality checks and the percentage of runs that passed.</span>
<span class="sd">    The report is printed to the console.</span>
<span class="sd">    1. Group the DataFrame by taxon_id and run_accession.</span>
<span class="sd">    2. Aggregate the final_fastqc_status and final_star_status columns.</span>
<span class="sd">    3. Create a new column passed_both to indicate if both quality checks passed.</span>
<span class="sd">    4. Group the DataFrame by taxon_id and aggregate the results.</span>
<span class="sd">    5. Calculate the percentage of runs that passed both quality checks.</span>
<span class="sd">    6. Print the report to the console.</span>
<span class="sd">    7. Return the summarized DataFrame.</span>
<span class="sd">    8. The report includes the following columns:</span>
<span class="sd">        - taxon_id: The taxon ID.</span>
<span class="sd">        - passed_fastqc: The number of runs that passed FastQC.</span>
<span class="sd">        - passed_star: The number of runs that passed STAR.</span>
<span class="sd">        - passed_both: The number of runs that passed both quality checks.</span>
<span class="sd">        - total_runs: The total number of runs for the taxon_id.</span>
<span class="sd">        - percentage_passed_both: The percentage of runs that passed both quality checks.</span>
<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): DataFrame containing the data to be processed.</span>
<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: DataFrame with the summarized report.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Collapse to one row per run_accession</span>
    <span class="n">run_summary</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span> <span class="s2">&quot;run_accession&quot;</span><span class="p">])</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
            <span class="n">final_fastqc_status</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="s2">&quot;first&quot;</span><span class="p">),</span>
            <span class="n">final_star_status</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">,</span> <span class="s2">&quot;first&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">run_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">run_summary</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span>
        <span class="n">run_summary</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span>
    <span class="p">)</span>
    <span class="c1"># REPORT</span>
    <span class="n">taxon_pass_summary</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">run_summary</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
            <span class="n">passed_fastqc</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
            <span class="n">passed_star</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
            <span class="n">passed_both</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;passed_both&quot;</span><span class="p">,</span> <span class="s2">&quot;sum&quot;</span><span class="p">),</span>
            <span class="n">total_runs</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">,</span> <span class="s2">&quot;nunique&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="c1"># Compute percentage outside of .agg()</span>
    <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;percentage_passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;total_runs&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_fastqc_only&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_fastqc&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_star_only&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_star&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">taxon_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="c1"># print(taxon_pass_summary.head())</span>
    <span class="c1"># print(&quot;\n=== Run Quality Summary by Taxon ===&quot;)</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">taxon_pass_summary</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Taxon ID: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">])</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;Total Runs: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;total_runs&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;Passed BOTH: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_both&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="c1"># f&quot;FastQC Passed: {row[&#39;passed_fastqc&#39;]} | &quot;</span>
            <span class="c1"># f&quot;STAR Passed: {row[&#39;passed_star&#39;]} | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;🧪 FastQC Only: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_fastqc_only&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;🚀 STAR Only: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_star_only&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;✔️ Both %: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;percentage_passed_both&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">%&quot;</span>
        <span class="p">)</span>
        <span class="c1"># Prepare tissue-level summary</span>
    <span class="k">if</span> <span class="s2">&quot;tissue_prediction&quot;</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">run_tissue_summary</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span> <span class="s2">&quot;tissue_prediction&quot;</span><span class="p">,</span> <span class="s2">&quot;run_accession&quot;</span><span class="p">])</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
                <span class="n">final_fastqc_status</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="s2">&quot;first&quot;</span><span class="p">),</span>
                <span class="n">final_star_status</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">,</span> <span class="s2">&quot;first&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="n">run_tissue_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">run_tissue_summary</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span>
        <span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">run_tissue_summary</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span>

        <span class="n">tissue_pass_summary</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">run_tissue_summary</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span> <span class="s2">&quot;tissue_prediction&quot;</span><span class="p">])</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
                <span class="n">passed_fastqc</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
                <span class="n">passed_star</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;Passed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
                <span class="n">passed_both</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;passed_both&quot;</span><span class="p">,</span> <span class="s2">&quot;sum&quot;</span><span class="p">),</span>
                <span class="n">total_runs</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">,</span> <span class="s2">&quot;nunique&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;percentage_passed_both&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;total_runs&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>

        <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_fastqc_only&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_fastqc&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_star_only&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_star&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">tissue_pass_summary</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="c1"># Save to text file</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">tissue_report_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">tissue_pass_summary</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Taxon ID: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">])</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;Tissue: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;tissue_prediction&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;Total Runs: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;total_runs&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;Passed BOTH: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_both&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="c1"># f&quot;FastQC Passed: {row[&#39;passed_fastqc&#39;]} | &quot;</span>
                    <span class="c1"># f&quot;STAR Passed: {row[&#39;passed_star&#39;]} | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;FastQC Only: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_fastqc_only&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;STAR Only: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;passed_star_only&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> | &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;BOTH %: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;percentage_passed_both&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">%</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.fastqc_quality" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">fastqc_quality</span><span class="p">(</span><span class="n">row</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.fastqc_quality" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate FastQC quality based on the criteria.
The function checks the FastQC quality criteria for each row and returns True if the criteria are met.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>row</code>
            </td>
            <td>
                  <code><span title="_type_">_type_</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe row</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the criteria are met, False otherwise.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fastqc_quality</span><span class="p">(</span><span class="n">row</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate FastQC quality based on the criteria.</span>
<span class="sd">    The function checks the FastQC quality criteria for each row and returns True if the criteria are met.</span>


<span class="sd">    Args:</span>
<span class="sd">        row (_type_): dataframe row</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the criteria are met, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fastqc_criteria</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;per_base_sequence_quality&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">,</span> <span class="s2">&quot;WARN&quot;</span><span class="p">],</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;overrepresented_sequences&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">,</span> <span class="s2">&quot;WARN&quot;</span><span class="p">],</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;per_base_n_content&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">,</span> <span class="s2">&quot;WARN&quot;</span><span class="p">],</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;per_sequence_quality_scores&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">,</span> <span class="s2">&quot;WARN&quot;</span><span class="p">],</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">fastqc_criteria</span><span class="p">)</span> <span class="o">==</span> <span class="mi">4</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.filter_data" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">filter_data</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.filter_data" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Filter the DataFrame to select tissue-annotated samples based on quality criteria.</p>
<p>Steps:
1. Remove duplicate run_accessions.
2. Filter out rows with null tissue predictions.
3. Exclude rows with tumor/cancer-related tissue labels.
4. Prioritize specific tissue types (e.g., brain, heart, lung).
5. Group by tissue and select samples up to 250M total reads.
6. If fewer than 500 samples selected, add more high-quality unannotated samples.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Input DataFrame with tissue predictions and QC info.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>debug</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, write intermediate CSVs for debugging.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[str]: List of selected run_accessions.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">filter_data</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter the DataFrame to select tissue-annotated samples based on quality criteria.</span>

<span class="sd">    Steps:</span>
<span class="sd">    1. Remove duplicate run_accessions.</span>
<span class="sd">    2. Filter out rows with null tissue predictions.</span>
<span class="sd">    3. Exclude rows with tumor/cancer-related tissue labels.</span>
<span class="sd">    4. Prioritize specific tissue types (e.g., brain, heart, lung).</span>
<span class="sd">    5. Group by tissue and select samples up to 250M total reads.</span>
<span class="sd">    6. If fewer than 500 samples selected, add more high-quality unannotated samples.</span>

<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): Input DataFrame with tissue predictions and QC info.</span>
<span class="sd">        debug (bool): If True, write intermediate CSVs for debugging.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[str]: List of selected run_accessions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;run_accession&quot;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>
    <span class="n">df_tissue</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">notnull</span><span class="p">()]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df_tissue</span> <span class="o">=</span> <span class="n">df_tissue</span><span class="p">[</span>
        <span class="o">~</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span>
            <span class="sa">r</span><span class="s2">&quot;tumor|cancer|melanoma|cancer/disease&quot;</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
    <span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Prioritise tissue types</span>
    <span class="n">priority_tissues</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;heart&quot;</span><span class="p">,</span>
        <span class="s2">&quot;lung&quot;</span><span class="p">,</span>
        <span class="s2">&quot;brain&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ovary&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ovaries&quot;</span><span class="p">,</span>
        <span class="s2">&quot;testes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;testis&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gonad&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gonads&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">df_tissue</span> <span class="o">=</span> <span class="n">prioritise_tissues</span><span class="p">(</span><span class="n">df_tissue</span><span class="p">,</span> <span class="n">priority_tissues</span><span class="p">)</span>
    <span class="n">df_tissue</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;tissueall.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">df_tissue</span> <span class="o">=</span> <span class="n">df_tissue</span><span class="p">[</span>
        <span class="o">~</span><span class="p">(</span>
            <span class="p">(</span><span class="o">~</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Failed&quot;</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;final_star_status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Failed&quot;</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">]</span>
    <span class="n">df_tissue</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;tissueless.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">df_no_tissue</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">run_accessions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Filter tissue-annotated samples</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">df_tissue</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">):</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">group</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
            <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;priority&quot;</span><span class="p">,</span> <span class="s2">&quot;passed_both&quot;</span><span class="p">,</span> <span class="s2">&quot;final_fastqc_status&quot;</span><span class="p">,</span> <span class="s2">&quot;final_star_status&quot;</span><span class="p">],</span>
            <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">group</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;group.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">total_length</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">group</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">total_length</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;total_sequences&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">250</span> <span class="o">*</span> <span class="mi">10</span><span class="o">**</span><span class="mi">6</span><span class="p">:</span>
                <span class="n">run_accessions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">])</span>
                <span class="n">total_length</span> <span class="o">+=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;total_sequences&quot;</span><span class="p">]</span>

    <span class="n">df_tissue</span><span class="p">[</span><span class="n">df_tissue</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">run_accessions</span><span class="p">)]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="s2">&quot;tissuedanio.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">run_accessions</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">500</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">df_no_tissue</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">):</span>
            <span class="n">total_length</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">group</span><span class="p">[</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;passed_both&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                    <span class="n">run_accessions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">run_accessions</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.main" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.main" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Module's entry-point.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Module&#39;s entry-point.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">prog</span><span class="o">=</span><span class="s2">&quot;llm_prediction.py&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Predict tissue using LLMs&quot;</span>
    <span class="p">)</span>

    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--taxon_id&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;10116&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Taxonomy ID&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--host&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;mysql-ens-genebuild-prod-1&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Host&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--user&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;ensadmin&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;User&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--password&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;ensembl&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Password&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--database&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;gb_transcriptomic_registry&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Database&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--port&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">4527</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Port&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--file_name&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output file name&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--read_type&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;short&quot;</span><span class="p">,</span> <span class="s2">&quot;long&quot;</span><span class="p">],</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;short&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Read type short/long&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--csv_for_main&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if true will produce the csv we need for main pipeline otherwise </span><span class="se">\</span>
<span class="s2">            the one for the full alignment pipeline&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="n">engine</span> <span class="o">=</span> <span class="n">create_engine</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;mysql+pymysql://</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">user</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">password</span><span class="si">}</span><span class="s2">@</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">host</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">port</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">database</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Load the rows to fix (e.g., 150k rows)</span>
    <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">read_type</span> <span class="o">==</span> <span class="s2">&quot;short&quot;</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;SELECT r.taxon_id, r.run_accession, r.sample_accession, r.platform, r.paired, </span><span class="se">\</span>
<span class="s2">            r.tissue_prediction, d.file_name, d.file_url,d.md5,d.per_base_sequence_quality, </span><span class="se">\</span>
<span class="s2">            d.per_sequence_quality_scores, d.per_base_n_content, d.overrepresented_sequences,</span><span class="se">\</span>
<span class="s2">            d.total_sequences, a.uniquely_mapped_reads_percentage, </span><span class="se">\</span>
<span class="s2">            a.percentage_reads_mapped_to_multiple_loci,a.percentage_reads_unmapped_too_short, </span><span class="se">\</span>
<span class="s2">            a.assembly_accession </span><span class="se">\</span>
<span class="s2">            FROM run r INNER JOIN data_files d ON r.run_id=d.run_id INNER JOIN align a ON r.run_id=a.run_id </span><span class="se">\</span>
<span class="s2">            WHERE  r.qc_status=&#39;ALIGNED&#39; AND r.paired=1 AND r.platform=&#39;ILLUMINA&#39; AND r.taxon_id = &quot;</span>
            <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">read_type</span> <span class="o">==</span> <span class="s2">&quot;long&quot;</span><span class="p">:</span>
        <span class="n">query</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;SELECT r.taxon_id, r.run_accession, r.sample_accession, r.platform, r.paired, </span><span class="se">\</span>
<span class="s2">            r.tissue_prediction, d.file_name, d.file_url,d.md5,d.per_base_sequence_quality, </span><span class="se">\</span>
<span class="s2">            d.per_sequence_quality_scores, d.per_base_n_content, d.overrepresented_sequences, </span><span class="se">\</span>
<span class="s2">            d.total_sequences, a.uniquely_mapped_reads_percentage, </span><span class="se">\</span>
<span class="s2">            a.percentage_reads_mapped_to_multiple_loci,a.percentage_reads_unmapped_too_short </span><span class="se">\</span>
<span class="s2">            FROM run r INNER JOIN data_files d ON r.run_id=d.run_id INNER JOIN align a ON r.run_id=a.run_id </span><span class="se">\</span>
<span class="s2">            WHERE  r.qc_status=&#39;ALIGNED&#39; AND r.paired=0 AND r.platform=&#39;PACBIO_SMRT&#39; AND r.taxon_id = &quot;</span>
            <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span>
        <span class="p">)</span>
    <span class="c1"># Connect to the database</span>
    <span class="c1"># db_connection = connect_to_db(**db_config)</span>
    <span class="c1"># Clean the input text</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># if db_connection:</span>
        <span class="c1"># df = pd.read_sql(query, db_connection)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_sql</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">engine</span><span class="p">)</span>
        <span class="c1"># print(f&quot;Loaded {len(df)} rows.&quot;)</span>
    <span class="k">except</span> <span class="n">pymysql</span><span class="o">.</span><span class="n">MySQLError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error connecting to MySQL: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">check_fastqc_star_quality</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="c1"># df[&quot;tissue_prediction&quot;] = df[&quot;tissue_prediction&quot;].apply(</span>
        <span class="c1">#    lambda x: re.search(r&quot;^(.*)&quot;, str(x)).group(1) if re.search(r&quot;^(.*)&quot;, str(x)) else x</span>
        <span class="c1"># )</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span>
        <span class="p">)</span>
        <span class="c1"># df[&quot;tissue_prediction&quot;] = df[&quot;tissue_prediction&quot;].apply(</span>
        <span class="c1">#    lambda x: (m.group(1) if (m := re.search(r&quot;^(.*)&quot;, str(x))) else x)</span>
        <span class="c1"># )</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span>
            <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Answer: &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Output: &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
                <span class="sa">r</span><span class="s2">&quot;[, ]+&quot;</span><span class="p">,</span>
                <span class="s2">&quot;_&quot;</span><span class="p">,</span>  <span class="c1"># replace space/comma with underscore</span>
                <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span>
                    <span class="sa">r</span><span class="s2">&quot;\b\d+\b|\bday\b|\b[a-zA-Z]\b&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;&quot;</span><span class="p">,</span>  <span class="c1"># remove numbers, &quot;day&quot;, and single letters</span>
                    <span class="n">clean_repeated_words</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)),</span>
                <span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span>
                <span class="s2">&quot;_&quot;</span>
            <span class="p">)</span>  <span class="c1"># remove leading/trailing underscores</span>
        <span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>

        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span>
            <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;NONE&quot;</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
            <span class="s2">&quot;tissue_prediction&quot;</span><span class="p">,</span>
        <span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span>
            <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
            <span class="s2">&quot;tissue_prediction&quot;</span><span class="p">,</span>
        <span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">create_report</span><span class="p">(</span>
            <span class="n">df</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">file_name</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">_tissue_summary.txt&quot;</span>
        <span class="p">)</span>
        <span class="c1"># print(df.head())</span>
        <span class="c1"># Remove duplicates based on &#39;run_accession&#39; while keeping the first row for each</span>
        <span class="n">df_original</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">run_accessions</span> <span class="o">=</span> <span class="n">filter_data</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="c1"># Build a regex pattern from run_accession list</span>
        <span class="n">selected_accessions</span> <span class="o">=</span> <span class="n">run_accessions</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">25000</span><span class="p">]</span>
        <span class="c1"># df_final = df_original[df_original[&quot;run_accession&quot;].isin(run_accessions[0:250])].copy()</span>
        <span class="n">df_final</span> <span class="o">=</span> <span class="n">df_original</span><span class="p">[</span>
            <span class="n">df_original</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">selected_accessions</span><span class="p">)</span>
        <span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">df_final</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
            <span class="n">df_final</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">],</span> <span class="n">categories</span><span class="o">=</span><span class="n">selected_accessions</span><span class="p">,</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="n">df_final</span> <span class="o">=</span> <span class="n">df_final</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
            <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;run_accession&quot;</span><span class="p">,</span> <span class="s2">&quot;file_name&quot;</span><span class="p">]</span>
        <span class="p">)</span>  <span class="c1"># &quot;run_accession&quot;)</span>

        <span class="n">df_final</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df_final</span><span class="p">[</span><span class="s2">&quot;predicted_tissue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_final</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="p">(</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;sample_accession&quot;</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">])</span>
                <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># print(df_final.head())</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">df_final</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">csv_for_main</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">df_final</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;file_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">df_final</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.fastq.gz&quot;</span>
                <span class="p">)</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;col1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;col_1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;col0&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">df_final</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;ENA&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;ENA&quot;</span>

                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">file_name</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">df_final</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                        <span class="n">f</span><span class="p">,</span>
                        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                        <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                        <span class="n">columns</span><span class="o">=</span><span class="p">[</span>
                            <span class="s2">&quot;predicted_tissue&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;run_accession&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;col1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;file_name&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;col_1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;col1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;col0&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;ENA&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;platform&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;sample_accession&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;file_url&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;md5&quot;</span><span class="p">,</span>
                        <span class="p">],</span>
                        <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                    <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># taxon_id,gca,platform,paired,tissue,run_accession,pair1,md5_1,pair2,md5_2</span>
                <span class="n">output_df</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">df_final</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;run_accession&quot;</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                        <span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
                            <span class="p">{</span>
                                <span class="s2">&quot;taxon_id&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;taxon_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;assembly_accession&quot;</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
                                    <span class="s2">&quot;assembly_accession&quot;</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="s2">&quot;NA&quot;</span><span class="p">])</span>
                                <span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span>
                                    <span class="mi">0</span>
                                <span class="p">],</span>  <span class="c1"># optional if gca exists</span>
                                <span class="s2">&quot;platform&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;platform&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;paired&quot;</span><span class="p">:</span> <span class="nb">bool</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;paired&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">])),</span>
                                <span class="c1"># &quot;paired&quot;: g[&quot;paired&quot;].iloc[0],</span>
                                <span class="s2">&quot;predicted_tissue&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;predicted_tissue&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;pair1&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;_1&quot;</span><span class="p">)][</span>
                                    <span class="s2">&quot;file_url&quot;</span>
                                <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;md5_1&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;_1&quot;</span><span class="p">)][</span>
                                    <span class="s2">&quot;md5&quot;</span>
                                <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;pair2&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;_2&quot;</span><span class="p">)][</span>
                                    <span class="s2">&quot;file_url&quot;</span>
                                <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="s2">&quot;md5_2&quot;</span><span class="p">:</span> <span class="n">g</span><span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="s2">&quot;file_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;_2&quot;</span><span class="p">)][</span>
                                    <span class="s2">&quot;md5&quot;</span>
                                <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="p">}</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                    <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
                <span class="p">)</span>
                <span class="n">output_df</span> <span class="o">=</span> <span class="n">output_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tissue&quot;</span><span class="p">],</span> <span class="n">errors</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                    <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;predicted_tissue&quot;</span><span class="p">:</span> <span class="s2">&quot;tissue&quot;</span><span class="p">}</span>
                <span class="p">)</span>

                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">file_name</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">output_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                        <span class="n">f</span><span class="p">,</span>
                        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span>
                        <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                        <span class="n">columns</span><span class="o">=</span><span class="p">[</span>
                            <span class="s2">&quot;taxon_id&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;assembly_accession&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;platform&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;paired&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;tissue&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;run_accession&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;pair1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;md5_1&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;pair2&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;md5_2&quot;</span><span class="p">,</span>
                        <span class="p">],</span>
                        <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.prioritise_tissues" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">prioritise_tissues</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">priority_tissues</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.prioritise_tissues" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Assign a numeric priority to tissue predictions based on a predefined list.</p>
<p>This function adds a 'priority' column to the input DataFrame:
- Rows where the 'tissue_prediction' matches any of the <code>priority_tissues</code>
(case-insensitive, partial match) are assigned a priority of 1.
- All other rows receive a priority of 2.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing a 'tissue_prediction' column.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>priority_tissues</code>
            </td>
            <td>
                  <code><span title="List">List</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of tissue names to prioritize (e.g., ['brain', 'liver']).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: Modified DataFrame with an added 'priority' column.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">prioritise_tissues</span><span class="p">(</span>
    <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">priority_tissues</span><span class="p">:</span> <span class="n">typing</span><span class="o">.</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Assign a numeric priority to tissue predictions based on a predefined list.</span>

<span class="sd">    This function adds a &#39;priority&#39; column to the input DataFrame:</span>
<span class="sd">    - Rows where the &#39;tissue_prediction&#39; matches any of the `priority_tissues`</span>
<span class="sd">    (case-insensitive, partial match) are assigned a priority of 1.</span>
<span class="sd">    - All other rows receive a priority of 2.</span>

<span class="sd">    Args:</span>
<span class="sd">        df (pd.DataFrame): DataFrame containing a &#39;tissue_prediction&#39; column.</span>
<span class="sd">        priority_tissues (List[str]): List of tissue names to prioritize (e.g., [&#39;brain&#39;, &#39;liver&#39;]).</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: Modified DataFrame with an added &#39;priority&#39; column.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">priority_regex</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
        <span class="sa">r</span><span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">escape</span><span class="p">,</span> <span class="n">priority_tissues</span><span class="p">)),</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span>
    <span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;priority&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;tissue_prediction&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">1</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="n">priority_regex</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="mi">2</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="ensembl.genes.transcriptomic_data.select_transcriptomic_data.star_quality" class="doc doc-heading">
<code class="doc-symbol doc-symbol-heading doc-symbol-function"></code>            <code class="highlight language-python"><span class="n">star_quality</span><span class="p">(</span><span class="n">row</span><span class="p">)</span></code>

<a href="#ensembl.genes.transcriptomic_data.select_transcriptomic_data.star_quality" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate STAR quality based on the criteria.
The function checks the STAR quality criteria for each row and returns True if the criteria are met.
Args:
    row (<em>type</em>): dataframe row
Returns:
    bool: True if the criteria are met, False otherwise.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>src/python/ensembl/genes/transcriptomic_data/select_transcriptomic_data.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">star_quality</span><span class="p">(</span><span class="n">row</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate STAR quality based on the criteria.</span>
<span class="sd">    The function checks the STAR quality criteria for each row and returns True if the criteria are met.</span>
<span class="sd">    Args:</span>
<span class="sd">        row (_type_): dataframe row</span>
<span class="sd">    Returns:</span>
<span class="sd">        bool: True if the criteria are met, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;uniquely_mapped_reads_percentage&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">50</span>
        <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;percentage_reads_mapped_to_multiple_loci&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">30</span>
        <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;percentage_reads_unmapped_too_short&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">20</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; [2016-<script>document.write(new Date().getFullYear())</script>] <a href="https://www.ebi.ac.uk">EMBL-European Bioinformatics Institute</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.tooltips", "navigation.expand", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "search.highlight", "search.suggest", "toc.follow"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
    
  </body>
</html>